---
title: "To do"
output: html_notebook
---

* Make report from Flexgsea results ^
* Clustering with other comparisons (not just inv vs rest)
* GSEA with "interesting gene list"
* Variable influence factor (VIF) - Maybe
* Diffex and KM short vs long-term inv

Open questions/theories from survival interaction analysis:

Do low B cell patients have a high fibrinogen signature (FGB/FGG/FGFBP1)? Could the fibrinogen be suppressing B cell activity, resulting in increased mortality?
What about Akt1 signaling (DUSP13 and FLT3) and cancer-testis antigens (MAGE family)?

Recreate code for mountain plots with flexgsea
Steal some ideas from ClusterProlifer?
Or use UpSet plots instead?
```{r}
library(clusterProfiler)

d <- rep_inv_rest$annotated_results[,c("ensembl_gene_id", "log2FoldChange")]

d <- d %>% 
  left_join(select(gx_annot, ensembl_gene_id),
               ., by = "ensembl_gene_id") %>%
  filter(!is.na(log2FoldChange))
dim(d)
#Sort by decreasing l2fc
d <- d %>% arrange(desc(log2FoldChange))

## feature 1: numeric vector
geneList <- d$log2FoldChange

## feature 2: named vector
names(geneList) <- as.character(d$ensembl_gene_id)

test = gseGO(geneList=geneList, ont="BP", keyType = "ENSEMBL", OrgDb=org.Hs.eg.db, verbose=T, nPerm = 5, pvalueCutoff = 1)

head(test)
gseaplot(test, geneSetID="GO:0000724")
gseaplot
```


```{r}
inv_res_c2cp = readRDS(here("results", "flexgsea", "deseq", "results", "inv_vs_rest_canonpath_c2_results_flexdeseq.Rds"))

#names(inv_res_c2cp$flexgsea_results)
inv_res_c2cp[["flexgsea_results"]][["table"]] %>%
  as.data.frame() %>% arrange(inv_vs_rest_ppbc_inv_vs_rest.fdr)


```
```{r}
names(inv_res_c2cp$flexgsea_results)
inv_res_c2cp[["flexgsea_results"]][["table"]] %>%
  as.data.frame() %>% filter(inv_vs_rest_ppbc_inv_vs_rest.GeneSet %in% testpath)
```



```{r}
testpath =  rep_inv_rest$pathways %>% bind_rows() %>%
  filter(fdr < 0.05 & collection=="c2_canon") %>% pull(pathway)
```

```{r}
inv_res_c2cp = readRDS(here("results", "flexgsea", "deseq", "results", "inv_vs_rest_canonpath_c2_results_flexdeseq.Rds"))

#names(inv_res_c2cp$flexgsea_results)
inv_res_c2cp[["flexgsea_results"]][["table"]] %>%
  as.data.frame() %>% arrange(inv_vs_rest_ppbc_inv_vs_rest.fdr)


```



ALTERNATE THRESHOLD possibility:

Performs a filtering pass on the dataset to remove any features (rows) without at least 1 read per million in n of the samples, where n is the size of the smallest group of replicates. Recommended in:
Anders S, McCarthy DJ, Chen Y, Okoniewski M, Smyth GK, Huber W, Robinson MD (2013). " [Count-based differential expression analysis of RNA sequencing data using R and Bioconductor." Nat. Protocols, 8, 1765-1786.](https://www.nature.com/articles/nprot.2013.099)

```{r}
library(DESeq2)
library(tidyverse)
colnames(colData(dds))
as.data.frame(colData(dds)[,c("study_group", "TIL_percent")]) %>% 
  filter(!is.na(TIL_percent)) %>%
  ggplot(aes(x = study_group, y = TIL_percent)) +
  geom_boxplot() +
  ggtitle("TIL percent")
```

```{r}
print("Fix")
```


