---
title: "To do"
output: html_notebook
---


* in notebook 10, redo breastfeeding over time with immune analysis, use pregnancy as timepoint 
* also look at t-sne in clustering
* To do: Add two column annotations for the heatmap, showing both group vs rest and all groups separately in notebook 8.
* Cox regression to identify genes associated with survival
** First make Kaplan meijer without covariates, then compare with covariates
** Covariates: grade, stage, molecular subtype, age at diagnosis and year of diagnosis (Also possibly therapy type)
* Clustering with other comparisons (not just inv vs nonprbc)
* GSEA with "interesting gene list"
* Formally test for correlation between batch and study group
* CHeck batch effect for androgen receptor
* Make correlation matrix between batch and study group
* Variable influence factor (VIF)
* Set standardized colors for study group etc in heatmaps across the entire dataset
* Debug the row annotation colors in deduplicated heatmaps
* Redo diffex looking at just Leuven data


ALTERNATE THRESHOLD possibility:

Performs a filtering pass on the dataset to remove any features (rows) without at least 1 read per million in n of the samples, where n is the size of the smallest group of replicates. Recommended in:
Anders S, McCarthy DJ, Chen Y, Okoniewski M, Smyth GK, Huber W, Robinson MD (2013). " [Count-based differential expression analysis of RNA sequencing data using R and Bioconductor." Nat. Protocols, 8, 1765-1786.](https://www.nature.com/articles/nprot.2013.099)

```{r}
library(DESeq2)
library(tidyverse)
colnames(colData(dds))
as.data.frame(colData(dds)[,c("study_group", "TIL_percent")]) %>% 
  filter(!is.na(TIL_percent)) %>%
  ggplot(aes(x = study_group, y = TIL_percent)) +
  geom_boxplot() +
  ggtitle("TIL percent")
```


