---
title: "Process metadata"
date: "`r Sys.Date()`"
author: "Kat Moore"
output: 
  html_document:
    toc: yes
    toc_float: yes
    toc_depth: 4
    highlight: tango
    df_print: paged
---

```{r, include=F}
rm(list = ls())
```

```{r, include=F}
library(tidyverse)
library(openxlsx)
library(tximport)
library(here)
```

# Process patient annotations

As of April 2020, we rerun the entire analysis pipeline using new metadata from this file:

/DATA/share/postpartumbc/data/Hercoderingslijst_v09032020_KM.xlsx

```{r}
new_metadata_path = "/DATA/share/postpartumbc/data/external/Hercoderingslijst_v09032020_KM.xlsx"
```

This book contains the following sheets.

```{r}
readxl::excel_sheets(new_metadata_path)
```

The second sheet has been rewritten to be as compatible with R as possible.

```{r}
meta_raw <- readxl::read_excel(new_metadata_path, sheet = "Recoding_Book_v09032020_R")
```
```{r}
colnames(meta_raw)
```
A glimpse at the data type shows us that `Ouderdom FFPE` and `extraction_date` should be a date but are being read as a character.

```{r}
glimpse(meta_raw)
```

We manually specify those columns are dates.

```{r}
colnames(meta_raw)[c(7, 67)]
coltypes = c(rep("guess", 6), "date", rep("guess", 67-8), "date", rep("guess", ncol(meta_raw) - 67))
coltypes[c(7, 67)]
```


```{r}
meta_raw <- readxl::read_excel(new_metadata_path, sheet = "Recoding_Book_v09032020_R",
                               col_types = coltypes, na = c("", "x", "999", "opvragen", "e999tracted"))
```


The issue with dates appears to be caused by mixed format. In OuderdomFFPE for example, we have

9/25/2013 *American format (mo, day, year)
5/9/2007 *ambiguous
...
21/09/2004 *European format (day, mo, year)

This appears to break openxlsx's date detection. Reading it as text is no solution, because excel handles encoding of dates oddly, writing them as integers instead of strings.


## Fix dates

When this issue occurs, the date is skipped

```{r}
patient_annotations = meta_raw %>% rename(ouderdomFFPE = `Ouderdom FFPE`,
                                          nanodrop_ng_per_ul = `Nanodrop_ng/µl`,
                                          bioanalyze_ng_per_ul = `Bioanalyser_ng/µl`)

patient_annotations %>% filter(Ref == "HL-FA-034") %>% select(Ref, ouderdomFFPE)
```

Fix these.

```{r}
#Expecting date in G20 / R20C7: got '01/09/1997'
patient_annotations[[19,"ouderdomFFPE"]] <- as.POSIXct("01/09/1997", tz="UTC")
#Expecting date in G22 / R22C7: got '04/02/2002'
patient_annotations[[21,"ouderdomFFPE"]] <- as.POSIXct("04/02/2002", tz="UTC")
#Expecting date in G23 / R23C7: got '07/01/2003'
patient_annotations[[22,"ouderdomFFPE"]] <- as.POSIXct("07/01/2003", tz="UTC")
#Expecting date in G26 / R26C7: got '12/03/2004'
patient_annotations[[25,"ouderdomFFPE"]] <- as.POSIXct("12/03/2004", tz="UTC")
#Expecting date in G27 / R27C7: got '02/04/2004'
patient_annotations[[26,"ouderdomFFPE"]] <- as.POSIXct("02/04/2004", tz="UTC")
#Expecting date in G29 / R29C7: got '21/09/2004'
patient_annotations[[28,"ouderdomFFPE"]] <- as.POSIXct("21/09/2004", tz="UTC")
#Expecting date in G30 / R30C7: got '13/10/2004'
patient_annotations[[29,"ouderdomFFPE"]] <- as.POSIXct("13/10/2004", tz="UTC")
#Expecting date in G32 / R32C7: got '16/11/2005'
patient_annotations[[31,"ouderdomFFPE"]] <- as.POSIXct("16/11/2005", tz="UTC")
#Expecting date in G33 / R33C7: got '23/11/2005'
patient_annotations[[32,"ouderdomFFPE"]] <- as.POSIXct("23/11/2005", tz="UTC")
#Expecting date in G34 / R34C7: got '28/11/2005'
patient_annotations[[33,"ouderdomFFPE"]] <- as.POSIXct("28/11/2005", tz="UTC")
#Expecting date in G35 / R35C7: got '07/02/2006'
patient_annotations[[34,"ouderdomFFPE"]] <- as.POSIXct("07/02/2006", tz="UTC")
#Expecting date in G36 / R36C7: got '11/06/2007'
patient_annotations[[35,"ouderdomFFPE"]] <- as.POSIXct("11/06/2007", tz="UTC")
#Expecting date in G38 / R38C7: got '21/10/2008'
patient_annotations[[37,"ouderdomFFPE"]] <- as.POSIXct("21/10/2008", tz="UTC")
#Expecting date in G40 / R40C7: got '29/04/2010'
patient_annotations[[39,"ouderdomFFPE"]] <- as.POSIXct("29/04/2010", tz="UTC")
#Expecting date in G41 / R41C7: got '09/07/2010'
patient_annotations[[40,"ouderdomFFPE"]] <- as.POSIXct("09/07/2010", tz="UTC")
#Expecting date in G42 / R42C7: got '05/08/2010'
patient_annotations[[41,"ouderdomFFPE"]] <- as.POSIXct("05/08/2010", tz="UTC")
#Expecting date in G44 / R44C7: got '22/12/2010'
patient_annotations[[43,"ouderdomFFPE"]] <- as.POSIXct("22/12/2010", tz="UTC")
```

Remaining NAs in the date columns are true NAs (marked as "x" in original document).

## Remove previously generated analyses columns

Remove to avoid "time travel" (incorporating prior results that haven't been run yet in this version of the pipeline)

IG_ID
IG_ID_integer
PAM50
PAM50_integer

These columns were added prior to the April 2020 metadata update and will be regenerated later.

```{r}
patient_annotations <- patient_annotations %>% select(-IG_ID, -IG_ID_integer,
                                                      -PAM50, -PAM50_integer)
```

## Standarize column names

The following critical columns have had their names changed:

patient_ref -> Ref
overall_survival -> death
months_overall_survival -> time_death_months
distant_recurrence -> DR
months_to_drs -> time_DR_months
metastasis_at_diagnosis -> DR_diagnosis
months_of_followup -> time_FU_months

Note: 
DR refers to the presence of a metastasis, either at diagnosis or at a later timepoint.
DR recurrence refers only to those patients that suffered an additional metastasis after diagnosis.

```{r}
patient_annotations <- patient_annotations %>%
  rename(patient_ref = Ref,
         overall_survival = death,
         months_overall_survival = time_death_months,
         distant_recurrence = DR,
         months_to_drs = time_DR_months,
         metastasis_at_diagnosis = DR_diagnosis,
         months_of_followup = time_FU_months
  )
```

Note: Many other columns have also changed names, but these are the ones most frequently referenced in the survival analysis.

## Filter on sequenced samples

We only want those samples for which the status is SEQUENCED.
It appears that only sequenced samples have been included in this version of the metadata.

```{r}
patient_annotations$Status...3 %>% unique()
patient_annotations$Status...4 %>% unique()
```

```{r}
stopifnot(sum(patient_annotations$Status...3 == "SEQUENCED") == length(patient_annotations$Status...3=="SEQUENCED"))
```

# Process sample names

Add a separate "sample_ref" column that will be used to create a unique identifier for those patients who have been sequenced multiple times.

```{r}
patient_annotations = patient_annotations %>%
  dplyr::mutate(
    patient_ref = str_replace_all(patient_ref, "_", "-"),
    old_ref    = str_replace_all(old_ref, "_", "-"),
    sample_ref = patient_ref)
```


```{r}
samples <- list.files(here("data", "RNA-seq", "salmon"))

#All samples begin with this string
samples <- samples[str_detect(samples,"^rHLE")]

head(samples)
```

Sample name consists of:

* an identifier string that corresponds to the patient_ref or old_ref
* the date
* the sequencer
* the lane 
* a gcap identifier
* the strand (always single stranded) 

Separate the fields of interest into a data frame.

```{r}
seq_annot <- 
  tibble("sample_id" = samples) %>%
  separate(
    sample_id,
    into = c("sample_ref", "date",
             "dummy1", #HiSeq4000
             "dummy2", #FCB
             "lane",
             "dummy3", #gcap_number
             "dummy4"), #R1
    sep = "\\.",
    remove = FALSE
  ) %>%
  select(-contains("dummy")) %>%
  mutate(sample_ref = str_replace_all(sample_ref, "_", "-")) %>%
  mutate(sample_ref = str_replace(sample_ref, "rHLE-FA", "HL-FA"))


head(seq_annot)
```

#Combine metadata with file names

Several samples were sequenced more than once, so there are more fastq files than metadata entries.

```{r}
print(paste("Metadata entries:", nrow(patient_annotations)))
print(paste("Fastq files:", length(samples)))
```

Identify those files which do not immediately match a metadata entry.

```{r}
setdiff(seq_annot$sample_ref, patient_annotations$sample_ref)
```

## Remove bad library prep

HL-FA-117-and-rHLE-FA-151-index14 is a mistake in library prep from which the sample will need to be discarded.

```{r}
excluded = seq_annot %>%
  filter(str_detect(sample_ref, pattern="117-and-")) %>%
  select(fastq_file = sample_id, sample_ref)
excluded = excluded %>% mutate(reason = "lib prep error")
excluded
```

Remove this sample from the metadata.

```{r}
seq_annot = seq_annot %>% filter(!str_detect(sample_ref, pattern="117-and-"))
setdiff(seq_annot$sample_ref, patient_annotations$sample_ref)
```

## Remove neo-adj patients

At this stage, these samples remain that lack matching metadata.

```{r}
setdiff(seq_annot$sample_ref, patient_annotations$sample_ref)
```

Some of these come from samples which were accidentally sequenced even though they had neo-adjuvant chemotherapy.
We can see this by looking at older versions of the metadata.

```{r}
old_patients = readxl::read_excel(here("data", "external", "Hecoderingslijst HL-FA_v26062019.xlsx"), skip = 2)

#Once again the file nomenclature has changed over time
old_patients = old_patients %>%
  mutate(`OLD REF` = str_replace(`OLD REF`, "rHLE_", "rHLE-"))

#Filter those original patients with this status and retrieve both old and new identifiers
all_ct = old_patients %>% filter(Status == "Neo-Adj CT")
#Either the new or the old ref may be in referenced by the fastq file name
all_ct = c(all_ct$`NEW REF`, all_ct$`OLD REF`) 

#Cross reference with our list of sample prefixes absent in the current metadata
had_ct = all_ct[all_ct %in% setdiff(seq_annot$sample_ref, patient_annotations$sample_ref)]

had_ct
```

Remove these from the samples to be read.

```{r}
excluded = rbind(
  excluded,
  mutate(select(
    filter(seq_annot, sample_ref %in% had_ct),
    fastq_file = sample_id, sample_ref), reason = "had ct")
)

seq_annot = seq_annot %>% filter(!sample_ref %in% had_ct)
```

Running tally of files excluded:

```{r}
table(excluded$reason)
```

Remaining samples with no matching metadata

```{r}
setdiff(seq_annot$sample_ref, patient_annotations$sample_ref)
```

## Remove samples above the age cutoff

A few samples were sequenced even though they come from patients who were too old.

```{r}
too_old = old_patients %>% filter(str_detect(Status, "OLD")) #Can be spelled "TOO OLD" or "TO OLD"
#Either the new or the old ref may be in referenced by the fastq file name
too_old = c(too_old$`NEW REF`, too_old$`OLD REF`)

#Cross reference with our list of sample prefixes absent in the current metadata
too_old = too_old[too_old %in% setdiff(seq_annot$sample_ref, patient_annotations$sample_ref)]
too_old
```

Remove these from the samples to be read.

```{r}
excluded = rbind(
  excluded,
  mutate(select(
    filter(seq_annot, sample_ref %in% too_old),
    fastq_file = sample_id, sample_ref), reason = "too old")
)

seq_annot = seq_annot %>% filter(!sample_ref %in% too_old)
```

Running tally of files excluded:

```{r}
table(excluded$reason)
```

Remaining samples with no matching metadata

```{r}
setdiff(seq_annot$sample_ref, patient_annotations$sample_ref)
```

## Add leading zeros

Any identifier with less than three digits in the final field is padded with a leading 0 in the metadata. 

For instance: `rHLE-FA-10` is in the current metadata as `HL-FA-010`

This is NOT true for older identifiers, which have 3 dashes instead of 2.

```{r}
#Separate older identifiers which don't need padding from newer identifiers which do
old_refs = seq_annot %>%
  filter(str_count(sample_ref, "-")==3)

new_refs = seq_annot %>%
  filter(str_count(sample_ref, "-")==2)

#Split the newer sample refs by "-"
split_ids = new_refs %>%
  pull(sample_ref) %>%
  str_split("-")

#Changes in sample nomenclature over time yield lists of various lengths
#Newer samples give 3 fields, older ones give 4
#No longer needed as we separate them now
#lapply(split_ids, length) %>% unlist() %>% table()

#The index is always the last field
new_refs$idx = unlist(lapply(split_ids, function(x) x[[length(x)]]))

#Add a leading 0 to any index with less than 3 digits
new_refs = new_refs %>%
  mutate(idx = str_pad(idx, width = 3, side = "left", pad = "0"),
         file_prefix = sample_ref #Keep the original file prefix
  )

#Get all fields EXCEPT the last one and collapse them to preserve the text prefix
new_refs$text_prefix = unlist(lapply(
  lapply(split_ids, function(x) x[-length(x)]),
  function(x) paste(x, collapse="-")
))

#Now we can paste the text prefix and the idx together to get a sample ref that matches the metadata
new_refs = new_refs %>%
  mutate(sample_ref = paste(text_prefix, idx, sep = "-"))

#Recombine into one data frame
seq_annot = rbind(select(new_refs, sample_id, sample_ref, date, lane),
                  old_refs)
```

```{r, include=F}
seq_annot %>% arrange(sample_ref) %>% head(30)
```

Remaining samples with no matching metadata

```{r}
setdiff(seq_annot$sample_ref, patient_annotations$sample_ref)
```

## Unique ID for fastqs with old identifiers

Some of the sequencing runs were done using the old identifiers, which contain 4 dashes instead of three and start with rHLE instead of HL-FA. 
These won't be found in the patient_ref column (from which sample_ref was built).
Older identifiers contain an LP or UZ-HL pattern. Set this as "sample_ref" so we can match them with their metadata.

```{r}
patient_annotations = bind_rows(patient_annotations,
                                mutate(filter(patient_annotations, str_detect(old_ref, "LP|UZ-HL")), sample_ref=old_ref))
stopifnot(sum(duplicated(patient_annotations$sample_ref)) == 0)
```

## Match metadata to fastq

Combine based on fastq file names

```{r}
sample_annot = left_join(select(seq_annot, sample_ref, sample_id,),
                         patient_annotations, by="sample_ref")

head(sample_annot)
```

## Duplicate summary

These patients have duplicate samples:

```{r}
duplicate_samples = sample_annot %>%
  filter(patient_ref %in% filter(sample_annot, duplicated(patient_ref))$patient_ref) %>%
  arrange(patient_ref) %>% select(patient_ref, sample_ref, sample_id, old_ref)

duplicate_samples
```

Most patients with more than one sample were sequenced twice, but 7 were sequenced three times.

```{r}
duplicate_samples %>% group_by(patient_ref) %>% count() %>%
  pull(n) %>% table()
```

Save the duplicate samples for future reference.

```{r}
write_csv(duplicate_samples,
          path = here("data", "metadata", "01_patients_with_multiple_fastqs.csv"))
```

## Samples excluded

The final tally of pre-excluded samples:

```{r}
table(excluded$reason)
```

```{r}
write_csv(excluded,
          here("data", "metadata", "01_pre_excluded_samples.csv"))
```


# Standardize study group and molecular subtype

Check the study group for irregular strings and dangling whitespace.

```{r}
sample_annot = sample_annot %>%
  mutate(study_group = str_replace_all(tolower(str_trim(study_group, side="both")), "-", "_")) %>%
  select(sample_ref, study_group, everything())

sample_annot %>% group_by(study_group) %>% summarise(n=n())
```

```{r}
stopifnot(nrow(filter(sample_annot, is.na(study_group))) == 0)
```

Standardize molecular subtype

```{r}
sample_annot = sample_annot %>% mutate(molecular_subtype = case_when(
  molecular_subtype == "Luminal B (HER2-)" ~ "LumB_Her2neg",
  molecular_subtype == "Triple Negative" ~ "TripleNegative",
  molecular_subtype == "Luminal A" ~ "LumA",
  molecular_subtype == "Luminal B (HER2+)" ~ "LumB_Her2pos",
  molecular_subtype == "HER2+ (non-luminal)" ~ "Her2pos_nonluminal",
  TRUE ~ molecular_subtype
))

sample_annot %>% group_by(molecular_subtype) %>% summarise(n=n())
```

## Individual sample checks

Also double check a few samples based on this info from Hanne:

> HL-FA-90 is PPBC-lac en dus niet PPBC-inv

Already corrected in the current metadata:

```{r}
sample_annot %>% filter(str_detect(patient_ref, "HL-FA-090")) %>% select(sample_ref, study_group)
```

> HL-FA-111: This is also a patient from Berlin. Here I seemed to have made a mistake. This patient is not HER2+, but Triple Negative. 

This sample has also been corrected.

```{r}
sample_annot %>% filter(str_detect(patient_ref, "HL-FA-111")) %>% select(sample_ref, molecular_subtype)
```

> HL-FA-38 is ppbc-inv instead of prbc

```{r}
sample_annot %>% filter(str_detect(patient_ref, "HL-FA-038")) %>% select(sample_ref, study_group)
```

# Sample identifiers

Add an easily readable sample name.

```{r}
sample_annot = sample_annot %>%
  mutate(sample_name = paste(sample_annot$study_group,
                             str_remove(string = sample_annot$patient_ref, pattern = "HL-FA-"),
                             sep="_"))

sample_annot = sample_annot %>%
  select(sample_id, sample_ref, patient_ref,
         sample_name, study_group, everything())

head(sample_annot)
```

```{r}
stopifnot(length(unique(sample_annot$patient_ref))==length(unique(sample_annot$sample_name)))
```


# Save the metadata.

Before we do that, let's remove the totally unnecessary status columns. They both contain the same value everywhere.

```{r, collapse=T}
unique(sample_annot$Status...3)
unique(sample_annot$Status...4)
```
```{r}
sample_annot = sample_annot %>% select(-Status...3, -Status...4)
```

```{r}
head(sample_annot)
```

```{r}
write_tsv(sample_annot, here("data", "metadata", "01_sample_annot.tsv"))
```

# Gene annotation processing

Biomart export contains unfriendly column names : fix that.

```{r}
tx_annot <- read_tsv(here("data", "external", "biomart_ensemblgenes_v94.txt")) %>%
  select(
    gene_id = `Gene stable ID`,
    transcript_id = `Transcript stable ID`,
    transcript_id_version = `Transcript stable ID version`,
    gene_name = `Gene name`,
    gene_type = `Gene type`,
    hgnc_symbol = `HGNC symbol`,
    gene_description = `Gene description`,
    chromosome_name = `Chromosome/scaffold name`,
    transcipt_start =  `Transcript start (bp)`,
    transcipt_end =  `Transcript end (bp)`,
    transcription_start_site = `Transcription start site (TSS)`
  )
write_tsv(tx_annot, here("data", "metadata", "01_tx_annot.tsv"))
```

```{r}
head(tx_annot)
```

# Tximport

Create a tximport object using our new metadata and all files.

```{r tximport}
quant_files <- here("data", "RNA-seq", "salmon", sample_annot$sample_id, "quant.sf")
names(quant_files) <- sample_annot$sample_id

tx2gene <- select(tx_annot, transcript_id_version, gene_id)
tx <- tximport(quant_files, type = "salmon", tx2gene = tx2gene)
```

Save the tx data for downstream analysis.

```{r save tx}
saveRDS(tx,file = here("data", "Rds", "01_tx.Rds"))
```

Save notebook data.

```{r save image}
save.image(here("reports","01_process_metadata_tximport.RData"))
```

# Session info

```{r}
sessionInfo()
```



