---
title: "Likelihood ratio test, standard count threshhold"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
---

```{r, include=F}
#library(devtools)
#install_github("jokergoo/ComplexHeatmap")
#install_github("mikelove/DESeq2")
#install_github("azhu513/apeglm")

library(DESeq2)
library(apeglm)
library(here)
library(tidyverse)
library(ggrepel)
library(ComplexHeatmap)
library(openxlsx)
library(scrime)
library(RColorBrewer)
```

In this notebook, we enact differential expression as a fourway comparison between all 4 postpartum breast cancer groups, using the DESeq2 package and a likelihood ratio test. We perform this using the standard count threshold of non-zero expression in at least one third of all samples.

# Load and pre-process data

```{r}
dds <- readRDS(file = here("data/Rds/05_dds_PAM50_tumorpurity_batch.Rds"))

gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
gx_annot = gx_annot %>% select(ensembl_gene_id = gene_id, gene_name, gene_type, description = gene_description) %>% distinct()
```


## Filtering

Require non-zero counts in 1/3 of the dataset.

```{r}

keep <- rowSums(counts(dds)!=0) >= ceiling(ncol(dds)/3)
table(keep)

#Hang on to what we discard for later analysis
dds.discarded <-dds[!keep,]

#Keep the rest
dds <- dds[keep,]

```


## Retrieve immune genes

For downstream analyses, we'd like to know which of the differentially expressed genes are immunologically relevant.

An immune gene is defined as follows:

Either immune/immuno/interleukin is part of the gene name and description OR the gene is part of the the ImmPort database. See https://www.innatedb.com/redirect.do?go=resourcesGeneLists

The Immunology Database and Analysis Portal (ImmPort) system was developed under the Bioinformatics Integration Support Contract (BISC) Phase II by the Northrop Grumman Information Technology Health Solutions team for the NIH, NIAID, and DAIT. The principal investigator of the BISC project is Dr. Richard Scheuermann at University of Texas Southwestern Medical Center. The list of immunologically related genes in ImmPort is a collection of ~6,000 human genes, which was formed with the goal of retrieving all genes that have immune system-related functions. This list was generated using automatic searches of EntrezGene and Gene Ontology records using immunology-related keywords. The list was then manually curated by immunology experts examining various literature sources. 

```{r}
immune_gene_list <- read_csv(here("data", "external","gene-sets","InnateDB_genes.csv"))
head(immune_gene_list)
```

## Helper functions

```{r}
source(here("src", "deseq_report_functions.R"))
```

## Variance stabilizing transformation

We want a transformation that is fully blind to the experimental design, but still uses the faster sampling method in the vst wrapper for varianceStabilizingTransformation.

This homoskedastic dataset will be used later on for heatmap visualization.


```{r, eval=F}
vsd = blind_vst(dds)
saveRDS(vsd, here("data","Rds","06_vsd_standardfilter.Rds"))
```

```{r}
vsd = readRDS(here("data","Rds","06_vsd_standardfilter.Rds"))
```



# Likelihood ratio test

The likelihood ratio test (LRT) is an ANOVA-like test design to check for genes which are significantly differentially expressed in at least one group. From the manual: "The LRT examines two models for the counts, a full model with a certain number of terms and a reduced model, in which some of the terms of the full model are removed. The test determines if the increased likelihood of the data using the extra terms in the full model is more than expected if those extra terms are truly zero."

Our design formula is as follows:

```{r}
design(dds)
```

We will test it versus a formula that lacks the study group for genes of interest.
To save time, run once, then load the object.

```{r, eval=F}
start <- Sys.time()
ddsLRT = DESeq(dds, test="LRT", reduced= ~ batch + PAM50 + tumor_purity)
end <- Sys.time()
end-start

saveRDS(ddsLRT, here("data/Rds/06_ddsLRT_standard.Rds"))
```

>estimating size factors
using 'avgTxLength' from assays(dds), correcting for library size
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing
Time difference of 2.100146 hours


```{r}
ddsLRT = readRDS(here("data/Rds/06_ddsLRT_standard.Rds"))
```

```{r}
resLRT <- results(ddsLRT)
```


# LRT report

## Summary

```{r}
repLRT = deseq_report(resLRT, dds=ddsLRT, absl2fc = 0.5)
```




## Top genes

```{r}
repLRT$significant_genes %>% head(30)
```

## Heatmap

```{r}
draw(repLRT$heatmap, column_title = "Likelihood ratio test")
```


## Box plots

Plot the top 5 most significant.

```{r}
repLRT$beehive_plots
```

# Specific genes of interest

##Significant genes

### Interleukin 18 receptor 1

ENSG00000115604

```{r}
plot_gene_beehive(ddsLRT, result_df = repLRT$annotated_results[repLRT$annotated_results$ensembl_gene_id == "ENSG00000115604",])
```

### NOS1

Nitric oxide synthase

```{r}
plot_gene_beehive(ddsLRT, result_df = repLRT$annotated_results[repLRT$annotated_results$ensembl_gene_id == "ENSG00000089250",])
```

### CD68

CD68 (Cluster of Differentiation 68) is a protein highly expressed by cells in the monocyte lineage (e.g., monocytic phagocytes, osteoclasts), by circulating macrophages, and by tissue macrophages (e.g., Kupffer cells, microglia).[5] 

```{r}
plot_gene_beehive(ddsLRT, result_df = repLRT$annotated_results[repLRT$annotated_results$ensembl_gene_id == "ENSG00000129226",])
```

### Granulin precursor
ENSG00000030582	GRN

```{r}
plot_gene_beehive(ddsLRT, result_df = repLRT$annotated_results[repLRT$annotated_results$ensembl_gene_id == "ENSG00000030582",])
```

### Major histocompatibility complex, class II, DQ alpha 1	

ENSG00000228284	HLA-DQA1

```{r}
plot_gene_beehive(ddsLRT, result_df = repLRT$annotated_results[repLRT$annotated_results$ensembl_gene_id == "ENSG00000228284",])
```

## Various immune markers

### CD20

Official name of CD20 is MS4A1 ENSG00000156738
http://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000156738;r=11:60455752-60470760

```{r}
plot_gene_beehive(ddsLRT, result_df = repLRT$annotated_results[repLRT$annotated_results$ensembl_gene_id == "ENSG00000156738",])
```

Other plasma cells markers: CD138, CD98, TACI, BCMA
https://www.ncbi.nlm.nih.gov/pubmed/28787106

### CD138

CD138 = SDC1 syndecan 1  ENSG00000115884
http://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000115884;r=2:20200797-20225433

```{r}
plot_gene_beehive(ddsLRT, result_df = repLRT$annotated_results[repLRT$annotated_results$ensembl_gene_id == "ENSG00000115884",])
```

### CD21

ENSG00000117322

```{r}
plot_gene_beehive(ddsLRT, result_df = repLRT$annotated_results[repLRT$annotated_results$ensembl_gene_id == "ENSG00000117322",])
```

### CD30

TNFRSF8 ENSG00000120949
```{r}
plot_gene_beehive(ddsLRT, result_df = repLRT$annotated_results[repLRT$annotated_results$ensembl_gene_id == "ENSG00000120949",])
```

```{r}
save.image(here("reports/06_diffex_lrt_standard.Rdata"))
```

