---
title: "Likelihood ratio test, standard count threshhold"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
---

```{r, include=F}
rm(list = ls())
```

```{r, include=F}

#library(devtools)
#install_github("jokergoo/ComplexHeatmap")
#install_github("mikelove/DESeq2")
#install_github("azhu513/apeglm")

library(DESeq2)
library(apeglm)
library(here)
library(tidyverse)
library(ggrepel)
library(ComplexHeatmap)
library(openxlsx)
library(scrime)
library(RColorBrewer)
```

In this notebook, we enact differential expression as a fourway comparison between all 4 postpartum breast cancer groups, using the DESeq2 package and a likelihood ratio test. We perform this using the standard count threshold of non-zero expression in at least one third of all samples.

# Load and pre-process data

## Gene and sample data

```{r}
dds <- readRDS(file = here("data/Rds/05_dds_PAM50_batch.Rds"))

gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
gx_annot = gx_annot %>% select(ensembl_gene_id = gene_id, gene_name, gene_type, description = gene_description) %>% distinct()
```

## Gene signatures

```{r}
gene_sets = list(
  go_bp = flexgsea::read_gmt(here("data","external","gmt","c5.bp.v7.0.symbols.gmt")),
  hallmark = flexgsea::read_gmt(here("data","external","gmt","h.all.v7.0.symbols.gmt")),
  c2_canon = flexgsea::read_gmt(here("data","external","gmt","c2.cp.v7.0.symbols.gmt")),
  c2_cgp = flexgsea::read_gmt(here("data","external","gmt","c2.cgp.v7.0.symbols.gmt"))
)
```

## Filtering

Require non-zero counts in 1/3 of the dataset.

```{r}
keep <- rowSums(counts(dds)!=0) >= ceiling(ncol(dds)/3)
table(keep)

#Keep the rest
dds <- dds[keep,]
```

```{r}
table(dds$study_group)
```


Create a version without the lactation group, as it has very few samples compared to the rest

```{r}
dds.nolac = dds[,colnames(dds)[colData(dds)$study_group != "ppbc_lac"]]
dds.nolac$study_group = droplevels(dds.nolac$study_group)
saveRDS(dds.nolac, here("data", "Rds", "06_dds_nolac.Rds"))

table(dds.nolac$study_group)
```


## Retrieve immune genes

For downstream analyses, we'd like to know which of the differentially expressed genes are immunologically relevant.

An immune gene is defined as follows:

Either immune/immuno/interleukin is part of the gene name and description OR the gene is part of the the [ImmPort database](https://www.innatedb.com/redirect.do?go=resourcesGeneLists).

The Immunology Database and Analysis Portal (ImmPort) system was developed under the Bioinformatics Integration Support Contract (BISC) Phase II by the Northrop Grumman Information Technology Health Solutions team for the NIH, NIAID, and DAIT. The list of immunologically related genes in ImmPort is a collection of ~6,000 human genes, which was formed with the goal of retrieving all genes that have immune system-related functions. This list was generated using automatic searches of EntrezGene and Gene Ontology records using immunology-related keywords. The list was then manually curated by immunology experts examining various literature sources. 

```{r}
immune_gene_list <- read_csv(here("data", "external","gene-sets","InnateDB_genes.csv"))
head(immune_gene_list)
```


## Helper functions

```{r}
source(here("src", "deseq_report_functions.R"))
```

## Heatmap color parameters
 
Function to visualize the colors we've chosen for each group.
Adapted version of show_col from the scales package to plot group labels

```{r}
#All annotation info
ann_top = select(as.data.frame(colData(dds)), sample_name, study_group, PAM50) %>%
  mutate(study_group = factor(study_group, levels=c("non_prbc", "ppbc_inv", "ppbc_lac", "prbc", "rest"))) %>%
  column_to_rownames("sample_name")

#Colors must be named lists with named elements

#Define study group colors
study_colors=suppressWarnings(colorRampPalette(brewer.pal(length(levels(ann_top$study_group)),"Spectral"))(length(levels(ann_top$study_group))))
names(study_colors) = levels(ann_top$study_group)

#Define PAM50 colors
pam_colors = suppressWarnings(colorRampPalette(brewer.pal(length(unique(ann_top$PAM50)),"Paired"))(length(unique(ann_top$PAM50))))
names(pam_colors)= unique(ann_top$PAM50)
top_cols=list(study_group=study_colors, PAM50 = pam_colors)

#Define row annotation colors
#Reduce the possible types of genes down to those defined in annotate_results()
gene_types = c("other noncoding", "immune protein coding", "protein coding", "immunoglobulin", "T cell receptor")
gene_colors = suppressWarnings(colorRampPalette(brewer.pal(length(unique(gene_types)),"Dark2"))(length(unique(gene_types))))
names(gene_colors) = gene_types

color_grid(study_colors)
color_grid(pam_colors)
color_grid(gene_colors)
```

```{r}
saveRDS(list(study_colors = study_colors, pam_colors = pam_colors, gene_colors = gene_colors),
        here("data", "Rds", "06_heatmap_colors.Rds"))
```





```{r}
#Annotation object for heatmap
colTop <- HeatmapAnnotation(df = ann_top, which="col", col = top_cols)
```


## Variance stabilizing transformation

We want a transformation that is fully blind to the experimental design, but still uses the faster sampling method in the vst wrapper for varianceStabilizingTransformation.

This homoskedastic dataset will be used later on for heatmap visualization.


```{r, eval=F}
vsd = blind_vst(dds)
saveRDS(vsd, here("data","Rds","06_vsd_standardfilter.Rds"))
```

```{r}
vsd = readRDS(here("data","Rds","06_vsd_standardfilter.Rds"))
```

A version without the lac group is also produced.

```{r, eval=F}
vsd.nolac = blind_vst(dds.nolac)
saveRDS(vsd.nolac, here("data", "Rds", "06_vsd_nolac.Rds"))
```

```{r}
vsd.nolac = readRDS(here("data", "Rds", "06_vsd_nolac.Rds"))
```


# Likelihood ratio test

The likelihood ratio test (LRT) is an ANOVA-like test design to check for genes which are significantly differentially expressed in at least one group. From the manual: "The LRT examines two models for the counts, a full model with a certain number of terms and a reduced model, in which some of the terms of the full model are removed. The test determines if the increased likelihood of the data using the extra terms in the full model is more than expected if those extra terms are truly zero."

Our design formula is as follows (tumor_purity has been removed on the basis of notebook 5):

```{r}
design(dds)
```

We will test it versus a formula that lacks the study group for genes of interest.
To save time, run once, then load the object.

```{r, eval=F}
start <- Sys.time()
ddsLRT = DESeq(dds, test="LRT", reduced= ~ batch + PAM50)
end <- Sys.time()
end-start

saveRDS(ddsLRT, here("data/Rds/06_ddsLRT_standard.Rds"))
```


```{r}
ddsLRT = readRDS(here("data/Rds/06_ddsLRT_standard.Rds"))
```

```{r}
resLRT <- results(ddsLRT)
```

```{r}
resLRT
```



# LRT report

## Summary

```{r}
repLRT = deseq_report(resLRT, dds=ddsLRT, absl2fc = 1, title="Likelihood ratio test")
```

## Top genes

```{r}
repLRT$significant_genes %>% head(30)
```

## Heatmap

```{r, fig.height=6, fig.width=10}
repLRT$heatmap
```


## Box plots

Plot the top 5 most significant.

```{r}
repLRT$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
 repLRT$pathways %>% bind_rows() %>%
  filter(fdr < 0.1) %>% select(pathway,everything()) %>% arrange(fdr)
```


```{r, fig.height=6, fig.width=8}
#repLRT$pathway_plots

bind_rows(repLRT$pathways) %>%
  #filter(fdr < 0.1) %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Sig pathways fdr < 0.1: LRT study group"
      )
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r}
repLRT$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 10:

```{r}
repLRT$outlier_bees
```

# LRT - without the lac group

Both upstream and downstream results indicate that lactation samples, by far the smallest group in the study, are contributing disproportionately to the model and, in this small selection, have a poorer prognosis than is represented in the larger clinical data. For this reason, we also rerun the analysis when excluding the lactation group.

```{r, eval=F}
if(file.exists(here("data", "Rds", "06_ddsLRT_nolac.Rds"))==F){
  print("Starting LRT vs reduced formula ~ batch + PAM50 without lac")
  start <- Sys.time()
  ddsLRT.nolac = DESeq(dds.nolac, test="LRT", reduced= ~ batch + PAM50)
  end <- Sys.time()
  print(end-start)
  saveRDS(ddsLRT.nolac, here("data", "Rds", "06_ddsLRT_nolac.Rds"))
} 
```

```{r}
ddsLRT.nolac = readRDS(here("data", "Rds", "06_ddsLRT_nolac.Rds"))
```

```{r}
resLRT.nolac = results(ddsLRT.nolac)
```

## Summary

```{r}
repLRT.nolac = deseq_report(resLRT.nolac, dds=ddsLRT.nolac,
                            variance_stabilized_dds = vsd.nolac, absl2fc = 1, title="LRT, lactation excluded")
```


## Top genes

```{r}
repLRT.nolac$significant_genes %>% head(30)
```

## Comparison with group including lac

```{r}
print(paste("Sig genes with lac:", nrow(repLRT$significant_genes)))
print(paste("Sig genes without lac:", nrow(repLRT.nolac$significant_genes)))
```

Overlap: Decidedly more TCR genes than was evident in the original list.

```{r}
#distinct(gx_annot[gx_annot$gene_name %in% intersect(unique(repLRT$significant_genes$gene_name), unique(repLRT.nolac$significant_genes$gene_name)),])
intersect(unique(repLRT$significant_genes$gene_name), unique(repLRT.nolac$significant_genes$gene_name))
```

Lost from diffex when excluding lac:

```{r}
unique(repLRT$significant_genes$gene_name)[!unique(repLRT$significant_genes$gene_name) %in% unique(repLRT.nolac$significant_genes$gene_name)]
```

Becomes significant after excluding lac, not in original list:

```{r}
unique(repLRT.nolac$significant_genes$gene_name)[!unique(repLRT.nolac$significant_genes$gene_name) %in% unique(repLRT$significant_genes$gene_name)]
```

Plot the padj of significant genes with and without lac against each other:

```{r}
sig.somewhere = unique(c(repLRT$significant_genes$ensembl_gene_id, repLRT.nolac$significant_genes$ensembl_gene_id))

tibble(
  lac.padj = arrange(repLRT$annotated_results[repLRT$annotated_results$ensembl_gene_id %in% sig.somewhere,c("ensembl_gene_id","padj")],
                     ensembl_gene_id)$padj,
  nolac.padj = arrange(repLRT.nolac$annotated_results[repLRT.nolac$annotated_results$ensembl_gene_id %in% sig.somewhere,c("ensembl_gene_id","padj")],
                       ensembl_gene_id)$padj
  ) %>%
  ggplot(aes(x = lac.padj, y = nolac.padj)) + geom_point() +
  ggtitle("Significant genes in LRT with and without lac samples")

tibble(
  lac.padj = repLRT$annotated_results %>% arrange(ensembl_gene_id) %>% pull(padj),
  nolac.padj = repLRT.nolac$annotated_results %>% arrange(ensembl_gene_id) %>% pull(padj)
  ) %>%
  ggplot(aes(x = lac.padj, y = nolac.padj)) + geom_point() +
  ggtitle("All genes in LRT with and without lac samples")
```

(WOW)

```{r, include=F}
#Save the above plots

tibble(
  lac.padj = arrange(repLRT$annotated_results[repLRT$annotated_results$ensembl_gene_id %in% sig.somewhere,c("ensembl_gene_id","padj")],
                     ensembl_gene_id)$padj,
  nolac.padj = arrange(repLRT.nolac$annotated_results[repLRT.nolac$annotated_results$ensembl_gene_id %in% sig.somewhere,c("ensembl_gene_id","padj")],
                       ensembl_gene_id)$padj
  ) %>%
  ggplot(aes(x = lac.padj, y = nolac.padj)) + geom_point() +
  ggtitle("Significant genes in LRT with and without lac samples")

ggsave(here("results", "diffex", "figs", "06_LRT", "06_siggenes_fdrvsexcludelac.pdf"), height=6, width=8)

tibble(
  lac.padj = repLRT$annotated_results %>% arrange(ensembl_gene_id) %>% pull(padj),
  nolac.padj = repLRT.nolac$annotated_results %>% arrange(ensembl_gene_id) %>% pull(padj)
  ) %>%
  ggplot(aes(x = lac.padj, y = nolac.padj)) + geom_point() +
  ggtitle("All genes in LRT with and without lac samples")

ggsave(here("results", "diffex", "figs", "06_LRT", "06_allgenes_fdrvsexcludelac.pdf"), height=6, width=8)
```


## Heatmap

```{r, fig.height=6, fig.width=10}
repLRT.nolac$heatmap
```


## Box plots

Plot the top 5 most significant.

```{r}
repLRT.nolac$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
 repLRT.nolac$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>% arrange(fdr)
```


```{r, fig.height=6, fig.width=8}
repLRT.nolac$pathway_plots
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r}
repLRT.nolac$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 10:

```{r}
repLRT.nolac$outlier_bees
```

# Batch

```{r}
ddsLRT = readRDS(here("data", "Rds", "06_ddsLRT_batch.Rds"))
resBatch = results(ddsLRT)
```


## Summary

```{r}
repBatch = deseq_report(resBatch, dds=ddsLRT, absl2fc = 1, title="LRT Batch")
```

## Top genes

```{r}
repBatch$significant_genes %>% head(30)
```

## Heatmap

```{r, fig.height=6, fig.width=10}
repBatch$heatmap
```


## Box plots

Plot the top 5 most significant.

```{r}
repBatch$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
repBatch$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>% arrange(fdr)
```


```{r, fig.height=6, fig.width=8}
repBatch$pathway_plots
```

# Batch - no lac

```{r}
ddsLRT.nolac = readRDS(here("data", "Rds", "06_ddsLRT_batch_nolac.Rds"))
resBatch.nolac = results(ddsLRT.nolac)
```

## Summary

```{r}
repBatch.nolac = deseq_report(resBatch.nolac, dds=ddsLRT.nolac,
                              variance_stabilized_dds = vsd.nolac, absl2fc = 1, title="LRT Batch - no lac")
```

## Top genes

```{r}
repBatch.nolac$significant_genes %>% head(30)
```

## Heatmap

```{r, fig.height=6, fig.width=10}
repBatch.nolac$heatmap
```


## Box plots

Plot the top 5 most significant.

```{r}
repBatch.nolac$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
repBatch.nolac$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>% arrange(fdr)
```


```{r, fig.height=6, fig.width=8}
repBatch.nolac$pathway_plots
```

# PAM50

```{r}
ddsLRT = readRDS(here("data", "Rds", "06_ddsLRT_pam.Rds"))
resPam = results(ddsLRT)
```


## Summary

```{r}
repPam = deseq_report(resPam, dds=ddsLRT, absl2fc = 1, title="LRT Pam50")
```

## Top genes

```{r}
repPam$significant_genes %>% head(30)
```

## Heatmap

```{r, fig.height=6, fig.width=10}
repPam$heatmap
```


## Box plots

Plot the top 5 most significant.

```{r}
repPam$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
repPam$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>% arrange(fdr)
```


```{r, fig.height=6, fig.width=8}
repPam$pathway_plots
```


# PAM50 - no lac

```{r}
ddsLRT.nolac = readRDS(here("data", "Rds", "06_ddsLRT_pam_nolac.Rds"))
resPam.nolac = results(ddsLRT.nolac)
```


## Summary

```{r}
repPam.nolac = deseq_report(resPam.nolac, dds=ddsLRT.nolac,
                            variance_stabilized_dds = vsd.nolac, absl2fc = 1, title="LRT Pam50 - no lac")
```

## Top genes

```{r}
repPam.nolac$significant_genes %>% head(30)
```

## Heatmap

```{r, fig.height=6, fig.width=10}
repPam.nolac$heatmap
```


## Box plots

Plot the top 5 most significant.

```{r}
repPam.nolac$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
repPam.nolac$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>% arrange(fdr) %>%
  head(30)
```


```{r, fig.height=6, fig.width=8}
repPam.nolac$pathway_plots
```

# Overlap LRTs

Generally a sign of colinearity if many genes are significant for multiple terms.
```{r}
hsb2 <- read.csv("https://stats.idre.ucla.edu/wp-content/uploads/2016/02/hsb2-3.csv") 
attach(hsb2)
hw <- (write >= 60)
hm <- (math >= 60)
hr <- (read >= 60)
c3 <- cbind(hw, hm, hr)

c3 %>% head()
```

## Venn genes

```{r}
stopifnot(all(sort(repLRT$annotated_results$ensembl_gene_id) ==
                sort(repPam$annotated_results$ensembl_gene_id)))
stopifnot(all(sort(repLRT$annotated_results$ensembl_gene_id) ==
                sort(repBatch$annotated_results$ensembl_gene_id)))

#Binary df for all genes to indicate comparison significance
background = sort(repLRT$annotated_results$ensembl_gene_id)
sig_genes = data.frame(
  PPBC = background %in% repLRT$significant_genes$ensembl_gene_id,
  PAM50 = background %in% repPam$significant_genes$ensembl_gene_id,
  Batch = background %in% repBatch$significant_genes$ensembl_gene_id)
rownames(sig_genes) = background

#Remove genes that weren't significant anwyhere
#sig_genes = sig_genes[rowSums(sig_genes) > 0,] #Not actually necessary
#dim(sig_genes)

vc = limma::vennCounts(sig_genes)

limma::vennDiagram(vc,
  names = colnames(sig_genes), 
  cex = 1,
  main = "Overlap significant genes from likelihood ratio tests")
```

```{r, include=F}
pdf(here("results", "diffex", "figs", "06_LRT", "06_venn_lrts.pdf"), height = 6, width=9)
limma::vennDiagram(vc,
  names = colnames(sig_genes), 
  cex = 1,
  main = "Overlap significant genes from likelihood ratio tests")
dev.off()
```

Which genes are unique to study group?

```{r}
unique_ppbc_genes = repLRT$significant_genes$ensembl_gene_id[
  !repLRT$significant_genes$ensembl_gene_id %in% repPam$significant_genes$ensembl_gene_id &
  !repLRT$significant_genes$ensembl_gene_id %in% repBatch$significant_genes$ensembl_gene_id]

repLRT$significant_genes %>% filter(ensembl_gene_id %in% unique_ppbc_genes)
```

Showing just the gene names from the unique group:

```{r}
repLRT$significant_genes %>% filter(ensembl_gene_id %in% unique_ppbc_genes) %>% pull(gene_name)
```

### Heatmap unique genes

```{r, fig.height=6, fig.width=9}
mat_unique_ppbc <- assay(vsd[rownames(vsd) %in% unique_ppbc_genes,])
gx_annot$gene_name[match(rownames(mat_unique_ppbc), gx_annot$ensembl_gene_id)]

stopifnot(all(repLRT$significant_genes %>% filter(ensembl_gene_id %in% unique_ppbc_genes) %>% pull(gene_name) %in%
                gx_annot$gene_name[match(rownames(mat_unique_ppbc), gx_annot$ensembl_gene_id)]))

rownames(mat_unique_ppbc) = gx_annot$gene_name[match(rownames(mat_unique_ppbc), gx_annot$ensembl_gene_id)]

quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(xs, probs = seq(0, 1, length.out = n))
  breaks[!duplicated(breaks)]
}

mat_breaks <- quantile_breaks(scrime::rowScales(mat_unique_ppbc), n = 10)

pheatmap::pheatmap(mat_unique_ppbc,
                   show_colnames = F,
                   scale="row",
                   annotation_col = as.data.frame(colData(vsd))[,c("PAM50", "study_group")],
                   annotation_colors = list(PAM50 = pam_colors, study_group = study_colors, Type = gene_colors),
                   annotation_row = column_to_rownames(select(filter(repLRT$significant_genes, ensembl_gene_id %in% unique_ppbc_genes),
                                           gene_name, Type),"gene_name"),
                   color = viridis::viridis(length(mat_breaks) - 1), #God it's hideous
                   breaks = mat_breaks,
                   main="Genes uniquely significant in LRT study group (vs batch and PAM50)")

```

```{r, include=F}
pheatmap::pheatmap(mat_unique_ppbc,
                   show_colnames = F,
                   scale="row",
                   annotation_col = as.data.frame(colData(vsd))[,c("PAM50", "study_group")],
                   annotation_colors = list(PAM50 = pam_colors, study_group = study_colors, Type = gene_colors),
                   annotation_row = column_to_rownames(select(filter(repLRT$significant_genes, ensembl_gene_id %in% unique_ppbc_genes),
                                           gene_name, Type),"gene_name"),
                   color = viridis::viridis(length(mat_breaks) - 1), #God it's hideous
                   breaks = mat_breaks,
                   main="Genes uniquely significant in LRT study group (vs batch and PAM50)",
                   filename = here("results", "diffex", "figs", "06_LRT", "heatmaps", "06_hm_unique_ppbc_lrt_genes.pdf"),
                   width= 9, height = 7)

```


## Venn pathways

We'll see a similar picture for pathways.

```{r}
#Binary df for all genes to indicate comparison significance

background = c(names(gene_sets$go_bp), names(gene_sets$hallmark),
               names(gene_sets$c2_canon), names(gene_sets$c2_cgp))

sig_paths = data.frame(
  PPBC = background %in% filter(bind_rows(repLRT$pathways), fdr < 0.1)$pathway,
  PAM50 = background %in% filter(bind_rows(repPam$pathways), fdr < 0.1)$pathway,
  Batch = background %in% filter(bind_rows(repBatch$pathways), fdr < 0.1)$pathway)
rownames(sig_paths) = background

#Remove genes that weren't significant anwyhere
#sig_genes = sig_genes[rowSums(sig_genes) > 0,] #Not actually necessary
#dim(sig_genes)

vp = limma::vennCounts(sig_paths)

limma::vennDiagram(vp,
  names = colnames(sig_paths), 
  cex = 1,
  main = "Overlap significant pathways from likelihood ratio tests")
```

Which two pathways from study group are not colinear?

```{r}
unique_ppbc_paths = filter(bind_rows(repLRT$pathways), fdr < 0.1)$pathway[
  !filter(bind_rows(repLRT$pathways), fdr < 0.1)$pathway %in% filter(bind_rows(repPam$pathways), fdr < 0.1)$pathway &
  !filter(bind_rows(repLRT$pathways), fdr < 0.1)$pathway %in% filter(bind_rows(repBatch$pathways), fdr < 0.1)$pathway]

bind_rows(repLRT$pathways) %>% filter(pathway %in% unique_ppbc_paths)
```


# Save data

Create a master results list.

```{r}
#Get all report objects from environment
reslist = mget(ls(pattern="rep"))
reslist = reslist[-1:-3] #Slightly too greedy
names(reslist)
```


## Significance of genes

Save the individual results lists, without the figures.

```{r}
#Save the results for all genes in an Excel file with multiple tabs
#The tab names will be the list names
resdf = lapply(reslist, function(x) x$annotated_results)
names(resdf) = paste(names(resdf),"all", sep="_")
openxlsx::write.xlsx(resdf, file = here("results", "diffex", "06_LRT_allgenes.xlsx"))

#Save the significant genes in a multi-tab excel file
resdf = lapply(reslist, function(x) x$significant_genes)
names(resdf) = paste("sig",names(resdf), sep="_")
#Also keep the thresholds used for significance
resdf[[length(resdf)+1]] = 
enframe(unlist(lapply(reslist, function(x) x$sig_threshold)), "name", "value") %>%
  mutate(groups = if_else(str_detect(name, "nolac"), "no lactation", "allgroups")) %>%
  mutate(name = str_remove(name, ".nolac")) %>%
  separate(name, into = c("comparison", "threshold"), sep="\\.")
names(resdf)[length(resdf)] = "thresholds"
openxlsx::write.xlsx(resdf, file = here("results", "diffex", "06_LRT_sig_genes.xlsx"))

rm(resdf)
```

## Pathway results

Fisher's exact pathway analyses

```{r}
resdf = lapply(reslist, function(x) x$pathways)

#Returns a list of aggregated path results per comparison
pathlist = list()
pathlist[[names(resdf)[1]]] = resdf %>% map(1, .depth=1) %>% bind_rows( .id = "comp")
for (i in 2:length(resdf[[1]])){
  path = resdf %>% map(i, .depth=1) %>% bind_rows( .id = "comp")
  pathlist[[names(resdf)[i]]] = path
}

#Fix names
names(pathlist) = lapply(pathlist, function(x) unique(x$collection)) %>% unlist(use.names = F)
pathlist = lapply(pathlist, function(x) x[order(x$fdr),])

openxlsx::write.xlsx(pathlist, file = here("results", "diffex", "06_LRT_pathways.xlsx"))

rm(resdf)
```

## Heatmaps

```{r}
dir.create(here("results","diffex", "figs", "06_LRT"), showWarnings = F)
dir.create(here("results","diffex", "figs", "06_LRT" ,"heatmaps"), showWarnings = F)

hmlist = lapply(reslist, function(x) x$heatmap)
names(hmlist) = names(hmlist) %>% str_replace_all("rep", "hm")

for (i in 1:length(hmlist)){
  pdf(here("results", "diffex", "figs", "06_LRT", "heatmaps", paste0(names(hmlist)[i],".pdf")), width = 10, height = 8)
  print(hmlist[[i]])
  dev.off()
}

rm(hmlist)

```

## Volcano plot sig genes

Use caution when interpreting volcano plots for more than 2 groups.

```{r}
volclist = lapply(reslist, function(x) x$volcano_plot)
names(volclist) = names(volclist) %>% str_replace("rep", "volc")

dir.create(here("results","diffex", "figs", "06_LRT" ,"volcano_plots"), showWarnings = F)
lapply(names(volclist), 
       function(x) ggsave(filename=here("results","diffex", "figs", "06_LRT","volcano_plots",
                                        paste(x,".jpeg",sep="")), plot=volclist[[x]],
                          height = 6, width = 8))
rm(volclist)
```



## Volcano plot outliers

Volcano plots of significant genes are hard to interpret in an LRT with multiple levels. Use outliers only for diagnostics.

```{r}
outvplist = lapply(reslist, function(x) x$volc_outliers)
names(outvplist) = names(outvplist) %>% str_replace("rep", "volc")

dir.create(here("results","diffex", "figs", "06_LRT" ,"volcano_plots"), showWarnings = F)
lapply(names(outvplist), 
       function(x) ggsave(filename=here("results","diffex", "figs", "06_LRT","volcano_plots",
                                        paste(x,".outliers.jpeg",sep="")), plot=outvplist[[x]],
                          height = 6, width = 8))
rm(outvplist)
```

## Pathway dotplots 

```{r}
dir.create(here("results","diffex", "figs", "06_LRT" ,"fisher_pathway"), showWarnings = F)
pathwaylist = lapply(reslist, function(x) x$pathway_plots)
for(i in 1:length(pathwaylist)){
  thiscomp = pathwaylist[[i]]
  compname = str_replace(str_remove(names(pathwaylist[i]), "rep_"),"_","_vs_")
  for (j in 1:length(thiscomp)){
    if (length(thiscomp) < 1){next}
    thissig = thiscomp[[j]]
    file=paste0(paste(compname, names(thiscomp[j]), sep="-"),".jpeg")
    #print(file)
    ggsave(filename=here("results","diffex", "figs",
                         "06_LRT","fisher_pathway",file),
           plot=thissig, height = 6, width = 8)
  }
}


rm(pathwaylist)
```


## Notebook data

Remove a few large objects first

```{r}
rm(reslist) #Because it's huge
rm(gene_sets)
```


```{r}
save.image(here("reports/06_diffex_lrt_standard.Rdata"))
```

