---
title: "Pairwise Wald tests, standard threshold"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
---

```{r, include=F}
#library(devtools)
#install_github("jokergoo/ComplexHeatmap")
#install_github("mikelove/DESeq2")
#install_github("azhu513/apeglm")

library(DESeq2)
library(apeglm)
library(here)
library(ggrepel)
library(ComplexHeatmap)
library(openxlsx)
library(scrime)
library(RColorBrewer)
library(circlize)
library(tidyverse)
```

# Load and pre-process data

```{r}
dds <- readRDS(file = here("data/Rds/05_dds_PAM50_tumorpurity_batch.Rds"))

gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
gx_annot = gx_annot %>% select(ensembl_gene_id = gene_id, gene_name, gene_type, description = gene_description) %>% distinct()
```


## Filtering


Minimum threshold is a nonzero count in at least a 3rd of all samples

```{r}

keep <- rowSums(counts(dds)!=0) >= ceiling(ncol(dds)/3)
table(keep)

#Hang on to what we discard for later analysis
dds.discarded <-dds[!keep,]

#Keep the rest
dds <- dds[keep,]

```


## Retrieve immune genes

For downstream analyses, we'd like to know which of the differentially expressed genes are immunologically relevant.

An immune gene is defined as follows:

Either immune/immuno/interleukin is part of the gene name and description OR the gene is part of the the ImmPort database. See https://www.innatedb.com/redirect.do?go=resourcesGeneLists

The Immunology Database and Analysis Portal (ImmPort) system was developed under the Bioinformatics Integration Support Contract (BISC) Phase II by the Northrop Grumman Information Technology Health Solutions team for the NIH, NIAID, and DAIT. The principal investigator of the BISC project is Dr. Richard Scheuermann at University of Texas Southwestern Medical Center. The list of immunologically related genes in ImmPort is a collection of ~6,000 human genes, which was formed with the goal of retrieving all genes that have immune system-related functions. This list was generated using automatic searches of EntrezGene and Gene Ontology records using immunology-related keywords. The list was then manually curated by immunology experts examining various literature sources. 

```{r}
immune_gene_list <- read_csv(here("data", "external","gene-sets","InnateDB_genes.csv"))
head(immune_gene_list)
```

## Helper functions

```{r}
source(here("src", "deseq_report_functions.R"))
```




## Variance stabilizing transformation

We want a transformation that is fully blind to the experimental design, but still uses the faster sampling method in the vst wrapper for varianceStabilizingTransformation.

This homoskedastic dataset will be used later on for heatmap visualization.
Previously performed in notebook 6.


```{r}
vsd = readRDS(here("data","Rds","06_vsd_standardfilter.Rds"))
```


# Diffex: non-prbc as reference group

In this call to DESeq, we're focusing in group-by-group contrasts and using a Wald test instead of a likelihood ratio test.

Ensure that non-prbc is the reference level

```{r}
dds$study_group = relevel(dds$study_group, ref="non_prbc")

levels(dds$study_group)
```

Perform differential expression analysis.

```{r, eval=F}
start = Sys.time()
dds <- DESeq(dds)
end = Sys.time()
end-start

saveRDS(dds,here("data/Rds/07_dds_apeglm_nonprbcref_standard.Rds"))
```

>estimating size factors
using 'avgTxLength' from assays(dds), correcting for library size
estimating dispersions
gene-wise dispersion estimates
Warning in doTryCatch(return(expr), name, parentenv, handler) :
  restarting interrupted promise evaluation
mean-dispersion relationship
final dispersion estimates
fitting model and testing
Time difference of 1.71209 hours

```{r}
dds = readRDS(here("data/Rds/07_dds_apeglm_nonprbcref_standard.Rds"))
```



## Coefficient overview

Make it easier to find the coefficient we need.

```{r}
cdf = enframe(colnames(coef(dds)), "coef", "name")

cdf
```


# Involution vs non-prbc

## Extract results


```{r}
ape_inv_nonprbc = shrinkRes(dds,  contrast=c("study_group", "ppbc_inv", "non_prbc"))
```

## Summary report

```{r}
rep_inv_nonprbc = deseq_report(ape_inv_nonprbc, dds=dds, absl2fc = 0.5, groups = c("ppbc_inv", "non_prbc"))
```

## Top genes

```{r}
rep_inv_nonprbc$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_inv_nonprbc$volcano_plot
```

## Heatmap


```{r}
draw(rep_inv_nonprbc$heatmap, column_title = "Involution vs non-prbc")
```


## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_inv_nonprbc$beehive_plots
```




# Pregnancy vs non-pregnancy

## Extract results

```{r}
ape_prbc_nonprbc = shrinkRes(dds,  contrast=c("study_group", "prbc", "non_prbc"))
```

## Summary report

```{r}
rep_prbc_nonprbc = deseq_report(ape_prbc_nonprbc, dds=dds, absl2fc = 0.5, groups = c("prbc", "non_prbc"))
```

## Top genes

```{r}
rep_prbc_nonprbc$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_prbc_nonprbc$volcano_plot
```

## Heatmap

```{r}
draw(rep_prbc_nonprbc$heatmap, column_title = "Prbc vs non-prbc")
```


## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_prbc_nonprbc$beehive_plots
```


# Lac vs non-prbc

## Extract results

```{r}
ape_lac_nonprbc = shrinkRes(dds,  contrast=c("study_group", "ppbc_lac", "non_prbc"))
```

## Summary report

```{r}
rep_lac_nonprbc = deseq_report(ape_lac_nonprbc, dds=dds, absl2fc = 0.5, groups = c("ppbc_lac", "non_prbc"))
```



## Top genes

```{r}
rep_lac_nonprbc$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_lac_nonprbc$volcano_plot
```

## Heatmap

```{r}
draw(rep_lac_nonprbc$heatmap, column_title = "Lac vs non-prbc")
```


## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_lac_nonprbc$beehive_plots
```

# Diffex: prbc as reference level

From the manual: 

Although apeglm cannot be used with contrast, we note that many designs can be easily rearranged such that what was a contrast becomes its own coefficient. In this case, the dispersion does not have to be estimated again, as the designs are equivalent, up to the meaning of the coefficients. Instead, one need only run nbinomWaldTest to re-estimate MLE coefficients – these are necessary for apeglm – and then run lfcShrink specifying the coefficient of interest in resultsNames(dds). We give some examples below of producing equivalent designs for use with coef. We show how the coefficients change with model.matrix, but the user would, for example, either change the levels of dds$condition or replace the design using design(dds)<-, then run nbinomWaldTest followed by lfcShrink.

```{r}
dds$study_group = relevel(dds$study_group, ref="prbc")
levels(dds$study_group)
```

```{r, eval=F}
start = Sys.time()
dds = nbinomWaldTest(dds)
end = Sys.time()
end - start

saveRDS(dds,here("data/Rds/07_dds_apeglm_prbcref_standard.Rds"))
```

>found results columns, replacing these
Warning in eval(formal.args[[as.character(substitute(arg))]]) :
  restarting interrupted promise evaluation
Warning in doTryCatch(return(expr), name, parentenv, handler) :
  restarting interrupted promise evaluation
2 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest
Time difference of 35.17508 mins

```{r}
dds = readRDS(here("data/Rds/07_dds_apeglm_prbcref_standard.Rds"))
```


## Coefficient summary

```{r}
resultsNames(dds)
```

# Involution vs prbc

## Extract results

```{r}
ape_inv_prbc = shrinkRes(dds,  contrast=c("study_group", "ppbc_inv", "prbc"))
```

## Summary report

```{r}
rep_inv_prbc = deseq_report(ape_inv_prbc, dds=dds, absl2fc = 0.5, groups = c("ppbc_inv", "prbc"))
```

## Top genes

```{r}
rep_inv_prbc$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_inv_prbc$volcano_plot
```

## Heatmap

```{r}
draw(rep_inv_prbc$heatmap, column_title = "Involution vs prbc")
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_inv_prbc$beehive_plots
```


# Lac vs prbc

## Extract results

```{r}
ape_lac_prbc = shrinkRes(dds,  contrast=c("study_group", "ppbc_lac", "prbc"))
```

## Summary report

```{r}
rep_lac_prbc = deseq_report(ape_lac_prbc, dds=dds, absl2fc = 0.5, groups = c("ppbc_lac", "prbc"))
```

## Top genes

```{r}
rep_lac_prbc$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_lac_prbc$volcano_plot
```

## Heatmap

```{r}
draw(rep_lac_prbc$heatmap, column_title = "Lactation vs prbc")
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_lac_prbc$beehive_plots
```

# Diffex: Lac as reference level

We have only one comparison remaining: involution vs lac. For this we need to set lactation as the reference group.

```{r}
dds$study_group = relevel(dds$study_group, ref="ppbc_lac")
levels(dds$study_group)
```

```{r, eval=F}
start = Sys.time()
dds = nbinomWaldTest(dds)
end = Sys.time()
end - start

saveRDS(dds,here("data/Rds/07_dds_apeglm_lacref_standard.Rds"))
```

>found results columns, replacing these
3 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest
Time difference of 1.431062 hours

```{r}
dds = readRDS(here("data/Rds/07_dds_apeglm_lacref_standard.Rds"))
```

## Coefficient summary

```{r}
resultsNames(dds)
```

# Involution vs lac

## Extract results

```{r}
ape_inv_lac = shrinkRes(dds, contrast=c("study_group", "ppbc_inv", "ppbc_lac"))
```

## Summary report

```{r}
rep_inv_lac = deseq_report(ape_inv_lac, dds=dds, absl2fc = 0.5, groups = c("ppbc_inv", "ppbc_lac"))
```

## Top genes

```{r}
rep_inv_lac$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_inv_lac$volcano_plot
```

## Heatmap

```{r}
draw(rep_inv_lac$heatmap, column_title = "Involution vs lactation")
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.
GPM6A padj 3.51 e-07, log2fc -3.4
```{r}
rep_inv_lac$beehive_plots
```

# Distance plots

Attempt to visualize the number of significant genes in each comparison.

```{r}
#Get all report objects from environment
reslist = mget(ls(pattern="rep_"))
```


```{r}
sign = lapply(lapply(reslist, function(x) x$significant_genes), function(x) length(unique(x$gene_name)))

#Create comparison matrix for number of significant genes
sigmat = sign %>% unlist %>% enframe("name", "sig") %>%
  separate(name, into=c(NA, "num", "denom")) %>%
  spread(key="denom", value="sig") %>%
  bind_rows(., tibble(num="nonprbc", lac=NA, nonprbc=NA, prbc=NA)) %>%
  mutate(inv = NA) %>% column_to_rownames("num")

#Standardize column orders
sigmat = sigmat[c("nonprbc", "prbc", "lac", "inv"), c("nonprbc", "prbc", "lac", "inv")]
sigmat = as.matrix(sigmat)
```

Replace the diagonal with 0.

```{r}
for (x in 1:length(rownames(sigmat))){
  for (y in 1:length(colnames(sigmat))){
    if (rownames(sigmat)[x] == rownames(sigmat)[y]){
      sigmat[x,y] = 0
    }
  }
}

sigmat
```

Fill in the NAs from redundant comparisons.

```{r}
for (x in 1:length(rownames(sigmat))){
  for (y in 1:length(colnames(sigmat))){
    if (is.na(sigmat[x,y])){
      sigmat[x,y] = sigmat[y,x]
    }
  }
}

sigmat
```

Easiest way to color the diagonal is just to turn it back to NA.

```{r}
for (x in 1:length(rownames(sigmat))){
  for (y in 1:length(colnames(sigmat))){
    if (sigmat[x,y]==0){
      sigmat[x,y] = NA
    }
  }
}

sigmat
```


```{r}

dir.create(here("results","diffex", "figs", "07_pairwise" ,"heatmaps"), showWarnings = F, recursive = T)

nsig_hm = Heatmap(sigmat, name = "significant genes", cell_fun = function(j, i, x, y, width, height, fill) 
        {
          grid.text(sprintf("%.0f", sigmat[i, j]), x, y, gp = gpar(fontsize = 12))
        },
        cluster_rows = F, cluster_columns = F, column_names_side = "bottom", column_names_rot = 0,
        row_names_side = "left", col = colorRamp2(c(0, 200, 400), c("blue", "purple", "red")),
        column_title = "Number of significant genes in pairwise comparisons")

pdf(here("results", "diffex", "figs", "07_pairwise", "heatmaps", "n_siggenes_pairwise_hm.pdf"), width = 10, height = 8)
nsig_hm
dev.off()

nsig_hm
```


# Save data

Save the primary comparison separately.

```{r}
saveRDS(rep_inv_nonprbc, here("data", "Rds", "07_report_inv_nonprbc.Rds"))
```

Save the individual results lists, without the figures.

```{r}

#Save the results for all genes in an Excel file with multiple tabs
#The tab names will be the list names
resdf = lapply(reslist, function(x) x$annotated_results)
names(resdf) = paste(names(resdf),"all", sep="_")
openxlsx::write.xlsx(resdf, file = here("results", "diffex", "07_pairwise_comparisons_allgenes.xlsx"))

#Save the significant genes in a multi-tab excel file
resdf = lapply(reslist, function(x) x$significant_genes)
names(resdf) = paste("sig",names(resdf), sep="_")
#Also keep the thresholds used for significance
resdf[[length(resdf)+1]] = enframe(unlist(lapply(reslist, function(x) x$sig_threshold)), "name", "value") %>%
                 separate(name, into = c("comparison", "threshold"), sep="\\.")
names(resdf)[length(resdf)] = "thresholds"
library(openxlsx)
openxlsx::write.xlsx(resdf, file = here("results", "diffex", "07_pairwise_comparisons_sig_genes.xlsx"))

rm(resdf)
```

Save the figures

```{r}
hmlist = lapply(reslist, function(x) x$heatmap)
names(hmlist) = names(hmlist) %>% str_replace_all("rep", "hm")

for (i in 1:length(hmlist)){
  pdf(here("results", "diffex", "figs", "07_pairwise", "heatmaps", paste0(names(hmlist)[i],".pdf")), width = 10, height = 8)
  draw(hmlist[[i]], column_title = str_replace(str_remove(names(hmlist)[i], "hm_"), "_", " vs "))
  dev.off()
}

rm(hm_list)
```

```{r}
volclist = lapply(reslist, function(x) x$volcano_plot)
names(volclist) = names(volclist) %>% str_replace("rep", "volc")

dir.create(here("results","diffex", "figs", "07_pairwise" ,"volcano_plots"), showWarnings = F)
lapply(names(volclist), 
       function(x) ggsave(filename=here("results","diffex", "figs", "07_pairwise","volcano_plots",
                                        paste(x,".jpeg",sep="")), plot=volclist[[x]],
                          height = 6, width = 8))
rm(volclist)
```
```{r}
rm(reslist)
```


Save the notebook data.

```{r}
save.image(here("reports/07_diffex_pairwise_standard.RData"))
```

