---
title: "Pairwise Wald tests, standard threshold"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
---

```{r, include=F}
rm(list = ls())
```


```{r, include=F}


#library(devtools)
#install_github("jokergoo/ComplexHeatmap")
#install_github("mikelove/DESeq2")
#install_github("azhu513/apeglm")
library(org.Hs.eg.db)
library(DESeq2)
library(apeglm)
library(here)
library(ggrepel)
library(openxlsx)
library(scrime)
library(msigdbr)
library(RColorBrewer)
library(clusterProfiler)
library(circlize)
library(ComplexHeatmap)
library(tidyverse)
```

# Load and pre-process data

```{r}
dds <- readRDS(file = here("data/Rds/05_dds_PAM50_batch.Rds"))

gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
gx_annot = gx_annot %>% select(ensembl_gene_id = gene_id, gene_name, gene_type, description = gene_description) %>% distinct()
```


## Filtering


Minimum threshold is a nonzero count in at least a 3rd of all samples

```{r}

keep <- rowSums(counts(dds)!=0) >= ceiling(ncol(dds)/3)
table(keep)

#Keep the rest
dds <- dds[keep,]

```


## Retrieve immune genes

For downstream analyses, we'd like to know which of the differentially expressed genes are immunologically relevant.

An immune gene is defined as follows:

Either immune/immuno/interleukin is part of the gene name and description OR the gene is part of the the [ImmPort database](https://www.innatedb.com/redirect.do?go=resourcesGeneLists)

The Immunology Database and Analysis Portal (ImmPort) system was developed under the Bioinformatics Integration Support Contract (BISC) Phase II by the Northrop Grumman Information Technology Health Solutions team for the NIH, NIAID, and DAIT. The list of immunologically related genes in ImmPort is a collection of ~6,000 human genes, which was formed with the goal of retrieving all genes that have immune system-related functions. This list was generated using automatic searches of EntrezGene and Gene Ontology records using immunology-related keywords. The list was then manually curated by immunology experts examining various literature sources. 

```{r}
immune_gene_list <- read_csv(here("data", "external","gene-sets","InnateDB_genes.csv"))
head(immune_gene_list)
```

## Helper functions

```{r}
source(here("src", "deseq_report_functions.R"))
```

## Heatmap colors

Created in notebook 6

```{r}
all_colors = readRDS(here("data","Rds", "06_heatmap_colors.Rds"))
study_colors = all_colors$study_colors
pam_colors = all_colors$pam_colors
gene_colors = all_colors$gene_colors
```



## Variance stabilizing transformation

We want a transformation that is fully blind to the experimental design, but still uses the faster sampling method in the vst wrapper for varianceStabilizingTransformation.

This homoskedastic dataset will be used later on for heatmap visualization.
Previously performed in notebook 6.


```{r}
vsd = readRDS(here("data","Rds","06_vsd_standardfilter.Rds"))
```


# Diffex: non-prbc as reference group

In this call to DESeq, we're focusing in group-by-group contrasts and using a Wald test instead of a likelihood ratio test.

Ensure that non-prbc is the reference level

```{r}
dds$study_group = relevel(dds$study_group, ref="non_prbc")

levels(dds$study_group)
```

Perform differential expression analysis.

```{r, eval=F}
start = Sys.time()
dds <- DESeq(dds)
end = Sys.time()
end-start

saveRDS(dds,here("data/Rds/07_dds_apeglm_nonprbcref_standard.Rds"))
```



```{r}
dds = readRDS(here("data/Rds/07_dds_apeglm_nonprbcref_standard.Rds"))
```



## Coefficient overview

Make it easier to find the coefficient we need.

```{r}
cdf = enframe(colnames(coef(dds)), "coef", "name")

cdf
```


# Involution vs non-prbc

## Extract results


```{r, eval=T}
ape_inv_nonprbc = shrinkRes(dds,  contrast=c("study_group", "ppbc_inv", "non_prbc"))
saveRDS(ape_inv_nonprbc, here("data","Rds", "07_ape_inv_nonprbc.Rds"))
```

```{r}
ape_inv_nonprbc = readRDS(here("data","Rds", "07_ape_inv_nonprbc.Rds"))
```


```{r}
design(dds)
levels(dds$study_group)
```


## Summary report

```{r}
rep_inv_nonprbc = deseq_report(ape_inv_nonprbc, dds=dds, absl2fc = 0.5, groups = c("ppbc_inv", "non_prbc"),
                               title = "Involution vs nulliparous")
```

## Top genes

```{r}
rep_inv_nonprbc$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_inv_nonprbc$volcano_plot
```

## Heatmap


```{r, fig.height=6, fig.width=10}
rep_inv_nonprbc$heatmap
```


## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_inv_nonprbc$beehive_plots
```




# Pregnancy vs non-pregnancy

## Extract results

```{r, eval=T}
ape_prbc_nonprbc = shrinkRes(dds,  contrast=c("study_group", "prbc", "non_prbc"))
saveRDS(ape_prbc_nonprbc, here("data", "Rds", "07_ape_prbc_nonprbc.Rds"))
```

```{r}
ape_prbc_nonprbc = readRDS(here("data", "Rds", "07_ape_prbc_nonprbc.Rds"))
```


## Summary report

```{r}
rep_prbc_nonprbc = deseq_report(ape_prbc_nonprbc, dds=dds, absl2fc = 0.5, groups = c("prbc", "non_prbc"),
                                title = "Pregnant vs nulliparous")
```

## Top genes

```{r}
rep_prbc_nonprbc$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_prbc_nonprbc$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_prbc_nonprbc$heatmap
```


## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_prbc_nonprbc$beehive_plots
```


# Lac vs non-prbc

## Extract results

```{r, eval=T}
ape_lac_nonprbc = shrinkRes(dds,  contrast=c("study_group", "ppbc_lac", "non_prbc"))
saveRDS(ape_lac_nonprbc, here("data", "Rds", "07_ape_lac_nonprbc.Rds"))
```

```{r}
ape_lac_nonprbc = readRDS(here("data", "Rds", "07_ape_lac_nonprbc.Rds"))
```


## Summary report

```{r}
rep_lac_nonprbc = deseq_report(ape_lac_nonprbc, dds=dds, absl2fc = 0.5, groups = c("ppbc_lac", "non_prbc"),
                               title = "lac vs nonprbc")
```



## Top genes

```{r}
rep_lac_nonprbc$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_lac_nonprbc$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_lac_nonprbc$heatmap
```


## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_lac_nonprbc$beehive_plots
```

# Diffex: prbc as reference level

From the manual: 

Although apeglm cannot be used with contrast, we note that many designs can be easily rearranged such that what was a contrast becomes its own coefficient. In this case, the dispersion does not have to be estimated again, as the designs are equivalent, up to the meaning of the coefficients. Instead, one need only run nbinomWaldTest to re-estimate MLE coefficients – these are necessary for apeglm – and then run lfcShrink specifying the coefficient of interest in resultsNames(dds). We give some examples below of producing equivalent designs for use with coef. We show how the coefficients change with model.matrix, but the user would, for example, either change the levels of dds$condition or replace the design using design(dds)<-, then run nbinomWaldTest followed by lfcShrink.

```{r}
dds$study_group = relevel(dds$study_group, ref="prbc")
levels(dds$study_group)
```

```{r, eval=F}
start = Sys.time()
dds = nbinomWaldTest(dds)
end = Sys.time()
end - start

saveRDS(dds,here("data/Rds/07_dds_apeglm_prbcref_standard.Rds"))
```



```{r}
dds = readRDS(here("data/Rds/07_dds_apeglm_prbcref_standard.Rds"))
```


## Coefficient summary

```{r}
resultsNames(dds)
```

# Involution vs prbc

## Extract results

```{r, eval=T}
ape_inv_prbc = shrinkRes(dds,  contrast=c("study_group", "ppbc_inv", "prbc"))

saveRDS(ape_inv_prbc, here("data", "Rds", "07_ape_inv_prbc.Rds"))
```

```{r}
ape_inv_prbc = readRDS(here("data", "Rds", "07_ape_inv_prbc.Rds"))
```


## Summary report

```{r}
rep_inv_prbc = deseq_report(ape_inv_prbc, dds=dds, absl2fc = 0.5, groups = c("ppbc_inv", "prbc"),
                             title = "Involution vs pregnant")
```

## Top genes

```{r}
rep_inv_prbc$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_inv_prbc$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_inv_prbc$heatmap
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_inv_prbc$beehive_plots
```


# Lac vs prbc

## Extract results

```{r, eval=T}
ape_lac_prbc = shrinkRes(dds,  contrast=c("study_group", "ppbc_lac", "prbc"))
saveRDS(ape_lac_prbc, here("data", "Rds", "07_ape_lac_prbc.Rds"))
```

```{r}
ape_lac_prbc = readRDS(here("data", "Rds", "07_ape_lac_prbc.Rds"))
```


## Summary report

```{r}
rep_lac_prbc = deseq_report(ape_lac_prbc, dds=dds, absl2fc = 0.5, groups = c("ppbc_lac", "prbc"),
                            title = "Lactation vs pregnant")
```

## Top genes

```{r}
rep_lac_prbc$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_lac_prbc$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_lac_prbc$heatmap
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_lac_prbc$beehive_plots
```

# Diffex: Lac as reference level

We have only one comparison remaining: involution vs lac. For this we need to set lactation as the reference group.

```{r}
dds$study_group = relevel(dds$study_group, ref="ppbc_lac")
levels(dds$study_group)
```

```{r, eval=F}
start = Sys.time()
dds = nbinomWaldTest(dds)
end = Sys.time()
end - start

saveRDS(dds,here("data/Rds/07_dds_apeglm_lacref_standard.Rds"))
```

>found results columns, replacing these
3 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest
Time difference of 1.431062 hours

```{r}
dds = readRDS(here("data/Rds/07_dds_apeglm_lacref_standard.Rds"))
```

## Coefficient summary

```{r}
resultsNames(dds)
```

# Involution vs lac

## Extract results

```{r, eval=T}
ape_inv_lac = shrinkRes(dds, contrast=c("study_group", "ppbc_inv", "ppbc_lac"))
saveRDS(ape_inv_lac, here("data", "Rds", "07_ape_inv_lac.Rds"))
```

```{r}
ape_inv_lac = readRDS(here("data", "Rds", "07_ape_inv_lac.Rds"))
```


## Summary report

```{r}
rep_inv_lac = deseq_report(ape_inv_lac, dds=dds, absl2fc = 0.5, groups = c("ppbc_inv", "ppbc_lac"),
                           title = "Involutoin vs lactation")
```

## Top genes

```{r}
rep_inv_lac$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_inv_lac$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_inv_lac$heatmap
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.
GPM6A padj 3.51 e-07, log2fc -3.4
```{r}
rep_inv_lac$beehive_plots
```

# Distance plots

Attempt to visualize the number of significant genes in each comparison.

```{r}
#Get all report objects from environment
reslist = mget(ls(pattern="rep_"))
```


```{r}
sign = lapply(lapply(reslist, function(x) x$significant_genes), function(x) length(unique(x$gene_name)))

#Create comparison matrix for number of significant genes
sigmat = sign %>% unlist %>% enframe("name", "sig") %>%
  separate(name, into=c(NA, "num", "denom")) %>%
  spread(key="denom", value="sig") %>%
  bind_rows(., tibble(num="nonprbc", lac=NA, nonprbc=NA, prbc=NA)) %>%
  mutate(inv = NA) %>% column_to_rownames("num")

#Standardize column orders
sigmat = sigmat[c("nonprbc", "prbc", "lac", "inv"), c("nonprbc", "prbc", "lac", "inv")]
sigmat = as.matrix(sigmat)
```

Replace the diagonal with 0.

```{r}
for (x in 1:length(rownames(sigmat))){
  for (y in 1:length(colnames(sigmat))){
    if (rownames(sigmat)[x] == rownames(sigmat)[y]){
      sigmat[x,y] = 0
    }
  }
}

sigmat
```

Fill in the NAs from redundant comparisons.

```{r}
for (x in 1:length(rownames(sigmat))){
  for (y in 1:length(colnames(sigmat))){
    if (is.na(sigmat[x,y])){
      sigmat[x,y] = sigmat[y,x]
    }
  }
}

sigmat
```

Easiest way to color the diagonal is just to turn it back to NA.

```{r}
for (x in 1:length(rownames(sigmat))){
  for (y in 1:length(colnames(sigmat))){
    if (sigmat[x,y]==0){
      sigmat[x,y] = NA
    }
  }
}

sigmat
```


```{r}

dir.create(here("results","diffex", "figs", "07_pairwise" ,"heatmaps"), showWarnings = F, recursive = T)

nsig_hm = Heatmap(sigmat, name = "significant genes", cell_fun = function(j, i, x, y, width, height, fill) 
        {
          grid.text(sprintf("%.0f", sigmat[i, j]), x, y, gp = gpar(fontsize = 12))
        },
        cluster_rows = F, cluster_columns = F, column_names_side = "bottom", column_names_rot = 0,
        row_names_side = "left", col = colorRamp2(c(0, 200, 400), c("blue", "purple", "red")),
        column_title = "Number of significant genes in pairwise comparisons")

pdf(here("results", "diffex", "figs", "07_pairwise", "heatmaps", "n_siggenes_pairwise_hm.pdf"), width = 10, height = 8)
nsig_hm
dev.off()

nsig_hm
```

# Diagnostics

Check on lactalbumin in pregnant vs nulliparous.

This gene was previously significant when using tumor_purity in the design formula and now is no longer significant. Why is that?

LABLA ENSG00000167531

```{r}
dds = readRDS(here("data/Rds/07_dds_apeglm_nonprbcref_standard.Rds"))
plot_gene_beehive(dds, result_df = rep_prbc_nonprbc$annotated_results[rep_prbc_nonprbc$annotated_results$ensembl_gene_id == "ENSG00000167531",],
                  groups_to_plot = c("prbc", "non_prbc"))
```

Although the log fold change is substantial, it would appear that padj has been set to NA. There are several reasons why this can happen, according to the [DESeq2 manual](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pvaluesNA):

>Note on p-values set to NA: some values in the results table can be set to NA for one of the following reasons:
* If within a row, all samples have zero counts, the baseMean column will be zero, and the log2 fold change estimates, p value and adjusted p value will all be set to NA.
* If a row contains a sample with an extreme count outlier then the p value and adjusted p value will be set to NA. These outlier counts are detected by Cook’s distance. Customization of this outlier filtering and description of functionality for replacement of outlier counts and refitting is described [below](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#outlier)
* If a row is filtered by automatic independent filtering, for having a low mean normalized count, then only the adjusted p value will be set to NA. Description and customization of independent filtering is described [below](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#indfilt)

It is clearly not the first reason. Let's look at the results for this gene in more detail.

```{r}
rep_prbc_nonprbc$annotated_results[rep_prbc_nonprbc$annotated_results$gene_name=="LALBA",]
```

Both p value and padj have been set to NA here, which means that Cook's distance is the reason this gene was filtered out.

# Experiments with filtering

Let's look at the overall distribution of Cook's distance in the prbc vs nonprbc dataset and compare where LALBA falls. 

```{r}
#General distribution of Cook's distances
mcols(dds)$maxCooks %>% enframe() %>% pull(value) %>% quantile(probs=c(0,0.25,0.5,0.75,0.9,0.95,0.99)) #Anything above 1 is very high

#Cutoff for this comparison
m <- nrow(attr(dds, "dispModelMatrix"))
p <- ncol(attr(dds, "dispModelMatrix"))
defaultCutoff <- qf(0.99, p, m - p)
print(paste("Cooks cutoff: ", defaultCutoff))

#Plot cooks for this comparison
mcols(dds)$maxCooks %>% enframe() %>%
  filter(name == "ENSG00000167531") %>% 
  pull(value)

rm(list=c("m", "p"))
```


```{r}

mcols(dds)$maxCooks %>% enframe() %>%
  mutate(color = if_else(name == "ENSG00000167531", "LALBA", "other")) %>%
  ggplot(aes(x = "", y = value)) + #geom_boxplot() + 
  geom_jitter(aes(color = color)) +
  ggtitle("Cooks distance pregnant vs nulliparous") +
  geom_hline(yintercept = defaultCutoff)
```

>When there are 7 or more replicates for a given sample, the DESeq function will automatically replace counts with large Cook’s distance with the trimmed mean over all samples, scaled up by the size factor or normalization factor for that sample.

Details of Cook's distance replacement are not documented in as much detail as desired, but [this thread](http://seqanswers.com/forums/showthread.php?t=45533) is helpful.

>If an outlier is replaced by DESeq(), there is a column: mcols(dds)$replace which notes these.

Note that these genes will NOT get a padj of NA because outlier values have already been replaced.

```{r}
outliers_replaced_prbc_nonprbc = mcols(dds)$replace %>% enframe("ensembl_gene_id", "outlier_replaced") %>% filter(outlier_replaced==T) %>%
  left_join(.,select(gx_annot, "ensembl_gene_id", "gene_name", "gene_type"), by = "ensembl_gene_id")

print(paste("Genes with outliers replaced by the trimmed mean: ", nrow(outliers_replaced_prbc_nonprbc)))
```

Showing which samples were ignored because they exceeded the Cook's distance even after outliers were replaced requires manually adding a few extra columns. We can do the same for independent filtering below a minimum threshold.

```{r}
#Show the independent filtering for this comparison
print("Min count threshold for independent filtering:")
metadata(ape_prbc_nonprbc)$filterThreshold

#Add columns to show genes that fall below either Cook's or independent filtering thresholds
exres_prbc_nonprbc = rep_prbc_nonprbc$annotated_results %>%
  mutate(cooks_outlier = if_else(baseMean > 0 & is.na(pvalue), T, F)) %>%
  mutate(indepfilt = if_else(!is.na(pvalue) & is.na(padj), T, F))

print(paste("Number of genes that exceed max Cook's distance: ",
            nrow(filter(exres_prbc_nonprbc, cooks_outlier == T))))

print(paste("Number of genes that below independent filtering threshold: ",
            nrow(filter(exres_prbc_nonprbc, indepfilt == T))))
         
```

## Independent filtering

For independent filtering, genes below this threshold are very lowly expressed anyway and probably do not drive major biological differences.

```{r}
temp = exres_prbc_nonprbc %>% filter(indepfilt == T) %>% pull(ensembl_gene_id)
temp.counts = counts(dds)[rownames(counts(dds)) %in% temp,]

Heatmap(temp.counts, show_row_names = F, show_column_names = F,
        column_title = "Genes below the independent filtering threshold (raw counts)")
rm(list = c("temp", "temp.counts"))
```

## Cook's distance

Genes above the Cook's threshold include:

```{r}
cooks_rejects = exres_prbc_nonprbc %>% filter(cooks_outlier == T)

ego <- enrichGO(gene  = cooks_rejects$ensembl_gene_id, 
                universe      = rep_prbc_nonprbc$annotated_results$ensembl_gene_id, #All genes
                OrgDb         = org.Hs.eg.db,
                keytype = "ENSEMBL",
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
        readable      = TRUE)

print("GO BPs for genes lost due to Cook's: ")
ego@result$Description
```

Have a look at some Wikipathways too:

```{r}
wp2gene <- read.gmt(here("data", "external", "gmt", "wikipathways-20191010-gmt-Homo_sapiens.gmt"))
wp2gene <- wp2gene %>% tidyr::separate(ont, c("name","version","wpid","org"), "%")
wpid2gene <- wp2gene %>% dplyr::select(wpid, gene) #TERM2GENE
wpid2name <- wp2gene %>% dplyr::select(wpid, name) #TERM2NAME

#gene.df <- bitr(cooks_rejects$ensembl_gene_id, fromType = "ENSEMBL",
#        toType = c("SYMBOL","ENTREZID"),
#        OrgDb = org.Hs.eg.db, drop = F)
#gene.df %>% filter(rowSums(is.na(gene.df)) > 0) %>%
#  left_join(.,gx_annot, by =c("ENSEMBL" = "ensembl_gene_id")) #NAs are largely not protein coding genes

gene.df <- bitr(cooks_rejects$ensembl_gene_id, fromType = "ENSEMBL",
        toType = c("SYMBOL","ENTREZID"),
        OrgDb = org.Hs.eg.db, drop = T)


ewp <- enricher(gene.df$ENTREZID, TERM2GENE = wpid2gene, TERM2NAME = wpid2name)
ewp@result
```

Heatmap of these genes

```{r}
#deseq_heatmap(mat = assay(vsd), sampledata = as.data.frame(colData(vsd)),
#              sig_results = cooks_rejects,  title = "Cook's rejects: Prbc vs non-prbc",
#              row_scale = F) #Glitchy with this input

#Heatmap annotation
ann_top = as.data.frame(colData(vsd))[,"study_group", drop=F]

#Top column annotation
colTop <- HeatmapAnnotation(df=ann_top, which="col",
                              col = list(study_group=study_colors))

#Bottom column annotation
ann_bottom = as.data.frame(colData(vsd))[,"PAM50", drop=F]
colBottom <- HeatmapAnnotation(df=ann_bottom, which="col", col = list(PAM50 = pam_colors))

#Raw counts
counts(dds)[rownames(counts(dds)) %in% cooks_rejects$ensembl_gene_id, ] %>%
  Heatmap(., show_column_names = F, show_row_names=F,
          column_title = "Cooks rejects, raw counts",
          top_annotation = colTop,
          bottom_annotation = colBottom)

temp = log2(counts(dds)[rownames(counts(dds)) %in% cooks_rejects$ensembl_gene_id, ]+1)
temp %>%
  Heatmap(., show_column_names = F, show_row_names=F,
          column_title = "Cooks rejects, log2 transformed",
          top_annotation = colTop,
          bottom_annotation = colBottom)
#Which of these genes are decently highly expressed
#tempgene = temp[names(sort(rowSums(temp), decreasing = T)),] %>% head(10) %>%
#  rownames() 
#filter(exres_prbc_nonprbc, ensembl_gene_id %in% tempgene)
  
rm(temp)
#rm(tempgene)

assay(vsd)[rownames(assay(vsd)) %in% cooks_rejects$ensembl_gene_id, ] %>%
  scrime::rowScales() %>%
  Heatmap(., show_column_names = F, show_row_names=F,
          column_title = "Cooks rejects, rowscaled",
          top_annotation = colTop,
          bottom_annotation = colBottom,)
 	
#cooks_rejects %>% dplyr::select(gene_name, Type) %>% distinct() %>% filter(!duplicated(gene_name)) %>% column_to_rownames("gene_name")             
```

```{r}
m <- nrow(attr(dds, "dispModelMatrix"))
p <- ncol(attr(dds, "dispModelMatrix"))
alt_cutoff <- qf(0.99999999999, p, m - p)
print(paste("Alternate Cooks cutoff: ", alt_cutoff))


#cooks_rejects %>% 
exres_prbc_nonprbc %>%
  left_join(.,enframe(mcols(dds)$maxCooks, "ensembl_gene_id", "max_cooks"), by = "ensembl_gene_id") %>%
  mutate(label = if_else(max_cooks > 25, gene_name, NULL) ) %>%
  ggplot(aes(x=log2FoldChange, y = max_cooks, color = -log10(padj),
             shape=cooks_outlier, size=log2(baseMean), label=label)) +
  geom_point() + ggrepel::geom_label_repel() +
  ggtitle("Cooks rejects by expression, fold change and padj") +
  geom_hline(yintercept = defaultCutoff, color = "black", linetype="dashed") +
  geom_hline(yintercept = alt_cutoff, color = "red", linetype="dashed")
  
```

```{r}

```



```{r, eval=F, include=F}
#Code for retrieving Cook's cutoff by default and identifying genes above it, from DESeq2 source.

m <- nrow(attr(dds, "dispModelMatrix"))
p <- ncol(attr(object, "dispModelMatrix"))
defaultCutoff <- qf(0.99, p, m - p)

performCooksCutoff <- (is.numeric(cooksCutoff) | cooksCutoff)
  if (performCooksCutoff) {
    cooksOutlier <- mcols(object)$maxCooks > cooksCutoff
    if (any(cooksOutlier, na.rm = TRUE) & is(design(object), 
                                             "formula")) {
      designVars <- all.vars(design(object))
      if (length(designVars) == 1) {
        var <- colData(object)[[designVars]]
        if (is(var, "factor") && nlevels(var) == 2) {
          dontFilter <- logical(sum(cooksOutlier, na.rm = TRUE))
          for (i in seq_along(dontFilter)) {
            ii <- which(cooksOutlier)[i]
            outCount <- counts(object)[ii, which.max(assays(object)[["cooks"]][ii, 
                                                                               ])]
            if (sum(counts(object)[ii, ] > outCount) >= 
                3) {
              dontFilter[i] <- TRUE
            }
          }
          cooksOutlier[which(cooksOutlier)][dontFilter] <- FALSE
        }
      }
    }
    res$pvalue[which(cooksOutlier)] <- NA
  }
```


# Save data


Save the individual results lists, without the figures.

```{r}

#Save the results for all genes in an Excel file with multiple tabs
#The tab names will be the list names
resdf = lapply(reslist, function(x) x$annotated_results)
names(resdf) = paste(names(resdf),"all", sep="_")
openxlsx::write.xlsx(resdf, file = here("results", "diffex", "07_pairwise_comparisons_allgenes.xlsx"))

#Save the significant genes in a multi-tab excel file
resdf = lapply(reslist, function(x) x$significant_genes)
names(resdf) = paste("sig",names(resdf), sep="_")
#Also keep the thresholds used for significance
resdf[[length(resdf)+1]] = enframe(unlist(lapply(reslist, function(x) x$sig_threshold)), "name", "value") %>%
                 separate(name, into = c("comparison", "threshold"), sep="\\.")
names(resdf)[length(resdf)] = "thresholds"
library(openxlsx)
openxlsx::write.xlsx(resdf, file = here("results", "diffex", "07_pairwise_comparisons_sig_genes.xlsx"))

rm(resdf)
```

Save the figures

```{r}
hmlist = lapply(reslist, function(x) x$heatmap)
names(hmlist) = names(hmlist) %>% str_replace_all("rep", "hm")

for (i in 1:length(hmlist)){
  pdf(here("results", "diffex", "figs", "07_pairwise", "heatmaps", paste0(names(hmlist)[i],".pdf")), width = 10, height = 8)
  print(hmlist[[i]])
  dev.off()
}

rm(hmlist)
```

```{r}
volclist = lapply(reslist, function(x) x$volcano_plot)
names(volclist) = names(volclist) %>% str_replace("rep", "volc")

dir.create(here("results","diffex", "figs", "07_pairwise" ,"volcano_plots"), showWarnings = F)
lapply(names(volclist), 
       function(x) ggsave(filename=here("results","diffex", "figs", "07_pairwise","volcano_plots",
                                        paste(x,".jpeg",sep="")), plot=volclist[[x]],
                          height = 6, width = 8))
rm(volclist)
```

```{r}
rm(reslist)
```


Save the notebook data.

```{r}
save.image(here("reports/07_diffex_pairwise_standard.RData"))
```

