---
title: "Pairwise Wald tests, standard threshold"
Author: Kat Moore
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: yes
    toc_float: yes
    toc_depth: 4
    highlight: tango
    df_print: paged
---

```{r, include=F}
rm(list = ls())
```

```{r, include=F}
library(DESeq2)
library(apeglm)
library(here)
library(ggrepel)
library(openxlsx)
library(scrime)
library(circlize)
library(RColorBrewer)
library(ComplexHeatmap)
library(UpSetR)
library(tidyverse)
```

# Load and pre-process data {.tabset}

## Gene and sample data

```{r}
gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
gx_annot = gx_annot %>% select(ensembl_gene_id = gene_id,
                               gene_name, gene_type,
                               description = gene_description) %>%
  distinct()
```

## Gene signatures

```{r}
gene_sets = list(
  go_bp = flexgsea::read_gmt(here("data","external","gmt","c5.bp.v7.0.symbols.gmt")),
  hallmark = flexgsea::read_gmt(here("data","external","gmt","h.all.v7.0.symbols.gmt")),
  c2_canon = flexgsea::read_gmt(here("data","external","gmt","c2.cp.v7.0.symbols.gmt")),
  c2_cgp = flexgsea::read_gmt(here("data","external","gmt","c2.cgp.v7.0.symbols.gmt"))
)
```

## Retrieve immune genes

For downstream analyses, we'd like to know which of the differentially expressed genes are immunologically relevant.

An immune gene is defined as follows:

Either immune/immuno/interleukin is part of the gene name and description OR the gene is part of the the [ImmPort database](https://www.innatedb.com/redirect.do?go=resourcesGeneLists)

The Immunology Database and Analysis Portal (ImmPort) system was developed under the Bioinformatics Integration Support Contract (BISC) Phase II by the Northrop Grumman Information Technology Health Solutions team for the NIH, NIAID, and DAIT. The list of immunologically related genes in ImmPort is a collection of ~6,000 human genes, which was formed with the goal of retrieving all genes that have immune system-related functions. This list was generated using automatic searches of EntrezGene and Gene Ontology records using immunology-related keywords. The list was then manually curated by immunology experts examining various literature sources. 

```{r}
immune_gene_list <- read_csv(here("data", "external","gene-sets","InnateDB_genes.csv"))
head(immune_gene_list)
```

## Helper functions

```{r}
source(here("src", "deseq_report_functions.R"))
```

## Heatmap colors

Created in notebook 6

```{r}
cp = readRDS(here("data", "Rds", "color_palettes.Rds"))
study_colors = cp$study_colors
pam_colors = cp$pam_colors
gene_colors = cp$gene_colors

sp = readRDS(here("data/Rds/survival_colors.Rds"))
```

## Variance stabilizing transformation

Previously performed in script 6.

```{r}
vsd = readRDS(here("data","Rds","06_vsd.Rds"))
```

# Diffex: non-prbc as reference group

Filtering and differential expression was performed in script 7.
Ensure that non-prbc is the reference level.

```{r}
dds = readRDS(here("data/Rds/07_dds_pairwise_ref_non_prbc.Rds"))
design(dds)
levels(dds$study_group)
nrow(dds)
```

# Involution vs non-prbc {.tabset}

Apeglm shrinkage of log fold changes was performed in script 7.

```{r}
ape_inv_nonprbc = readRDS(here("data", "Rds", "07_ape_ppbc_inv_vs_non_prbc.Rds"))

ape_inv_nonprbc@elementMetadata$description
```

## Summary report

```{r}
dds$death <- dds$overall_survival
vsd$death <- vsd$overall_survival

rep_inv_nonprbc = deseq_report(ape_inv_nonprbc, dds=dds, absl2fc = 0.5, 
                               variance_stabilized_dds = vsd,
                               groups = c("ppbc_inv", "non_prbc"),
                               top_vars = c("PPBC", "death"),
                               top_colors = list(PPBC = cp$ppbc_colors,
                                                 death = sp$overall_survival),
                               beehive_groups = "comparison",
                               title = "Involution vs nulliparous")

```

## Top genes

```{r}
rep_inv_nonprbc$significant_genes %>%
  select(gene_name, gene_type, padj, log2FoldChange,
         baseMean, description, everything()) %>%
  head(30)
```

## Volcano plot

```{r}
rep_inv_nonprbc$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_inv_nonprbc$heatmap
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_inv_nonprbc$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
rep_inv_nonprbc$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>%
  arrange(fdr)
```


```{r, fig.height=6, fig.width=8}
rep_inv_nonprbc$pathway_plots
```

```{r,fig.height=6, fig.width=8}
rep_inv_nonprbc$pathways %>%
  bind_rows() %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Top 20 sig pathways fdr < 0.1: Inv. vs nulliparous"
      )
```

## Cook's outliers

Have a look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r}
rep_inv_nonprbc$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 10:

```{r}
rep_inv_nonprbc$outlier_bees
```

# Pregnancy vs non-pregnancy {.tabset}

```{r}
ape_prbc_nonprbc = readRDS(here("data", "Rds", "07_ape_prbc_vs_non_prbc.Rds"))
ape_prbc_nonprbc@elementMetadata$description
```

## Summary report

```{r}
rep_prbc_nonprbc = deseq_report(ape_prbc_nonprbc, dds=dds, absl2fc = 0.5, 
                               variance_stabilized_dds = vsd,
                               groups = c("prbc", "non_prbc"),
                               top_vars = c("PPBC", "death"),
                               top_colors = list(PPBC = cp$ppbc_colors,
                                                 death = sp$overall_survival),
                               beehive_groups = "comparison",
                               title = "Pregnant vs nulliparous")


```

## Top genes

```{r}
rep_prbc_nonprbc$significant_genes %>%
  select(gene_name, gene_type, padj, log2FoldChange,
         baseMean, description, everything()) %>%
  head(30)
```

## Volcano plot

```{r}
rep_prbc_nonprbc$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_prbc_nonprbc$heatmap
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_prbc_nonprbc$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
rep_prbc_nonprbc$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>% arrange(fdr)
```


```{r, fig.height=6, fig.width=8}
rep_prbc_nonprbc$pathway_plots
```

## Cook's outliers

```{r}
rep_prbc_nonprbc$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 10:

```{r}
rep_prbc_nonprbc$outlier_bees
```

# Lac vs non-prbc {.tabset}

```{r}
ape_lac_nonprbc = readRDS(here("data", "Rds", "07_ape_ppbc_lac_vs_non_prbc.Rds"))
ape_lac_nonprbc@elementMetadata$description
```

## Summary report

```{r}
rep_lac_nonprbc = deseq_report(ape_lac_nonprbc, dds=dds, absl2fc = 0.5, 
                               variance_stabilized_dds = vsd,
                               groups = c("ppbc_lac", "non_prbc"),
                               top_vars = c("PPBC", "death"),
                               top_colors = list(PPBC = cp$ppbc_colors,
                                                 death = sp$overall_survival),
                               beehive_groups = "comparison",
                               title = "Lactating vs nulliparous")

```

## Top genes

```{r}
rep_lac_nonprbc$significant_genes %>%
  select(gene_name, gene_type, padj, log2FoldChange,
         baseMean, description, everything()) %>%
  head(30)
```

## Volcano plot

```{r}
rep_lac_nonprbc$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_lac_nonprbc$heatmap
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_lac_nonprbc$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
rep_lac_nonprbc$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>% arrange(fdr)
```

```{r, fig.height=6, fig.width=8}
rep_lac_nonprbc$pathway_plots
```

## Cook's outliers

```{r}
rep_lac_nonprbc$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 10:

```{r}
rep_lac_nonprbc$outlier_bees
```

# Diffex: prbc as reference level

```{r}
dds = readRDS(here("data/Rds/07_dds_pairwise_ref_prbc.Rds"))
design(dds)
levels(dds$study_group)
nrow(dds)
```

# Involution vs prbc {.tabset}

```{r}
ape_inv_prbc = readRDS(here("data", "Rds", "07_ape_ppbc_inv_vs_prbc.Rds"))
ape_inv_nonprbc@elementMetadata$description
```

## Summary report

```{r}
dds$death <- dds$overall_survival
vsd$death <- vsd$overall_survival

rep_inv_prbc = deseq_report(ape_inv_prbc, dds=dds, absl2fc = 0.5, 
                            variance_stabilized_dds = vsd,
                            groups = c("ppbc_inv", "prbc"),
                            top_vars = c("PPBC", "death"),
                            top_colors = list(PPBC = cp$ppbc_colors,
                                              death = sp$overall_survival),
                            beehive_groups = "comparison",
                            title = "Involuting vs pregnant")
```

## Top genes

```{r}
rep_inv_prbc$significant_genes %>%
  select(gene_name, gene_type, padj, log2FoldChange,
         baseMean, description, everything()) %>%
  head(30)
```

## Volcano plot

```{r}
rep_inv_prbc$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_inv_prbc$heatmap
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_inv_prbc$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
rep_inv_prbc$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>% arrange(fdr)
```


```{r, fig.height=6, fig.width=8}
rep_inv_prbc$pathway_plots
```

```{r, fig.height=6, fig.width=8}
rep_inv_prbc$pathways %>%
  bind_rows() %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Top 20 sig pathways fdr < 0.1: Involution vs pregnant"
      )
```

## Cook's outliers

```{r}
rep_inv_prbc$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 10:

```{r}
rep_inv_prbc$outlier_bees
```

# Lac vs prbc {.tabset}

```{r}
ape_lac_prbc = readRDS(here("data", "Rds", "07_ape_prbc_vs_ppbc_lac.Rds"))
ape_lac_prbc@elementMetadata$description
```
## Summary report

```{r}
rep_lac_prbc = deseq_report(ape_lac_prbc, dds=dds, absl2fc = 0.5, 
                            variance_stabilized_dds = vsd,
                            groups = c("ppbc_lac", "prbc"),
                            top_vars = c("PPBC", "death"),
                            top_colors = list(PPBC = cp$ppbc_colors,
                                              death = sp$overall_survival),
                            beehive_groups = "comparison",
                            title = "Lactation vs pregnant")

```

## Top genes

```{r}
rep_lac_prbc$significant_genes %>%
  select(gene_name, gene_type, padj, log2FoldChange,
         baseMean, description, everything()) %>%
  head(30)
```

## Volcano plot

```{r}
rep_lac_prbc$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_lac_prbc$heatmap
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_lac_prbc$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
rep_lac_prbc$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>% arrange(fdr)
```

```{r, fig.height=6, fig.width=8}
rep_lac_prbc$pathway_plots
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r}
rep_lac_prbc$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 10:

```{r}
rep_lac_prbc$outlier_bees
```

# Diffex: Lac as reference level

```{r}
dds = readRDS(here("data/Rds/07_dds_pairwise_ref_ppbc_lac.Rds"))
design(dds)
levels(dds$study_group)
nrow(dds)
```

# Involution vs lac {.tabset}

## Extract results

```{r}
ape_inv_lac = readRDS(here("data", "Rds", "07_ape_ppbc_inv_vs_ppbc_lac.Rds"))
ape_inv_lac@elementMetadata$description
```


## Summary report

```{r}
dds$death <- dds$overall_survival
vsd$death <- vsd$overall_survival

rep_inv_lac = deseq_report(ape_inv_lac, dds=dds, absl2fc = 0.5, 
                           variance_stabilized_dds = vsd,
                           groups = c("ppbc_inv", "ppbc_lac"),
                           top_vars = c("PPBC", "death"),
                           top_colors = list(PPBC = cp$ppbc_colors,
                                             death = sp$overall_survival),
                           beehive_groups = "comparison",
                           title = "Involution vs lactation")

```

## Top genes

```{r}
rep_inv_lac$significant_genes %>%
  select(gene_name, gene_type, padj, log2FoldChange,
         baseMean, description, everything()) %>%
  head(30)
```

## Volcano plot

```{r}
rep_inv_lac$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_inv_lac$heatmap
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
rep_inv_lac$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>% arrange(fdr)
```

```{r, fig.height=6, fig.width=8}
rep_inv_lac$pathway_plots
```

```{r, fig.height=6, fig.width=8}
rep_inv_lac$pathways %>%
  bind_rows() %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Top 20 sig pathways fdr < 0.1: Involution vs lactation"
      )
```

## Cook's outliers

```{r}
rep_inv_lac$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 10:

```{r}
rep_inv_lac$outlier_bees
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_inv_lac$beehive_plots
```

# Distance plots

Attempt to visualize the number of significant genes in each comparison.

```{r}
#Get all report objects from environment
reslist = mget(ls(pattern="rep_"))
```


```{r}
sign = lapply(lapply(reslist, function(x) x$significant_genes), function(x) length(unique(x$gene_name)))

#Create comparison matrix for number of significant genes
sigmat = sign %>% unlist %>% enframe("name", "sig") %>%
  separate(name, into=c(NA, "num", "denom")) %>%
  spread(key="denom", value="sig") %>%
  bind_rows(., tibble(num="nonprbc", lac=NA, nonprbc=NA, prbc=NA)) %>%
  mutate(inv = NA) %>% column_to_rownames("num")

#Standardize column orders
sigmat = sigmat[c("nonprbc", "prbc", "lac", "inv"), c("nonprbc", "prbc", "lac", "inv")]
sigmat = as.matrix(sigmat)
```

Replace the diagonal with 0.

```{r}
for (x in 1:length(rownames(sigmat))){
  for (y in 1:length(colnames(sigmat))){
    if (rownames(sigmat)[x] == rownames(sigmat)[y]){
      sigmat[x,y] = 0
    }
  }
}

sigmat
```

Fill in the NAs from redundant comparisons.

```{r}
for (x in 1:length(rownames(sigmat))){
  for (y in 1:length(colnames(sigmat))){
    if (is.na(sigmat[x,y])){
      sigmat[x,y] = sigmat[y,x]
    }
  }
}

sigmat
```

Easiest way to color the diagonal is just to turn it back to NA.

```{r}
for (x in 1:length(rownames(sigmat))){
  for (y in 1:length(colnames(sigmat))){
    if (sigmat[x,y]==0){
      sigmat[x,y] = NA
    }
  }
}

sigmat
```


```{r}
dir.create(here("results","diffex", "figs", "07_pairwise" ,"heatmaps"), showWarnings = F, recursive = T)

nsig_hm = Heatmap(sigmat, name = "significant genes", cell_fun = function(j, i, x, y, width, height, fill) 
        {
          grid.text(sprintf("%.0f", sigmat[i, j]), x, y, gp = gpar(fontsize = 12))
        },
        cluster_rows = F, cluster_columns = F, column_names_side = "bottom", column_names_rot = 0,
        row_names_side = "left", col = colorRamp2(c(0, 200, 400), c("blue", "purple", "red")),
        column_title = "Number of significant genes in pairwise comparisons")

pdf(here("results", "diffex", "figs", "07_pairwise", "heatmaps", "n_siggenes_pairwise_hm.pdf"), width = 10, height = 8)
nsig_hm
dev.off()

nsig_hm
```

# Involution summary {.tabset}

```{r}
#All genes considered:
background = sort(rep_inv_nonprbc$annotated_results$ensembl_gene_id)

inv_overview = data.frame(
  vs_nulliparous = background %in% rep_inv_nonprbc$significant_genes$ensembl_gene_id,
  vs_pregnant = background %in% rep_inv_prbc$significant_genes$ensembl_gene_id,
  vs_lactation = background %in% rep_inv_lac$significant_genes$ensembl_gene_id)
rownames(inv_overview) = background

#Remove genes that weren't significant anwyhere
#sig_genes = sig_genes[rowSums(sig_genes) > 0,] #Not actually necessary
#dim(sig_genes)

vp = limma::vennCounts(inv_overview)

limma::vennDiagram(vp,
  names = colnames(inv_overview), 
  cex = 1,
  main = "Overlap significant genes from pairwise involution tests")
```

## Overlap pathways

```{r}
#Binary df for all genes to indicate comparison significance

background = c(names(gene_sets$go_bp), names(gene_sets$hallmark),
               names(gene_sets$c2_canon), names(gene_sets$c2_cgp))

sig_paths = data.frame(
  vs_nulliparous = background %in% filter(bind_rows(rep_inv_nonprbc$pathways), fdr < 0.1)$pathway,
  vs_pregnant = background %in% filter(bind_rows(rep_inv_prbc$pathways), fdr < 0.1)$pathway,
  vs_lactation = background %in% filter(bind_rows(rep_inv_lac$pathways), fdr < 0.1)$pathway)
rownames(sig_paths) = background

#Remove genes that weren't significant anwyhere
#sig_genes = sig_genes[rowSums(sig_genes) > 0,] #Not actually necessary
#dim(sig_genes)

vp = limma::vennCounts(sig_paths)

limma::vennDiagram(vp,
  names = colnames(sig_paths), 
  cex = 1,
  main = "Overlap significant pathways from pairwise involution tests")
```


```{r}

unique_inv_paths = Reduce(intersect, list(filter(bind_rows(rep_inv_nonprbc$pathways), fdr < 0.1)$pathway,
                                          filter(bind_rows(rep_inv_prbc$pathways), fdr < 0.1)$pathway,
                                          filter(bind_rows(rep_inv_lac$pathways), fdr < 0.1)$pathway))

bind_rows(rep_inv_nonprbc$pathways) %>% filter(pathway %in% unique_inv_paths) %>% pull(pathway) %>%
  #paste(., collapse=", ") 
  tolower() %>% str_replace_all("_", " ")
```

## Heatmap common pathways

```{r}
members_common_paths <- bind_rows(rep_inv_nonprbc$pathways) %>% filter(pathway %in% unique_inv_paths) %>% pull(members_in_sample)

members_common_paths = sapply(members_common_paths, function(x) unlist(strsplit(x, split=";"))) %>% unlist()
names(members_common_paths) = NULL

members_common_paths = unique(members_common_paths)

#members_common_paths

members_common_paths %>% paste(., collapse=", ")
```

```{r, fig.height=6, fig.width=9}
mat <- rownames_to_column(as.data.frame(assay(vsd)), "ensembl_gene_id")
mat <- right_join(select(gx_annot, ensembl_gene_id, gene_name),
                               mat, by = "ensembl_gene_id") %>%
  select(-ensembl_gene_id)

mat <- summarize_expression_duplicate_ids(mat, id_column = "gene_name", verbose = T)
```

```{r}
mat_common_paths <- mat[rownames(mat) %in% members_common_paths,]
dim(mat_common_paths)
stopifnot(nrow(mat_common_paths) == length(members_common_paths))
```

```{r}
ann_row = 
select(filter(rep_inv_nonprbc$significant_genes, gene_name %in% members_common_paths),
       gene_name, Type) %>% distinct() %>%
  inner_join(., data.frame(gene_name = members_common_paths,
       vs_nulliparous = members_common_paths %in% rep_inv_nonprbc$significant_genes$gene_name,
       vs_pregnant = members_common_paths %in% rep_inv_prbc$significant_genes$gene_name,
       vs_lactation = members_common_paths %in% rep_inv_lac$significant_genes$gene_name),
       by="gene_name")%>%
  column_to_rownames("gene_name")

ann_row
```

```{r, fig.height=6, fig.width=9}
mat_common_paths <- as.matrix(mat_common_paths)

quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(xs, probs = seq(0, 1, length.out = n))
  breaks[!duplicated(breaks)]
}

mat_breaks <- quantile_breaks(scrime::rowScales(mat_common_paths), n = 5)

pheatmap::pheatmap(mat_common_paths,
                   show_colnames = F,
                   scale="row",
                   annotation_col = as.data.frame(colData(vsd))[,c("PAM50", "study_group")],
                   annotation_colors = list(PAM50 = pam_colors, study_group = study_colors, Type = gene_colors),
                   annotation_row = ann_row[,"Type", drop=F],
                   #color = viridis::viridis(length(mat_breaks) - 1), 
                   #breaks = mat_breaks,
                   main="Members of pathways common to all pairwise involution tests")

```

```{r, eval=F}
pheatmap::pheatmap(mat_common_paths,
                   show_colnames = F,
                   scale="row",
                   annotation_col = as.data.frame(colData(vsd))[,c("PAM50", "study_group")],
                   annotation_colors = list(PAM50 = pam_colors, study_group = study_colors, Type = gene_colors),
                   annotation_row = ann_row,
                   #color = viridis::viridis(length(mat_breaks) - 1), 
                   #breaks = mat_breaks,
                   main="Members of pathways common to all pairwise involution tests",
                   filename = here("results", "diffex", "figs", "07_pairwise", "heatmaps", "07_commoninvpaths_hm.pdf"))
```

## PCA common pathways

```{r}
nc.pca <- prcomp(t(mat_common_paths))
#pca.df <- as.data.frame(nc.pca$x[, 1:17])
pca.df <- as.data.frame(nc.pca$x)
pca.df$sample_name <- rownames(nc.pca$x)
pca.df <- inner_join(pca.df, as.data.frame(colData(dds)), by='sample_name')
pca.df = pca.df %>% select(sample_name, PPBC, everything())
#head(pca.df)
```

```{r}
ggplot(pca.df, aes(x=PC1, y=PC2, color=PPBC)) +
  geom_point()  + 
  scale_color_manual(values = cp$ppbc_colors) +
  ggtitle("PCA genes belonging to common pathways") +
  theme_bw()
```

# UpSet plots

An alternative to Venn diagrams.

Existing combinations:

```{r}
t(combn(unique(dds$study_group), 2))
```

Create list of significant genes from these combinations

```{r}
sig_genes <- list(
  "inv vs nonprbc" = rep_inv_nonprbc$significant_genes$ensembl_gene_id,
  "inv vs prbc" = rep_inv_prbc$significant_genes$ensembl_gene_id,
  "inv vs lac" = rep_inv_lac$significant_genes$ensembl_gene_id,
  "prbc vs nonprbc" = rep_prbc_nonprbc$significant_genes$ensembl_gene_id,
  "lac vs nonprbc" = rep_lac_nonprbc$significant_genes$ensembl_gene_id,
  "lac vs prbc" = rep_lac_prbc$significant_genes$ensembl_gene_id
)
```

```{r, fig.width=8}
upset(fromList(sig_genes), order.by = "freq", #empty.intersections = "on",
      nsets = 6, set_size.show = TRUE, mainbar.y.label = "Differentially expressed genes",
      sets.bar.color = "blue", main.bar.color = "black", set_size.scale_max = 600,
      sets.x.label = "Sig genes pairwise comparisons")
```

```{r, echo=F}
pdf(here("results", "diffex", "figs", "07_pairwise", "upset_pairwise.pdf"),
    onefile = F, width = 10, height = 7)
upset(fromList(sig_genes), order.by = "freq", #empty.intersections = "on",
      nsets = 6, set_size.show = TRUE, mainbar.y.label = "Differentially expressed genes",
      sets.bar.color = "blue", main.bar.color = "black", set_size.scale_max = 600,
      sets.x.label = "Sig genes pairwise comparisons")
dev.off()
```

# Save data {.tabset}

## Significance of genes

Save the individual results lists, without the figures.

```{r}

#Save the results for all genes in an Excel file with multiple tabs
#The tab names will be the list names
resdf = lapply(reslist, function(x) x$annotated_results)
names(resdf) = paste(names(resdf),"all", sep="_")
openxlsx::write.xlsx(resdf, file = here("results", "diffex", "07_pairwise_comparisons_allgenes.xlsx"))

#Save the significant genes in a multi-tab excel file
resdf = lapply(reslist, function(x) x$significant_genes)
names(resdf) = paste("sig",names(resdf), sep="_")
#Also keep the thresholds used for significance
resdf[[length(resdf)+1]] = enframe(unlist(lapply(reslist, function(x) x$sig_threshold)), "name", "value") %>%
                 separate(name, into = c("comparison", "threshold"), sep="\\.")
names(resdf)[length(resdf)] = "thresholds"
library(openxlsx)
openxlsx::write.xlsx(resdf, file = here("results", "diffex", "07_pairwise_comparisons_sig_genes.xlsx"))

rm(resdf)
```

## Pathway results

Fisher's exact pathway analyses

```{r}
resdf = lapply(reslist, function(x) x$pathways)

#Returns a list of aggregated path results per comparison
pathlist = list()
pathlist[[names(resdf)[1]]] = resdf %>% map(1, .depth=1) %>% bind_rows( .id = "comp")
for (i in 2:length(resdf[[1]])){
  path = resdf %>% map(i, .depth=1) %>% bind_rows( .id = "comp")
  pathlist[[names(resdf)[i]]] = path
}

#Fix names
names(pathlist) = lapply(pathlist, function(x) unique(x$collection)) %>% unlist(use.names = F)
pathlist = lapply(pathlist, function(x) x[order(x$fdr),])

openxlsx::write.xlsx(pathlist, file = here("results", "diffex", "07_pairwise_comparisons_pathways.xlsx"))

rm(resdf)
```


## Heatmaps

```{r}
hmlist = lapply(reslist, function(x) x$heatmap)
names(hmlist) = names(hmlist) %>% str_replace_all("rep", "hm")

for (i in 1:length(hmlist)){
  pdf(here("results", "diffex", "figs", "07_pairwise", "heatmaps", paste0(names(hmlist)[i],".pdf")), width = 10, height = 8)
  print(hmlist[[i]])
  dev.off()
}
```

## Volcano plots sig hits

```{r}
volclist = lapply(reslist, function(x) x$volcano_plot)
names(volclist) = names(volclist) %>% str_replace("rep", "volc")

dir.create(here("results","diffex", "figs", "07_pairwise" ,"volcano_plots"), showWarnings = F)
lapply(names(volclist), 
       function(x) ggsave(filename=here("results","diffex", "figs", "07_pairwise","volcano_plots",
                                        paste(x,".jpeg",sep="")), plot=volclist[[x]],
                          height = 6, width = 8))
```

## Volcano plot outliers

```{r}
outvplist = lapply(reslist, function(x) x$volc_outliers)
names(outvplist) = names(outvplist) %>% str_replace("rep", "volc")

dir.create(here("results","diffex", "figs", "07_pairwise" ,"volcano_plots"), showWarnings = F)
lapply(names(outvplist), 
       function(x) ggsave(filename=here("results","diffex", "figs", "07_pairwise","volcano_plots",
                                        paste(x,".outliers.jpeg",sep="")), plot=outvplist[[x]],
                          height = 6, width = 8))
```

## Pathway dotplots 

```{r}
dir.create(here("results","diffex", "figs", "07_pairwise" ,"fisher_pathway"), showWarnings = F)
pathwaylist = lapply(reslist, function(x) x$pathway_plots)
for(i in 1:length(pathwaylist)){
  thiscomp = pathwaylist[[i]]
  compname = str_replace(str_remove(names(pathwaylist[i]), "rep_"),"_","_vs_")
  for (j in 1:length(thiscomp)){
    if (length(thiscomp) < 1){next}
    thissig = thiscomp[[j]]
    file=paste0(paste(compname, names(thiscomp[j]), sep="-"),".jpeg")
    #print(file)
    ggsave(filename=here("results","diffex", "figs",
                         "07_pairwise","fisher_pathway",file),
           plot=thissig, height = 6, width = 8)
  }
}


rm(pathwaylist)
```

## Session Info

```{r}
sessionInfo()
```

