---
title: "One vs rest Wald tests, standard threshold"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
---

```{r, include=F}
rm(list = ls())
```


```{r, include=F}
#library(devtools)
#install_github("jokergoo/ComplexHeatmap")
#install_github("mikelove/DESeq2") 
#install_github("azhu513/apeglm")
library(flexgsea) #For read_gmt()
library(ggrepel)
library(ComplexHeatmap)
library(openxlsx)
library(scrime)
library(RColorBrewer)
library(here)
library(DESeq2)
library(apeglm)
library(tidyverse)
```

# Load and pre-process data

## Gene data

```{r}
dds <- readRDS(file = here("data/Rds/05_dds_PAM50_batch.Rds"))

gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
#head(gx_annot)
gx_annot = gx_annot %>% select(ensembl_gene_id = gene_id, gene_name, hgnc_symbol,
                               gene_type, description = gene_description) %>% distinct()
```

## Gene signatures

```{r}
gene_sets = list(
  go_bp = flexgsea::read_gmt(here("data","external","gmt","c5.bp.v7.0.symbols.gmt")),
  hallmark = flexgsea::read_gmt(here("data","external","gmt","h.all.v7.0.symbols.gmt")),
  c2_canon = flexgsea::read_gmt(here("data","external","gmt","c2.cp.v7.0.symbols.gmt")),
  c2_cgp = flexgsea::read_gmt(here("data","external","gmt","c2.cgp.v7.0.symbols.gmt"))
)
```


## Filtering

Minimum threshold is a nonzero count in at least a 3rd of all samples

```{r}
keep <- rowSums(counts(dds)!=0) >= ceiling(ncol(dds)/3)
table(keep)

dds <- dds[keep,]
```


## Retrieve immune genes

For downstream analyses, we'd like to know which of the differentially expressed genes are immunologically relevant.

An immune gene is defined as follows:

Either immune/immuno/interleukin is part of the gene name and description OR the gene is part of the the ImmPort database. See https://www.innatedb.com/redirect.do?go=resourcesGeneLists

The Immunology Database and Analysis Portal (ImmPort) system was developed under the Bioinformatics Integration Support Contract (BISC) Phase II by the Northrop Grumman Information Technology Health Solutions team for the NIH, NIAID, and DAIT. The principal investigator of the BISC project is Dr. Richard Scheuermann at University of Texas Southwestern Medical Center. The list of immunologically related genes in ImmPort is a collection of ~6,000 human genes, which was formed with the goal of retrieving all genes that have immune system-related functions. This list was generated using automatic searches of EntrezGene and Gene Ontology records using immunology-related keywords. The list was then manually curated by immunology experts examining various literature sources. 

```{r}
immune_gene_list <- read_csv(here("data", "external","gene-sets","InnateDB_genes.csv"))
head(immune_gene_list)
```



## Helper functions

```{r}
source(here("src", "deseq_report_functions.R"))
```


Slight tweak to report function to show all groups in the beehive plots and heat maps. 


## Heatmap colors

Created in notebook 6

```{r}
all_colors = readRDS(here("data","Rds", "06_heatmap_colors.Rds"))
study_colors = all_colors$study_colors
pam_colors = all_colors$pam_colors
gene_colors = all_colors$gene_colors
```

## Variance stabilizing transformation

We want a transformation that is fully blind to the experimental design, but still uses the faster sampling method in the vst wrapper for varianceStabilizingTransformation.

This homoskedastic dataset will be used later on for heatmap visualization.
Previously performed in notebook 6.


```{r}
vsd = readRDS(here("data","Rds","06_vsd_standardfilter.Rds"))
```

# One vs rest formula levels

```{r}
design(dds)
```

```{r, collapse=T}
dds$inv_vs_rest  = colData(dds) %>% as.data.frame() %>% mutate(inv_vs_rest = if_else(study_group == "ppbc_inv", "ppbc_inv", "rest")) %>%
  pull(inv_vs_rest) %>% factor(levels=c("rest", "ppbc_inv"))

dds$prbc_vs_rest  = colData(dds) %>% as.data.frame() %>% mutate(prbc_vs_rest = if_else(study_group == "prbc", "prbc", "rest")) %>%
  pull(prbc_vs_rest) %>% factor(levels=c("rest", "prbc"))

dds$lac_vs_rest  = colData(dds) %>% as.data.frame() %>% mutate(lac_vs_rest = if_else(study_group == "ppbc_lac", "ppbc_lac", "rest")) %>%
  pull(lac_vs_rest) %>% factor(levels=c("rest", "ppbc_lac"))

dds$nonprbc_vs_rest  = colData(dds) %>% as.data.frame() %>% mutate(nonprbc_vs_rest = if_else(study_group == "non_prbc", "non_prbc", "rest")) %>%
  pull(nonprbc_vs_rest) %>% factor(levels=c("rest", "non_prbc"))

table(dds$inv_vs_rest, dds$study_group)
table(dds$prbc_vs_rest, dds$study_group)
table(dds$lac_vs_rest, dds$study_group)
table(dds$nonprbc_vs_rest, dds$study_group)
```


Do the same to our vsd data.

```{r}
vsd$study_group_orig = vsd$study_group
vsd$inv_vs_rest  = colData(vsd) %>% as.data.frame() %>% mutate(inv_vs_rest = if_else(study_group == "ppbc_inv", "ppbc_inv", "rest")) %>%
  pull(inv_vs_rest) %>% factor(levels=c("rest", "ppbc_inv"))

vsd$prbc_vs_rest  = colData(vsd) %>% as.data.frame() %>% mutate(prbc_vs_rest = if_else(study_group == "prbc", "prbc", "rest")) %>%
  pull(prbc_vs_rest) %>% factor(levels=c("rest", "prbc"))

vsd$lac_vs_rest  = colData(vsd) %>% as.data.frame() %>% mutate(lac_vs_rest = if_else(study_group == "ppbc_lac", "ppbc_lac", "rest")) %>%
  pull(lac_vs_rest) %>% factor(levels=c("rest", "ppbc_lac"))

vsd$nonprbc_vs_rest  = colData(vsd) %>% as.data.frame() %>% mutate(nonprbc_vs_rest = if_else(study_group == "non_prbc", "non_prbc", "rest")) %>%
  pull(nonprbc_vs_rest) %>% factor(levels=c("rest", "non_prbc"))

saveRDS(vsd, here("data", "Rds", "09_vsd_standard_vsrest.Rds"))
```


# Diffex: involution vs rest

In this call to DESeq, we're focusing one one vs rest contrasts and using a Wald test instead of a likelihood ratio test.

```{r}
dds$group = dds$inv_vs_rest

levels(dds$group)
```

```{r}
design(dds) = ~batch + PAM50 + group
```

Perform differential expression analysis.

```{r, eval=F}
start = Sys.time()
dds <- DESeq(dds)
end = Sys.time()
end-start

saveRDS(dds,here("data/Rds/08_dds_inv_vs_rest_standard.Rds"))
```

>estimating size factors
using 'avgTxLength' from assays(dds), correcting for library size
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing
-- replacing outliers and refitting for 165 genes
-- DESeq argument 'minReplicatesForReplace' = 7 
-- original counts are preserved in counts(dds)
estimating dispersions
fitting model and testing
Time difference of 14.03538 hours

```{r}
dds = readRDS(here("data/Rds/08_dds_inv_vs_rest_standard.Rds"))
```

Double check the design

```{r}
design(dds)
```


## Coefficient overview

Make it easier to find the coefficient we need.

```{r}
cdf = enframe(colnames(coef(dds)), "coef", "name")

cdf
```


## Extract results

```{r, eval=F}
ape_inv_rest = shrinkRes(dds,  contrast=c("group", "ppbc_inv", "rest"))
saveRDS(ape_inv_rest, here("data","Rds", "08_ape_inv_rest.Rds"))
```

```{r}
ape_inv_rest = readRDS(here("data","Rds", "08_ape_inv_rest.Rds"))
```


## Summary report

The reporting function requires vsd to have the same levels as dds in study_group

```{r}
vsd$study_group = vsd$inv_vs_rest
```


A log2 fold change cutoff of 0.5 is equivalent to 1.414.

```{r}
rep_inv_rest = deseq_report(ape_inv_rest, dds=dds, absl2fc = 0.5, groups = c("ppbc_inv", "rest"),
                            beehive_groups = "all", title = "Involution vs rest")
```


## Top genes

```{r}
rep_inv_rest$significant_genes
```

## Volcano plot


```{r}
rep_inv_rest$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_inv_rest$heatmap
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
 rep_inv_rest$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>% arrange(fdr)
```


```{r, fig.height=6, fig.width=8}
rep_inv_rest$pathway_plots
```


## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_inv_rest$beehive_plots
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r}
rep_inv_rest$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 10:

```{r}
rep_inv_rest$outlier_bees
```

Other outliers with known cancer functions but without a high fold change:

[IRS4](https://www.nature.com/articles/ncomms13567) is of potential interest here : "IRS4 induces mammary tumorigenesis and confers resistance to HER2-targeted therapy through constitutive PI3K/AKT-pathway hyperactivation".

However, it appears that it has been correctly labelled as an outlier.

```{r}
plot_gene_beehive(dds, rep_inv_rest$annotated_results[rep_inv_rest$annotated_results$ensembl_gene_id=="ENSG00000133124",])
```

[MAEL](https://www.spandidos-publications.com/or/31/1/342) may interact with stress granules in breast cancer.

Its plausibilty in inv vs rest is more substantial.

```{r}
plot_gene_beehive(dds, rep_inv_rest$annotated_results[rep_inv_rest$annotated_results$ensembl_gene_id=="ENSG00000143194",])
```

# Diffex: Pregnancy vs rest

```{r}
dds$group = dds$prbc_vs_rest
vsd$study_group = vsd$prbc_vs_rest
levels(dds$group)
```

Perform differential expression analysis.

From the manual: 

Although apeglm cannot be used with contrast, we note that many designs can be easily rearranged such that what was a contrast becomes its own coefficient. In this case, the dispersion does not have to be estimated again, as the designs are equivalent, up to the meaning of the coefficients. Instead, one need only run nbinomWaldTest to re-estimate MLE coefficients – these are necessary for apeglm – and then run lfcShrink specifying the coefficient of interest in resultsNames(dds). We give some examples below of producing equivalent designs for use with coef. We show how the coefficients change with model.matrix, but the user would, for example, either change the levels of dds$condition or replace the design using design(dds)<-, then run nbinomWaldTest followed by lfcShrink.

```{r, eval=F}
start = Sys.time()
dds <- nbinomWaldTest(dds)
end = Sys.time()
end-start

saveRDS(dds,here("data/Rds/08_dds_prbc_vs_rest_standard.Rds"))
```

>found results columns, replacing these
Time difference of 49.49288 mins

```{r}
dds = readRDS(here("data/Rds/08_dds_prbc_vs_rest_standard.Rds"))
```


## Extract results

```{r, eval=F}
ape_prbc_rest = shrinkRes(dds,  contrast=c("group", "prbc", "rest"))

saveRDS(ape_prbc_rest, here("data","Rds", "08_ape_prbc_rest.Rds"))
```

```{r}
ape_prbc_rest = readRDS(here("data","Rds","08_ape_prbc_rest.Rds"))
```


## Summary report

A log2 fold change cutoff of 0.5 is equivalent to 1.414.

```{r}
rep_prbc_rest = deseq_report(ape_prbc_rest, dds=dds, absl2fc = 0.5, groups = c("prbc", "rest"),
                             title="Pregnancy vs rest", beehive_groups = "all")
```

## Top genes

There's a noticeable chemokine/innate immune enrichment among the top hits.

```{r}
rep_prbc_rest$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_prbc_rest$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_prbc_rest$heatmap
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
 rep_prbc_rest$pathways %>% bind_rows() %>%
  filter(fdr < 0.05)%>% select(pathway,everything())%>% arrange(fdr)
```


```{r, fig.height=6, fig.width=8}
rep_prbc_rest$pathway_plots
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_prbc_rest$beehive_plots
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r}
rep_prbc_rest$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold:

```{r}
rep_prbc_rest$outlier_bees
```

# Diffex: Lac vs rest

```{r}
dds$group = dds$lac_vs_rest
vsd$study_group = vsd$lac_vs_rest
levels(dds$group)
```

Perform differential expression analysis.

From the manual: 

Although apeglm cannot be used with contrast, we note that many designs can be easily rearranged such that what was a contrast becomes its own coefficient. In this case, the dispersion does not have to be estimated again, as the designs are equivalent, up to the meaning of the coefficients. Instead, one need only run nbinomWaldTest to re-estimate MLE coefficients – these are necessary for apeglm – and then run lfcShrink specifying the coefficient of interest in resultsNames(dds). We give some examples below of producing equivalent designs for use with coef. We show how the coefficients change with model.matrix, but the user would, for example, either change the levels of dds$condition or replace the design using design(dds)<-, then run nbinomWaldTest followed by lfcShrink.

```{r, eval=F}
start = Sys.time()
dds <- nbinomWaldTest(dds)
end = Sys.time()
end-start

saveRDS(dds,here("data/Rds/08_dds_lac_vs_rest_standard.Rds"))
```

>found results columns, replacing these
1 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest
Time difference of 22.55606 mins

```{r}
dds = readRDS(here("data/Rds/08_dds_lac_vs_rest_standard.Rds"))
```


## Extract results

```{r, eval=F}
ape_lac_rest = shrinkRes(dds,  contrast=c("group", "ppbc_lac", "rest"))
saveRDS(ape_lac_rest, here("data", "Rds", "08_ape_lac_rest.Rds"))
```

```{r}
ape_lac_rest = readRDS(here("data", "Rds", "08_ape_lac_rest.Rds"))
```


## Summary report

A log2 fold change cutoff of 0.5 is equivalent to 1.414.

```{r}
rep_lac_rest = deseq_report(ape_lac_rest, dds=dds, absl2fc = 0.5, groups = c("ppbc_lac", "rest"),
                            title="Lactation vs rest", beehive_groups = "all")
```

## Top genes

```{r}
rep_lac_rest$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_lac_rest$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_lac_rest$heatmap
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
 rep_lac_rest$pathways %>% bind_rows() %>%
  filter(fdr < 0.05)%>% select(pathway,everything())%>%
  arrange(fdr)
```


```{r, fig.height=6, fig.width=8}
rep_lac_rest$pathway_plots
```
## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_lac_rest$beehive_plots
```
## Cook's outliers

Have a look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r}
rep_lac_rest$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold:

```{r}
rep_lac_rest$outlier_bees
```


# Diffex: Non-prbc vs rest

```{r}
dds$group = dds$nonprbc_vs_rest
vsd$study_group = vsd$nonprbc_vs_rest
levels(dds$group)
```

Perform differential expression analysis.

From the manual: 

Although apeglm cannot be used with contrast, we note that many designs can be easily rearranged such that what was a contrast becomes its own coefficient. In this case, the dispersion does not have to be estimated again, as the designs are equivalent, up to the meaning of the coefficients. Instead, one need only run nbinomWaldTest to re-estimate MLE coefficients – these are necessary for apeglm – and then run lfcShrink specifying the coefficient of interest in resultsNames(dds). We give some examples below of producing equivalent designs for use with coef. We show how the coefficients change with model.matrix, but the user would, for example, either change the levels of dds$condition or replace the design using design(dds)<-, then run nbinomWaldTest followed by lfcShrink.

```{r, eval=F}
start = Sys.time()
dds <- nbinomWaldTest(dds)
end = Sys.time()
end-start

saveRDS(dds,here("data/Rds/08_dds_nonprbc_vs_rest_standard.Rds"))
```

>found results columns, replacing these
2 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest
Time difference of 15.83548 mins

```{r}
dds = readRDS(here("data/Rds/08_dds_nonprbc_vs_rest_standard.Rds"))
```


## Extract results

```{r, eval=F}
ape_nonprbc_rest = shrinkRes(dds,  contrast=c("group", "non_prbc", "rest"))

saveRDS(ape_nonprbc_rest, here("data", "Rds", "08_ape_nonprbc_rest.Rds"))
```

```{r}
ape_nonprbc_rest = readRDS(here("data", "Rds", "08_ape_nonprbc_rest.Rds"))
```


## Summary report

A log2 fold change cutoff of 0.5 is equivalent to 1.414.

```{r}
rep_nonprbc_rest = deseq_report(ape_nonprbc_rest, dds=dds, absl2fc = 0.5, groups = c("non_prbc", "rest"),
                                title = "Nulliparous vs rest", beehive_groups = "all")
```



## Top genes

```{r}
rep_nonprbc_rest$significant_genes %>% head(30)
```

## Volcano plot

```{r}
rep_nonprbc_rest$volcano_plot
```

## Heatmap

```{r, fig.width=10}
rep_nonprbc_rest$heatmap
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
 rep_prbc_rest$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>% arrange(fdr)
```


```{r, fig.height=6, fig.width=8}
rep_nonprbc_rest$pathway_plots
```

## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_nonprbc_rest$beehive_plots
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r}
rep_nonprbc_rest$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold:

```{r}
rep_nonprbc_rest$outlier_bees
```

# Save data

```{r}
res_list = mget(ls(pattern="rep_"))
names(res_list)
```

## Significance of genes

Save the significant genes and results from each comparison.

```{r}
#Save the results for all genes in an Excel file with multiple tabs

#The tab names will be the list names
resdf = lapply(res_list, function(x) x$annotated_results)
names(resdf) = paste(names(resdf),"all", sep="_")
openxlsx::write.xlsx(resdf, file = here("results", "diffex", "08_one_vs_rest_allgenes.xlsx"))

#Save the significant genes in a multi-tab excel file
resdf = lapply(res_list, function(x) x$significant_genes)
names(resdf) = paste("sig",names(resdf), sep="_")
#Also keep the thresholds used for significance
resdf[[length(resdf)+1]] = enframe(unlist(lapply(res_list, function(x) x$sig_threshold)), "name", "value") %>% #signficance and l2fc
                 separate(name, into = c("comparison", "threshold"), sep="\\.") %>%
  bind_rows(.,enframe(unlist(lapply(res_list, function(x) x$cooks_threshold)), #Cooks
                      "comparison", "value") %>% mutate(threshold = "cooks"))
names(resdf)[length(resdf)] = "thresholds"
write.xlsx(resdf, file = here("results", "diffex", "08_one_vs_rest_sig_genes.xlsx"))
rm(resdf)
```

## Pathway results

Fisher's exact pathway analyses

```{r}

resdf = lapply(res_list, function(x) x$pathways)

#Is there a way to extract a dynamic column name with bind rows for two levels of a nested df list?
#
#resdf %>% map(1) %>% bind_rows( .id = "comp") #will work but loses the gene signature name

#Returns one giant dataframe:
#pathlist = resdf %>% map(1) %>% bind_rows( .id = "comp")
#for (i in 2:length(resdf)){
#  path = resdf %>% map(i) %>% bind_rows( .id = "comp")
#  pathlist = bind_rows(pathlist, path)
#}

#do.call(Map, c(f = bind_rows, resdf)) #Not quite. Binds at second layer.

#Returns a list of aggregated path results per comparison
pathlist = list()
pathlist[[names(resdf)[1]]] = resdf %>% map(1, .depth=1) %>% bind_rows( .id = "comp")
for (i in 2:length(resdf[[1]])){
  path = resdf %>% map(i, .depth=1) %>% bind_rows( .id = "comp")
  pathlist[[names(resdf)[i]]] = path
}

#Fix names
names(pathlist) = lapply(pathlist, function(x) unique(x$collection)) %>% unlist(use.names = F)
pathlist = lapply(pathlist, function(x) x[order(x$fdr),])

openxlsx::write.xlsx(pathlist, file = here("results", "diffex", "08_one_vs_rest_pathways.xlsx"))

rm(resdf)
```



## Heatmaps

```{r}
hmlist = lapply(res_list, function(x) x$heatmap)
names(hmlist) = names(hmlist) %>% str_replace_all("rep", "hm")

dir.create(here("results","diffex", "figs", "08_onevsrest" ,"heatmaps"), showWarnings = F, recursive = T)
for (i in 1:length(hmlist)){
  pdf(here("results", "diffex", "figs", "08_onevsrest", "heatmaps", paste0(names(hmlist)[i],".pdf")), width = 10, height = 8)
  print(hmlist[[i]])
  dev.off()
}

rm(hm_list)
```

## Volcano plot sig hits

```{r}
volclist = lapply(res_list, function(x) x$volcano_plot)
names(volclist) = names(volclist) %>% str_replace("rep", "volc")

dir.create(here("results","diffex", "figs", "08_onevsrest" ,"volcano_plots"), showWarnings = F)
lapply(names(volclist), 
       function(x) ggsave(filename=here("results","diffex", "figs", "08_onevsrest","volcano_plots",
                                        paste(x,".jpeg",sep="")), plot=volclist[[x]],
                          height = 6, width = 8))
rm(volclist)
```

## Volcano plot outliers

```{r}
outvplist = lapply(res_list, function(x) x$volc_outliers)
names(outvplist) = names(outvplist) %>% str_replace("rep", "volc")

dir.create(here("results","diffex", "figs", "08_onevsrest" ,"volcano_plots"), showWarnings = F)
lapply(names(outvplist), 
       function(x) ggsave(filename=here("results","diffex", "figs", "08_onevsrest","volcano_plots",
                                        paste(x,".outliers.jpeg",sep="")), plot=outvplist[[x]],
                          height = 6, width = 8))
rm(outvplist)
```

## Pathway dotplots 

```{r}
dir.create(here("results","diffex", "figs", "08_onevsrest" ,"fisher_pathway"), showWarnings = F)
pathwaylist = lapply(res_list, function(x) x$pathway_plots)
for(i in 1:length(pathwaylist)){
  thiscomp = pathwaylist[[i]]
  compname = str_replace(str_remove(names(pathwaylist[i]), "rep_"),"_","_vs_")
  for (j in 1:length(thiscomp)){
    thissig = thiscomp[[j]]
    file=paste0(paste(compname, names(thiscomp[j]), sep="-"),".jpeg")
    #print(file)
    ggsave(filename=here("results","diffex", "figs",
                         "08_onevsrest","fisher_pathway",file),
           plot=thissig, height = 6, width = 8)
  }
}


rm(pathwaylist)
```

## Notebook data.

```{r}
#These objects are huge
rm(res_list)
rm(gene_sets)
save.image(here("reports/08_diffex_onevsrest_standard.RData"))
```

