---
title: "Genewise reports"
output: html_notebook
---

```{r, include=F}
library(DESeq2)
library(tidyverse)
library(here)
```

# Load data

Count matrix from LRT. Counts will be the same between all comparisons.

```{r}
dds = readRDS(here("data", "Rds","06_ddsLRT_standard.Rds"))
```

Gene annotation

```{r}
gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
gx_annot = gx_annot %>% select(ensembl_gene_id = gene_id, gene_name, gene_type, description = gene_description) %>% distinct()
```


```{r}
source(here("src", "deseq_report_functions.R"))
```

Likelihood ratio tests, pairwise comparisons, and one vs rest.

For the LRT, significant and all genes are kept in the same Excel output. Pairwise and one vs rest comparisons report significant genes separately. Here, we load all genes so that we can plot expression regardless of significance within a given comparison.

```{r}
lrt <- here("results", "diffex", "06_LRT.xlsx") %>% 
  readxl::excel_sheets() %>% 
  set_names() %>% 
  map(readxl::read_excel, path = here("results", "diffex", "06_LRT.xlsx"))
names(lrt)

pw <- here("results", "diffex", "07_pairwise_comparisons_allgenes.xlsx") %>% 
  readxl::excel_sheets() %>% 
  set_names() %>% 
  map(readxl::read_excel, path = here("results", "diffex", "07_pairwise_comparisons_allgenes.xlsx"))
names(pw)

ovr = here("results", "diffex", "08_one_vs_rest_allgenes.xlsx") %>% 
  readxl::excel_sheets() %>% 
  set_names() %>% 
  map(readxl::read_excel, path = here("results", "diffex", "08_one_vs_rest_allgenes.xlsx"))
names(ovr)

```

# Process and combine

Substitute lrt down to all genes and rename it for convenience.

```{r}
lrt = lrt["all_genes"]

names(lrt) = "LRT"
```

Combine all results into single list.

```{r}
res_list = append(lrt, pw)
res_list = append(res_list, ovr)
names(res_list) = str_remove_all(str_remove_all(names(res_list), "rep_"), "_all")
names(res_list) = str_replace(names(res_list), "_", "_vs_")
names(res_list)
```



# Differentially expressed genes

## Inv vs rest

### [JChain](https://www.genecards.org/cgi-bin/carddisp.pl?gene=JCHAIN&keywords=jchain)

Frequently reoccuring significant hit (inv vs rest)

>Serves to link two monomer units of either IgM or IgA. In the case of IgM, the J chain-joined dimer is a nucleating unit for the IgM pentamer, and in the case of IgA it induces larger polymers. It also help to bind these immunoglobulins to secretory component.

```{r}
gene_report(res_list,"ENSG00000132465")
```

```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000132465",])
```

### [POTEC](https://www.genecards.org/cgi-bin/carddisp.pl?gene=POTEC&keywords=potec)

Inv vs rest significant hit. Part of the the cancer/testis (CT) antigen family.

Cancer/testis (CT) genes represent a unique class of genes, which are expressed by germ cells, normally silenced in somatic cells, but activated in various cancers. We have previously reported that CTs are relatively commonly expressed in estrogen receptor (ER) negative, high risk carcinomas. In this study, we examined the expression of selected CT genes in ductal carcinoma in situ (DCIS), lobular carcinoma in situ (LCIS) and benign proliferative lesions of the breast. ER negative DCIS were found to be associated with significant CT gene expression together with HER2 positivity and a marked stromal immune response. See also [this article](https://www.ncbi.nlm.nih.gov/pubmed/19719775/).

```{r}
gene_report(res_list, "ENSG00000183206")
```

```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000183206",])
```

Is POTEC more expressed in ER- breast cancers?
ESR1 = ENSG00000091831
ESR2 = ENSG00000140009

```{r}
dds = readRDS(here("data/Rds/08_dds_inv_vs_rest_standard.Rds"))
```

Here the opposite appears to be true : POTEC is more highly expressed in ESR postiive tumors.

```{r}
rbind(
  plotCounts(dds, gene="ENSG00000091831", intgroup=c("study_group","PAM50"), normalize=T, returnData=TRUE) %>%
    rownames_to_column("sample_name") %>% mutate(ensembl_gene_id = "ENSG00000091831", gene_name = "ESR1"),
  #Expression of ESR2 is quite low
  #plotCounts(dds, gene="ENSG00000140009", intgroup=c("study_group","PAM50"), normalize=T, returnData=TRUE) %>%
   # rownames_to_column("sample_name") %>% mutate(ensembl_gene_id = "ENSG00000183206", gene_name = "ESR2"),
  plotCounts(dds, gene="ENSG00000183206", intgroup=c("study_group","PAM50"), normalize=T, returnData=TRUE) %>%
    rownames_to_column("sample_name") %>% mutate(ensembl_gene_id = "ENSG00000183206", gene_name = "POTEC")
  ) %>% select(-ensembl_gene_id) %>%
  spread(key = gene_name, value=count) %>%
  ggplot(aes(x = ESR1, y= POTEC, color=study_group, shape=PAM50)) + geom_jitter(height=0) +
  scale_y_log10() + scale_x_log10() +
  ggtitle("Relationship between POTEC and ESR1, study group")

rbind(
  plotCounts(dds, gene="ENSG00000091831", intgroup=c("study_group","PAM50"), normalize=T, returnData=TRUE) %>%
    rownames_to_column("sample_name") %>% mutate(ensembl_gene_id = "ENSG00000091831", gene_name = "ESR1"),
  plotCounts(dds, gene="ENSG00000183206", intgroup=c("study_group","PAM50"), normalize=T, returnData=TRUE) %>%
    rownames_to_column("sample_name") %>% mutate(ensembl_gene_id = "ENSG00000183206", gene_name = "POTEC")
  ) %>% select(-ensembl_gene_id) %>%
  spread(key = gene_name, value=count) %>%
  ggplot(aes(x = ESR1, y= POTEC, color=PAM50, shape=study_group)) + geom_jitter(height=0) +
  scale_y_log10() + scale_x_log10() +
  ggtitle("Relationship between POTEC and ESR1, PAM50")
```

### [POTED](https://www.genecards.org/cgi-bin/carddisp.pl?gene=POTED&keywords=poted)

As POTEC above.

```{r}
gene_report(res_list, "ENSG00000166351")
```

```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000166351",])
```

```{r}
rbind(
  plotCounts(dds, gene="ENSG00000091831", intgroup=c("study_group","PAM50"), normalize=T, returnData=TRUE) %>%
    rownames_to_column("sample_name") %>% mutate(ensembl_gene_id = "ENSG00000091831", gene_name = "ESR1"),
  #Expression of ESR2 is quite low
  #plotCounts(dds, gene="ENSG00000140009", intgroup=c("study_group","PAM50"), normalize=T, returnData=TRUE) %>%
   # rownames_to_column("sample_name") %>% mutate(ensembl_gene_id = "ENSG00000183206", gene_name = "ESR2"),
  plotCounts(dds, gene="ENSG00000166351", intgroup=c("study_group","PAM50"), normalize=T, returnData=TRUE) %>%
    rownames_to_column("sample_name") %>% mutate(ensembl_gene_id = "ENSG00000166351", gene_name = "POTED")
  ) %>% select(-ensembl_gene_id) %>%
  spread(key = gene_name, value=count) %>%
  ggplot(aes(x = ESR1, y= POTED, color=study_group, shape=PAM50)) + geom_jitter(height=0) +
  scale_y_log10() + scale_x_log10() +
  ggtitle("Relationship between POTED and ESR1, study group")

rbind(
  plotCounts(dds, gene="ENSG00000091831", intgroup=c("study_group","PAM50"), normalize=T, returnData=TRUE) %>%
    rownames_to_column("sample_name") %>% mutate(ensembl_gene_id = "ENSG00000091831", gene_name = "ESR1"),
  plotCounts(dds, gene="ENSG00000166351", intgroup=c("study_group","PAM50"), normalize=T, returnData=TRUE) %>%
    rownames_to_column("sample_name") %>% mutate(ensembl_gene_id = "ENSG00000166351", gene_name = "POTED")
  ) %>% select(-ensembl_gene_id) %>%
  spread(key = gene_name, value=count) %>%
  ggplot(aes(x = ESR1, y= POTED, color=PAM50, shape=study_group)) + geom_jitter(height=0) +
  scale_y_log10() + scale_x_log10() +
  ggtitle("Relationship between POTED and ESR1, PAM50")
```

### [CXADRP3](https://www.genecards.org/cgi-bin/carddisp.pl?gene=CXADRP3)

Inv vs rest significant hit. One of many genes typically expressed in the prostate or testis that is surprisingly found to be differentially expressed in ppbc.

Coxsackie Virus And Adenovirus Receptor Pseudogene 3 

```{r}
gene_report(res_list, "ENSG00000265766")
```

```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000265766",])
```

### [ESR2](https://www.genecards.org/cgi-bin/carddisp.pl?gene=ESR2&keywords=ESR2)

Estrogen receptor 2. Lowly expressed, but significant in inv vs rest.

Oddly, shows a similar expression pattern as BCMA, but is reported as much more significant. Possibly due to the presence of an outlier in the involution group?

```{r}
gene_report(res_list,"ENSG00000140009")
```

```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000140009",])
```

## LRT

### [LABLA](https://www.genecards.org/cgi-bin/carddisp.pl?gene=LALBA&keywords=lalba)

Common milk protein, upregulated in pregnancy. Sometimes subject to Cook's distance filtering due to extreme outliers. Technically not diffex in LRT due to Cook's distance, but would be if that feature is disabled.

```{r}
gene_report(res_list, "ENSG00000167531")
```

```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000167531",])
```
### [GRN](https://www.genecards.org/cgi-bin/carddisp.pl?gene=GRN&keywords=GRN) 

Granulin precursor, mostly upregulated in pregnancy.

>Secreted proteins that act as key regulator of lysosomal function and as a growth factor involved in inflammation, wound healing and cell proliferation (PubMed:28541286, PubMed:28073925, PubMed:18378771, PubMed:28453791, PubMed:12526812). Functions as regulator of proteins trafficking to lysosome and activity of lysosomal enzymes (PubMed:28453791, PubMed:28541286). Facilitates also the acidification of lysosomes, causing degradation of mature CTSD by CTSB (PubMed:28073925). In addition, functions as wound-related growth factor that acts directly on dermal fibroblasts and endothelial cells to promote division, migration and the formation of capillary-like tubule structure (By similarity). Also promotes epithelial cell proliferation by blocking TNF-mediated neutrophil activation preventing release of oxidants and proteases (PubMed:12526812). Moreover, modulates inflammation in neuron by preserving neuron survival, axonal outgrowth and neuronal integrity (PubMed:18378771).

```{r}
gene_report(res_list, "ENSG00000030582")
```


```{r}
plot_gene_beehive(dds, result_df = res_list$LRT[res_list$LRT$ensembl_gene_id == "ENSG00000030582",])
```

### [IL18R1](https://www.genecards.org/cgi-bin/carddisp.pl?gene=IL18R1&keywords=ENSG00000115604)

Interleukin 18 receptor 1. Downregulated in pregnancy. Roughly equivalent in np and inv.

>FN-alpha and IL12 are reported to induce the expression of this receptor in NK and T cells. 

```{r}
gene_report(res_list, "ENSG00000115604")
```

```{r}
plot_gene_beehive(dds, result_df = res_list$LRT[res_list$LRT$ensembl_gene_id == "ENSG00000115604",])
```
# Immune cell markers

## B cells markers 

### [CD20](https://www.genecards.org/cgi-bin/carddisp.pl?gene=MS4A1&keywords=CD20)

Official name of CD20 is MS4A1 ENSG00000156738

```{r}
gene_report(res_list,"ENSG00000156738")
```

Appears as though there might be a difference in inv/nonprbc vs prbc? Basals in particular seem to have less CD20 in prbc than in inv/nonprbc.

```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000156738",])
```

### [IL10](https://www.genecards.org/cgi-bin/carddisp.pl?gene=IL10&keywords=IL-10)

B-reg marker IL10: ENSG00000136634

```{r}
gene_report(res_list,"ENSG00000136634")
```

```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000136634",])
```

### [SIRPA](https://www.genecards.org/cgi-bin/carddisp.pl?gene=SIRPA&keywords=SIRPA)

Previously significant in pregnancy vs rest, above threshold after removing tumor purity from design formula.

```{r}
gene_report(res_list,"ENSG00000198053")
```

```{r}
plot_gene_beehive(dds, res_list$prbc_vs_nonprbc[res_list$prbc_vs_nonprbc$ensembl_gene_id=="ENSG00000198053",])
```

Other plasma cells markers: CD138, CD98, TACI, BCMA
https://www.ncbi.nlm.nih.gov/pubmed/28787106

### [CD138](https://www.genecards.org/cgi-bin/carddisp.pl?gene=SDC1&keywords=CD138)

CD138 = SDC1 syndecan 1
[Plasma cell marker](https://www.ncbi.nlm.nih.gov/pubmed/28787106)

```{r}
gene_report(res_list,"ENSG00000115884")
```

```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000115884",])
```

### [CD98](https://www.genecards.org/cgi-bin/carddisp.pl?gene=SLC3A2&keywords=SLC3A2)

Official name SLC3A2 ENSG00000168003

>CD98 is an important amino acid transporter, also termed LAT1, which consists of heavy (SLC3A2) and light (SLC7A5) subunits. The increased expression of CD98 is most likely a reflection of the high metabolic requirements of ASCs (Antibody secreting cells).


```{r}
gene_report(res_list,"ENSG00000168003")
```

```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000168003",])
```

### *[TACI](https://www.genecards.org/cgi-bin/carddisp.pl?gene=TNFRSF13B&keywords=taci)

Official name TNFRSF13B ENSG00000240505

```{r}
gene_report(res_list,"ENSG00000240505")
```

```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000240505",])
```

### *[BCMA](https://www.genecards.org/cgi-bin/carddisp.pl?gene=TNFRSF17&keywords=BCMA)

Official name TNFRSF17 ENSG00000048462

A good example of a gene which "looks" significant but is reported with a lower log2fc and higher p value than the graph would indicate. Similar to ESR2, which was reported with both higher fold changes and lower padj.

```{r}
gene_report(res_list,"ENSG00000048462")
```

```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000048462",])
```

## T cell markers

### [CD30](https://www.genecards.org/cgi-bin/carddisp.pl?gene=TNFRSF8&keywords=ENSG00000120949)

TNFRSF8 ENSG00000120949

>The protein encoded by this gene is a member of the TNF-receptor superfamily. This receptor is expressed by activated, but not by resting, T and B cells. TRAF2 and TRAF5 can interact with this receptor, and mediate the signal transduction that leads to the activation of NF-kappaB. This receptor is a positive regulator of apoptosis, and also has been shown to limit the proliferative potential of autoreactive CD8 effector T cells and protect the body against autoimmunity.

```{r}
gene_report(res_list, "ENSG00000120949")
```


```{r}
plot_gene_beehive(dds, result_df = res_list$LRT[res_list$LRT$ensembl_gene_id == "ENSG00000120949",])
```

## Macrophages

### [CD68](https://www.genecards.org/cgi-bin/carddisp.pl?gene=CD68&keywords=cd68)

>CD68 (Cluster of Differentiation 68) is a protein highly expressed by cells in the monocyte lineage (e.g., monocytic phagocytes, osteoclasts), by circulating macrophages, and by tissue macrophages (e.g., Kupffer cells, microglia). This gene encodes a 110-kD transmembrane glycoprotein that is highly expressed by human monocytes and tissue macrophages. It is a member of the lysosomal/endosomal-associated membrane glycoprotein (LAMP) family. The protein primarily localizes to lysosomes and endosomes with a smaller fraction circulating to the cell surface.

Significant padj, but below the l2fc threshold.

```{r}
gene_report(res_list, "ENSG00000129226")
```


```{r}
plot_gene_beehive(dds, result_df = res_list$LRT[res_list$LRT$ensembl_gene_id == "ENSG00000129226",])
```

### [CD163](https://www.genecards.org/cgi-bin/carddisp.pl?gene=CD163&keywords=cd163)

>Compared to CD68 that is commonly used as a pan-macrophage marker (Holness & Simmons, 1993), CD163 is regarded as a specific monocyte/macrophage marker for M2 macrophages (Ambarus et al., 2012; Lau et al., 2004; Qian & Pollard, 2010). The presence of CD163+ macrophages was suggested to have a stronger association with less favorable clinicopathological features than CD68+ macrophages (Medrek, Ponten, Jirstrom, & Leandersson, 2012). 

See [this overview](https://www.sciencedirect.com/topics/immunology-and-microbiology/cd163)

```{r}
gene_report(res_list, "ENSG00000177575")
```


```{r}
plot_gene_beehive(dds, result_df = res_list$LRT[res_list$LRT$ensembl_gene_id == "ENSG00000177575",])
```

# Other genes of interest



## *[NFKB1](https://www.genecards.org/cgi-bin/carddisp.pl?gene=NFKB1&keywords=ENSG00000109320)

>NF-kappa-B is a pleiotropic transcription factor present in almost all cell types and is the endpoint of a series of signal transduction events that are initiated by a vast array of stimuli related to many biological processes such as inflammation, immunity, differentiation, cell growth, tumorigenesis and apoptosis.

Gene appears as though it should be more differentially expressed than it is.

```{r}
gene_report(res_list, "ENSG00000109320")
```

```{r}
plot_gene_beehive(dds, result_df = res_list$LRT[res_list$LRT$ensembl_gene_id == "ENSG00000109320",])
```

In particular, it looks as though this gene should be significant in inv vs rest. Did it maybe get replaced due to Cook's distance outliers?

```{r}
dds = readRDS(here("data/Rds/08_dds_inv_vs_rest_standard.Rds"))

outliers_replaced_inv_vs_rest = mcols(dds)$replace %>% enframe("ensembl_gene_id", "outlier_replaced") %>% filter(outlier_replaced==T) %>%
  left_join(.,select(gx_annot, "ensembl_gene_id", "gene_name", "gene_type"), by = "ensembl_gene_id")

print(paste("Genes with outliers replaced by the trimmed mean: ", nrow(outliers_replaced_inv_vs_rest)))
```

Evidently not...

```{r}
outliers_replaced_inv_vs_rest %>% filter(ensembl_gene_id == "ENSG00000109320")
```


### [FOXP3](https://www.genecards.org/cgi-bin/carddisp.pl?gene=FOXP3&keywords=foxp3)

```{r}
gene_report(res_list,"ENSG00000049768")
```

```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000049768",])
```

### [Androgren receptor AR](https://www.genecards.org/cgi-bin/carddisp.pl?gene=AR&keywords=AR)


```{r}
gene_report(res_list,"ENSG00000049768")
```


```{r}
plot_gene_beehive(dds, res_list$inv_vs_rest[res_list$inv_vs_rest$ensembl_gene_id=="ENSG00000169083",])
```



### [NOS1](https://www.genecards.org/cgi-bin/carddisp.pl?gene=NOS1&keywords=nos1)

Nitric oxide synthase

```{r}
gene_report(res_list, ensembl_id = "ENSG00000089250")
```


```{r}
plot_gene_beehive(dds, result_df = res_list$LRT[res_list$LRT$ensembl_gene_id == "ENSG00000089250",])
```


### [HLA-DQA1](https://www.genecards.org/cgi-bin/carddisp.pl?gene=HLA-DQA1&keywords=HLA-DQA1)

>HLA-DQA1 belongs to the HLA class II alpha chain paralogues.

Reported as significant in inv vs prbc, but shows a bimodal distribution in most groups with several extreme outliers in each.

```{r}
gene_report(res_list, "ENSG00000228284")
```


```{r}
plot_gene_beehive(dds, result_df = res_list$LRT[res_list$LRT$ensembl_gene_id == "ENSG00000228284",])
```

```{r}
save.image(here("reports", "09_genewise_diffex_reports.RData"))
```

