---
title: "PPBC: Survival analyses with interaction terms"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
    toc_depth: 4
    highlight: tango
---

# Introduction

We specifically want to identify genes which are altered in the involution group uniquely. Can we achieve this using a multivariate Cox model containing interaction terms?

PPBC (study group) is a categorical variable with four levels. 

* ppbc_inv - Involution
* ppbc_lac - Lactation
* prbc - Pregnancy
* non_prbc - Nulliparous (reference)

We discovered in notebook 12 that the inclusion of (some) clinical covariates (espcially stage and surgery) is critical. The formula there was:

> survival ~ clinical covariates + gene

PPBC cannot meaningfully be coerced to a numerical/ordinal variable. To incorporate a gene:ppbc interaction term, we will need to incorporate separate terms for each level of PPBC. 

> survival ~ clinical covariates + study group + gene + study group * gene

Or, written out:

> survival ~ clinical covariates + non_prbc + prbc + ppbc_lac + ppbc_inv + gene + gene * non_prbc + gene * prbc + gene * ppbc_lac + gene * ppbc_inv


# Data prep

```{r}
rm(list = ls())
```

```{r, include=F}
library(DESeq2)
library(survival)
library(ggthemes)
library(survminer)
library(pheatmap)
library(RColorBrewer)
library(edgeR)
library(glmnet)
library(openxlsx)
library(here)
library(ggpubr)
library(tidyverse)
```

# Pre-process and load data

```{r}
source(here("src", "deseq_report_functions.R"))
```

## Count and sample data

```{r}
dds = readRDS(here("data/Rds/08_dds_inv_vs_rest_standard.Rds"))
gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
gx_annot = gx_annot %>% select(ensembl_gene_id = gene_id, gene_name, gene_type, description = gene_description) %>% distinct()
```


Variance stabilized transformed dds, for visualizations. Converted to gene names and deduplicated.

```{r}
vsd = readRDS(here("data","Rds","08_vsd_standard_vsrest.Rds"))

vsd = rownames_to_column(as.data.frame(assay(vsd)), "ensembl_gene_id")
vsd = right_join(select(gx_annot, gene_name, ensembl_gene_id), vsd, by = "ensembl_gene_id") %>%
    select(-ensembl_gene_id)
vsd = summarize_expression_duplicate_ids(vsd, id_column = "gene_name", verbose = T)
```

## Gene sets

For Fisher's exact tests

```{r}
gene_sets = list(
  go_bp = flexgsea::read_gmt(here("data","external","gmt","c5.bp.v7.0.symbols.gmt")),
  hallmark = flexgsea::read_gmt(here("data","external","gmt","h.all.v7.0.symbols.gmt")),
  c2_canon = flexgsea::read_gmt(here("data","external","gmt","c2.cp.v7.0.symbols.gmt")),
  c2_cgp = flexgsea::read_gmt(here("data","external","gmt","c2.cgp.v7.0.symbols.gmt"))
)
```

## ImmPort genes

To easily identify genes with a known immune function

```{r}
immune_gene_list <- read_csv(here("data", "external","gene-sets","InnateDB_genes.csv"))
```


## Survival metadata

As described in notebook 12, we have 183 samples with adequate data.

```{r}
metadata = readxl::read_excel(here("data", "metadata", "12_survival_metadata.xlsx"))
nrow(metadata)
```

## Transposed matrix for Cox regressions

Also created in notebook 12.

```{r}
coxdata = readRDS(here("data", "Rds", "12_coxdata.Rds"))
colnames(coxdata)[1:30]
```

## Separate columns per PPBC level

```{r}
levels(coxdata$study_group)
```

`



```{r}
coxdata = coxdata %>%
  mutate(non_prbc = as.numeric(study_group == "non_prbc"),
         prbc = as.numeric(study_group == "prbc"),
         ppbc_lac = as.numeric(study_group == "ppbc_lac"),
         ppbc_inv = as.numeric(study_group == "ppbc_inv")) %>%
  select(sample_name:study_group, non_prbc:ppbc_inv, everything())

head(coxdata)
```

```{r}
colnames(coxdata)[1:30]
#Genes start at position 25
colnames(coxdata)[25]
```


## Heatmap colors

Created in notebook 6

```{r}
all_colors = readRDS(here("data","Rds", "06_heatmap_colors.Rds"))
study_colors = all_colors$study_colors
pam_colors = all_colors$pam_colors
gene_colors = all_colors$gene_colors
shroom_colors = c(gene_colors,"n.s."="gray") #For mushroom plots
```


Define colors for survival groups

```{r}
surv_df <- coxdata %>% 
  select(sample_name,study_group, PAM50,
         months_overall_survival, overall_survival,
         months_to_drs, distant_recurrence) %>%
  mutate(survival_group = case_when(
    overall_survival == 1 & months_overall_survival < 50 ~ "death_50mo",
    overall_survival == 1 & months_overall_survival < 100 ~ "death_100mo",  
    overall_survival == 1 & months_overall_survival < 150 ~ "death_150mo",
    overall_survival == 1 & months_overall_survival < 200 ~ "death_200mo",
    overall_survival == 1 & months_overall_survival < 250 ~ "death_200mo",
    overall_survival == 0 ~ "survived",
  TRUE ~ "no_data"
)) %>%
  mutate(drs_group = case_when(
  distant_recurrence == 1 & months_to_drs < 50 ~ "met_50mo",
  distant_recurrence == 1 & months_to_drs < 100 ~ "met_100mo",  
  distant_recurrence == 1 & months_to_drs < 150 ~ "met_150mo",
  distant_recurrence == 1 & months_to_drs < 200 ~ "met_200mo",
  distant_recurrence == 1 & months_to_drs < 250 ~ "met_200mo",
  distant_recurrence == 0 ~ "no_metastasis",
  TRUE ~ "no_data"
))

rownames(surv_df) = surv_df$sample_name

#head(surv_df)

#ann_df = [,c("study_group", "survival_group"), drop=F]
os_colors = viridis::viridis(length(unique(surv_df$survival_group)))
#unique(sd$survival_group) %>% paste(collapse=", ")

names(os_colors) = c("survived","death_200mo", "death_150mo" , "death_100mo", "death_50mo")
color_grid(os_colors)

drs_colors = viridis::viridis(length(unique(surv_df$drs_group))) 
#Ordered from best to worst
names(drs_colors) =  c("no_metastasis","met_200mo","met_150mo","met_100mo", "met_50mo")
color_grid(drs_colors)
```

## TMM normalization of gene expression

Normally we use vst normalization for visualization, but since we used TMM for the cox regression, we should also use TMM for the visualizations.

```{r}
#The pre-loaded Cox data is already TMM normalized
tmmnorm <- t(coxdata[,25:ncol(coxdata)]) #Genes start at position 25
colnames(tmmnorm) <- coxdata$sample_name
tmmnorm <- rownames_to_column(as.data.frame(tmmnorm), "ensembl_gene_id")
tmmnorm <- as.data.frame(tmmnorm) %>% left_join(.,select(gx_annot, gene_name, ensembl_gene_id), by = "ensembl_gene_id") #%>%
tmmnorm <- tmmnorm %>%  select(-ensembl_gene_id) %>% select(gene_name, everything())
tmmnorm <-  summarize_expression_duplicate_ids(tmmnorm, "gene_name", verbose = T)
saveRDS(tmmnorm, here("data", "Rds", "12b_tmmnorm_geneEx_survival.Rds"))
```

# Functions for interaction term Cox regressions

## Extract results from models

```{r}
extract_interaction_cox_results <- function(cox_models, type = "full", interaction_coef = "ppbc_inv",
                                            anno_df = gx_annot, time, event){
  
  stopifnot(type == "full") #"interaction_only" not yet implemented
  
  
  if(type == "full"){
    gene_results <- lapply(cox_models,
                           function(x){
                             x <- summary(x)
                             
                             #Overall model significance
                             logrank_p <- x$logtest[[3]] #p value logrank test
                             wald_p <- x$waldtest[[3]] #p value wald test
                             lrt_p <- x$sctest[[3]] #p value likelihood ratio test
                             
                             #Save formula
                             coefs <- as.data.frame(x$coefficients)
                             form = paste0(paste0('Surv(time=',time,', event=',event,')~'),
                                           paste0(rownames(coefs), collapse = "+"))
                             
                             #Save gene from formula
                             coefs <- rownames_to_column(coefs, "feature")
                             gene = coefs$feature[str_detect(coefs$feature, "^ENSG")]
                                      
                             #Extract results for interaction coefficient
                                      
                             int <- coefs[coefs$feature==paste0(interaction_coef, ":", gene), , drop=F]
                             colnames(int) <- c("feature", "beta", "HR", "se(beta)", "z", "p.value")
                             int <- int %>% mutate(p.value = signif(p.value, 2),
                                                   beta = signif(beta, 2),
                                                   HR = signif(HR, 2),
                                                   ensembl_gene_id = gene,
                                                   interaction = paste0(interaction_coef, ":gene"),
                                                   formula_type = type,
                                                   logrank_p = signif(logrank_p, 2),
                                                   wald_p = signif(logrank_p, 2),
                                                   lrt_p = signif(logrank_p, 2),
                                                   formula = form)
                          
                             #Do the same for conf intervals
                             conf = rownames_to_column(as.data.frame(x$conf.int), "feature")
                             conf = conf[conf$feature == paste0(interaction_coef, ":", gene), , drop=F]
                             HR.confint.lower <- signif(conf[,"lower .95"], 2)
                             HR.confint.upper <- signif(conf[,"upper .95"],2)
                             int$HR <- paste0(int$HR, " (", 
                                              HR.confint.lower, "-", HR.confint.upper, ")")
                             res<- int %>% select(ensembl_gene_id, beta, HR, p.value, everything())
                             res <- res %>% rename(`HR (95% CI for HR)` = HR)
                             res <- res %>% select(-feature, -`se(beta)`, -z) #redundant
                             res <- as.vector(res[1,,drop=T])
                             
                             return(res)
                             })
    #res_df <- t(as.data.frame(gene_results, check.names = FALSE))
  
    res_df <- bind_rows(gene_results)
    #return(res_df)
    }
  
  df = res_df %>%
    left_join(., anno_df, by="ensembl_gene_id") %>%
      mutate(fdr = p.adjust(p.value, method = "fdr")) %>%
      select(gene_name, fdr, beta,`HR (95% CI for HR)`, p.value, gene_type, description, everything()) %>%
      arrange(fdr, p.value)
  
  return(df)
}
```

## Apply formula per gene

```{r}
genewise_cox_interactions <- function(gene_list, time, event, data = coxdata, type="full",
                                      interaction_coef = "ppbc_inv", show_runtime=T){
  stopifnot(type == "full") #"interaction_only" Not yet implemented 
  stopifnot(time %in% colnames(data))
  stopifnot(event %in% colnames(data))

  start <- Sys.time()
  
  #time = "months_overall_survival"
  #event = "overall_survival"
  #data = coxdata
  study_levels = levels(data$study_group)
  
  clin_cov = 'age_at_diagnosis+year_of_diagnosis+stage+grade+surgery+radiotherapy+hormonetherapy+chemotherapy+herceptin+PAM50+'

  formula = paste0('Surv(time=',time,', event=',event,')~',clin_cov)

  if(type == "full"){
    formula = paste0(formula,paste(study_levels, collapse="+"),"+")
    if(show_runtime){print(formula)}
    gene_formulas <-  sapply(gene_list,
                             function(x)
                               as.formula(paste(formula, paste0(levels(coxdata$study_group),
                                                                "*", x, collapse = "+"))))
    
    gene_models <- lapply(gene_formulas, function(x){coxph(x, data = data)})
    
    res <- extract_interaction_cox_results(gene_models, type=type, interaction_coef = interaction_coef,
                                        time = time, event = event) #Comment out to test
  }
  
  end <- Sys.time()

  if(show_runtime){print(end-start)}
  
  return(res)
}

```

Test on 5 genes

```{r}
test = genewise_cox_interactions(gene_list = colnames(coxdata)[25:30], #Full: colnames(coxdata)[21:ncol(coxdata)],
                                 time = "months_overall_survival",
                                 event = "overall_survival",
                                 type = "full")

test$formula %>% head(1) %>% as.formula()
```




# Overall survival

```{r}
#Results directory
resDir = here("data", "Rds")
dir.create(resDir, showWarnings = F)
stopifnot(file.exists(resDir))
print("Genes start at position 25")
print(colnames(coxdata)[25])
```

Run in script 12b.

```{r, eval=F}
resPath = file.path(resDir, "12b_interaction_full_os.Rds")

if(file.exists(resPath) == F| overwrite == T){
  print("Interaction ppbc_inv:gene for OS, full formula")
  res <- genewise_cox_interactions(gene_list = colnames(coxdata)[25:ncol(coxdata)],
                                   time = "months_overall_survival",
                                   event = "overall_survival",
                                   data = coxdata,
                                   type="full",
                                   interaction_coef = "ppbc_inv",
                                   show_runtime=T)
  saveRDS(res, resPath)
}
```


```{r}
os <- readRDS(file.path(resDir,"12b_interaction_full_os.Rds"))

head(os)
```

Example formula (gene varies by row):

```{r}
as.formula(os$formula[1])
```

Stats reported for coefficient:

```{r}
os$interaction[1]
```


## Total hits
 
```{r}
print(paste("Total number of OS-associated genes with fdr ~ 0:", nrow(os[os$fdr == 0,])))

print(paste("Total number of OS-associated genes with fdr < 0.05:", nrow(os[os$fdr < 0.05,])))

print(paste("Total number of OS-associated genes with fdr < 0.1:", nrow(os[os$fdr < 0.1,])))

print(
  paste(
    "Total number of OS-associated genes with fdr < 0.05 for which the overall model (logrank, Wald, lrt) is also significant:",
    (os %>% filter(fdr < 0.05 & logrank_p <0.05 & wald_p < 0.05 & lrt_p < 0.05) %>% nrow())
         ))
```




## Beta > 0 : Higher gene expression = Likelier death

```{r, collapse=T}
print(paste("Overall survival Beta > 0, fdr < 0.1:",
      nrow(os %>% filter(beta > 0 & fdr < 0.1))))

os %>% filter(beta > 0 & fdr < 0.1) %>%
  pull(gene_name)
```

## Beta < 0 : Lower gene expression = Likelier death


```{r}
print(paste("Overall survival multivariate Beta < 0, fdr < 0.1:",
      nrow(os %>% filter(beta < 0 & fdr < 0.1))))
os %>% filter(beta < 0 & fdr < 0.1) %>%
  pull(gene_name)
```

## Pathway analysis (Fisher's)

Input: Genes with fdr < 0.05
Output: Pathways significant with an fdr of 0.1

```{r}
path_os = fisher_pathways(
  sig_genes = pull(filter(os, fdr < 0.05),gene_name),
      background_genes = os$gene_name,
      list_signatures = gene_sets,
      fdr_cutoff = Inf,
      verbose = F,
      collapse_rows=F)

print("Sig pathways fdr < 0.1: ")

lapply(path_os, function(x) nrow(filter(x, fdr < 0.1)))
```

Top 20 pathways:

```{r, fig.height=6, fig.width=9}
bind_rows(path_os, .id = "gene_set") %>%
  arrange(fdr, pval) %>% head(20)

#Plotting one pathway not worth it
#plot_enrichment(
#        enrich_res = .,
#        fdr = 0.1,
#        max_nchar_path = 40,
#        max_path = 20,
#        title = "Sig pathways fdr < 0.1: Overall Survival Interactions"
#      )
```



```{r, fig.width=8, include=F}
#Mushroom plot

#Doesn't really work because of fdr/pvalue being 0 (infinite scale)

os %>% mutate(Type=case_when( #Order is hierarchical
        str_detect(string=gene_type,pattern="^IG") ~ "immunoglobulin",
        str_detect(string=gene_type,pattern="^TR") ~ "T cell receptor",
        ensembl_gene_id %in% immune_gene_list$ensembl ~ "immune protein coding",
        gene_type == "protein_coding" ~ "protein coding",
        TRUE ~ "other noncoding")) %>%
  mutate(Color = if_else(fdr < 0.05, Type, "n.s.")) %>%
  ggplot(aes(x = beta, y = -log10(fdr), color = Color)) +
  geom_point() +
  scale_color_manual(values=shroom_colors) +
  ylim(c(0, 300))#scale_y_continuous()


```


## Heatmap

### Involution only:

No real pattern here.

```{r, fig.width=8}
ann_df = surv_df[surv_df$study_group=="ppbc_inv", ,drop=F]
ann_df = ann_df[,"survival_group", drop=F]
#Sort by survival group
ann_df = ann_df[order(ann_df$survival_group),,drop=F]

#Match
#ovsd = vsd[,colnames(vsd)[match(rownames(ann_df), colnames(vsd))]]
stopifnot(identical(sort(colnames(tmmnorm)), sort(colnames(tmmnorm))))
otmm = tmmnorm[,colnames(tmmnorm)[match(rownames(ann_df), colnames(tmmnorm))]]

#Quantile breaks
quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(as.matrix(xs), probs = seq(0, 1, length.out = n))
  breaks[!duplicated(breaks)]
}

mat_breaks <- quantile_breaks(scrime::rowScales(otmm), n = 10)

pheatmap::pheatmap(as.matrix(otmm[rownames(otmm) %in% filter(os, fdr < 0.05)$gene_name,]),
         show_colnames = F, show_rownames = F,
         scale="row", annotation_col = ann_df,
         cluster_cols = F,
         #color = viridis::cividis(length(mat_breaks) - 1), 
         color = RColorBrewer::brewer.pal(length(mat_breaks) - 1, "Blues"),
         breaks=mat_breaks,
         annotation_colors = list(survival_group = os_colors),
         main = "Genes associated with OS interaction gene:ppbc_inv (fdr <0.05): Involution samples only")

#Not better without row scaling
#mat_breaks <- quantile_breaks(ovsd, n = 10)

#pheatmap::pheatmap(as.matrix(ovsd[rownames(ovsd) %in% filter(os, fdr < 00.5)$gene_name,]),
#         show_colnames = F, show_rownames = F,
#         scale="none", annotation_col = ann_df,
#         cluster_cols = F,
#         color = viridis::cividis(length(mat_breaks) - 1),
#         breaks=mat_breaks,
#         annotation_colors = list(survival_group = os_colors),
#         main = "Genes associated with OS interaction gene:ppbc_inv (fdr < 0.05)")
```

### All samples

No real pattern here either.

```{r, fig.width= 8}
ann_df = surv_df[,c("sample_name", "study_group", "overall_survival"), drop=F]

#Sort by survival group
ann_df = ann_df %>% arrange(study_group, overall_survival)
ann_df = column_to_rownames(ann_df, "sample_name")

#Match
otmm = tmmnorm[,colnames(tmmnorm)[match(rownames(ann_df), colnames(tmmnorm))]]

#Quantile breaks
quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(as.matrix(xs), probs = seq(0, 1, length.out = n))
  breaks[!duplicated(breaks)]
}

#Two color scale
colfunc <- colorRampPalette(c("white", "steelblue"))

mat_breaks <- quantile_breaks(scrime::rowScales(otmm), n = 10)

pheatmap::pheatmap(otmm[rownames(otmm) %in% filter(os, fdr < 0.05)$gene_name,],
         show_colnames = F, show_rownames = F,
         scale="row", annotation_col = ann_df,
         cluster_cols = F,
         #color = viridis::cividis(length(mat_breaks) - 1), 
         #color = colfunc(length(mat_breaks) - 1),
         color = RColorBrewer::brewer.pal(length(mat_breaks) - 1, "Blues"),
         breaks=mat_breaks,
         annotation_colors = list(study_group = study_colors[-5],
                                  survival_group = os_colors),
         main = "Genes associated with OS interaction gene:ppbc_inv- genewise fdr <0.05") 
```

## Genewise heatmaps

It may be that looking at individual genes makes it more obvious why they are selected as significant for the interaction term ppbc_inv:gene. Select a few and plot them with covariates.

### Up

An up-gene (beta > 0):

```{r}
up_gene <- os %>% filter(beta > 0) %>% arrange(fdr) %>% head(1)

up_gene
```



```{r, fig.width=7, fig.height=14}
#colnames(coxdata)[1:30]
ann_df = coxdata[,c("study_group", "overall_survival", "grade", "stage",
                    "PAM50", "year_of_diagnosis", "age_at_diagnosis",
                    "surgery","radiotherapy","hormonetherapy","chemotherapy",
                    "herceptin","sample_name"), drop=F]
#Sort by survival group and study group
#ann_df = ann_df[order(ann_df$overall_survival),]
ann_df = ann_df %>% arrange(study_group, overall_survival)

rownames(ann_df) <- NULL
ann_df = column_to_rownames(ann_df, "sample_name")

#Match
otmm = tmmnorm[,colnames(tmmnorm)[match(rownames(ann_df), colnames(tmmnorm))]]

#Quantile breaks
quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(as.matrix(xs), probs = seq(0, 1, length.out = n))
  breaks[!duplicated(breaks)]
}

#Two color scale
colfunc <- colorRampPalette(c("white", "steelblue"))

mat_breaks <- quantile_breaks(scrime::rowScales(otmm), n = 10)

pheatmap::pheatmap(t(otmm[rownames(otmm) == up_gene$gene_name,]),
         show_colnames = T, show_rownames = F,
         scale="column", annotation_row = ann_df,
         cluster_cols = F, cluster_rows=F,
         #color = viridis::cividis(length(mat_breaks) - 1), 
         #color = colfunc(length(mat_breaks) - 1),
         color = RColorBrewer::brewer.pal(length(mat_breaks) - 1, "Blues"),
         breaks=mat_breaks,
         annotation_colors = list(study_group = study_colors[-5],
                                  survival_group = os_colors),
         main = paste(up_gene$gene_name, ", fdr: ", up_gene$fdr, ", beta: ", up_gene$beta)) 
```


Here we can see that the expression of DUSP13 is generally higher in involution patients who died than in involution patients who survived. 

A different up gene

```{r}
up_gene <- os %>% filter(beta > 0) %>% arrange(fdr) %>% slice(2)

up_gene
```

```{r, fig.width=7, fig.height=14}
#colnames(coxdata)[1:30]
ann_df = coxdata[,c("study_group", "overall_survival", "grade", "stage",
                    "PAM50", "year_of_diagnosis", "age_at_diagnosis",
                    "surgery","radiotherapy","hormonetherapy","chemotherapy",
                    "herceptin","sample_name"), drop=F]
#Sort by survival group and study group
#ann_df = ann_df[order(ann_df$overall_survival),]
ann_df = ann_df %>% arrange(study_group, overall_survival)

rownames(ann_df) <- NULL
ann_df = column_to_rownames(ann_df, "sample_name")

#Match
otmm = tmmnorm[,colnames(tmmnorm)[match(rownames(ann_df), colnames(tmmnorm))]]

#Quantile breaks
quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(as.matrix(xs), probs = seq(0, 1, length.out = n))
  breaks[!duplicated(breaks)]
}

#Two color scale
colfunc <- colorRampPalette(c("white", "steelblue"))

mat_breaks <- quantile_breaks(scrime::rowScales(otmm), n = 10)

pheatmap::pheatmap(t(otmm[rownames(otmm) == up_gene$gene_name,]),
         show_colnames = T, show_rownames = F,
         scale="column", annotation_row = ann_df,
         cluster_cols = F, cluster_rows=F,
         #color = viridis::cividis(length(mat_breaks) - 1), 
         #color = colfunc(length(mat_breaks) - 1),
         color = RColorBrewer::brewer.pal(length(mat_breaks) - 1, "Blues"),
         breaks=mat_breaks,
         annotation_colors = list(study_group = study_colors[-5],
                                  survival_group = os_colors),
         main = paste(up_gene$gene_name, ", fdr: ", up_gene$fdr, ", beta: ", up_gene$beta)) 
```

### Down

A down-gene (beta < 0):

```{r}
#down_gene <- os %>% filter(beta < 0) %>% arrange(fdr) %>% head(1) #Some pseudogene
#More interesting:

down_gene = os %>% filter(gene_name == "FLT3")

down_gene
```

Here we see that FLT3 is  generally lower expressed in involution patients who died.

```{r, fig.width=7, fig.height=14}
#colnames(coxdata)[1:30]
ann_df = coxdata[,c("study_group", "overall_survival", "grade", "stage",
                    "PAM50", "year_of_diagnosis", "age_at_diagnosis",
                    "surgery","radiotherapy","hormonetherapy","chemotherapy",
                    "herceptin","sample_name"), drop=F]
#Sort by survival group and study group
#ann_df = ann_df[order(ann_df$overall_survival),]
ann_df = ann_df %>% arrange(study_group, overall_survival)

rownames(ann_df) <- NULL
ann_df = column_to_rownames(ann_df, "sample_name")

#Match
otmm = tmmnorm[,colnames(tmmnorm)[match(rownames(ann_df), colnames(tmmnorm))]]

#Quantile breaks
quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(as.matrix(xs), probs = seq(0, 1, length.out = n))
  breaks[!duplicated(breaks)]
}

#Two color scale
colfunc <- colorRampPalette(c("white", "steelblue"))

mat_breaks <- quantile_breaks(scrime::rowScales(otmm), n = 10)

pheatmap::pheatmap(t(otmm[rownames(otmm) == down_gene$gene_name,]),
         show_colnames = T, show_rownames = F,
         scale="column", annotation_row = ann_df,
         cluster_cols = F, cluster_rows=F,
         #color = viridis::cividis(length(mat_breaks) - 1), 
         #color = colfunc(length(mat_breaks) - 1),
         color = RColorBrewer::brewer.pal(length(mat_breaks) - 1, "Blues"),
         breaks=mat_breaks,
         annotation_colors = list(study_group = study_colors[-5],
                                  survival_group = os_colors),
         main = paste(down_gene$gene_name, ", fdr: ", down_gene$fdr, ", beta: ", down_gene$beta)) 
```

## Dot plots

Hopefully these will be easier to read than the heat maps


```{r}
cox_dotplot <- function(gene_id,  cox_results, id_type = "gene_name", cox_events = "os",
                        data = coxdata, gene_annotation = gx_annot){
  
  stopifnot(id_type %in% c("gene_name", "ensembl_gene_id"))
  stopifnot(cox_events %in% c("os", "drs")) 
  
  if(cox_events == "os"){
    event_data = coxdata %>% select(sample_name, study_group, months_overall_survival, overall_survival)
    title_string = "OS for interaction "
    legend_label = "Death"
  } else {
    event_data = coxdata %>% select(sample_name, study_group, months_to_drs, distant_recurrence)
    title_string = "DRS for interaction "
    legend_label = "Metastasis"
  }
  
  event_type = colnames(event_data)[3:4]
  colnames(event_data)[3:4] <- c("time_months", "event")
  
  if(id_type == "gene_name"){
    gene_annot <- gx_annot[gx_annot$gene_name==gene_id,,drop=F]
  } else {
    gene_annot <- gx_annot[gx_annot$ensembl_gene_id==gene_id,,drop=F]
  }
  
  stopifnot(nrow(gene_annot) == 1)
  
  gene_data = coxdata[,c(gene_annot$ensembl_gene_id), drop=F]
  gene_data = cbind(gene_data, event_data)
  
  colnames(gene_data)[1] = "TMM_lognorm_expression"
  gene_data$gene_name = gene_annot$gene_name
  
  gene_results = cox_results[cox_results$ensembl_gene_id == gene_annot$ensembl_gene_id, ]
  stopifnot(nrow(gene_results) == 1)
  
  title_string = paste0(title_string, paste0("(involution*", gene_annot$gene_name), ")\n") #Response type for gene name
  title_string = paste(title_string,
                       paste(paste(colnames(gene_results)[2:4], gene_results[2:4], sep=": "), #FDR, beta and HR
                             collapse = ", "))
  
  gene_data %>%
    mutate(event = as.factor(event),
           study_group = recode(study_group,
                                non_prbc = "Nulliparous",
                                ppbc_inv = "Involution",
                                ppbc_lac = "Lactation",
                                prbc = "Pregnancy",
                                .default = levels(study_group))) %>%
    ggplot(aes(x =time_months, y = TMM_lognorm_expression, color = event)) +
    geom_point() +
    geom_rug()+
    #geom_smooth(method='lm',se=F)+
    facet_wrap(~study_group)+
    scale_color_colorblind() + theme_few() +
    ggtitle(title_string) +
    ylab("log(TMM)") +
    xlab("Time (Months)") +
    labs(color=legend_label)
  
  }
```

### [DUSP13](https://www.genecards.org/cgi-bin/carddisp.pl?gene=DUSP13&keywords=DUSP13)

> Probable protein tyrosine phosphatase. Has phosphatase activity with synthetic substrates (PubMed:15252030, PubMed:29106959). Has a phosphatase activity-independent regulatory role in MAP3K5/ASK1-mediated apoptosis, preventing MAP3K5/ASK1 inhibition by AKT1. Shows no phosphatase activity on MAPK1/ERK2, MAPK8/JNK, MAPK14/p38 and MAP3K5/ASK1


```{r}
cox_dotplot(gene_id = "DUSP13", os)
```

### [FGL1](https://www.genecards.org/cgi-bin/carddisp.pl?gene=FGL1&keywords=FGL1)

>Immune suppressive molecule that inhibits antigen-specific T-cell activation by acting as a major ligand of LAG3 (PubMed:30580966). Responsible for LAG3 T-cell inhibitory function (PubMed:30580966). Binds LAG3 independently from MHC class II (MHC-II) (PubMed:30580966). In addition to its role in inhibiting inflammatory immune responses, also plays a role in hepatocytes (PubMed:11470158, PubMed:19880967). Under normal physiological conditions, primarily secreted from hepatocytes and contributes to its mitogenic and metabolic functions 

```{r}
cox_dotplot(gene_id = "FGL1", os)
```

### Let7 pathway enriched genes


```{r, collapse=T}
bind_rows(path_os) %>% filter(fdr < 0.05) %>% pull(pathway)

bind_rows(path_os) %>% filter(fdr < 0.05) %>% pull(fdr)

l7_genes = bind_rows(path_os) %>% filter(fdr < 0.05) %>% pull(members_in_sample) %>% strsplit(split=";") %>% unlist()

l7_genes
```


```{r}

for (i in l7_genes){
  print(cox_dotplot(i, os))
}

```

#### [FGB](https://www.genecards.org/cgi-bin/carddisp.pl?gene=FGB&keywords=fgb) and [FGG](https://www.genecards.org/cgi-bin/carddisp.pl?gene=FGG&keywords=FGG)

>Together with fibrinogen alpha (FGA) and fibrinogen beta (FGB), polymerizes to form an insoluble fibrin matrix. Has a major function in hemostasis as one of the primary components of blood clots. In addition, functions during the early stages of wound repair to stabilize the lesion and guide cell migration during re-epithelialization. Was originally thought to be essential for platelet aggregation, based on in vitro studies using anticoagulated blood. However, subsequent studies have shown that it is not absolutely required for thrombus formation in vivo. Enhances expression of SELP in activated platelets via an ITGB3-dependent pathway. Maternal fibrinogen is essential for successful pregnancy. Fibrin deposition is also associated with infection, where it protects against IFNG-mediated hemorrhage. May also facilitate the antibacterial immune response via both innate and T-cell mediated pathways. 

FGG and FGB are [known prognostic factors in breast cancer](https://www.ncbi.nlm.nih.gov/pubmed/11460495). See also [this reference](https://onlinelibrary.wiley.com/doi/full/10.1111/j.1538-7836.2007.02808.x).

Also [poor prognostic factors for liver cancer](https://www.proteinatlas.org/ENSG00000171564-FGB/pathology)

#### [MAGEA2](https://www.genecards.org/cgi-bin/carddisp.pl?gene=MAGEA2&keywords=MAGEA2) and [MAGEA6](https://www.genecards.org/cgi-bin/carddisp.pl?gene=MAGEA6&keywords=magea6)

>Reduces p53/TP53 transactivation function through recruitment of HDAC3 to p53/TP53 transcription sites. Also represses p73/TP73 activity. Proposed to enhance ubiquitin ligase activity of RING-type zinc finger-containing E3 ubiquitin-protein ligases. In vitro enhances ubiquitin ligase activity of TRIM28 and stimulates p53/TP53 ubiquitination by TRIM28 potentially in presence of Ubl-conjugating enzyme UBE2H. Proposed to act through recruitment and/or stabilization of the Ubl-conjugating enzyme (E2) at the E3:substrate complex. May play a role in embryonal development and tumor transformation or aspects of tumor progression. In vitro promotes cell viability in melanoma cell lines. Antigen recognized on a melanoma by autologous cytolytic T-lymphocytes. Negatively regulates acetylation and sumoylation of PML and represses PML-induced p53/TP53 acetylation and activation

From: [Identification of MAGEA antigens as causal players in the development of tamoxifen-resistant breast cancer](https://www.nature.com/articles/onc201445)

>The antiestrogen tamoxifen is a well-tolerated, effective treatment for estrogen receptor-α-positive (ER+) breast cancer, but development of resistance eventually limits its use. Here we show that expression of MAGEA2, and related members of this cancer-testis antigen family, is upregulated in tamoxifen-resistant tumor cells. Expression of MAGEA2 in tumor lines grown in vitro or as xenografts led to continued proliferation in the presence of tamoxifen. At the molecular level, we demonstrate that MAGEA2 protein localizes to the nucleus and forms complexes with p53 and ERα, resulting in repression of the p53 pathway but increased ER-dependent signaling. In a series of ER+, tamoxifen-treated breast cancer patients, we show a highly significant (P=0.006) association between MAGEA (melanoma-associated antigen) expression and reduced overall survival, confirming the clinical significance of our observations.

From [MAGE-A Cancer/Testis Antigens Inhibit MDM2 Ubiquitylation Function and Promote Increased Levels of MDM4](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0127713)

>Melanoma antigen A (MAGE-A) proteins comprise a structurally and biochemically similar sub-family of Cancer/Testis antigens that are expressed in many cancer types and are thought to contribute actively to malignancy. MAGE-A proteins are established regulators of certain cancer-associated transcription factors, including p53, and are activators of several RING finger-dependent ubiquitin E3 ligases. Here, we show that MAGE-A2 associates with MDM2, a ubiquitin E3 ligase that mediates ubiquitylation of more than 20 substrates including mainly p53, MDM2 itself, and MDM4, a potent p53 inhibitor and MDM2 partner that is structurally related to MDM2. We find that MAGE-A2 interacts with MDM2 via the N-terminal p53-binding pocket and the RING finger domain of MDM2 that is required for homo/hetero-dimerization and for E2 ligase interaction. Consistent with these data, we show that MAGE-A2 is a potent inhibitor of the E3 ubiquitin ligase activity of MDM2, yet it does not have any significant effect on p53 turnover mediated by MDM2. Strikingly, however, increased MAGE-A2 expression leads to reduced ubiquitylation and increased levels of MDM4. Similarly, silencing of endogenous MAGE-A expression diminishes MDM4 levels in a manner that can be rescued by the proteasomal inhibitor, bortezomid, and permits increased MDM2/MDM4 association. These data suggest that MAGE-A proteins can: (i) uncouple the ubiquitin ligase and degradation functions of MDM2; (ii) act as potent inhibitors of E3 ligase function; and (iii) regulate the turnover of MDM4. We also find an association between the presence of MAGE-A and increased MDM4 levels in primary breast cancer, suggesting that MAGE-A-dependent control of MDM4 levels has relevance to cancer clinically.

See also [this review on the MAGE family and cancer](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4688208/)

#### [KRT85](https://www.genecards.org/cgi-bin/carddisp.pl?gene=KRT85&keywords=krt85)

>The protein encoded by this gene is a member of the keratin gene family. As a type II hair keratin, it is a basic protein which heterodimerizes with type I keratins to form hair and nails. The type II hair keratins are clustered in a region of chromosome 12q13 and are grouped into two distinct subfamilies based on structure similarity. One subfamily, consisting of KRTHB1, KRTHB3, and KRTHB6, is highly related. The other less-related subfamily includes KRTHB2, KRTHB4, and KRTHB5.

Upregulated in [response to ESR1](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4987164/)


### [FGFBP1](https://www.genecards.org/Search/Keyword?queryString=FGFBP)

>Acts as a carrier protein that release fibroblast-binding factors (FGFs) from the extracellular matrix (EM) storage and thus enhance the mitogenic activity of FGFs. Enhances FGF2 signaling during tissue repair, angiogenesis and in tumor growth

```{r}
cox_dotplot("FGFBP1", os)
```


See [Tumor growth and angiogenesis induced by a secreted binding protein for fibroblast growth factors](http://www.jbc.org/content/269/45/28243?ijkey=e0f55f14fb0fdc17575bfca543bea30f92b6d1b3&keytype2=tf_ipsecsha) and [Expression of a Fibroblast Growth Factor–Binding Protein during the Development of Adenocarcinoma of the Pancreas and Colon](https://cancerres.aacrjournals.org/content/66/2/1191#ref-20)

### [CD1B](https://www.genecards.org/cgi-bin/carddisp.pl?gene=CD1B&keywords=cd1b)

> This gene encodes a member of the CD1 family of transmembrane glycoproteins, which are structurally related to the major histocompatibility complex (MHC) proteins and form heterodimers with beta-2-microglobulin. The CD1 proteins mediate the presentation of primarily lipid and glycolipid antigens of self or microbial origin to T cells. The human genome contains five CD1 family genes organized in a cluster on chromosome 1. The CD1 family members are thought to differ in their cellular localization and specificity for particular lipid ligands. The protein encoded by this gene localizes to late endosomes and lysosomes via a tyrosine-based motif in the cytoplasmic tail, and requires vesicular acidification to bind lipid antigens.


```{r}
cox_dotplot("CD1B", os)
```

### [BCAR4](https://www.genecards.org/cgi-bin/carddisp.pl?gene=BCAR4&keywords=bcar4)

>This gene produces a spliced long non-coding RNA (lncRNA) that has been implicated in breast cancer metastasis. It was originally identified in a screen for genes responsible for the development of resistance to anti-estrogens in breast cancer cells. It is thought that release of CCL21 enables this lncRNA to bind to the SNIP1 and PNUTS transcription factors, thereby activating a non-canonical GLI-dependent hedgehog signaling pathway that promotes cancer cell migration and invasion. A similar gene in cow expresses a protein in mature oocytes and preimplantation embryos


```{r}
cox_dotplot("BCAR4", os)
```

### [CCL21](https://www.genecards.org/cgi-bin/carddisp.pl?gene=CCL21&keywords=ccl7)

>This antimicrobial gene is one of several CC cytokine genes clustered on the p-arm of chromosome 9. Cytokines are a family of secreted proteins involved in immunoregulatory and inflammatory processes. The CC cytokines are proteins characterized by two adjacent cysteines. Similar to other chemokines the protein encoded by this gene inhibits hemopoiesis and stimulates chemotaxis. This protein is chemotactic in vitro for thymocytes and activated T cells, but not for B cells, macrophages, or neutrophils. The cytokine encoded by this gene may also play a role in mediating homing of lymphocytes to secondary lymphoid organs. It is a high affinity functional ligand for chemokine receptor 7 that is expressed on T and B lymphocytes and a known receptor for another member of the cytokine family (small inducible cytokine A19).

```{r}
cox_dotplot("CCL21", os)
```

### [MMP2](https://www.genecards.org/cgi-bin/carddisp.pl?gene=MMP2&keywords=mmp2)

>This gene is a member of the matrix metalloproteinase (MMP) gene family, that are zinc-dependent enzymes capable of cleaving components of the extracellular matrix and molecules involved in signal transduction. The protein encoded by this gene is a gelatinase A, type IV collagenase, that contains three fibronectin type II repeats in its catalytic site that allow binding of denatured type IV and V collagen and elastin. Unlike most MMP family members, activation of this protein can occur on the cell membrane. This enzyme can be activated extracellularly by proteases, or, intracellulary by its S-glutathiolation with no requirement for proteolytical removal of the pro-domain. This protein is thought to be involved in multiple pathways including roles in the nervous system, endometrial menstrual breakdown, regulation of vascularization, and metastasis. Mutations in this gene have been associated with Winchester syndrome and Nodulosis-Arthropathy-Osteolysis (NAO) syndrome. Alternative splicing results in multiple transcript variants encoding different isoforms.

```{r}
cox_dotplot("MMP2", os)
```

### [CCL7](https://www.genecards.org/cgi-bin/carddisp.pl?gene=CCL7&keywords=ccl7)

>This gene encodes monocyte chemotactic protein 3, a secreted chemokine which attracts macrophages during inflammation and metastasis. It is a member of the C-C subfamily of chemokines which are characterized by having two adjacent cysteine residues. The protein is an in vivo substrate of matrix metalloproteinase 2, an enzyme which degrades components of the extracellular matrix. This gene is part of a cluster of C-C chemokine family members on chromosome 17q. [provided by RefSeq, Jul 2008]

```{r}
cox_dotplot("CCL7", os)
```

## [FLT3](https://www.genecards.org/cgi-bin/carddisp.pl?gene=FLT3&keywords=flt3)

>Tyrosine-protein kinase that acts as cell-surface receptor for the cytokine FLT3LG and regulates differentiation, proliferation and survival of hematopoietic progenitor cells and of dendritic cells. Promotes phosphorylation of SHC1 and AKT1, and activation of the downstream effector MTOR. Promotes activation of RAS signaling and phosphorylation of downstream kinases, including MAPK1/ERK2 and/or MAPK3/ERK1. Promotes phosphorylation of FES, FER, PTPN6/SHP, PTPN11/SHP-2, PLCG1, and STAT5A and/or STAT5B. Activation of wild-type FLT3 causes only marginal activation of STAT5A or STAT5B. Mutations that cause constitutive kinase activity promote cell proliferation and resistance to apoptosis via the activation of multiple signaling pathways

## Save data

```{r}

openxlsx::write.xlsx(x = list("os_inv*gene" = os,
                              "fisher_paths" = bind_rows(path_os) %>% arrange(fdr, pval) %>%
                                select(pathway,fdr, pval, everything())),
                    file = here("results", "survival", "12b_os_gene:inv_interaction.xlsx"))
```


# DRS

```{r, eval=F}
resPath = file.path(resDir, "12b_interaction_full_drs.Rds")

if(file.exists(resPath) == F| overwrite == T){
  print("Interaction ppbc_inv:gene for DRS, full formula")
  res <- genewise_cox_interactions(gene_list = colnames(coxdata)[25:ncol(coxdata)],
                                   time = "months_to_drs",
                                   event = "distant_recurrence",
                                   data = coxdata,
                                   type="full",
                                   interaction_coef = "ppbc_inv",
                                   show_runtime=T)
  saveRDS(res, resPath)
}
```

```{r}
drs = readRDS(file.path(resDir, "12b_interaction_full_drs.Rds"))
```

Example formula (gene varies by row):

```{r}
as.formula(drs$formula[1])
```

Stats reported for coefficient:

```{r}
drs$interaction[1]
```


## Total hits
 
```{r}
print(paste("Total number of drs-associated genes with fdr ~ 0:", nrow(drs[drs$fdr == 0,])))

print(paste("Total number of drs-associated genes with fdr < 0.05:", nrow(drs[drs$fdr < 0.05,])))

print(paste("Total number of drs-associated genes with fdr < 0.1:", nrow(drs[drs$fdr < 0.1,])))

print(
  paste(
    "Total number of drs-associated genes with fdr < 0.05 for which the overall model (logrank, Wald, lrt) is also significant:",
    (drs %>% filter(fdr < 0.05 & logrank_p <0.05 & wald_p < 0.05 & lrt_p < 0.05) %>% nrow())
         ))
```


## Beta > 0 : Higher gene expression = Likelier death

```{r, collapse=T}
print(paste("Overall survival Beta > 0, fdr < 0.1:",
      nrow(drs %>% filter(beta > 0 & fdr < 0.1))))

drs %>% filter(beta > 0 & fdr < 0.1) %>%
  pull(gene_name)
```

## Beta < 0 : Lower gene expression = Likelier death


```{r}
print(paste("Overall survival multivariate Beta < 0, fdr < 0.1:",
      nrow(drs %>% filter(beta < 0 & fdr < 0.1))))
drs %>% filter(beta < 0 & fdr < 0.1) %>%
  pull(gene_name)
```

## Pathway analysis (Fisher's)

Input: Genes with fdr < 0.05
Output: Pathways significant with an fdr of 0.1

```{r}
path_drs = fisher_pathways(
  sig_genes = pull(filter(drs, fdr < 0.05),gene_name),
      background_genes = drs$gene_name,
      list_signatures = gene_sets,
      fdr_cutoff = Inf,
      verbose = F,
      collapse_rows=F)

print("Sig pathways fdr < 0.1: ")

lapply(path_drs, function(x) nrow(filter(x, fdr < 0.1)))
```

Top 20 pathways:

```{r, fig.height=6, fig.width=9}
bind_rows(path_drs, .id = "gene_set") %>%
  arrange(fdr, pval) %>% head(20)

#Plotting one pathway not worth it
#plot_enrichment(
#        enrich_res = .,
#        fdr = 0.1,
#        max_nchar_path = 40,
#        max_path = 20,
#        title = "Sig pathways fdr < 0.1: Overall Survival Interactions"
#      )
```


## Heatmap

Involution only:

```{r, fig.width=8}
ann_df = surv_df[surv_df$study_group=="ppbc_inv", ,drop=F]
ann_df = ann_df[,"drs_group", drop=F]
#Sort by drs group
ann_df = ann_df[order(ann_df$drs_group),,drop=F]

#Match
ovsd = vsd[,colnames(vsd)[match(rownames(ann_df), colnames(vsd))]]

#Quantile breaks
quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(as.matrix(xs), probs = seq(0, 1, length.out = n))
  breaks[!duplicated(breaks)]
}

mat_breaks <- quantile_breaks(scrime::rowScales(ovsd), n = 10)

pheatmap::pheatmap(as.matrix(ovsd[rownames(ovsd) %in% filter(drs, fdr < 0.05)$gene_name,]),
         show_colnames = F, show_rownames = T,
         scale="row", annotation_col = ann_df,
         cluster_cols = F,
         #color = viridis::cividis(length(mat_breaks) - 1), 
         color = RColorBrewer::brewer.pal(length(mat_breaks) - 1, "Blues"),
         breaks=mat_breaks,
         annotation_colors = list(drs_group = drs_colors),
         main = "Genes associated with DRS interaction gene:ppbc_inv (fdr <0.05)")

#Not better without row scaling
#mat_breaks <- quantile_breaks(ovsd, n = 10)

#pheatmap::pheatmap(as.matrix(ovsd[rownames(ovsd) %in% filter(os, fdr < 00.5)$gene_name,]),
#         show_colnames = F, show_rownames = F,
#         scale="none", annotation_col = ann_df,
#         cluster_cols = F,
#         color = viridis::cividis(length(mat_breaks) - 1),
#         breaks=mat_breaks,
#         annotation_colors = list(survival_group = os_colors),
#         main = "Genes associated with OS interaction gene:ppbc_inv (fdr < 0.05)")
```

All samples

```{r, fig.width= 8}
ann_df = surv_df[,c("study_group", "drs_group"), drop=F]
#Sort by drs group
ann_df = ann_df[order(ann_df$drs_group),]

#Match
ovsd = vsd[,colnames(vsd)[match(rownames(ann_df), colnames(vsd))]]

#Quantile breaks
quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(as.matrix(xs), probs = seq(0, 1, length.out = n))
  breaks[!duplicated(breaks)]
}

#Two color scale
colfunc <- colorRampPalette(c("white", "steelblue"))

mat_breaks <- quantile_breaks(scrime::rowScales(ovsd), n = 10)

pheatmap::pheatmap(ovsd[rownames(ovsd) %in% filter(drs, fdr < 0.05)$gene_name,],
         show_colnames = F, show_rownames = T,
         scale="row", annotation_col = ann_df,
         cluster_cols = F,
         #color = viridis::cividis(length(mat_breaks) - 1), 
         #color = colfunc(length(mat_breaks) - 1),
         color = RColorBrewer::brewer.pal(length(mat_breaks) - 1, "Blues"),
         breaks=mat_breaks,
         annotation_colors = list(study_group = study_colors[-5],
                                  drs_group = drs_colors),
         main = "Genes associated with DRS interaction gene:ppbc_inv (fdr <0.05)") 
```
## Save data

```{r}
openxlsx::write.xlsx(x = list("drs_inv*gene" = drs,
                              "fisher_paths" = bind_rows(path_drs) %>% arrange(fdr, pval) %>%
                                select(pathway,fdr, pval, everything())),
                    file = here("results", "survival", "12b_drs_gene:inv_interaction.xlsx"))
```


```{r}
save.image(here("reports", "12b_survival_interaction.RData"))
```






