---
title: "Kaplan Meier plots for gene tertiles: New interaction formula"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
    toc_depth: 4
    highlight: tango
---

```{r}
rm(list = ls())
```

```{r, include=F}
library(DESeq2)
library(survival)
library(ggthemes)
library(survminer)
library(pheatmap)
library(RColorBrewer)
library(edgeR)
library(glmnet)
library(openxlsx)
library(here)
library(ggpubr)
library(tidyverse)
```

# Set up

```{r}
source(here("src", "survival_tools.R"))
```

## Raw counts

```{r}
dds = readRDS(here("data/Rds/08_dds_inv_vs_rest_standard.Rds"))
```


## Gene metadata

```{r}
gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
gx_annot = gx_annot %>%
  select(ensembl_gene_id = gene_id, gene_name, gene_type, description = gene_description) %>%
  distinct() %>%
  mutate(description = str_remove_all(description, " \\[.*\\]"))

head(gx_annot)

saveRDS(gx_annot, here("data","Rds","12e_gx_annot.Rds"))
```


## Survival metadata

As described in notebook 12, we have 183 samples with adequate data.

```{r}
metadata = readxl::read_excel(here("data", "metadata", "12_survival_metadata.xlsx"))
nrow(metadata)
```


## Previous results

### Diffex



```{r, collapse=T}
lrt <- here("results", "diffex", "06_LRT_allgenes.xlsx") %>% 
  readxl::excel_sheets() %>% 
  purrr::set_names() %>% 
  map(readxl::read_excel, path = here("results", "diffex", "06_LRT_allgenes.xlsx"))
names(lrt)

pw <- here("results", "diffex", "07_pairwise_comparisons_allgenes.xlsx") %>% 
  readxl::excel_sheets() %>% 
  purrr::set_names() %>% 
  map(readxl::read_excel, path = here("results", "diffex", "07_pairwise_comparisons_allgenes.xlsx"))
names(pw)

ovr = here("results", "diffex", "08_one_vs_rest_allgenes.xlsx") %>% 
  readxl::excel_sheets() %>% 
  purrr::set_names() %>% 
  map(readxl::read_excel, path = here("results", "diffex", "08_one_vs_rest_allgenes.xlsx"))
names(ovr)

#Substitute lrt down to all genes and rename it for convenience.
lrt = lrt["repLRT_all"]
names(lrt) = "LRT"

#Combine all results into single list.
res_list = append(lrt, pw)
res_list = append(res_list, ovr)
names(res_list) = str_remove_all(str_remove_all(names(res_list), "rep_"), "_all")
names(res_list) = str_replace(names(res_list), "_", "_vs_")
names(res_list)

saveRDS(res_list, here("data", "Rds", "12e_diffex_res_list.Rds"))
```

### Overall survival

```{r}

os <- readxl::read_excel(path = here("results", "survival", "12_cox_allgenes.xlsx"),
                         sheet = "multi_cox_surv")
head(filter(os, gene_type == "protein_coding"), 20)
```

```{r}
os_path = readxl::read_excel(path = here("results", "survival", "12_cox_allgenes.xlsx"),
                         sheet = "path_multi_os")
os_path
```


### Distant recurrence

```{r}
drs <- readxl::read_excel(path = here("results", "survival", "12_cox_allgenes.xlsx"),
                         sheet = "multi_cox_dr")
head(filter(drs, gene_type == "protein_coding"), 20)
```


### Interaction model

```{r}
inv_int_os <- readxl::read_excel(here("results", "survival", "12d_os_gene:inv_newformula.xlsx"))

head(inv_int_os)
```

```{r}
inv_int_drs <- readxl::read_excel(here("results", "survival", "12d_drs_gene:inv_newformula.xlsx"))

head(inv_int_drs)
```

## Transposed matrix for Cox regressions

Also created in notebook 12. First 20 columns are metadata, after that there is one column per gene. Gene expression columns are TMM-log2 normalized.

```{r}
coxdata = readRDS(here("data", "Rds", "12_coxdata.Rds"))
colnames(coxdata)[1:30]
```

Exclusively sample data. Clean up sample labels a little.

```{r}
sample_data = coxdata[,1:20]

#head(sample_data)

sample_data = sample_data %>%
  mutate(PPBC = case_when(
    study_group == "non_prbc" ~ "nulliparous",
    study_group == "ppbc_inv" ~ "involution",
    study_group == "ppbc_lac" ~ "lactation",
    study_group == "prbc" ~ "pregnant",
    
  ))

table(sample_data$study_group, sample_data$PPBC)

```


In gene x sample format:

```{r}
ens_mat <- t(coxdata[21:ncol(coxdata)])
dim(ens_mat)
#ens_mat[1:10,1:10]

saveRDS(ens_mat, here("data", "Rds", "12e_ensembl_tmmnorm_genesxsample.Rds"))
```


A count matrix in genes x samples format, containing the same TMM/log normalized expression information as coxdata, but with gene symbols and repeat IDs de-duplicated:

```{r}
sym_mat <- as.matrix(readRDS(here("data", "Rds", "12b_tmmnorm_geneEx_survival.Rds")))
dim(sym_mat)
#sym_mat[1:10,1:10]

saveRDS(sym_mat, here("data", "Rds", "12e_symbol_tmmnorm_genesxsample.Rds"))
```

## Involution vs rest encoding

```{r}
levels(coxdata$study_group)
```

```{r, eval=F}
#Takes a long time for some reason
coxdata = coxdata %>%
  mutate(involution = as.numeric(study_group == "ppbc_inv")) %>%
  select(sample_name:study_group, involution, everything())

saveRDS(coxdata, here("data", "Rds", "12e_coxdata_invgroup.Rds"))

```

```{r}
coxdata <- readRDS(here("data", "Rds", "12e_coxdata_invgroup.Rds"))
table(coxdata$study_group, coxdata$involution)
```


```{r}
colnames(coxdata)[1:30]
#Genes start at position 22
colnames(coxdata)[22]
```
```{r}
gene_dump <- function(id, id_type = "symbol"){
  
  res <- list()
  
  res$surv_df <- gene_summary(id = id, id_type = id_type)$survival_report
  
  res$os_km <- km_ntiles_ovr(gene_id = id, id_type = id_type, survival_type = "os")
  
  res$drs_km <- km_ntiles_ovr(gene_id = id, id_type = id_type, survival_type = "os")
  
  res$diffex_df <- gene_summary(id = id, id_type = id_type)$diffex_report
  
  res$beehive <- tmm_plots(id = id, id_type = id_type)$beehive
  
  return(res)
}
```

## Sample data

```{r}
sample_data = coxdata[,1:21]

#head(sample_data)

sample_data = sample_data %>%
  mutate(PPBC = case_when(
    study_group == "non_prbc" ~ "nulliparous",
    study_group == "ppbc_inv" ~ "involution",
    study_group == "ppbc_lac" ~ "lactation",
    study_group == "prbc" ~ "pregnant",
    
  ))

head(sample_data)

saveRDS(sample_data, here("data","Rds","12e_survival_sample_data.Rds"))
```

# Genes of interest from general OS/DRS

Or background literature

## CD20 (MS4A1)

### Summary survival

```{r}
gene_summary("MS4A1")$survival_report
```


### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("MS4A1", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("MS4A1"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("MS4A1", ovr_column = "involution", survival_type = "drs")
```



```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("MS4A1", survival_type = "drs"), ncol = 2, nrow = 1)
```



### Diffex

```{r}
gene_summary("MS4A1")$diffex_report
```


### Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
tmm_plots("MS4A1")$beehive
```

## CD27

The protein encoded by this gene is a member of the TNF-receptor superfamily. This receptor is required for generation and long-term maintenance of T cell immunity. It binds to ligand CD70, and plays a key role in regulating B-cell activation and immunoglobulin synthesis. This receptor transduces signals that lead to the activation of NF-κB and MAPK8/JNK. Adaptor proteins TRAF2 and TRAF5 have been shown to mediate the signaling process of this receptor. CD27-binding protein (SIVA), a proapoptotic protein, can bind to this receptor and is thought to play an important role in the apoptosis induced by this receptor.[5]

In murine γδ T cells its expression has been correlated with the secretion of IFNγ.[6]
Varlilumab is an antibody that binds to CD27 and is an experimental cancer treatment.
CD27 has been shown to interact with SIVA1,[7] TRAF2[8][9] and TRAF3.[8][9] 

[The immunobiology of CD27 and OX40 and their potential as targets for cancer immunotherapy](https://www.ncbi.nlm.nih.gov/pubmed/29118006)

### Summary survival

```{r}
gene_summary("CD27")$survival_report
```

### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("CD27", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("CD27"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("CD27", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("MS4A1", survival_type = "drs"), ncol = 2, nrow = 1)
```

### Diffex

```{r}
gene_summary("CD27")$diffex_report
```


### Expression

```{r, fig.width=8}
tmm_plots("CD27")$beehive
```

#### TRAF2

Downstream of CD27

##### Summary survival

```{r}
gene_summary("TRAF2")$survival_report
```

##### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("TRAF2", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("TRAF2"), ncol = 2, nrow = 1)
```

##### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("TRAF2", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("MS4A1", survival_type = "drs"), ncol = 2, nrow = 1)
```

##### Diffex

```{r}
gene_summary("TRAF2")$diffex_report
```


##### Expression

```{r, fig.width=8}
tmm_plots("TRAF2")$beehive
```

##### Traf3?
```{r}
gene_summary("TRAF3")$survival_report
```

## CD7

This gene encodes a transmembrane protein which is a member of the immunoglobulin superfamily. This protein is found on thymocytes and mature T cells. It plays an essential role in T-cell interactions and also in T-cell/B-cell interaction during early lymphoid development.

### Summary survival

```{r}
gene_summary("CD7")$survival_report
```

### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("CD7", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("CD7"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("CD7", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("CD7", survival_type = "drs"), ncol = 2, nrow = 1)
```

### Diffex

```{r}
gene_summary("CD7")$diffex_report
```

### Expression

```{r, fig.width=8}
tmm_plots("CD7")$beehive
```

## FoxP3

### Summary survival

```{r}
gene_summary("FOXP3")$survival_report
```


### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("FOXP3", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("FOXP3"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("FOXP3", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("FOXP3", survival_type = "drs"), ncol = 2, nrow = 1)
```


### Diffex

```{r}
gene_summary("FOXP3")$diffex_report
```


### Expression

```{r, fig.width=8}
tmm_plots("FOXP3")$beehive
```

## CD69

This gene encodes a member of the calcium dependent lectin superfamily of type II transmembrane receptors. Expression of the encoded protein is induced upon activation of T lymphocytes, and may play a role in proliferation. Furthermore, the protein may act to transmit signals in natural killer cells and platelets.

CD69 expression has been associated with both regulatory T cell (Treg), memory T cell and Bcl6 loCD69 hiLZ GC B plasmablast precursors.[11] Treg precursors exit the thymus expressing CD69 and complete differentiation into Treg cells in peripheral tissues when they encounter antigens and other cytokines, like IL-2.[12] Through the JAK/STAT signaling pathway, CD69 activation also induces the production of TGF-β as well as IL-2, which contribute to the differentiation of Treg cells as mentioned above.[8] Furthermore, CD69 is also known to be upregulated by NF-κB signaling at the onset of an immune response. A prolonged immune response is then maintained by the non-canonical NF-κB pathway, which in turn is associated with Treg differentiation.[7]

In addition to Treg differentiation, CD69 is a common marker of precursor and mature resident memory T cells (TRMs) that are localized in peripheral tissues.[13][9] TGF-β is also responsible for the development of TRMs, thus promoting TRM differentiation in a manner similar to Treg differentiation.[14] 

### Summary survival

```{r}
gene_summary("CD69")$survival_report
```

### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("CD69", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("CD69"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("CD69", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("CD69", survival_type = "drs"), ncol = 2, nrow = 1)
```

### Diffex

```{r}
gene_summary("CD69")$diffex_report
```

### Expression

```{r, fig.width=8}
tmm_plots("CD69")$beehive
```


# Genes of interest from interaction models

```{r,collapse=T}
print("Overall survival")
inv_int_os %>% filter(fdr < 0.25 & gene_type == "protein_coding") %>% pull(gene_name)

print("Distant recurrence")
inv_int_drs %>% filter(fdr < 0.25 & gene_type == "protein_coding") %>% pull(gene_name)
```

```{r}
inv_int_os %>% filter(fdr < 0.25 & gene_type == "protein_coding") %>% pull(description)
```
```{r}
inv_int_drs %>% filter(fdr < 0.25 & gene_type == "protein_coding") %>% pull(description)
```

## LILRA3

Leukocyte Immunoglobulin Like Receptor A3

This gene encodes a member of a family of immunoreceptors that are expressed predominantly in monocytes and B cells, and at lower levels in dendritic cells and natural killer cells. The encoded protein lacks the transmembrane region found in other members of this family. It acts as a soluble receptor for class I major histocompatibility complex (MHC) antigens. Alternatively spliced transcript variants encoding different isoforms have been found. This gene is located in a cluster of related genes on chromosome 19 and is polymorphic in human populations, with many individuals containing a deletion of this genomic region.

Multiple ensembl IDs retrieved with wildly different p values!

This gene is very lowly expressed.

```{r}
inv_int_os %>% filter(gene_name == "LILRA3")
```


### Summary survival

```{r}
gene_summary("ENSG00000275841", id_type = "ensembl")$survival_report
```


### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("ENSG00000275841", id_type = "ensembl", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("ENSG00000275841", id_type = "symbol",), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("ENSG00000275841", id_type = "ensembl", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("ENSG00000275841", id_type = "ensembl", survival_type = "drs"), ncol = 2, nrow = 1)
```


### Diffex

```{r}
gene_summary("ENSG00000275841", id_type = "ensembl")$diffex_report
```


### Expression

```{r, fig.width=8}
tmm_plots("ENSG00000275841", id_type = "ensembl")$beehive
```


## ING4

This gene encodes a tumor suppressor protein that contains a PHD-finger, which is a common motif in proteins involved in chromatin remodeling. This protein can bind TP53 and EP300/p300, a component of the histone acetyl transferase complex, suggesting its involvement in the TP53-dependent regulatory pathway. Multiple alternatively spliced transcript variants have been observed that encode distinct proteins.

### Summary survival

```{r}
gene_summary("ING4")$survival_report
```

### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("ING4", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("ING4"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("ING4", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("ING4", survival_type = "drs"), ncol = 2, nrow = 1)
```

### Diffex

```{r}
gene_summary("ING4")$diffex_report
```

### Expression

```{r, fig.width=8}
tmm_plots("ING4")$beehive
```

## EIF4E3

Binds to 5' cap during translation. Known tumor suppressor.

[eIF4E3 acts as a tumor suppressor by utilizing an atypical mode of methyl-7-guanosine cap recognition](https://www.pnas.org/content/110/10/3877)
[eIF4E3, a new actor in mRNA metabolism and tumor suppression](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3674077/)

### Summary survival

```{r}
gene_summary("EIF4E3")$survival_report
```

### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("EIF4E3", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("EIF4E3"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("EIF4E3", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("EIF4E3", survival_type = "drs"), ncol = 2, nrow = 1)
```

### Diffex

```{r}
gene_summary("EIF4E3")$diffex_report
```

### Expression

```{r, fig.width=8}
tmm_plots("EIF4E3")$beehive
```

## FGFRL1

The protein encoded by this gene is a member of the fibroblast growth factor receptor (FGFR) family, where amino acid sequence is highly conserved between members and throughout evolution. FGFR family members differ from one another in their ligand affinities and tissue distribution. A full-length representative protein would consist of an extracellular region, composed of three immunoglobulin-like domains, a single hydrophobic membrane-spanning segment and a cytoplasmic tyrosine kinase domain. The extracellular portion of the protein interacts with fibroblast growth factors, setting in motion a cascade of downstream signals, ultimately influencing mitogenesis and differentiation. A marked difference between this gene product and the other family members is its lack of a cytoplasmic tyrosine kinase domain. The result is a transmembrane receptor that could interact with other family members and potentially inhibit signaling. Multiple alternatively spliced transcript variants encoding the same isoform have been found for this gene. [provided by RefSeq, Jul 2008]

GeneCards Summary for FGFRL1 Gene

FGFRL1 (Fibroblast Growth Factor Receptor Like 1) is a Protein Coding gene. Diseases associated with FGFRL1 include Wolf-Hirschhorn Syndrome and Familiar Ovarian Carcinoma. Among its related pathways are Signaling by GPCR and VEGF Signaling Pathway. Gene Ontology (GO) annotations related to this gene include heparin binding and fibroblast growth factor-activated receptor activity. An important paralog of this gene is MCRIP2.


### Summary survival

```{r}
gene_summary("FGFRL1")$survival_report
```

### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("FGFRL1", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("FGFRL1"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("FGFRL1", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("FGFRL1", survival_type = "drs"), ncol = 2, nrow = 1)
```

### Diffex

```{r}
gene_summary("FGFRL1")$diffex_report
```

### Expression

```{r, fig.width=8}
tmm_plots("FGFRL1")$beehive
```


## LENG1

Multiple ensembl IDs retrieved. "ENSG00000105617" is associated with interaction term for survival.

```{r}
inv_int_os %>% filter(gene_name == "LENG1")
```


### Summary survival

```{r}
gene_summary("ENSG00000105617", id_type = "ensembl")$survival_report
```

### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("ENSG00000105617", id_type = "ensembl", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("ENSG00000105617", id_type = "ensembl", ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("ENSG00000105617", id_type = "ensembl", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("ENSG00000105617", id_type = "ensembl", survival_type = "drs"), ncol = 2, nrow = 1)
```

### Diffex

```{r}
gene_summary("ENSG00000105617", id_type = "ensembl")$diffex_report
```

### Expression

```{r}
as.data.frame(assay(dds)) %>% rownames_to_column("ensembl_gene_id") %>%
  filter(ensembl_gene_id == "ENSG00000105617") %>% select(-ensembl_gene_id) %>%
  t() %>% summary()
```


```{r, fig.width=8}
tmm_plots("ENSG00000105617", id_type = "ensembl")$beehive
```

## Rad52


### Summary survival

```{r}
gene_summary("RAD52")$survival_report
```

### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("RAD52", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("CD69"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("RAD52", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("CD69", survival_type = "drs"), ncol = 2, nrow = 1)
```

### Diffex

```{r}
gene_summary("RAD52")$diffex_report
```


```{r, fig.width=8}
tmm_plots("RAD52")$beehive
```

## GNB3

G protein subunit beta 3

https://www.ncbi.nlm.nih.gov/pubmed/?term=GNB3+cancer

Polymorphisms associated with altered survival in several cancer types.

[Reduced prevalence of the C825T polymorphism of the G-protein beta subunit gene in women with breast cancer.](https://www.ncbi.nlm.nih.gov/pubmed/22034052)

```{r}
gene_dump("GNB3")
```


## ANKS1A

[Anks1a regulates COPII-mediated anterograde transport of receptor tyrosine kinases critical for tumorigenesis](https://www.ncbi.nlm.nih.gov/pubmed/27619642)

ErbB2 (HER2) signalling, which is amplified by EphA2 binding, is an important therapeutic target for breast cancer. Despite the importance of the EphA2/ErbB2 complex in promoting breast tumorigenesis, the mechanism by which these receptor tyrosine kinases (RTKs) are exported from the endoplasmic reticulum (ER) remains poorly understood. Here we report that the PTB adaptor Anks1a is specifically localized to the ER on its own serine phosphorylation. Once there, Anks1a acts as an important regulator of COPII-mediated EphA2 ER export. The Anks1a ankyrin repeat domain binds EphA2 and causes it to accumulate at sites of ER exit. Simultaneously, the Anks1a PTB domain binds Sec23. This induces internalization of EphA2 via COPII vesicles, while Anks1a remains behind on the ER membrane. EphA2 also binds ErbB2 in the ER and seems to load ErbB2 into growing COPII carriers. Together, our study reveals a novel mechanism that regulates the loading of RTKs into COPII vesicles.

```{r}
gene_dump("ANKS1A")
```

## YTHDF3

Specifically recognizes and binds N6-methyladenosine (m6A)-containing RNAs and promotes RNA translation
efficiency (PubMed:28106072, PubMed:28106076, PubMed:28281539). M6A is a modification present at
internal sites of mRNAs and some non-coding RNAs and plays a role in the efficiency of mRNA splic-
ing, processing and stability (PubMed:22575960, PubMed:24284625, PubMed:28106072, PubMed:28106076,
1PubMed:28281539). Shares m6A-containing mRNAs targets with YTHDF1 and YTHDF2, and regulates dif-
ferent processes depending on the context (PubMed:28106072, PubMed:28106076). Facilitates the translation
of targeted mRNAs in cooperation with YTHDF1 by binding to m6A-containing mRNAs and interacting
with 40S and 60S ribosome subunits (PubMed:28106072, PubMed:28106076). Acts as a negative regulator of
type I interferon response by down-regulating interferon-stimulated genes (ISGs) expression: acts by binding
to FOXO3 mRNAs and promoting their translation. Binds to FOXO3 mRNAs independently of METTL3-
mediated m6A modification (By similarity). Can also act as a regulator of mRNA stability in cooperation with
YTHDF2 by binding to m6A-containing mRNA and promoting their degradation (PubMed:28106072). Recog-
nizes and binds m6A-containing circular RNAs (circRNAs) and promotes their translation (PubMed:28281539).
circRNAs are generated through back-splicing of pre-mRNAs, a non-canonical splicing process promoted by
dsRNA structures across circularizing exons (PubMed:28281539).

[N6-methyladenosine-related Genomic Targets are Altered in Breast Cancer Tissue and Associated with Poor Survival](https://www.ncbi.nlm.nih.gov/pubmed/31632489)

> Purpose: The ectopic expression of N6-methyladenosine (m6A) associated genes is a common feature of multiple tumors. However, little is known about the expression status and the prognostic value of these genes in human breast cancer (BRC). Herein, we conducted a comprehensive analysis to identify the expression profiling and clinical significance of m6A-related genomic targets in BRC. Materials and Methods: The expression data including 1109 BRC tissues and 113 normal breast tissues were obtained from The Cancer Genome Atlas (TCGA) database to evaluate the mRNA expression levels of m6A-related genomic targets. In addition, 6 independent BRCA cohorts retrieved from the Gene Expression Omnibus (GEO) database were enrolled to further ascertain the expression profiling of m6A-related genomic targets. Meanwhile, the immunohistochemical (IHC) staining data from BRC tissue microarray (TMA) cohort and the Human Protein Atlas (HPA) database were used to evaluate the proteomic expression of m6A-related genomic targets. Immunofluorescence (IF) analysis was performed to validate the subcellular location of m6A-related genomic targets. Moreover, the prognostic value of m6A-related genomic targets in BRC was analyzed by Kaplan-Meier analysis and Cox regression models. Results: m6A-related genomic targets were differentially expressed in BRC tissues. TMA IHC staining showed that most of the m6A-related genomic targets were significantly altered at the protein level (either upregulated or downregulated), consistent with their changes in the genomic profile. IF analysis showed the subcellular location of m6A-related genomic targets in BRC cell lines. Furthermore, we demonstrated that overexpression of YTHDF1 (P=0.049), YTHDF3 (P<0.001) and KIAA1429 (P=0.032) predicted poor prognosis in terms of overall survival (OS). Upregulation of YTHDF3 was an independent prognostic factor for OS in patients with BRC (P=0.036). Conclusion: m6A-related genomic targets are significantly altered in BRC and predict poor prognosis. These m6A-related genomic targets could serve as novel prognostic biomarkers for BRC.

```{r}
gene_dump("YTHDF3")
```


## AHRR

aryl-hydrocarbon receptor repressor

The protein encoded by this gene participates in the aryl hydrocarbon receptor (AhR) signaling cascade, which mediates dioxin toxicity, and is involved in regulation of cell growth and differentiation. It functions as a feedback modulator by repressing AhR-dependent gene expression. Alternatively spliced transcript variants encoding different isoforms have been described for this gene. [provided by RefSeq, Jun 2011]

Normally a tumor repressor gene.

May be a marker of environmental toxicity

```{r}
gene_dump("AHRR")
```

## ACRBP

The protein encoded by this gene is similar to proacrosin binding protein sp32 precursor found in mouse, guinea pig, and pig. This protein is located in the sperm acrosome and is thought to function as a binding protein to proacrosin for packaging and condensation of the acrosin zymogen in the acrosomal matrix. This protein is a member of the *cancer/testis* family of antigens and it is found to be immunogenic. In normal tissues, this mRNA is expressed only in testis, whereas it is detected in a range of different tumor types such as bladder, breast, lung, liver, and colon. [provided by RefSeq, Jul 2008]

```{r}
gene_dump("ACRBP")
```

## PCGF3

Component of a Polycomb group (PcG) multiprotein PRC1-like complex, a complex class required to maintain the transcriptionally repressive state of many genes, including Hox genes, throughout development. PcG PRC1 complex acts via chromatin remodeling and modification of histones; it mediates monoubiquitination of histone H2A 'Lys-119', rendering chromatin heritably changed in its expressibility. Within the PRC1-like complex, regulates RNF2 ubiquitin ligase activity (PubMed:26151332). Plays a redundant role with PCGF5 as part of a PRC1-like complex that mediates monoubiquitination of histone H2A 'Lys-119' on the X chromosome and is required for normal silencing of one copy of the X chromosome in XX females.

[Response Predictors of S-1, Cisplatin, and Docetaxel Combination Chemotherapy for Metastatic Gastric Cancer: Microarray Analysis of Whole Human Genes.](https://www.ncbi.nlm.nih.gov/pubmed/?term=PCGF3+cancer)

```{r}
gene_dump("PCGF3")
```

## P3H3

The protein encoded by this gene belongs to the leprecan family of proteoglycans, which function as collagen prolyl hydroxylases that are required for proper collagen biosynthesis, folding and assembly. This protein, like other family members, is thought to reside in the endoplasmic reticulum. Epigenetic inactivation of this gene is associated with breast and other cancers, suggesting that it may function as a tumor suppressor. [provided by RefSeq, Aug 2013]

```{r}
gene_dump("P3H3")
```

## CLIC2

May be involved in metastasis to the liver.

[Chloride intracellular channel protein 2 in cancer and non-cancer human tissues: relationship with tight junctions](https://www.ncbi.nlm.nih.gov/pubmed/30929599)

Chloride intracellular channel protein 2 (CLIC2) belongs to the CLIC family of conserved metazoan proteins. Although CLICs have been identified as chloride channels, they are currently considered multifunctional proteins. CLIC2 is the least studied family member. We investigated CLIC2 expression and localization in human hepatocellular carcinoma, metastatic colorectal cancer in the liver, and colorectal cancer. Significant expression of mRNAs encoding CLIC1, 2, 4, and 5 were found in the human tissues, but only CLIC2 was predominantly expressed in non-cancer tissues surrounding cancer masses. Fibrotic or dysfunctional (aspartate aminotransferase ≥40) non-cancer liver tissues and advanced stage HCC tissues expressed low levels of CLIC2. Endothelial cells lining blood vessels but not lymphatic vessels in non-cancer tissues expressed CLIC2 as well as high levels of the tight junction proteins claudins 1 and 5, occludin, and ZO-1. Most endothelial cells in blood vessels in cancer tissues had very low expressions of CLIC2 and tight junction proteins. CD31+/CD45- endothelial cells isolated from non-cancer tissues expressed mRNAs encoding CLIC2, claudin 1, occludin and ZO-1, while similar cell fractions from cancer tissues had very low expressions of these molecules. Knockdown of CLIC2 expression in human umbilical vein endothelial cells (HUVECs) allowed human cancer cells to transmigrate through a HUVEC monolayer. These results suggest that CLIC2 may be involved in the formation and/or maintenance of tight junctions and that cancer tissue vasculature lacks CLIC2 and tight junctions, which allows the intravasation of cancer cells necessary for hematogenous metastasis.

```{r}
gene_dump("CLIC2")
```

## DPYSL4

[p53-inducible DPYSL4 associates with mitochondrial supercomplexes and regulates energy metabolism in adipocytes and cancer cells](https://www.ncbi.nlm.nih.gov/pubmed/30061407)

```{r}
gene_dump("DPYSL4")
```

## KANSL2

May promote tumorigenesis and metastasis

[The NSL Chromatin-Modifying Complex Subunit KANSL2 Regulates Cancer Stem-like Properties in Glioblastoma That Contribute to Tumorigenesis.](https://www.ncbi.nlm.nih.gov/pubmed/27406830)


```{r}
gene_dump("KANSL2")
```

# Genes of interest from diffex

## IGHD

Multiple ensembl IDs reported

```{r}
gx_annot %>% filter(gene_name == "IGHD")
```
Which was the differentially expressed gene?

```{r}
res_list$inv_vs_rest[res_list$inv_vs_rest$gene_name == "IGHD", ]
```


### Summary survival

Check for both isotypes.

```{r}
gene_summary("ENSG00000211898", id_type = "ensembl")$survival_report
#gene_summary("ENSG00000278801", id_type = "ensembl")$survival_report
#gene_survival("ENSG00000278801", id_type = "ensembl")
```

### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("IGHD", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("IGHD"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("IGHD", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("IGHD", survival_type = "drs"), ncol = 2, nrow = 1)
```

### Diffex

```{r}
gene_summary("IGHD")$diffex_report
```

### Expression

```{r, fig.width=8}
tmm_plots("IGHD")$beehive
```

## IGHA1

### Summary survival

```{r}
gene_summary("IGHA1")$survival_report
```

### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("IGHA1", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("IGHA1"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("IGHA1", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("IGHA1", survival_type = "drs"), ncol = 2, nrow = 1)
```

### Diffex

```{r}
gene_summary("IGHA1")$diffex_report
```

### Expression

```{r, fig.width=8}
tmm_plots("IGHA1")$beehive
```

## JCHAIN

### Summary survival

```{r}
gene_summary("JCHAIN")$survival_report
```

### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("JCHAIN", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("IGHA1"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("JCHAIN", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("IGHA1", survival_type = "drs"), ncol = 2, nrow = 1)
```

### Diffex

```{r}
gene_summary("JCHAIN")$diffex_report
```

### Expression

```{r, fig.width=8}
tmm_plots("JCHAIN")$beehive +
  geom_hline(yintercept = 5.3179857, linetype = "dashed") +
  geom_hline(yintercept = 6.7924900, linetype = "dashed")
```

```{r}
tmm_plots("JCHAIN")$violin +
  geom_hline(yintercept = 5.3179857, linetype = "dashed") +
  geom_hline(yintercept = 6.7924900, linetype = "dashed")
```


How many of the lower ntile samples are there in the involution group?

```{r}
gene_ntile("JCHAIN", id_type = "symbol") %>%
  left_join(., select(sample_data, sample_name, study_group), by = "sample_name") %>%
  select(study_group, ntile) %>% table()
```
```{r, eval=F, include=F}
dplyr::ntile(sym_mat[rownames(sym_mat) %in% c("JCHAIN"), ], n = 3)

quantile(sym_mat[rownames(sym_mat) %in% c("JCHAIN"), ], probs = seq(0, 1, 0.333333), na.rm = FALSE)
```



# From literature

## IFNG

### Summary survival

```{r}
gene_summary("ENSG00000111537", id_type = "ensembl")$survival_report
```

### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("IFNG", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("IGHA1"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("IFNG", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("IGHA1", survival_type = "drs"), ncol = 2, nrow = 1)
```

### Diffex

```{r}
gene_summary("IFNG")$diffex_report
```

### Expression

```{r, fig.width=8}
tmm_plots("IFNG")$beehive
```

## IGHG1

### Summary survival

```{r}
gene_summary("IGHG1")$survival_report
```

### OS

One vs rest survival depications

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("IGHG1", ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, eval=F, include=F, fig.height=5, fig.width=10}
#Separated study groups
ggarrange(plotlist = km_ntiles("IGHA1"), ncol = 2, nrow = 1)
```

### DRS

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("IGHG1", ovr_column = "involution", survival_type = "drs")
```

```{r, warning=F, include=F, eval=F, fig.height=5, fig.width=10}
#Separated study groups.
ggarrange(plotlist = km_ntiles("IGHA1", survival_type = "drs"), ncol = 2, nrow = 1)
```

### Diffex

```{r}
gene_summary("IGHG1")$diffex_report
```

### Expression

```{r, fig.width=8}
tmm_plots("IGHG1")$beehive
```

### Save data

```{r}
save.image(here("reports", "12e_genewise_km_new_formula.RData"))
```












































































