---
title: "Involution vs rest, Belgian samples only"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
---

```{r, include=F}
rm(list = ls())
```


```{r, include=F}
#library(devtools)
#install_github("jokergoo/ComplexHeatmap")
#install_github("mikelove/DESeq2") 
#install_github("azhu513/apeglm")
library(flexgsea) #For read_gmt()
library(ggrepel)
library(ComplexHeatmap)
library(openxlsx)
library(scrime)
library(RColorBrewer)
library(here)
library(DESeq2)
library(apeglm)
library(tidyverse)
```

For staining, we have better access to Belgian patients than any other location. Check to see whether the IG signature is still present in the Belgium-only samples.

# Load gene and sample data

```{r}
dds <- readRDS(file = here("data/Rds/05_dds_PAM50_batch.Rds"))

gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
gx_annot = gx_annot %>% select(ensembl_gene_id = gene_id, gene_name, gene_type, description = gene_description) %>% distinct()
```

## Filter

```{r}
#Minimum threshold is a nonzero count in at least a 3rd of all samples
keep <- rowSums(counts(dds)!=0) >= ceiling(ncol(dds)/3)
table(keep)

dds <- dds[keep,]
```

```{r}
#Remove non Belgian samples
be_samples = as.data.frame(colData(dds)) %>%
  filter(country == "BE") %>%
  pull(sample_name)

print(paste("All samples: ", ncol(dds)))
dds <- dds[,be_samples]
print(paste("Belgian samples: ", ncol(dds)))

stopifnot(all(dds$country=="BE"))

table(dds$study_group)
```

## Retrieve immune genes

For downstream analyses, we'd like to know which of the differentially expressed genes are immunologically relevant.
An immune gene is defined as follows:


Either immune/immuno/interleukin is part of the gene name and description OR the gene is part of the the ImmPort database.
See https://www.innatedb.com/redirect.do?go=resourcesGeneLists

The Immunology Database and Analysis Portal (ImmPort) system was developed under the Bioinformatics Integration Support Contract (BISC) Phase II by the Northrop Grumman Information Technology Health Solutions team for the NIH, NIAID, and DAIT. The principal investigator of the BISC project is Dr. Richard Scheuermann at University of Texas Southwestern Medical Center. The list of immunologically related genes in ImmPort is a collection of ~6,000 human genes, which was formed with the goal of retrieving all genes that have immune system-related functions. This list was generated using automatic searches of EntrezGene and Gene Ontology records using immunology-related keywords. The list was then manually curated by immunology experts examining various literature sources.

```{r}
immune_gene_list <- read_csv(here("data", "external","gene-sets","InnateDB_genes.csv"))
head(immune_gene_list)
```

## Helper functions

```{r}
source(here("src", "deseq_report_functions.R"))
```

## VST

Recalculate vst in absence of other countries


```{r}
vsd = blind_vst(dds)
saveRDS(vsd, here("data", "Rds", "13_BE_vsd.Rds"))
```

```{r}
vsd = readRDS(here("data", "Rds", "13_BE_vsd.Rds"))
```


## Gene signatures

```{r}
gene_sets = list(
  go_bp = flexgsea::read_gmt(here("data","external","gmt","c5.bp.v7.0.symbols.gmt")),
  hallmark = flexgsea::read_gmt(here("data","external","gmt","h.all.v7.0.symbols.gmt")),
  c2_canon = flexgsea::read_gmt(here("data","external","gmt","c2.cp.v7.0.symbols.gmt")),
  c2_cgp = flexgsea::read_gmt(here("data","external","gmt","c2.cgp.v7.0.symbols.gmt"))
)
```

## Heatmap colors

Created in notebook 6

```{r}
all_colors = readRDS(here("data","Rds", "06_heatmap_colors.Rds"))
study_colors = all_colors$study_colors
pam_colors = all_colors$pam_colors
gene_colors = all_colors$gene_colors
```

## One vs rest formula levels

```{r, collapse=T}
print(design(dds))

dds$inv_vs_rest  = colData(dds) %>% as.data.frame() %>% mutate(inv_vs_rest = if_else(study_group == "ppbc_inv", "ppbc_inv", "rest")) %>%
  pull(inv_vs_rest) %>% factor(levels=c("rest", "ppbc_inv"))

dds$prbc_vs_rest  = colData(dds) %>% as.data.frame() %>% mutate(prbc_vs_rest = if_else(study_group == "prbc", "prbc", "rest")) %>%
  pull(prbc_vs_rest) %>% factor(levels=c("rest", "prbc"))

dds$lac_vs_rest  = colData(dds) %>% as.data.frame() %>% mutate(lac_vs_rest = if_else(study_group == "ppbc_lac", "ppbc_lac", "rest")) %>%
  pull(lac_vs_rest) %>% factor(levels=c("rest", "ppbc_lac"))

dds$nonprbc_vs_rest  = colData(dds) %>% as.data.frame() %>% mutate(nonprbc_vs_rest = if_else(study_group == "non_prbc", "non_prbc", "rest")) %>%
  pull(nonprbc_vs_rest) %>% factor(levels=c("rest", "non_prbc"))

table(dds$inv_vs_rest, dds$study_group) %>% print()
table(dds$prbc_vs_rest, dds$study_group) %>% print()
table(dds$lac_vs_rest, dds$study_group) %>% print()
table(dds$nonprbc_vs_rest, dds$study_group) %>% print()
```

# Diffex: involution vs rest

```{r}
dds$group = dds$inv_vs_rest

print(levels(dds$group))

design(dds) = ~batch + PAM50 + group

print(design(dds))
resFile = here("data/Rds/13_dds_inv_vs_rest_BE.Rds")
```

Perform differential expression analysis.
Run in notebook 13.

```{r, eval=F}
overwrite <- F

if(file.exists(resFile)==F | overwrite == T){
  start = Sys.time()
  dds <- DESeq(dds)
  end = Sys.time()
  print(end-start)
  saveRDS(dds,resFile)
}
```

```{r}
dds=readRDS(resFile)
levels(dds$group) %>% print()
```

# Shrink fold changes

```{r, eval=F}
resFile = here("data","Rds", "13_ape_inv_rest_BE.Rds")
if(file.exists(resFile)==F | overwrite == T){
  start = Sys.time()
  ape_inv_rest = shrinkRes(dds,  contrast=c("group", "ppbc_inv", "rest"))
  saveRDS(ape_inv_rest, resFile)
  end = Sys.time()
  print(end-start)
}

```

```{r}
ape_inv_rest <- readRDS( here("data","Rds", "13_ape_inv_rest_BE.Rds"))
```


# Summary report

The reporting function requires vsd to have the same levels as dds in study_group

```{r}
vsd$study_group = vsd$inv_vs_rest
```


A log2 fold change cutoff of 0.5 is equivalent to 1.414.
```{r}
colData(vsd) %>% colnames()
```



```{r}
rep_inv_rest = deseq_report(ape_inv_rest, dds=dds, absl2fc = 0.5, groups = c("ppbc_inv", "rest"),
                            beehive_groups = "all", title = "BE: Involution vs rest")
```


## Top genes

```{r}
rep_inv_rest$significant_genes
```

## Volcano plot


```{r}
rep_inv_rest$volcano_plot
```

## Heatmap

```{r, fig.height=6, fig.width=10}
rep_inv_rest$heatmap
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r}
 rep_inv_rest$pathways %>% bind_rows() %>%
  filter(fdr < 0.05) %>% select(pathway,everything()) %>% arrange(fdr)
```


```{r, fig.height=6, fig.width=8}
rep_inv_rest$pathway_plots
```



## Beehive plots

Beehive plot of top 5 most significantly expressed genes.

```{r}
rep_inv_rest$beehive_plots
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r}
rep_inv_rest$volc_outliers
```

