---
title: "Alternate deconvolution algorithms"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
    toc_depth: 4
    highlight: tango
---

```{r, include=F}
rm(list = ls())
```


```{r, include=F}
library(here)
library(RColorBrewer)
library(pheatmap)
library(DESeq2)
library(tidyverse)
```


# Introduction

This report summarizes two deconvolution methods aside from Cibersort: DWLS and NMF.

DWLS - Dampened weighted least squares, from the [DWLS](https://bitbucket.org/yuanlab/dwls/src/default/) package.

>DWLS is an estimation method for gene expression deconvolution, in which the cell-type composition of a bulk RNA-seq data set is computationally inferred. This method corrects common biases towards cell types that are characterized by highly expressed genes and/or are highly prevalent, to provide accurate detection across diverse cell types. To begin, the user must input a bulk RNA-seq data set, along with a labeled representative single-cell RNA-seq data set that will serve to generate cell-type-specific gene expression profiles. Ideally, the single-cell data set will contain cells from all cell types that may be found in the bulk data. DWLS will return the cell-type composition of the bulk data.

More background in publication: [Accurate estimation of cell-type composition from gene expression data](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6611906/#!po=81.5789)

For DWLS we use a signature matrix and code provided by Lennart Kester.

NOT YET IMPLEMENTED NMF - Non-Negative Matrix Factorization, from the [NNLM](https://cran.r-project.org/web/packages/NNLM/index.html) (Non-Negative Linear Models) package.

# Import data/settings

## Sample and gene annotation

```{r}
dds = readRDS(here("data/Rds/08_dds_inv_vs_rest_standard.Rds"))
vsd = readRDS(here("data","Rds","12_vst_mat_gene_symbol.Rds")) #Vst for visualization

gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
gx_annot = gx_annot %>% select(ensembl_gene_id = gene_id, gene_name, gene_type, description = gene_description) %>% distinct()

ig_colclust <- readRDS(here("data", "Rds", "11_ig_column_clusters.Rds"))
```

## Heatmap annotation and colors

```{r}
all_colors = readRDS(here("data","Rds", "06_heatmap_colors.Rds"))
study_colors = all_colors$study_colors
pam_colors = all_colors$pam_colors
gene_colors = all_colors$gene_colors
IG_colors = readRDS(here("data", "Rds", "11_IG_cluster_colors.Rds"))


anndf = as.data.frame(colData(dds)) %>%
  select(sample_name, PPBC = study_group, PAM50) %>% 
  left_join(., select(ig_colclust, sample_name, `IG cluster` = IG_col_cluster), by = "sample_name") %>%
  column_to_rownames("sample_name")

stopifnot(all(rowSums(is.na(anndf)) == 0))

head(anndf)

ann_colors = list(
    PPBC = study_colors[names(study_colors) != "rest"],
    PAM50 = pam_colors,
    `IG cluster` = IG_colors
)
```

# DWLS

Relevant scripts are:
* Master script "src/dwls_forKat.R"
* "src/dwls_functions.R".

## Input preview

Input matrices are: 
* Signature matrix "data/dwls/20200106_refProfiles_STARres0.1CellTypes.csv"
* Marker genes  "data/dwls/20200106_markerGenes_STARres0.1CellTypes.csv"
* Gene exression matrix from tumor samples "data/dwls/ppbc_samplesxgenes.csv"

```{r}
refProfiles <- read.table(here("data/dwls/20200106_refProfiles_STARres0.1CellTypes.csv"))
dim(refProfiles)
refProfiles[1:10,1:10]
```

```{r}
markerGenes <- read.table(here("data/dwls/20200106_markerGenes_STARres0.1CellTypes.csv"))
markerGenes <- markerGenes[,1] #Should be a vector
head(markerGenes)


```


```{r}
target <- read_csv(here("data/dwls/ppbc_samplesxgenes.csv"))
#revert to genes x samples format, also done in dwls_forKat.R
target <- as.data.frame(target)
rownames(target) <- target$sample_name
target <- target[,-1]
target <- t(target)
dim(target)
target[1:10,1:10]
```

## Results

Results files are:
* data/dwls/dwls_results.csv
* data/dwls/results.Rds (lacks permuations and p values)

```{r}
results <- read_csv(here("data/dwls/dwls_results.csv"))
colnames(results)[1] <- "sample_name"

head(results)
```

Split results in cell tpes and stats separately

```{r}
colnames(results)
rmat <- results %>% select(sample_name:TC_IV)
rmat <- column_to_rownames(rmat, "sample_name")
rmat <- t(as.matrix(rmat))
rstat <- results %>% select(sample_name, pearsonCor:dwlsPval)
```

## Heatmap cell type

TC = "tumor cluster"
Cell abundances are relative (colSums are 1)

### All cell types

```{r, fig.width=10}
pheatmap(rmat, color = colorRampPalette(brewer.pal(9,"Reds"))(100),#rev(heat.colors(10)),#viridis::plasma(10),
         annotation_col = anndf, annotation_colors = ann_colors,
         show_colnames = F, main = "DWLS, relative cell fraction")
```

```{r}
rownames(rmat)
```

### Tumor clusters

```{r, fig.width=10}
pheatmap(rmat[str_detect(rownames(rmat), "TC_"), ],
         color = colorRampPalette(brewer.pal(9,"Reds"))(100),#rev(heat.colors(10)),#viridis::plasma(10),
         annotation_col = anndf, annotation_colors = ann_colors,
         show_colnames = F, main = "DWLS,  relative abundance tumor clusters")
```

### Immune cells

```{r, fig.width=10}
pheatmap(rmat[rownames(rmat) %in% c("Tcells", "Bcells", "Macrophages", "Mastcells"), ],
         color = colorRampPalette(brewer.pal(9,"Reds"))(100),#rev(heat.colors(10)),#viridis::plasma(10),
         annotation_col = anndf, annotation_colors = ann_colors,
         show_colnames = F, main = "DWLS, relative abundance immune cells")

```

```{r,  fig.width=10}
pheatmap(rmat[rownames(rmat) %in% c("Tcells", "Bcells", "Macrophages", "Mastcells"), ],
         #color = viridis::plasma(100),
         #color = colorRampPalette(brewer.pal(9,"BrBG"))(100),
         color = rev(colorRampPalette(brewer.pal(9,"RdBu"))(100)),
         #annotation_col = anndf,
         annotation_col = anndf[,-2],
         #annotation_colors = ann_colors,
         annotation_colors = ann_colors[-2],
         scale="row",
         show_colnames = F, main = "DWLS, Z-score immune cells")
```

```{r,  fig.width=8, fig.height = 4}
for (cell in c("Tcells", "Bcells", "Macrophages", "Mastcells")){
 pheatmap(rmat[rownames(rmat) == cell, ,drop = F],
         #color = viridis::plasma(100),
         #color = colorRampPalette(brewer.pal(9,"BrBG"))(100),
         color = rev(colorRampPalette(brewer.pal(9,"RdBu"))(100)),
         annotation_col = anndf,
         #annotation_col = anndf[,-2],
         annotation_colors = ann_colors,
         #annotation_colors = ann_colors[-2],
         scale="row", cluster_rows = F,
         show_colnames = F, main = "DWLS, Z-score immune cells") 
}


```

### Other cell types

```{r,  fig.width=8, fig.height = 4}
pheatmap(rmat[rownames(rmat) %in% c("Myoepithelial", "CAF", "Muscle"), ],
         color = colorRampPalette(brewer.pal(9,"Reds"))(100),#rev(heat.colors(10)),#viridis::plasma(10),
         annotation_col = anndf, annotation_colors = ann_colors,
         show_colnames = F, main = "DWLS,  relative abundance other cell types")

for (cell in c("Myoepithelial", "CAF", "Muscle")){
 pheatmap(rmat[rownames(rmat) == cell, ,drop = F],
         #color = viridis::plasma(100),
         #color = colorRampPalette(brewer.pal(9,"BrBG"))(100),
         color = colorRampPalette(brewer.pal(9,"Reds"))(100),
         annotation_col = anndf,
         #annotation_col = anndf[,-2],
         annotation_colors = ann_colors,
         #annotation_colors = ann_colors[-2],
         scale="none", cluster_rows = F,
         show_colnames = F, main = "DWLS, relative abundance other cell types") 
}


```

## Bar plots cell types

Wrangle into tidy format.

```{r}
rdf = rmat %>% as.data.frame() %>%
  rownames_to_column("cell_type") %>%
  gather(key = "sample_name", -cell_type, value = "dwls") %>%
  left_join(., rownames_to_column(anndf, "sample_name"), by = "sample_name")
```

Separate cell types

```{r}
rdf$cell_type %>% unique()
tumor_clusters = sort(unique(rdf$cell_type)[str_detect(unique(rdf$cell_type), "TC_")])
immune_cells = c("Tcells", "Bcells", "Macrophages", "Mastcells")
other_cells = c("Myoepithelial", "Endothelial", "CAF", "Muscle")
```


## Immune cells

```{r, fig.height=6}
r = rdf %>%
  filter(cell_type %in% immune_cells) %>%
  arrange(`PPBC`) %>%
  mutate(sample_name = as.factor(sample_name))

r %>%
  filter(cell_type %in% immune_cells) %>%
  ggplot(aes(x = factor(sample_name, levels = unique(sample_name)), y = dwls, fill = PPBC))+
  geom_bar(stat = "identity") +
  facet_wrap(~cell_type, ncol=1) +
  theme(axis.text.x = element_blank()) +
  xlab("samples")+
  ylab("relative fraction") +
  ggtitle("PPBC and dwls detection of immune cells") +
  theme_minimal() +
  ggthemes::scale_fill_colorblind()
```

```{r, fig.height=6}
r = rdf %>%
  filter(cell_type %in% immune_cells) %>%
  arrange(PAM50) %>%
  mutate(sample_name = as.factor(sample_name))

r %>%
  filter(cell_type %in% immune_cells) %>%
  ggplot(aes(x = factor(sample_name, levels = unique(sample_name)), y = dwls, fill = PAM50))+
  geom_bar(stat = "identity") +
  facet_wrap(~cell_type, ncol=1) +
  theme(axis.text.x = element_blank()) +
  xlab("samples")+
  ylab("relative fraction") +
  ggtitle("Pam50 and dwls detection of immune cells")+
  theme_minimal() +
  ggthemes::scale_fill_colorblind()
```


```{r, fig.height=6}
r = rdf %>%
  filter(cell_type %in% immune_cells) %>%
  arrange(`IG cluster`) %>%
  mutate(sample_name = as.factor(sample_name))

r %>%
  filter(cell_type %in% immune_cells) %>%
  arrange(`IG cluster`) %>%
  ggplot(aes(x = factor(sample_name, levels = unique(sample_name)), y = dwls, fill = `IG cluster`))+
  geom_bar(stat = "identity") +
  facet_wrap(~cell_type, ncol=1) +
  theme(axis.text.x = element_blank()) +
  xlab("samples")+
  ylab("relative fraction") +
  ggtitle("Relationship IG expression and dwls detection of immune cells") +
  theme_minimal() +
  ggthemes::scale_fill_colorblind()
```


## Correlation cell types and cell markers

```{r}

correlate_marker_cell <- function(gene_symbol, cell_type, genEx = vsd, dwls = rmat,
                                  stats = rstat, cormet = "spearman"){
  
  stopifnot(identical(colnames(genEx), colnames(dwls)))
  stopifnot(identical(colnames(genEx), stats$sample_name))
  genEx = as.matrix(genEx)
  g = genEx[rownames(genEx) == gene_symbol, , drop=T]
  c = dwls[rownames(dwls) == cell_type, , drop=T]
  #print(class(g))
  #print(class(c))
  cg = signif(cor(c,g, method = cormet), 3)
  
  d = tibble(`Sample name` = colnames(genEx),
             `Vst gene expression` = g,
             `Relative cell abundance` = c,
             `Dwls Spearman Cor` = stats$spearmanCor,
             `Dwls Spearman Pval` = stats$spearmanCorPval,
             `Dwls Pval` = stats$dwlsPval)
  
  title = paste0("Correlation ", gene_symbol, " and relative abundance of ", cell_type,
                 " from dwls: ", cg)
  
  ggplot(d, aes(x = `Vst gene expression`, y = `Relative cell abundance`,
                color = `Dwls Spearman Pval`,
                #size = `Dwls Pval`
                )) +
    geom_point() +
    geom_smooth(method = "lm", color= "red") +
    #ggthemes::theme_few()
    theme_minimal() +
    ggtitle(title) +
    xlab(paste("Vst gene expression", gene_symbol)) +
    ylab(paste("Relative dwls cell abundance", cell_type))
  
}
#vsd[rownames(vsd) == "MS4A1", , drop=F]
#rownames(rmat)
#rmat[rownames(rmat) == "Bcells", , drop=T]
correlate_marker_cell(gene_symbol = "MS4A1", cell_type = "Bcells")

```

```{r}
correlate_marker_cell(gene_symbol = "CD3G", cell_type = "Tcells")
```

```{r, fig.height=8, fig.width=7}
ggpubr::ggarrange(correlate_marker_cell(gene_symbol = "MS4A1", cell_type = "Bcells"),
          correlate_marker_cell(gene_symbol = "CD3G", cell_type = "Tcells"),
          nrow = 2)
```


## P values and significance

```{r}
rstat
```

### Correlation

```{r}
rstat %>% left_join(.,rownames_to_column(anndf, "sample_name"), by= "sample_name") %>%
  ggplot(aes(x=pearsonCor, y = spearmanCor, color = PPBC)) + geom_point() +
  ggtitle("DWLS relationship spearman and pearson cor")+
  ggthemes::theme_few()
```

### P values

```{r}
rstat %>% left_join(.,rownames_to_column(anndf, "sample_name"), by= "sample_name") %>%
  ggplot(aes(x=pearsonCorPval, y = spearmanCorPval, size = dwlsPval, color=PPBC)) + geom_point() +
  ggtitle("DWLS relationship spearman and pearson p values") +
  ggthemes::theme_few()
```

```{r}
rstat
```



Stats are calculated as:

```{r, eval=F, echo=T}
sig <- as.matrix(apply(refProfiles,2,function(x) x/sum(x))[markerGenes,])

results$pearsonCor <- sapply(c(1:nrow(results)),function(x) cor(sig %*% as.numeric(results[x,c(1:ncol(sig))]),rt[,x]))

results$pearsonCorPval <- sapply(c(1:nrow(results)),
                                 function(x) cor.test(sig %*% as.numeric(results[x,c(1:ncol(sig))]),rt[,x])$p.value)

results$spearmanCor <- sapply(c(1:nrow(results)),
                              function(x) cor(sig %*% as.numeric(results[x,c(1:ncol(sig))]),rt[,x],method = "spearman"))

results$spearmanCorPval <- sapply(c(1:nrow(results)),
                                  function(x) cor.test(sig %*% as.numeric(results[x,c(1:ncol(sig))]),
                                                       rt[,x],method = "spearman")$p.value)

results$rmse <- sapply(c(1:nrow(results)),function(x) sqrt(mean((sig %*% as.numeric(results[x,c(1:ncol(sig))])-rt[,x])**2)))
```


dwlsPval calculated as:

```{r, eval=F, echo=T}
corDist <- foreach ( i = c(1:perm), .combine=c , .options.snow=opts) %dopar% {
  library(DWLS)
  set.seed(i)
  rtlist <- as.list(data.matrix(rt))
  randomMixture <- as.numeric(rtlist[sample(length(rtlist),nrow(sig))])
  randomMixture <- randomMixture/sum(randomMixture)
  cf <- solveDampenedWLS(sig,randomMixture)
  return(cor(sig %*% cf,randomMixture))
}

results$dwlsPval <- sapply(c(1:nrow(results)),function(x) sum(corDist > results$pearsonCor[x]))
```

