---
title: "Gene unity report"
author: "Kat Moore"
date: "`r Sys.Date()`"
output: 
   html_document:
    toc: yes
    toc_float: yes
    toc_depth: 4
    highlight: tango
    df_print: paged
---

This notebook will unify the various differential expression and survival analyses to identify a list of interesting genes for discussion in the manuscript.


```{r, include=F}
library(DESeq2)
library(survival)
library(ggthemes)
library(survminer)
library(pheatmap)
library(RColorBrewer)
library(edgeR)
library(glmnet)
library(openxlsx)
library(here)
library(ggpubr)
library(tidyverse)
```

# Load and summarize results {.tabset}

We aggregate all of the previous results for summary and print the number of significant hits per analysis (where applicable).

## Diffex results

Main diffex comparisons:

```{r, collapse=T}
lrt <- list(readxl::read_excel(here("results", "diffex", "06_LRT_allgenes.xlsx"),
                          sheet = "repLRT_all"))
names(lrt) <- "LRT"
pw <- here("results", "diffex", "07_pairwise_comparisons_allgenes.xlsx") %>% 
  readxl::excel_sheets() %>% 
  purrr::set_names() %>% 
  map(readxl::read_excel, path = here("results", "diffex", "07_pairwise_comparisons_allgenes.xlsx"))
#names(pw)

ovr = here("results", "diffex", "08_one_vs_rest_allgenes.xlsx") %>% 
  readxl::excel_sheets() %>% 
  purrr::set_names() %>% 
  map(readxl::read_excel, path = here("results", "diffex", "08_one_vs_rest_allgenes.xlsx"))
#names(ovr)

#Combine all results into single list.
res_list = append(lrt, pw)
res_list = append(res_list, ovr)
names(res_list) = str_remove_all(str_remove_all(names(res_list), "rep_"), "_all")
names(res_list) = str_replace(names(res_list), "_", "_vs_")
names(res_list)

```

Number of hits per group:

```{r}
lapply(res_list, function(x){
  nrow(x[x$padj <= 0.05 & abs(x$log2FoldChange) > 0.5,])
}) %>% unlist() %>% enframe("comparison", "significant genes")
```

Subgroup analysis:

```{r}
read_excel_tabs = function(path){
  tab_names = readxl::excel_sheets(path)
  ldf = lapply(tab_names, function(x) readxl::read_excel(path, sheet = x))
  names(ldf) = tab_names
  ldf <- ldf[names(ldf)!="thresholds"]
  ldf <- bind_rows(ldf, .id = "subgroup")
  return(ldf)
}

sub_reslist <- list.files(here("results/diffex"),
           pattern = "14_subgroup.*allgenes.xlsx") %>%
  set_names(str_remove(str_remove(., "14_"),
            "_allgenes.xlsx")) %>%
  map(~read_excel_tabs(here("results/diffex",.)))

names(sub_reslist)
#Manual fix
names(sub_reslist) <- str_replace(
  str_remove_all(names(sub_reslist), "diffex_|ppbc_"),
  "non_prbc", "nonprbc")
names(sub_reslist)
```

Rename to keep consistent (for functions).

```{r}
sub_reslist <- lapply(sub_reslist, function(x){
  rename(x, log2FoldChange = l2fc)
})
```


Hits per subgroup:

```{r}
lapply(sub_reslist, function(y){
  sapply(c("Basal", "Her2", "LumA", "LumB"),
         function(x) {
           sub = y[y$subgroup == x,]
           nrow(sub[sub$padj <= 0.05 & abs(sub$log2FoldChange) > 0.5,])
         })
})

```

## Genewise survival

Excluding models run on involution samples only, as there were no hits.

```{r}
gs_sheets = readxl::excel_sheets(here("results", "survival", "12_cox_allgenes.xlsx"))
gs_sheets = gs_sheets[!str_detect(gs_sheets,"inv_only")]

gw_surv <- lapply(gs_sheets, function(x) readxl::read_excel(
  here("results", "survival", "12_cox_allgenes.xlsx"), sheet = x))
names(gw_surv) = gs_sheets

gw_surv <- bind_rows(gw_surv, .id = "model")
gw_surv <- gw_surv %>%
  mutate(
    formula = if_else(
      str_detect(model, "univ"), "univariate", "multivariate"),
    outcome = if_else(
      str_detect(model, "surv"), "overall survival", "distant recurrence"),
    family = "genewise"
    ) %>%
  mutate(fdr = signif(fdr, 2),
         p.value = signif(p.value, 2),
         padj.inv_vs_rest = signif(padj.inv_vs_rest, 2),
         padj.inv_vs_nonprbc = signif(padj.inv_vs_nonprbc, 2),
         l2fc.inv_vs_rest = signif(l2fc.inv_vs_rest, 2),
         l2fc.inv_vs_nonprbc = signif(l2fc.inv_vs_nonprbc, 2))

gw_surv = gw_surv %>%
  select(model:fdr,p.value, everything())

gw_surv = gw_surv %>%
  mutate(model = recode(model,
    "univ_cox_dr" = "univDRS",
    "univ_cox_surv" = "univOS",
    "multi_cox_dr" = "multiDRS",
    "multi_cox_surv" = "multiOS"
  ))

head(gw_surv)
```

Number of hits at various thresholds:

```{r}
thresh = seq(0.05, 0.3, by = 0.05)
model = unique(gw_surv$model)


sapply(thresh, function(y){
  sapply(model, function(x){
    z = gw_surv[gw_surv$model == x,]
    nrow(z[z$fdr <= y,])
  })
}) %>% set_colnames(
  as.character(thresh)
)

rm(thresh)
rm(model)

```

## Interaction model

```{r}
int_surv <- bind_rows(
  read_csv(here("results/survival", "13_multi_interaction_os.csv")) %>%
    mutate(formula = "multivariate",
    outcome = "overall survival",
    family = "interaction"),
  read_csv(here("results/survival", "13_multi_interaction_drs.csv")) %>%
    mutate(formula = "multivariate",
    outcome = "distant recurrence",
    family = "interaction")  
)

int_surv = int_surv %>% mutate(
  model = paste0(
    if_else(formula == "multivariate",
            "multi", "uni"),
    if_else(outcome == "overall survival",
            "OS", "DRS"),
    "_Inv:Gene"
  )
) %>%
  select(model, gene_name:fdr, p.value, everything()) %>%
  mutate(fdr = signif(fdr, 2),
         p.value = signif(p.value, 2),
         padj.inv_vs_rest = signif(padj.inv_vs_rest, 2),
         padj.inv_vs_nonprbc = signif(padj.inv_vs_nonprbc, 2),
         l2fc.inv_vs_rest = signif(l2fc.inv_vs_rest, 2),
         l2fc.inv_vs_nonprbc = signif(l2fc.inv_vs_nonprbc, 2))


head(int_surv)
```

Significant hits at various thresholds:

```{r}
sapply(unique(int_surv$analysis), function(x){
  sapply(seq(0.05, 0.3, by = 0.05), function(y){
    z = int_surv[int_surv$analysis == x,]
    nrow(z[z$fdr <= y,])
  })
}) %>% t() %>%
  set_colnames(seq(0.05, 0.3, by = 0.05))
```

## Elastic net Cox

We have several versions of this analysis, predicting outcome in:

a) All study groups, using the top 1000 and top 5000 most variable genes as input
b) Involution samples only, using the top 1000 and top 5000 most variable genes as input
c) Involution samples only, using all genes as input

These are the Spearman correlations between the relative risk predicted by the model and the outcome.
The p value was created from the t stat relating Spearman's rho and the sampe size (see notebook 15).

(Remember that n is smaller for drs than for os, since samples with metastasis at diagnosis must be excluded from drs analyses.)

```{r}
enet.pvals = 
  lapply(list.files(here("results/survival"),
                  pattern = "15.*.csv", full.names = T),
       read_csv) %>%
  set_names(list.files(here("results/survival"),
                  pattern = "15.*.csv")) %>%
  bind_rows(.id = "samples") %>%
  mutate(samples = if_else(
    str_detect(samples, "15[bc]"),
    "involution only", "all samples"
  )) %>% arrange(desc(cor), pval) %>%
  select(samples, outcome, input, everything()) %>%
  mutate(tstat = signif(tstat, 2))

enet.pvals
```

Models predicting *distant recurrence* generally perform better than those predicting overall survival.
We can see that the model performs better on *involution samples only* than it does on all samples.
For os, limiting input to the top 1000 or 5000 most variable genes improves performance.
For drs, all genes perform almost as well as the top variable genes.

Based on this, and because we are particular interested in genes that are responsible for the poor outcome in involution,
the reporting here is limited to features from the involution-only analyses.

```{r}
read_excel_tabs = function(path){
  tab_names = readxl::excel_sheets(path)
  ldf = lapply(tab_names, function(x) readxl::read_excel(path, sheet = x))
  names(ldf) = tab_names
  return(ldf)
}

glm_features =
append(read_excel_tabs(here("results/survival", "15b_inv_elastic_cox_features.xlsx")),
       read_excel_tabs(here("results/survival", "15c_all_inv_elastic_cox_features.xlsx")))
```

## Feature selection visualization

```{r, fig.height = 8}
glm_df = glm_features %>%
  bind_rows(.id = "input") %>%
  mutate(outcome = if_else(str_detect(input, "os"),
                            "overall survival",
                            "distant recurrence")) %>%
  mutate(gene_input = case_when(
    str_detect(input, "1000") ~ "1000",
    str_detect(input, "5000") ~ "5000",
    str_detect(input, "all") ~ "all",
  ))

glm_df %>%
  ggplot(aes(x = mean_beta, y = n, color = gene_input, label = feature)) +
  geom_point() +
  ggrepel::geom_label_repel(show.legend = F,
    data = glm_df %>%
      filter(n > 30 & feature_type == "protein_coding")
    ) +
  facet_wrap(~outcome, ncol=1, nrow = 2)
```

Depending on the haplotype, MARVELD2 can be either protective or negative for DRS.

```{r}
glm_df %>%
  filter(outcome == "distant recurrence") %>%
  filter(feature == "MARVELD2")
```


# Load count and metadata {.tabset}

Raw counts:

```{r}
dds = readRDS(here("data/Rds/08_dds_ovr_inv_vs_rest.Rds"))
```

Gene annotation:

```{r}
gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
gx_annot = gx_annot %>%
  select(ensembl_gene_id = gene_id, gene_name, gene_type, description = gene_description) %>%
  distinct() %>%
  mutate(description = str_remove_all(description, " \\[.*\\]"))

head(gx_annot)
```

A larger gene annotation data frame that includes uniprot and entrez ids, for webscraping:

```{r}
bx_annot <- readRDS(here("shinyApp","VisualizePPBCgene","data","app_gx_annot.Rds"))
head(bx_annot)
```



## Transposed matrix for Cox regressions

Also created in notebook 12. Gene expression columns are TMM-log2 normalized.

```{r}
coxdata = readRDS(here("data", "Rds", "12_coxdata.Rds"))
colnames(coxdata)[1:30]
```

## Involution vs rest encoding

```{r}
coxdata$involution = as.numeric(coxdata$study_group == "ppbc_inv")

coxdata = coxdata %>% select(sample_name:study_group, involution, everything())

table(coxdata$involution, coxdata$PPBC)
```

## Sample data

Separate data frame containing exclusively the sample data without gene expression.

```{r}
genecol = which(colnames(coxdata)=="ENSG00000000003")
sample_data = coxdata[,1:(genecol - 1)]

head(sample_data)
```

## Gene expression df

For plotting

```{r}
geneEx = rownames_to_column(as.data.frame(coxdata[,genecol:ncol(coxdata)]),"sample_name")

head(geneEx[,1:10])

```
In gene x sample format:

```{r}
ens_mat <- t(coxdata[,genecol:ncol(coxdata)])
#dim(ens_mat)
ens_mat[1:10,1:10]
```

# Functions {.tabset}

## Gene lookup

```{r}
gene_lookup <- function(gene){
  
  if(gene %in% gx_annot$ensembl_gene_id){
    return(gene)
  } else {
    id = filter(gx_annot, gene_name == !!gene) %>%
      pull(ensembl_gene_id)
    if(length(id) == 0){
      stop("Gene name or ID not found")
    }
    return(id)
  }
  
}

gene_lookup("MS4A1")
gene_lookup("ENSG00000156738")
gene_lookup("MARVELD2")
#gene_lookup("Steve")
```

## Gene survival

```{r}
genewise_survival_results <- function(id, df = gw_surv, return_first = F){
  
  i = gene_lookup(id)
  
  if(length(i) == 1){
    return(arrange(filter(df, ensembl_gene_id == i), fdr))
  }
  
  if (length(i > 1) & return_first == T){
    i = i[1]
    warning(paste0("Multiple ensembl IDs found for ",
                   id, "returning first in series: ", i))
    return(arrange(filter(df, ensembl_gene_id == i), fdr))
  } 
  
  if (length(i > 1 & return_first == F)){
    lapply(i, function(x){
      df[df$ensembl_gene_id == x, ]
    }) %>% bind_rows() %>%
      select(ensembl_gene_id, everything()) %>%
      arrange(fdr)
  }
  
}

genewise_survival_results("MARVELD2", return_first = T)
genewise_survival_results("MARVELD2", return_first = F)
genewise_survival_results("MS4A1", return_first = F)
#genewise_survival_results("Steve", return_first = F)
```

Can also be use to extract interaction results:

```{r}
genewise_survival_results("MARVELD2", df = int_surv, return_first = F)
```

## Gene diffex

TODO: Option to return all results when multiple ensembl ids are found

```{r}
diffex_report = function(id, subgroup = F, list_reports = res_list, subgroup_reports = sub_reslist,
                         pthresh = 0.05, abslogfcthresh = 0.5){
  
  #print(paste("Adjusted p value threshold:", pthresh, ", abs(log2FoldChange) threshold:", abslogfcthresh))
  
  ensembl_id = gene_lookup(id)
  
  if(length(ensembl_id) > 1){
    ensembl_id = ensembl_id[1]
    warning(paste0("Multiple ensembl IDs found for ",
                   id, "returning first in series: ", ensembl_id))
  }
  
  if(subgroup == T){
    lr = subgroup_reports
  } else {
    lr = list_reports  
    }
  

  comp = lapply(lr,
                function(x) x[x$ensembl_gene_id==ensembl_id,]) %>%
    bind_rows(.id = "comparison")
  
  comp = comp %>% mutate(
    padj = signif(padj, 2),
    log2FoldChange = signif(log2FoldChange, 2),
    baseMean = round(baseMean, 1),
    sig = (padj <= !!pthresh & abs(log2FoldChange) > !!abslogfcthresh)
  ) %>% arrange(padj)
  
  if (subgroup == T){
    comp = comp %>% select(comparison, subgroup, sig,
                           padj, log2FoldChange, baseMean,
                           gene_name, ensembl_gene_id, gene_type)
  } else {
    comp = comp %>% select(comparison, sig,
                           padj, log2FoldChange, baseMean,
                           gene_name, ensembl_gene_id, gene_type)
  }
  
  comp  
}

diffex_report("MARVELD2", subgroup = F)
diffex_report("MARVELD2", subgroup = T)
```

## Boxplot expression

TODO: Auto-handle multiple ensembl IDs

```{r}
box_plots <- function(id, gEx = geneEx, d=dds, s = sample_data){
  
  #stopifnot(norm_type %in% c("size factor", "log2 TMM"))
  
  ensembl_id <- gene_lookup(id)
  
  if(length(ensembl_id) > 1){
    ensembl_id = ensembl_id[1]
    warning(paste0("Multiple ensembl IDs found for ",
                   id, ", returning first in series: ", ensembl_id))
  }
  
  #TMM
  df <- gEx[,c("sample_name",  ensembl_id)]
    colnames(df)[2] <- "geneEx"
  
  df <- left_join(df, s, by = "sample_name") %>%
    mutate(id = id, ensembl_id = ensembl_id)
  
  #return(df)
  
  df = df %>%
  select(sample_name, id, ensembl_id, geneEx, PPBC, PAM50,
         death = overall_survival,
         metastasis = distant_recurrence) %>%
  gather(key = "outcome", value = "value", death, metastasis) %>%
    mutate(value = as.factor(value))
  
  tmm = df %>%
    ggplot(aes(x = PPBC, y = geneEx)) +
    geom_boxplot(alpha=0, notch=F) +
    geom_jitter(width = 0.1, aes(color = value)) +
    facet_wrap(~outcome) +
    scale_color_colorblind() +
    ylab(paste0("TMM-log2 normalized expression")) +
    ggtitle(id)
  
  #size factor
  df <- plotCounts(d, gene = ensembl_id,
                   intgroup=c("PPBC", "PAM50",
                              "overall_survival",
                              "distant_recurrence"),
                   returnData = T, normalized = T)
  
  colnames(df)[colnames(df)=="count"] <- "geneEx"
  df <- rownames_to_column(df, "sample_name") %>%
    mutate(id = id, ensembl_id = ensembl_id)

  df = df %>%
  select(sample_name, id, ensembl_id, geneEx, PPBC, PAM50,
         death = overall_survival,
         metastasis = distant_recurrence) 
  df = df %>%
  gather(key = "outcome", value = "value", death, metastasis) %>%
    mutate(value = as.factor(value)) %>%
    na.omit()
  #return(df)
  sf = df %>%
    ggplot(aes(x = PPBC, y = geneEx)) +
    geom_boxplot(alpha=0, notch=F) +
    geom_jitter(width = 0.1, aes(color = value)) +
    facet_wrap(~outcome) +
    scale_color_colorblind() +
    ylab(paste0("Size-factor normalized counts")) +
    ggtitle(id)

  list(tmm = tmm, sf = sf)
}

box_plots("MARVELD2")
```

## Gene ntiles

```{r}
gene_ntile <- function(gene_id, gEx = geneEx, gene_dict = gx_annot,
                       n = 3){
  
  stopifnot(n %in% c(2,3,4))
  
  #TODO update gene_lookup this wya
  if(str_detect(gene_id, "^ENSG")){
    id <- gene_id
    name <- gene_dict[gene_dict$ensembl_gene_id == gene_id,]$gene_name
  } else {
    id <- gene_dict[gene_dict$gene_name == gene_id,]$ensembl_gene_id
    name <- gene_id
  }
  
  labels = switch(as.character(n),
                  "2" = c("low", "high"),
                  "3" = c("low", "medium", "high"),
                  "4" = c("Q1", "Q2", "Q3", "Q4")
  )
  samples <- geneEx[,"sample_name"]
  g <- geneEx[, colnames(gEx) %in% id , drop = F]
  g <- as.matrix(g)
  rownames(g) <- samples
  #return(g)
  if(ncol(g) == 0){
    stop("Gene not found in provided count matrix")
  }
  
  if(ncol(g) > 1){
    warning("Multiple ensembl ids mapped to single gene symbol. Averaging expression.")
    gene <- as.matrix(rowMeans(g))
    #return(gene)
  }
  
  #rownames(gene) <- name #Not strictly necessary, useful for debugging
  
  if(length(id) > 1){
    id = paste(id, collapse = ";")
  }
  
  gene_df <- tibble::tibble(sample_name = rownames(g),
                            gene_symbol = name,
                            ensembl_gene_id = id,
                            tmm = g[,1],
                            ntile = dplyr::ntile(g[,1], n),
                            labels = factor(dplyr::ntile(g[,1], n), labels = labels)
  )
  
  #gene_df <- dplyr::mutate(gene_df, labels = as.character(labels)) #Or survminer will complain
  
  return(gene_df)
  

}

gene_ntile("MARVELD2")%>%head()
gene_ntile("MS4A1")%>%head()

```


## OVR Kaplan Meiers

One vs rest survival curves

```{r, fig.height=7, fig.width=10}
km_ntiles_ovr <- function(gene_id, gene_dict = gx_annot, gEX = geneEx,
                          n = 3, ovr_column = "involution", line_colors = viridis::scale_color_viridis(discrete = T),
                          sampledata = sample_data, survival_type = "os", legend_positions = c("none","none", "bottom","bottom"),
                          p_method = "anova", return_list = F){
  
  stopifnot(survival_type %in% c("os", "drs"))
  stopifnot(n %in% c(2,3,4))
  stopifnot(p_method %in% c("anova", "coef"))
  stopifnot(ovr_column %in% colnames(sampledata))
  
  ntiles <- gene_ntile(gene_id = gene_id, gene_dict = gene_dict,
                       gEx = gEX, n = n)
  
  sd <- dplyr::left_join(ntiles, sampledata, by = "sample_name")
  #  return(sd)
  
  #Toggle for survival type
  if(survival_type == "os"){
    ti <- "months_overall_survival"
    ev <- "overall_survival"
    ylab <- "Overall survival probability"
    type <- "Overall survival"
  }
  
  if(survival_type == "drs"){
    ti <- "months_to_drs"
    ev <- "distant_recurrence"
    ylab <- "Distant recurrence probability"
    type <- "Distant recurrence"
  }
  
  covariates <- c("age_at_diagnosis", "grade",
                  "stage", "surgery", "radiotherapy", "strata(hormonetherapy)",
                  "chemotherapy", "herceptin", "PAM50")
  
  cov_legend <- c("age", "year of diagnosis", "grade",
                  "stage", "treatment", "PAM50")
  
  cov <- paste(covariates, collapse = "+")
  svf = paste0('Surv(time=',ti,', event=',ev,')~')
  facet_formula <- as.formula(paste0(svf, "labels"))
  #ntile_formula <- as.formula(paste0(svf, "ntile"))
  reduced =  as.formula(paste0(svf, cov))
  formula = as.formula(paste0(svf, paste0(cov, "+ntile")))
  
  reslist <- list()
  gn <- unique(sd$gene_symbol)
  
  #Split sample data into those samples belonging to the group of interest (or not)
  sd1 = sd[sd[[ovr_column]] == 1,]
  sd0 = sd[sd[[ovr_column]] == 0,]
  
  
  curv1 <- survminer::ggsurvplot(
    fit = survminer::surv_fit(facet_formula, data = as.data.frame(sd1)), 
    xlab = "Months", 
    ylab = ylab,
    legend = legend_positions[1],
    #title = paste(type, "(unadjusted), with", gn, "\nexpression in", ovr_column, "samples"),
    title = paste("Unadjusted curve for", ovr_column, "samples"),
    pval = T, #Logrank method
    )
  
  curv1 = curv1$plot #Otherwise we get a ggsurvplot object, which cannot be used with ggarrange
  
  curv1 <- curv1 +
    theme(plot.title = element_text(size = 12),
          axis.title.x=element_blank(),
          axis.title.y=element_blank()
    ) +
    line_colors
  
  reslist$curv1 <- curv1
  
  #Unadajusted kaplan meier plot in other samples
  curv0 <- survminer::ggsurvplot(
    fit = survminer::surv_fit(facet_formula, data = as.data.frame(sd0)), 
    xlab = "Months", 
    ylab = ylab,
    legend = legend_positions[2],
    #title = paste(type, "(unadjusted), with", gn, "\nexpression in other samples"),
    title = paste("Unadjusted curve for other samples"),
    pval = T, #Logrank method
    ) 
  
  curv0 <- curv0$plot #Otherwise we get a ggsurvplot object, which cannot be used with ggarrange
  
  curv0 <- curv0 +
    theme(plot.title = element_text(size = 12),
          axis.title.x=element_blank(),
          axis.title.y=element_blank()
    ) +
    line_colors
  
  reslist$curv0 <- curv0
  
  #return(reslist)
  
  #Cox models for involution group, adjusted for all covariates and ntile gene expression
  c <- survival::coxph(formula, data = sd1)
  #print(c)
  
  #Store p value associated with the coefficient of the ntile
  fit_ntile <- as.data.frame(summary(c)$coefficients)
  p_ntile <- signif(fit_ntile[rownames(fit_ntile) == "ntile", ]$`Pr(>|z|)`, 2)
  if (p_ntile < 0.001){
    p_ntile <- "< 0.001"
  }
  
  #print(p_ntile)
  #Anova p value vs reduced model
  d <- survival::coxph(reduced, data = as.data.frame(sd1))
  aod <- anova(d,c) #Conventional to list models from smallest to largest but order does not matter
  aod_p <- signif(aod[4][2,], 2)
  if (aod_p < 0.001){
    aod_p <- "< 0.001"
  }
  
  if (p_method == "anova"){
    p <- aod_p
  } else {
    p <- p_ntile
  }
  
  #Adjusted survival curve for samples belonging to group of interest
  #Use surv_adjustedcurves to get data and the then manually plot for more control over appearance
  adj_curv1 <- survminer::surv_adjustedcurves(fit = c, variable = "ntile", data = as.data.frame(sd1), method = "conditional") %>%
    left_join(select(mutate(sd1, ntile = as.factor(ntile)), ntile, labels), by = c("variable" = "ntile")) %>%
    ggplot(., aes(x = time, y = surv, color = labels)) + 
    geom_step(size = 1) +
    theme_survminer() +
    line_colors + 
    scale_y_continuous(limits = c(0,1)) +
    #scale_x_continuous(limits = c(0,250)) +
    ylab(label =  ylab) +
    labs(color = paste(gn, "expression")) +
    #ggtitle(paste(type, "for gene", gn, "in", ovr_column, "\nsamples, adjusted for clinical covariates")) +
    ggtitle(paste("Adjusted curve for", ovr_column,"samples")) +
    #annotate(geom = "text", label = paste0("Coef pval for ntile: ", p_ntile), x= 25, y =0.1) +
    #annotate(geom = "text", label = paste0("Anova pval vs reduced: ", aod_p), x= 25, y =0.2) +
    annotate(geom = "text", label = paste("p =", p), x= 25, y =0.1) +
    theme(plot.title = element_text(size = 12),
          axis.title.x=element_blank(),
          axis.title.y=element_blank(),
          legend.position = legend_positions[3]
    )
  
  reslist$adj_curv1 <- adj_curv1
  
  #Cox models for rest group, adjusted for all covariates and ntile gene expression
  c <- survival::coxph(formula, data = sd0)
  #print(c)
  
  #Store p value associated with the coefficient of the ntile
  fit_ntile <- as.data.frame(summary(c)$coefficients)
  p_ntile <- signif(fit_ntile[rownames(fit_ntile) == "ntile", ]$`Pr(>|z|)`, 2)
  if (p_ntile < 0.001){
    p_ntile <- "< 0.001"
  }
  
  #Anova p value vs reduced model
  d <- survival::coxph(reduced, data = as.data.frame(sd0))
  aod <- anova(d,c) #Conventional to list models from smallest to largest but order does not matter
  aod_p <- signif(aod[4][2,], 2)
  if (aod_p < 0.001){
    aod_p <- "< 0.001"
  }
  
  if (p_method == "anova"){
    p <- aod_p
  } else {
    p <- p_ntile
  }
  
  #Adjusted survival curve for samples belonging to group of interest
  #Use surv_adjustedcurves to get data and the then manually plot for more control over appearance
  adj_curv0 <- survminer::surv_adjustedcurves(fit = c, variable = "ntile", data = as.data.frame(sd0), method = "conditional") %>%
    left_join(select(mutate(sd0, ntile = as.factor(ntile)), ntile, labels), by = c("variable" = "ntile")) %>%
    ggplot(., aes(x = time, y = surv, color = labels)) + 
    geom_step(size = 1) + 
    theme_survminer() + 
    line_colors +
    scale_y_continuous(limits = c(0,1)) + 
    #scale_x_continuous(limits = c(0,250)) +
    ylab(label =  ylab) +
    labs(color = paste(gn, "expression")) +
    #ggtitle(paste(type, "for gene", gn, "\nin other samples, adjusted for clinical covariates")) +
    ggtitle(paste("Adjusted curve for other samples")) +
    #annotate(geom = "text", label = paste0("Coef pval for ntile: ", p_ntile), x= 25, y =0.1) +
    #annotate(geom = "text", label = paste0("Anova pval vs reduced: ", aod_p), x= 25, y =0.2) +
    annotate(geom = "text", label = paste("p =", p), x= 25, y =0.1) +
    theme(plot.title = element_text(size = 12),
          axis.title.x=element_blank(),
          axis.title.y=element_blank(),
          legend.position = legend_positions[4]
          )
  
  reslist$adj_curv0 <- adj_curv0
  
  if(return_list == T){
    return(reslist)  
  }
  
  # Arrange them all together
  arplot <- ggpubr::ggarrange(plotlist = reslist, nrow = 2, ncol = 2)
  
  arplot <- annotate_figure(arplot,
                  top = text_grob(paste(type, "for gene", gn), face = "bold", size = 14),
                  #bottom = text_grob("Time (Months)", size = 12),
                  bottom = text_grob(paste0("Time (Months)\n\nClinical covariates: ",
                                            paste(cov_legend, collapse=", ")),
                                     size = 12),
                  left = text_grob(ylab, rot = 90),
                  #right = "I'm done, thanks :-)!",
                  #fig.lab = "Figure 1", fig.lab.face = "bold"
  )
  
  return(arplot)
  
}

km_ntiles_ovr("MARVELD2")

```

## Retrieve gene summaries

Webscrape the entrez and uniprot summaries for a gene.

```{r}
get_entrez_summary <- function(id, dict){
  
  ent <- unique(bx_annot[bx_annot$gene_name == id, ]$entrez_id)
  
  if(length(ent) == 0){
    print("No ID found")
    return(NA)
  }
  
  gene_summary <- rentrez::entrez_summary(db = "gene", id = ent)
  
  res <- list()
  
  res$name <- rentrez::extract_from_esummary(gene_summary, "name")
  res$description <- rentrez::extract_from_esummary(gene_summary, "description")
  res$otheraliases <- rentrez::extract_from_esummary(gene_summary, "otheraliases")
  #res$otherdesignations <- rentrez::extract_from_esummary(gene_summary, "otherdesignations")
  res$entrez_summary <- rentrez::extract_from_esummary(gene_summary, "summary")
  return(res)
}

get_entrez_summary("MARVELD2")
```

```{r}
#TODO Make html_text work better somehow

get_uniprot_summary <- function(id){
  
  uni <- unique(bx_annot[bx_annot$gene_name == id, ]$uniprot_id)
  
  if(length(uni) == 0){
    print("No ID found")
    return(NA)
  }
  
  base_url = "https://www.uniprot.org/uniprot/"
  url = paste0(base_url, uni)
  
  uniprot_summary <- tryCatch(
  xml2::read_html(url) %>%
    rvest::html_node(xpath='/html/body/main/div/div[3]/div[3]/div[1]') %>%
    rvest::html_text(),
    error = function(e){NA} 
  )
  
  uniprot_summary <- uniprot_summary %>% str_split("[:digit:] Publications")
  uniprot_summary <- uniprot_summary[[1]][[1]]
  return(uniprot_summary)
}

get_uniprot_summary("MARVELD2")
```



# Genes of interest from genewise Cox

Or background literature.

Genes associated with negative outcome:

```{r}
gw_surv %>%
  bind_rows() %>%
  filter(fdr <= 0.05) %>%
  filter(beta > 0)
```

Genes associated with positive outcome:

```{r}
gw_surv %>%
  bind_rows() %>%
  filter(fdr <= 0.05) %>%
  filter(beta < 0)
```


# CD20 (MS4A1) {.tabset}

Summary survival

Genewise models:

```{r}
gene = "MS4A1"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```

# CD27

The protein encoded by this gene is a member of the TNF-receptor superfamily. This receptor is required for generation and long-term maintenance of T cell immunity. It binds to ligand CD70, and plays a key role in regulating B-cell activation and immunoglobulin synthesis. This receptor transduces signals that lead to the activation of NF-κB and MAPK8/JNK. Adaptor proteins TRAF2 and TRAF5 have been shown to mediate the signaling process of this receptor. CD27-binding protein (SIVA), a proapoptotic protein, can bind to this receptor and is thought to play an important role in the apoptosis induced by this receptor.[5]

In murine γδ T cells its expression has been correlated with the secretion of IFNγ.[6]
Varlilumab is an antibody that binds to CD27 and is an experimental cancer treatment.
CD27 has been shown to interact with SIVA1,[7] TRAF2[8][9] and TRAF3.[8][9] 

[The immunobiology of CD27 and OX40 and their potential as targets for cancer immunotherapy](https://www.ncbi.nlm.nih.gov/pubmed/29118006)

Summary survival

Genewise models:

```{r}
gene = "CD27"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```

# CD7

This gene encodes a transmembrane protein which is a member of the immunoglobulin superfamily. This protein is found on thymocytes and mature T cells. It plays an essential role in T-cell interactions and also in T-cell/B-cell interaction during early lymphoid development.

Summary survival

Genewise models:

```{r}
gene = "CD7"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```
# FoxP3

Summary survival

Genewise models:

```{r}
gene = "FOXP3"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```

# CD69

This gene encodes a member of the calcium dependent lectin superfamily of type II transmembrane receptors. Expression of the encoded protein is induced upon activation of T lymphocytes, and may play a role in proliferation. Furthermore, the protein may act to transmit signals in natural killer cells and platelets.

CD69 expression has been associated with both regulatory T cell (Treg), memory T cell and Bcl6 loCD69 hiLZ GC B plasmablast precursors.[11] Treg precursors exit the thymus expressing CD69 and complete differentiation into Treg cells in peripheral tissues when they encounter antigens and other cytokines, like IL-2.[12] Through the JAK/STAT signaling pathway, CD69 activation also induces the production of TGF-β as well as IL-2, which contribute to the differentiation of Treg cells as mentioned above.[8] Furthermore, CD69 is also known to be upregulated by NF-κB signaling at the onset of an immune response. A prolonged immune response is then maintained by the non-canonical NF-κB pathway, which in turn is associated with Treg differentiation.[7]

In addition to Treg differentiation, CD69 is a common marker of precursor and mature resident memory T cells (TRMs) that are localized in peripheral tissues.[13][9] TGF-β is also responsible for the development of TRMs, thus promoting TRM differentiation in a manner similar to Treg differentiation.[14] 

Summary survival

Genewise models:

```{r}
gene = "CD69"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```

# UACA

Uveal Autoantigen With Coiled-Coil Domains And Ankyrin Repeats

This gene encodes a protein that contains ankyrin repeats and coiled coil domains and likely plays a role in apoptosis. Studies in rodents have implicated the encoded protein in the stimulation of apoptosis and the regulation of mammary gland involution, in which the mammary gland returns to its pre-pregnant state. This protein has also been proposed to negatively regulate apoptosis based on experiments in human cell lines in which the protein was shown to interact with PRKC apoptosis WT1 regulator protein, also known as PAR-4, and inhibit translocation of the PAR-4 receptor. Autoantibodies to this protein have been identified in human patients with panuveitis and Graves' disease. Differential expression of this gene has been observed in various human cancers. 

Summary survival

Genewise models:

```{r}
gene = "UACA"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```

# ZNF131 {.tabset}

Plays a role during development and organogenesis as well as in the function of the adult central nervous system (By similarity). May be involved in transcriptional regulation as a repressor of ESR1/ER-alpha signaling

Summary survival

Genewise models:

```{r}
gene = "ZNF131"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```

# AHNAK2 {.tabset}

Summary survival

Genewise models:

```{r}
gene = "AHNAK2"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```

# Genes of interest from elastic net models

```{r}
glm_df %>% filter(n > 40) %>%
  arrange(desc(n), desc(abs(mean_beta)))
```

```{r, fig.height = 8}
glm_df = glm_features %>%
  bind_rows(.id = "input") %>%
  mutate(outcome = if_else(str_detect(input, "os"),
                            "overall survival",
                            "distant recurrence")) %>%
  mutate(gene_input = case_when(
    str_detect(input, "1000") ~ "1000",
    str_detect(input, "5000") ~ "5000",
    str_detect(input, "all") ~ "all",
  ))

glm_df %>%
  ggplot(aes(x = mean_beta, y = n, color = gene_input, label = feature)) +
  geom_point() +
  ggrepel::geom_label_repel(show.legend = F,
    data = glm_df %>%
      filter(n > 30 & feature_type == "protein_coding")
    ) +
  facet_wrap(~outcome, ncol=1, nrow = 2)
```



# MARVELD2 {.tabset}

Summary survival

Genewise models:

```{r}
gene = "MARVELD2"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```

# COPS7A {.tabset}

Summary survival

Genewise models:

```{r}
gene = "COPS7A"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```



# ANKRD46 {.tabset}

Summary survival

Genewise models:

```{r}
gene = "ANKRD46"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```


# FAM133A {.tabset}

Summary survival

Genewise models:

```{r}
gene = "FAM133A"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```

# CFAP73 {.tabset}

Summary survival

Genewise models:

```{r}
gene = "CFAP73"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```

# LPA {.tabset}

Summary survival

Genewise models:

```{r}
gene = "LPA"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```


# DXO {.tabset}

Summary survival

Genewise models:

```{r}
gene = "DXO"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```

```{r}
box_plots("ENSG00000204348")$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr("ENSG00000204348", ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```

# Notch4 {.tabset}

Summary survival

Genewise models:

```{r}
gene = "NOTCH4"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```



# HLA-DQB1 {.tabset}

Summary survival

Genewise models:

```{r}
gene = "HLA-DQB1"
genewise_survival_results(gene, df = gw_surv, return_first = F)
```

Interaction models:

```{r}
genewise_survival_results(gene, df = int_surv, return_first = F)
```

Summary diffex

```{r}
diffex_report(gene, subgroup = F)
```

Subgroup analysis:

```{r}
diffex_report(gene, subgroup = T)
```

## Expression

```{r, fig.width=8}
#source(here("src", "survival_tools.R"))
box_plots(gene)$tmm
```


## OS Kaplan-Meiers

One vs rest survival depictions

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "os")
```

## DRS Kaplan-Meiers

```{r, warning=F, fig.height=7, fig.width=10}
km_ntiles_ovr(gene, ovr_column = "involution", survival_type = "drs")
```

## Gene function

```{r}
get_entrez_summary(gene)
```

```{r}
get_uniprot_summary(gene)
```

















































