---
title: "Differential expression with simplified formula based on ER and HER2 status"
date: "`r Sys.Date()`"
author: "Kat Moore"
output: 
  html_document:
    toc: yes
    toc_float: yes
    toc_depth: 4
    highlight: tango
---

* Redo diffex with ER status instead of PAM50 and without batch (assuming batch and ER status are not correlated)


```{r clear}
rm(list = ls())
```

```{r libraries, include=F}
library(flexgsea) #For read_gmt()
library(ggrepel)
library(ComplexHeatmap)
library(openxlsx)
library(scrime)
library(RColorBrewer)
library(here)
library(tictoc)
library(DESeq2)
library(UpSetR)
library(apeglm)
library(tidyverse)
```

# Load data

## Gene data

```{r gene data}
#DDS with filtering and one vs rest groups pre-defined, see notebook 8
dds <- readRDS(here("data/Rds/08_dds_standard_vs_rest.Rds"))

#Variance stabilized transformed data for visualizations
vsd <- readRDS(here("data", "Rds", "08_vsd_standard_vsrest.Rds"))

gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
#head(gx_annot)
gx_annot = gx_annot %>% select(ensembl_gene_id = gene_id, gene_name, hgnc_symbol,
                               gene_type, description = gene_description) %>% distinct()
```

## Gene signatures

```{r gene signatures}
gene_sets = list(
  go_bp = flexgsea::read_gmt(here("data","external","gmt","c5.bp.v7.0.symbols.gmt")),
  hallmark = flexgsea::read_gmt(here("data","external","gmt","h.all.v7.0.symbols.gmt")),
  c2_canon = flexgsea::read_gmt(here("data","external","gmt","c2.cp.v7.0.symbols.gmt")),
  c2_cgp = flexgsea::read_gmt(here("data","external","gmt","c2.cgp.v7.0.symbols.gmt"))
)
```


## Filtering

Minimum threshold is a nonzero count in at least a 3rd of all samples

```{r filtering}
keep <- rowSums(counts(dds)!=0) >= ceiling(ncol(dds)/3)
table(keep)

dds <- dds[keep,]
```
## Retrieve immune genes

For downstream analyses, we'd like to know which of the differentially expressed genes are immunologically relevant.

An immune gene is defined as follows:

Either immune/immuno/interleukin is part of the gene name and description OR the gene is part of the the ImmPort database. See https://www.innatedb.com/redirect.do?go=resourcesGeneLists

The Immunology Database and Analysis Portal (ImmPort) system was developed under the Bioinformatics Integration Support Contract (BISC) Phase II by the Northrop Grumman Information Technology Health Solutions team for the NIH, NIAID, and DAIT. The principal investigator of the BISC project is Dr. Richard Scheuermann at University of Texas Southwestern Medical Center. The list of immunologically related genes in ImmPort is a collection of ~6,000 human genes, which was formed with the goal of retrieving all genes that have immune system-related functions. This list was generated using automatic searches of EntrezGene and Gene Ontology records using immunology-related keywords. The list was then manually curated by immunology experts examining various literature sources. 

```{r immune genes}
immune_gene_list <- read_csv(here("data", "external","gene-sets","InnateDB_genes.csv"))
head(immune_gene_list)
```

## Helper functions

```{r source}
source(here("src", "deseq_report_functions.R"))
```

## Survival categories in sample data

```{r survival categories}
sample_data <- as.data.frame(colData(dds))

print(paste("Max follow up overall survival:",
            max(sample_data$months_overall_survival, na.rm = T)))

print(paste("Max follow up distant recurrence:",
            max(sample_data$months_to_drs, na.rm = T)))


#Define classes for overall survival plotting
sample_data <- sample_data %>% mutate(survival = case_when(
  overall_survival == 1 & months_overall_survival < 50 ~ "death at 50 months",
  overall_survival == 1 & months_overall_survival < 100 ~ "death at 100 months",  
  overall_survival == 1 & months_overall_survival < 150 ~ "death at 150 months",
  overall_survival == 1 & months_overall_survival < 200 ~ "death at 200 months",
  overall_survival == 1 & months_overall_survival < 250 ~ "death at 250 months",
  overall_survival == 0 ~ "survived",
  TRUE ~ "no_data"
)) %>%
  mutate(survival = factor(survival,
                           levels = c("survived",
                                      "death at 50 months",
                                      "death at 100 months",  
                                      "death at 150 months",
                                      "death at 200 months",
                                      "death at 250 months")))

stopifnot(identical(sample_data$sample_name, colnames(dds)))
dds$survival <- sample_data$survival
stopifnot(identical(sample_data$sample_name, colnames(vsd)))
vsd$survival <- sample_data$survival

#Define classes for distant recurrence plotting
sample_data <- sample_data %>% mutate(recurrence = case_when(
  distant_recurrence == 1 & months_to_drs < 50 ~ "recurrence at 50 months",
  distant_recurrence == 1 & months_to_drs < 100 ~ "recurrence at 100 months",  
  distant_recurrence == 1 & months_to_drs < 150 ~ "recurrence at 150 months",
  distant_recurrence == 1 & months_to_drs < 200 ~ "recurrence at 200 months",
  distant_recurrence == 1 & months_to_drs < 250 ~ "recurrence at 250 months",
  distant_recurrence == 0 ~ "no recurrence",
  TRUE ~ "no_data"
)) %>%
  mutate(recurrence = factor(recurrence,
                            levels = c("no recurrence",
                                       "recurrence at 50 months",
                                       "recurrence at 100 months",  
                                       "recurrence at 150 months",
                                       "recurrence at 200 months",
                                       "recurrence at 250 months")))

dds$recurrence <- sample_data$recurrence
vsd$recurrence <- sample_data$recurrence

#dds and vsd are saved below, under "combined receptor status"
```

## Heatmap colors

Created in notebook 6

```{r load colors}
all_colors = readRDS(here("data","Rds", "06_heatmap_colors.Rds"))
study_colors = all_colors$study_colors
pam_colors = all_colors$pam_colors
gene_colors = all_colors$gene_colors
```

### Receptor colors

We also need new colors to represent IHC receptor status.

```{r receptor colors}
#RColorBrewer::brewer.pal.info
#er_colors <- suppressWarnings(colorRampPalette(brewer.pal(length(unique(sample_data$ER)), "PiYG"))(length(unique(sample_data$ER))))
er_colors <- c("lightgray", "midnightblue")
#names(er_colors) <- unique(sample_data$ER)
names(er_colors) <- c(0,1)

#her2_colors <- suppressWarnings(colorRampPalette(brewer.pal(length(unique(sample_data$HER2)), "PuOr"))(length(unique(sample_data$HER2))))
her2_colors <- c("gray", "maroon")
#names(her2_colors) <- unique(sample_data$HER2)
names(her2_colors) <- c(0,1)

#pr_colors <- suppressWarnings(colorRampPalette(brewer.pal(length(unique(sample_data$PR)), "RdGy"))(length(unique(sample_data$PR))))
pr_colors <- c("lightsteelblue1", "olivedrab")
#names(pr_colors) <- unique(sample_data$PR)
names(pr_colors) <- c(0,1)

color_grid(er_colors)
color_grid(her2_colors)
color_grid(pr_colors)

#Saved below, under "combined receptor status"
```
### Survival colors

We should also standardize the overall survival and distant recurrence colors

```{r survival colors}
os_colors = viridis::viridis(length(unique(dds$survival)))
#unique(sd$survival_group) %>% paste(collapse=", ")

names(os_colors) = c("survived", "death at 250 months", "death at 200 months",
                     "death at 150 months", "death at 100 months", "death at 50 months")
color_grid(os_colors)

drs_colors = viridis::viridis(length(unique(dds$recurrence))) 
#Ordered from best to worst
names(drs_colors) =  c("no recurrence","recurrence at 250 months","recurrence at 200 months",
                       "recurrence at 150 months","recurrence at 100 months", "recurrence at 50 months")
color_grid(drs_colors)

saveRDS(list(os_colors = os_colors,
             drs_colors = drs_colors),
        file = here("data", "Rds", "17_survival_colors.Rds"))
```
## PPBC group names

For prettier graphs

```{r}
sample_data <- sample_data %>%
  mutate(PPBC = case_when(
    study_group == "non_prbc" ~ "nulliparous",
    study_group == "prbc" ~ "pregnant",
    study_group == "ppbc_lac" ~ "lactating",
    study_group == "ppbc_inv" ~ "involuting"
  )) %>%
  mutate(PPBC = factor(PPBC,
                       levels = c(
                          "nulliparous",
                          "pregnant",
                          "lactating",
                          "involuting"
                       )))

stopifnot(identical(sample_data$sample_name, colnames(dds)))
dds$PPBC <- sample_data$PPBC
stopifnot(identical(sample_data$sample_name, colnames(vsd)))
vsd$PPBC <- sample_data$PPBC

ppbc_colors <- study_colors
names(ppbc_colors) <- c("nulliparous", "involuting", "lactating", "pregnant", "rest")

color_grid(ppbc_colors)
```

# Relationship ER/HER2/PR status and study group

We want the p value between ER status and study group to be low to avoid colinearity within the differential expression design formula. Using ER status, we can hopefully avoid colinearity without completely ignoring PAM50 subtypes.

```{r chisq new}
tbl <- table(sample_data$ER, sample_data$study_group)
chires =chisq.test(sample_data$ER, sample_data$study_group, simulate.p.value = T)
v = sqrt(chires$statistic/sum(tbl))
print(paste("ChiSq pval between ER status and study group:", round(chires$p.value, 4), ", V:", round(v,2)))
```
This is substantially better than the relationship between study group and PAM50, or between study group and batch.

```{r chisq old}
corvars = t(combn(c("batch", "PAM50", "study_group"), 2))

for (i in 1:nrow(corvars)){
  thisx = corvars[i,1]
  thisy = corvars[i,2]
  tbl = table(sample_data[,thisx], sample_data[,thisy])
  chires =chisq.test(sample_data[,thisx], sample_data[,thisy], simulate.p.value = T)
  v = sqrt(chires$statistic / sum(tbl))
  print(paste("ChiSq pval between", thisx, "and", thisy, ":", round(chires$p.value, 4), ", V:", round(v,2)))
  
}
```

Should we consider using HER2 and PR status as well?

```{r chisq receptors}
corvars = t(combn(c("ER", "HER2", "PR", "study_group"), 2))

for (i in 1:nrow(corvars)){
  thisx = corvars[i,1]
  thisy = corvars[i,2]
  tbl = table(sample_data[,thisx], sample_data[,thisy])
  chires =chisq.test(sample_data[,thisx], sample_data[,thisy], simulate.p.value = T)
  v = sqrt(chires$statistic / sum(tbl))
  print(paste("ChiSq pval between", thisx, "and", thisy, ":", round(chires$p.value, 4), ", V:", round(v,2)))
  
}
```

Looks like only ER and PR are highly associated. Let's use both ER and HER2.

```{r design}
design(dds) <- ~ER + HER2 + study_group
```

## Combined receptor status

Define category that includes both ER and HER2 status

```{r}
sample_data <- sample_data %>%
  mutate(IHC = case_when(
    ER == 0 & HER2 == 0 ~ "ER- HER2-",
    ER == 1 & HER2 == 0 ~ "ER+ HER2-",
    ER == 0 & HER2 == 1 ~ "ER- HER2+",
    ER == 1 & HER2 == 1 ~ "ER+ HER2+",
  )) %>% mutate(
    IHC = factor(IHC,
                    levels = c("ER- HER2-",
                               "ER+ HER2-",
                               "ER- HER2+",
                               "ER+ HER2+")))

stopifnot(identical(sample_data$sample_name, colnames(dds)))
stopifnot(identical(sample_data$sample_name, colnames(vsd)))

dds$IHC <- sample_data$IHC
vsd$IHC <- sample_data$IHC

saveRDS(object = dds, file = here("data","Rds","17_dds_receptorstatus.Rds"))
saveRDS(object = vsd, file = here("data","Rds","17_vsd_receptorstatus.Rds"))
```

## Combined colors

```{r}
ihc_colors <- c("lightgray", "midnightblue", "maroon", "purple")
names(ihc_colors) <- levels(sample_data$IHC)
color_grid(ihc_colors)
saveRDS(list(er_colors = er_colors, her2_colors = her2_colors, pr_colors = pr_colors,
             ihc_colors = ihc_colors), 
        here("data", "Rds", "17_receptor_colors.Rds"))
```


# Likelihood ratio test {.tabset}

Perform diffex in script 17. Load and display reports here.


```{r datadir}
dataDir = here::here("data", "Rds")
lrtPath = file.path(dataDir, "17_dds_LRT.Rds")
```


```{r lrt deseq, eval=F}
#Via script 17

if(file.exists(lrtPath) == F| overwrite == T){
  print("Likelihood ratio test vs reduced formula ~ER + HER2")
  tic("LRT")
  ddsLRT <- DESeq(dds, test="LRT", reduced= ~ ER + HER2)
  saveRDS(ddsLRT, lrtPath)
  toc()
}
```


```{r load LRT}
ddsLRT = readRDS(lrtPath)
resLRT <- results(ddsLRT)

design(ddsLRT)
resLRT@elementMetadata$description[4]
```
## Output log

>estimating size factors
using 'avgTxLength' from assays(dds), correcting for library size
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing
-- replacing outliers and refitting for 556 genes
-- DESeq argument 'minReplicatesForReplace' = 7 
-- original counts are preserved in counts(dds)
estimating dispersions
fitting model and testing
Likelihood ratio test vs reduced formula ~ER + HER2: 1077.463 sec elapsed
```r lubridate::seconds_to_period(1077.463) ```

## Summary

```{r LRT report}
source(here("src", "deseq_report_functions.R"))
#stopifnot(identical(colnames(ddsLRT), sample_data$sample_name))
#ddsLRT$PPBC <- sample_data$PPBC
#ddsLRT$survival <- sample_data$survival

repLRT = deseq_report(resLRT, dds=ddsLRT, absl2fc = 1, title="Likelihood ratio test",
                      intgroup="PPBC", groups = levels(ddsLRT$PPBC), 
                      colorby="survival",
                      top_vars = c("PPBC", "survival"),
                      top_colors = list(PPBC=ppbc_colors,
                                        survival = os_colors),
                      bottom_vars = c("ER", "HER2"),
                      bottom_colors = list(ER = er_colors,
                                           HER2 = her2_colors))

```

## Top genes 

```{r lrt top 30}
repLRT$significant_genes %>%
  select(gene_name, Type, padj, log2FoldChange, baseMean, description, everything()) %>%
  head(30)
```

Protein coding hits:

```{r lrt protein coding hits}
repLRT$significant_genes %>%
  filter(Type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name)
  
```

## Heatmap

```{r lrt heatmap, fig.height=6, fig.width=10}
repLRT$heatmap
```

## Box plots

Plot the top 5 most significant.

```{r LRT beehives}
repLRT$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r n lrt pathways}
print(paste("Number of pathways significant at fdr < 0.1:",
            (repLRT$pathways %>% bind_rows() %>%
               filter(fdr < 0.1) %>% nrow())))
```
Top 30 preview

```{r top 30 lrt pathways}
repLRT$pathways %>% bind_rows() %>%
  filter(fdr < 0.1) %>% select(pathway,everything()) %>%
  arrange(fdr) %>% head(30)
```

Plot top pathways

```{r lrt pathway plot, fig.height=6, fig.width=8}
#repLRT$pathway_plots

bind_rows(repLRT$pathways) %>%
  #filter(fdr < 0.1) %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Sig pathways fdr < 0.1: LRT study group"
      )
```

## Cook's outliers {.tabset}

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r lrt outlier volcano}
repLRT$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 6:

```{r lrt outlier bees}
repLRT$outlier_bees[1:6]
```


# Pairwise comparisons setup {.tabset}

Set up all possible comparisons:

```{r, possible comparisons, collapse=T}
combos <- t(combn(levels(dds$study_group), 2))

#Ensure that nulliparous is always the reference (2nd group) where possible
#And that involution is always the first group where possible
for (i in 1:nrow(combos)){
  first <- combos[i,1]
  second <- combos[i,2]
  if (first == "non_prbc" | second == "ppbc_inv"){
    combos[i,1] <- second
    combos[i, 2] <- first
  }
}
combos <- as.data.frame(combos)
colnames(combos) <- c("primary", "reference")
print("Pairwise comparisons:")
print(as.matrix(combos))
```

Run pairwise comparisons. (Not evaluated, was performed in script 17.)

```{r pairwise diffex, eval=F}
#Keep unanalyzed copy of dds separate
ddsR <- dds

#Pairwise loop
for (ref in unique(combos$reference)){
  
  pw_dds_file <- file.path(resDir, paste0("17_dds_pairwise_ref_", ref, ".Rds"))
  
  if (file.exists(pw_dds_file) == F | overwrite == T){
    
    #Relevel for current reference
    print(paste("Pairwise comparisons with ref:", ref))
    ddsR$study_group <- relevel(ddsR$study_group, ref = as.character(ref))
    print(design(ddsR))
    print(paste("Levels:", paste(levels(ddsR$study_group), collapse= ", ")))
    
    print("Results dds at:")
    print(pw_dds_file)
    
    #Differential expression
    #If we haven't run DESeq2 yet, or if this is the first time with this design, run it
    if(ref == unique(combos$reference)[1] | is.null(dispersions(ddsR))){
      tic(paste("DESeq2 with ref", ref))
      ddsR <- DESeq2::DESeq(ddsR)
      saveRDS(object = ddsR, file = pw_dds_file)
      toc()
    } else {
      tic(paste("NbinomWaldTest with ref", ref))
      ddsR <- DESeq2::nbinomWaldTest(ddsR)
      toc()
      saveRDS(object = ddsR, file = pw_dds_file)
    }
    

  }
  
  #Select comparisons that involve the current reference
  comps <- combos[combos$reference==ref,,drop=F]
  
  for (primary in unique(comps$primary)){
   
    pw_ape_file <- file.path(resDir, paste0("17_ape_", primary, "_vs_", ref, ".Rds"))
    
    if (file.exists(pw_ape_file) == F | overwrite == T){
      print("Reloading results from file:")
      print(pw_dds_file)
      ddsR <- readRDS(pw_dds_file)
      print(paste("Apeglm fold shrinkage for", primary, "vs", ref))
      print("Apeglm results at:")
      print(pw_ape_file)
      
      #Fold shrinkage with apeglm
      tic("apeglm")
      ape = shrinkRes(ddsR,  contrast=c("study_group", primary, ref))
      saveRDS(object = ape, file = pw_ape_file)
      toc()
      
    }
  }
}
```

## Output log

>[1] "Pairwise comparisons:"
   primary reference
1 ppbc_inv  non_prbc
2 ppbc_lac  non_prbc
3     prbc  non_prbc
4 ppbc_inv  ppbc_lac
5 ppbc_inv      prbc
6 ppbc_lac      prbc
[1] "Pairwise comparisons with ref: non_prbc"
~ER + HER2 + study_group
[1] "Levels: non_prbc, ppbc_inv, ppbc_lac, prbc"
[1] "Results dds at:"
[1] "/DATA/share/postpartumbc/data/Rds/17_dds_pairwise_ref_non_prbc.Rds"
estimating size factors
using 'avgTxLength' from assays(dds), correcting for library size
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing
-- replacing outliers and refitting for 556 genes
-- DESeq argument 'minReplicatesForReplace' = 7 
-- original counts are preserved in counts(dds)
estimating dispersions
fitting model and testing
DESeq2 with ref non_prbc: 827.901 sec elapsed
[1] "Reloading results from file:"
[1] "/DATA/share/postpartumbc/data/Rds/17_dds_pairwise_ref_non_prbc.Rds"
[1] "Apeglm fold shrinkage for ppbc_inv vs non_prbc"
[1] "Apeglm results at:"
[1] "/DATA/share/postpartumbc/data/Rds/17_ape_ppbc_inv_vs_non_prbc.Rds"
[1] "Detecting coefficient corresponding to contrast:"
[1] "Contrast name: study_group_ppbc_inv_vs_non_prbc"
[1] "Coefficient: 4"
[1] "Retrieving results from deseq dataset with contrast..."
[1] "Shrinking log2foldchanges via method apeglm"
using 'apeglm' for LFC shrinkage. If used in published research, please cite:
    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for
    sequence count data: removing the noise and preserving large differences.
    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895
apeglm: 54.433 sec elapsed
[1] "Reloading results from file:"
[1] "/DATA/share/postpartumbc/data/Rds/17_dds_pairwise_ref_non_prbc.Rds"
[1] "Apeglm fold shrinkage for ppbc_lac vs non_prbc"
[1] "Apeglm results at:"
[1] "/DATA/share/postpartumbc/data/Rds/17_ape_ppbc_lac_vs_non_prbc.Rds"
[1] "Detecting coefficient corresponding to contrast:"
[1] "Contrast name: study_group_ppbc_lac_vs_non_prbc"
[1] "Coefficient: 5"
[1] "Retrieving results from deseq dataset with contrast..."
[1] "Shrinking log2foldchanges via method apeglm"
using 'apeglm' for LFC shrinkage. If used in published research, please cite:
    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for
    sequence count data: removing the noise and preserving large differences.
    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895
apeglm: 45.86 sec elapsed
[1] "Reloading results from file:"
[1] "/DATA/share/postpartumbc/data/Rds/17_dds_pairwise_ref_non_prbc.Rds"
[1] "Apeglm fold shrinkage for prbc vs non_prbc"
[1] "Apeglm results at:"
[1] "/DATA/share/postpartumbc/data/Rds/17_ape_prbc_vs_non_prbc.Rds"
[1] "Detecting coefficient corresponding to contrast:"
[1] "Contrast name: study_group_prbc_vs_non_prbc"
[1] "Coefficient: 6"
[1] "Retrieving results from deseq dataset with contrast..."
[1] "Shrinking log2foldchanges via method apeglm"
using 'apeglm' for LFC shrinkage. If used in published research, please cite:
    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for
    sequence count data: removing the noise and preserving large differences.
    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895
apeglm: 43.669 sec elapsed
[1] "Pairwise comparisons with ref: ppbc_lac"
~ER + HER2 + study_group
[1] "Levels: ppbc_lac, non_prbc, ppbc_inv, prbc"
[1] "Results dds at:"
[1] "/DATA/share/postpartumbc/data/Rds/17_dds_pairwise_ref_ppbc_lac.Rds"
found results columns, replacing these
[1] "Reloading results from file:"
[1] "/DATA/share/postpartumbc/data/Rds/17_dds_pairwise_ref_ppbc_lac.Rds"
[1] "Apeglm fold shrinkage for ppbc_inv vs ppbc_lac"
[1] "Apeglm results at:"
[1] "/DATA/share/postpartumbc/data/Rds/17_ape_ppbc_inv_vs_ppbc_lac.Rds"
[1] "Detecting coefficient corresponding to contrast:"
[1] "Contrast name: study_group_ppbc_inv_vs_ppbc_lac"
[1] "Coefficient: 5"
[1] "Retrieving results from deseq dataset with contrast..."
[1] "Shrinking log2foldchanges via method apeglm"
using 'apeglm' for LFC shrinkage. If used in published research, please cite:
    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for
    sequence count data: removing the noise and preserving large differences.
    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895
apeglm: 43.99 sec elapsed
[1] "Pairwise comparisons with ref: prbc"
~ER + HER2 + study_group
[1] "Levels: prbc, non_prbc, ppbc_inv, ppbc_lac"
[1] "Results dds at:"
[1] "/DATA/share/postpartumbc/data/Rds/17_dds_pairwise_ref_prbc.Rds"
estimating size factors
using 'avgTxLength' from assays(dds), correcting for library size
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing
-- replacing outliers and refitting for 557 genes
-- DESeq argument 'minReplicatesForReplace' = 7 
-- original counts are preserved in counts(dds)
estimating dispersions
fitting model and testing
DESeq2 with ref prbc: 967.081 sec elapsed
[1] "Reloading results from file:"
[1] "/DATA/share/postpartumbc/data/Rds/17_dds_pairwise_ref_prbc.Rds"
[1] "Apeglm fold shrinkage for ppbc_inv vs prbc"
[1] "Apeglm results at:"
[1] "/DATA/share/postpartumbc/data/Rds/17_ape_ppbc_inv_vs_prbc.Rds"
[1] "Detecting coefficient corresponding to contrast:"
[1] "Contrast name: study_group_ppbc_inv_vs_prbc"
[1] "Coefficient: 5"
[1] "Retrieving results from deseq dataset with contrast..."
[1] "Shrinking log2foldchanges via method apeglm"
using 'apeglm' for LFC shrinkage. If used in published research, please cite:
    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for
    sequence count data: removing the noise and preserving large differences.
    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895
apeglm: 46.604 sec elapsed
[1] "Reloading results from file:"
[1] "/DATA/share/postpartumbc/data/Rds/17_dds_pairwise_ref_prbc.Rds"
[1] "Apeglm fold shrinkage for ppbc_lac vs prbc"
[1] "Apeglm results at:"
[1] "/DATA/share/postpartumbc/data/Rds/17_ape_ppbc_lac_vs_prbc.Rds"
[1] "Detecting coefficient corresponding to contrast:"
[1] "Contrast name: study_group_ppbc_lac_vs_prbc"
[1] "Coefficient: 6"
[1] "Retrieving results from deseq dataset with contrast..."
[1] "Shrinking log2foldchanges via method apeglm"
using 'apeglm' for LFC shrinkage. If used in published research, please cite:
    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for
    sequence count data: removing the noise and preserving large differences.
    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895

# Involuting vs nulliparous {.tabset}

```{r load inv vs nonprbc, collapse=T}
#list.files(here("data", "Rds"))[str_detect(list.files(here("data", "Rds")), "17_")]

dds = readRDS(here("data/Rds/17_dds_pairwise_ref_non_prbc.Rds"))
ape = readRDS(here("data/Rds/17_ape_ppbc_inv_vs_non_prbc.Rds"))

design(dds)
dds$study_group %>% levels()
ape@elementMetadata$description[4]
```


## Summary

```{r inv vs nonprbc report}
source(here("src", "deseq_report_functions.R"))

rep_inv_nonprbc = deseq_report(ape, dds=dds, absl2fc = 0.5, title="Involuting vs nulliparous",
                               intgroup="PPBC", groups = c("involuting", "nulliparous"), 
                               colorby="survival",
                               top_vars = c("PPBC", "survival"),
                               top_colors = list(PPBC=ppbc_colors,
                                                 survival = os_colors),
                               bottom_vars = c("ER", "HER2"),
                               bottom_colors = list(ER = er_colors,
                                                    HER2 = her2_colors))

```

## Top genes 

```{r inv vs nonprbc top 30}
rep_inv_nonprbc$significant_genes %>%
  select(gene_name, Type, padj, log2FoldChange, baseMean, description, everything()) %>%
  head(30)
```
Protein coding hits:

```{r inv vs nonprbc protein coding hits}
rep_inv_nonprbc$significant_genes %>%
  filter(Type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name)
  
```

## Heatmap

```{r inv vs nonprbc heatmap, fig.height=6, fig.width=10}
rep_inv_nonprbc$heatmap
```

## Volcano plot

```{r inv vs nonprbc volcano}
rep_inv_nonprbc$volcano_plot
```


## Box plots

Plot the top 5 most significant.

```{r inv vs nonprbc beehives}
rep_inv_nonprbc$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r n inv vs nonprbc pathways}
print(paste("Number of pathways significant at fdr < 0.1:",
            (rep_inv_nonprbc$pathways %>% bind_rows() %>%
               filter(fdr < 0.1) %>% nrow())))
```
Top 30 preview

```{r top 30 inv vs nonprbc pathways}
rep_inv_nonprbc$pathways %>% bind_rows() %>%
  filter(fdr < 0.1) %>% select(pathway,everything()) %>%
  arrange(fdr) %>% head(30)
```

Plot top pathways

```{r inv vs nonprbc pathway plot, fig.height=6, fig.width=8}
#repLRT$pathway_plots

bind_rows(rep_inv_nonprbc$pathways) %>%
  #filter(fdr < 0.1) %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Sig pathways fdr < 0.1: Involution vs nulliparous"
      )
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r inv vs nonprbc outlier volcano}
rep_inv_nonprbc$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 6:

```{r inv vs nonpbrc outlier bees}
rep_inv_nonprbc$outlier_bees[1:6]
```

# Lactating vs nulliparous {.tabset}

```{r load lac vs nonprbc}
#list.files(here("data", "Rds"))[str_detect(list.files(here("data", "Rds")), "17_")]

dds = readRDS(here("data/Rds/17_dds_pairwise_ref_non_prbc.Rds"))
ape = readRDS(here("data/Rds/17_ape_ppbc_lac_vs_non_prbc.Rds"))

design(dds)
dds$study_group %>% levels()
ape@elementMetadata$description[4]
```


## Summary

```{r lac vs nonprbc report}
source(here("src", "deseq_report_functions.R"))

rep_lac_nonprbc = deseq_report(ape, dds=dds, absl2fc = 0.5, title="Lactating vs nulliparous",
                               intgroup="PPBC", groups = c("lactating", "nulliparous"), 
                               colorby="survival",
                               top_vars = c("PPBC", "survival"),
                               top_colors = list(PPBC=ppbc_colors,
                                                 survival = os_colors),
                               bottom_vars = c("ER", "HER2"),
                               bottom_colors = list(ER = er_colors,
                                                    HER2 = her2_colors))

```

## Top genes 

```{r lac vs nonprbc top 30}
rep_lac_nonprbc$significant_genes %>%
  select(gene_name, Type, padj, log2FoldChange, baseMean, description, everything()) %>%
  head(30)
```
Protein coding hits:

```{r lac vs nonprbc protein coding hits}
rep_lac_nonprbc$significant_genes %>%
  filter(Type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name)
  
```

## Heatmap

```{r lac vs nonprbc heatmap, fig.height=6, fig.width=10}
rep_lac_nonprbc$heatmap
```

## Volcano plot

```{r lac vs nonprbc volcano}
rep_lac_nonprbc$volcano_plot
```


## Box plots

Plot the top 5 most significant.

```{r lac vs nonprbc beehives}
rep_lac_nonprbc$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r n lac vs nonprbc pathways}
print(paste("Number of pathways significant at fdr < 0.1:",
            (rep_lac_nonprbc$pathways %>% bind_rows() %>%
               filter(fdr < 0.1) %>% nrow())))
```
Top 30 preview

```{r top 30 lac vs nonprbc pathways}
rep_lac_nonprbc$pathways %>% bind_rows() %>%
  filter(fdr < 0.1) %>% select(pathway,everything()) %>%
  arrange(fdr) %>% head(30)
```

Plot top pathways

```{r lac vs nonprbc pathway plot, fig.height=6, fig.width=8}
#repLRT$pathway_plots

bind_rows(rep_lac_nonprbc$pathways) %>%
  #filter(fdr < 0.1) %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Sig pathways fdr < 0.1: Lactating vs nulliparous"
      )
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r lac vs nonprbc outlier volcano}
rep_lac_nonprbc$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 6:

```{r lac vs nonpbrc outlier bees}
rep_lac_nonprbc$outlier_bees[1:6]
```

# Pregnant vs nulliparous {.tabset}

```{r load prbc vs nonprbc}
#combos
#list.files(here("data", "Rds"))[str_detect(list.files(here("data", "Rds")), "17_")]

dds = readRDS(here("data/Rds/17_dds_pairwise_ref_non_prbc.Rds"))
ape = readRDS(here("data/Rds/17_ape_prbc_vs_non_prbc.Rds"))

design(dds)
dds$study_group %>% levels()
ape@elementMetadata$description[4]
```


## Summary

```{r prbc vs nonprbc report}
source(here("src", "deseq_report_functions.R"))

rep_prbc_nonprbc = deseq_report(ape, dds=dds, absl2fc = 0.5, title="Pregnant vs nulliparous",
                               intgroup="PPBC", groups = c("pregnant", "nulliparous"), 
                               colorby="survival",
                               top_vars = c("PPBC", "survival"),
                               top_colors = list(PPBC=ppbc_colors,
                                                 survival = os_colors),
                               bottom_vars = c("ER", "HER2"),
                               bottom_colors = list(ER = er_colors,
                                                    HER2 = her2_colors))

```

## Top genes 

```{r prbc vs nonprbc top 30}
rep_prbc_nonprbc$significant_genes %>%
  select(gene_name, Type, padj, log2FoldChange, baseMean, description, everything()) %>%
  head(30)
```

Protein coding hits:

```{r prbc vs nonprbc protein coding hits}
rep_prbc_nonprbc$significant_genes %>%
  filter(Type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name)
```

## Heatmap

```{r prbc vs nonprbc heatmap, fig.height=6, fig.width=10}
rep_prbc_nonprbc$heatmap
```

## Volcano plot

```{r prbc vs nonprbc volcano}
rep_prbc_nonprbc$volcano_plot
```


## Box plots

Plot the top 5 most significant.

```{r prbc vs nonprbc beehives}
rep_prbc_nonprbc$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r n prbc vs nonprbc pathways}
print(paste("Number of pathways significant at fdr < 0.1:",
            (rep_prbc_nonprbc$pathways %>% bind_rows() %>%
               filter(fdr < 0.1) %>% nrow())))
```
Top 30 preview

```{r top 30 prbc vs nonprbc pathways}
rep_prbc_nonprbc$pathways %>% bind_rows() %>%
  filter(fdr < 0.1) %>% select(pathway,everything()) %>%
  arrange(fdr) %>% head(30)
```

Plot top pathways

```{r prbc vs nonprbc pathway plot, fig.height=6, fig.width=8}
#repLRT$pathway_plots

bind_rows(rep_prbc_nonprbc$pathways) %>%
  #filter(fdr < 0.1) %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Sig pathways fdr < 0.1: Pregnant vs nulliparous"
      )
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r prbc vs nonprbc outlier volcano}
rep_prbc_nonprbc$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 6:

```{r prbc vs nonpbrc outlier bees}
rep_prbc_nonprbc$outlier_bees[1:6]
```

# Involuting vs lactating {.tabset}

```{r load inv vs lac}
#combos
#list.files(here("data", "Rds"))[str_detect(list.files(here("data", "Rds")), "17_")]

dds = readRDS(here("data/Rds/17_dds_pairwise_ref_ppbc_lac.Rds"))
ape = readRDS(here("data/Rds/17_ape_ppbc_inv_vs_ppbc_lac.Rds"))

design(dds)
dds$study_group %>% levels()
ape@elementMetadata$description[4]
```


## Summary

```{r inv vs lac report}
source(here("src", "deseq_report_functions.R"))

rep_inv_lac = deseq_report(ape, dds=dds, absl2fc = 0.5, title="Involuting vs lactating",
                           intgroup="PPBC", groups = c("involuting", "lactating"), 
                           colorby="survival",
                           top_vars = c("PPBC", "survival"),
                           top_colors = list(PPBC=ppbc_colors,
                                             survival = os_colors),
                           bottom_vars = c("ER", "HER2"),
                           bottom_colors = list(ER = er_colors,
                                                HER2 = her2_colors))

```

## Top genes 

```{r inv vs lac top 30}
rep_inv_lac$significant_genes %>%
  select(gene_name, Type, padj, log2FoldChange, baseMean, description, everything()) %>%
  head(30)
```

Protein coding hits:

```{r inv vs lac protein coding hits}
rep_inv_lac$significant_genes %>%
  filter(Type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name)
```

## Heatmap

```{r inv vs lac heatmap, fig.height=6, fig.width=10}
rep_inv_lac$heatmap
```

## Volcano plot

```{r inv vs lac volcano}
rep_inv_lac$volcano_plot
```


## Box plots

Plot the top 5 most significant.

```{r inv vs lac beehives}
rep_inv_lac$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r n inv vs lac pathways}
print(paste("Number of pathways significant at fdr < 0.1:",
            (rep_inv_lac$pathways %>% bind_rows() %>%
               filter(fdr < 0.1) %>% nrow())))
```
Top 30 preview

```{r top 30 inv vs lac pathways}
rep_inv_lac$pathways %>% bind_rows() %>%
  filter(fdr < 0.1) %>% select(pathway,everything()) %>%
  arrange(fdr) %>% head(30)
```

Plot top pathways

```{r inv vs lac pathway plot, fig.height=6, fig.width=8}
#repLRT$pathway_plots

bind_rows(rep_inv_lac$pathways) %>%
  #filter(fdr < 0.1) %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Sig pathways fdr < 0.1: inv vs lac"
      )
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r inv vs lac outlier volcano}
rep_inv_lac$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 6:

```{r inv vs lac outlier bees}
rep_inv_lac$outlier_bees[1:6]
```

# Involuting vs pregnant {.tabset}

```{r load inv vs prbc}
#combos
#list.files(here("data", "Rds"))[str_detect(list.files(here("data", "Rds")), "17_")]

dds = readRDS(here("data/Rds/17_dds_pairwise_ref_prbc.Rds"))
ape = readRDS(here("data/Rds/17_ape_ppbc_inv_vs_prbc.Rds"))

design(dds)
dds$study_group %>% levels()
ape@elementMetadata$description[4]
```


## Summary

```{r inv vs prbc report}
source(here("src", "deseq_report_functions.R"))

rep_inv_prbc = deseq_report(ape, dds=dds, absl2fc = 0.5, title="Involuting vs pregnant",
                               intgroup="PPBC", groups = c("involuting", "pregnant"), 
                               colorby="survival",
                               top_vars = c("PPBC", "survival"),
                               top_colors = list(PPBC=ppbc_colors,
                                                 survival = os_colors),
                               bottom_vars = c("ER", "HER2"),
                               bottom_colors = list(ER = er_colors,
                                                    HER2 = her2_colors))

```

## Top genes 

```{r inv vs prbc top 30}
rep_inv_prbc$significant_genes %>%
  select(gene_name, Type, padj, log2FoldChange, baseMean, description, everything()) %>%
  head(30)
```

Protein coding hits:

```{r inv vs prbc protein coding hits}
rep_inv_prbc$significant_genes %>%
  filter(Type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name)
```

## Heatmap

```{r inv vs prbc heatmap, fig.height=6, fig.width=10}
rep_inv_prbc$heatmap
```

## Volcano plot

```{r inv vs prbc volcano}
rep_inv_prbc$volcano_plot
```


## Box plots

Plot the top 5 most significant.

```{r inv vs prbc beehives}
rep_inv_prbc$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r n inv vs prbc pathways}
print(paste("Number of pathways significant at fdr < 0.1:",
            (rep_inv_prbc$pathways %>% bind_rows() %>%
               filter(fdr < 0.1) %>% nrow())))
```

Top 30 preview

```{r top 30 inv vs prbc pathways}
rep_inv_prbc$pathways %>% bind_rows() %>%
  filter(fdr < 0.1) %>% select(pathway,everything()) %>%
  arrange(fdr) %>% head(30)
```

Plot top pathways

```{r inv vs prbc pathway plot, fig.height=6, fig.width=8}
#repLRT$pathway_plots

bind_rows(rep_inv_prbc$pathways) %>%
  #filter(fdr < 0.1) %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Sig pathways fdr < 0.1: Involuting vs pregnant"
      )
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r inv vs prbc outlier volcano}
rep_inv_prbc$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 6:

```{r inv vs prbc outlier bees}
rep_inv_prbc$outlier_bees[1:6]
```

# Lactating vs pregnant {.tabset}

```{r load lac vs prbc}
#combos
#list.files(here("data", "Rds"))[str_detect(list.files(here("data", "Rds")), "17_")]

dds = readRDS(here("data/Rds/17_dds_pairwise_ref_prbc.Rds"))
ape = readRDS(here("data/Rds/17_ape_ppbc_lac_vs_prbc.Rds"))

design(dds)
dds$study_group %>% levels()
ape@elementMetadata$description[4]
```

## Summary

```{r lac vs prbc report}
source(here("src", "deseq_report_functions.R"))

rep_lac_prbc = deseq_report(ape, dds=dds, absl2fc = 0.5, title="Lactating vs pregnant",
                               intgroup="PPBC", groups = c("lactating", "pregnant"), 
                               colorby="survival",
                               top_vars = c("PPBC", "survival"),
                               top_colors = list(PPBC=ppbc_colors,
                                                 survival = os_colors),
                               bottom_vars = c("ER", "HER2"),
                               bottom_colors = list(ER = er_colors,
                                                    HER2 = her2_colors))

```

## Top genes 

```{r lac vs prbc top 30}
rep_lac_prbc$significant_genes %>%
  select(gene_name, Type, padj, log2FoldChange, baseMean, description, everything()) %>%
  head(30)
```

Protein coding hits:

```{r lac vs prbc protein coding hits}
rep_lac_prbc$significant_genes %>%
  filter(Type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name)
```

## Heatmap

```{r lac vs prbc heatmap, fig.height=6, fig.width=10}
rep_lac_prbc$heatmap
```

## Volcano plot

```{r lac vs prbc volcano}
rep_lac_prbc$volcano_plot
```


## Box plots

Plot the top 5 most significant.

```{r lac vs prbc beehives}
rep_lac_prbc$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r n lac vs prbc pathways}
print(paste("Number of pathways significant at fdr < 0.1:",
            (rep_lac_prbc$pathways %>% bind_rows() %>%
               filter(fdr < 0.1) %>% nrow())))
```

Top 30 preview

```{r top 30 lac vs prbc pathways}
rep_lac_prbc$pathways %>% bind_rows() %>%
  filter(fdr < 0.1) %>% select(pathway,everything()) %>%
  arrange(fdr) %>% head(30)
```

Plot top pathways

```{r lac vs prbc pathway plot, fig.height=6, fig.width=8}
#repLRT$pathway_plots

bind_rows(rep_lac_prbc$pathways) %>%
  #filter(fdr < 0.1) %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Sig pathways fdr < 0.1: Lactating vs pregnant"
      )
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r lac vs prbc outlier volcano}
rep_lac_prbc$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 6:

```{r lac vs prbc outlier bees}
rep_lac_prbc$outlier_bees[1:6]
```

# UpSet plots for pairwise comparisons

An alternative to Venn diagrams.

Existing combinations:

```{r}
combos
```

Create list of significant genes from these combinations

```{r}
sig_genes <- list(
  "involuting vs nulliparous" = rep_inv_nonprbc$significant_genes$ensembl_gene_id,
  "involuting vs pregnant" = rep_inv_prbc$significant_genes$ensembl_gene_id,
  "involuting vs lactating" = rep_inv_lac$significant_genes$ensembl_gene_id,
  "pregnant vs nulliparous" = rep_prbc_nonprbc$significant_genes$ensembl_gene_id,
  "lactating vs nulliparous" = rep_lac_nonprbc$significant_genes$ensembl_gene_id,
  "lactating vs pregnant" = rep_lac_prbc$significant_genes$ensembl_gene_id
)
```

```{r, fig.width=8}
upset_pairwise = upset(fromList(sig_genes), order.by = "freq", #empty.intersections = "on",
                       nsets = 6, set_size.show = TRUE, mainbar.y.label = "Differentially expressed genes",
                       sets.bar.color = "blue", main.bar.color = "black", 
                       #set_size.scale_max = 600,
                       sets.x.label = "Sig genes pairwise comparisons")

upset_pairwise
```

```{r, echo=F}
pairwise_figdir <- here("results", "diffex", "figs", "17_pairwise_ERHER2")

dir.create(pairwise_figdir, showWarnings = F)

pdf(file.path(pairwise_figdir, "17_upset_pairwise_ERHER2.pdf"),
    onefile = F, width = 10, height = 7)
upset_pairwise
dev.off()
```


# One vs rest setup {.tabset}

```{r, ovr overview, collapse=T}
overwrite <- F
ovr_comps <- colnames(colData(dds))[str_detect(colnames(colData(dds)), "vs_rest")]

print("One vs rest comparisons:")
print(paste(ovr_comps, collapse = ", "))
```
Loop through each comparison. (Not evaluated: run in script 17).

```{r, eval=F}
#Keep unanalyzed copy separate
ddsOVR <- dds

for (ovr in ovr_comps){

  design(ddsOVR) <- as.formula(paste("~ ", paste(c("ER", "HER2", ovr), collapse= "+")))
  ovr_dds_file <- file.path(resDir, paste0("17_dds_ovr_", ovr, ".Rds"))

  if(file.exists(ovr_dds_file) ==F | overwrite == T){
    
    print(paste("One vs rest comparison with design:"))
    print(design(ddsOVR))
    print("Results file at:")
    print(ovr_dds_file)
    
    #Run DESeq2 if this is the first run with this design, or if dispersions are lacking
    if(ovr == ovr_comps[1] | is.null(dispersions(ddsOVR))){
      tic(paste("DESeq with:", ovr))
      ddsOVR <- DESeq2::DESeq(ddsOVR)
      saveRDS(object = ddsOVR, file = ovr_dds_file)
      toc()
    } else {
      tic(paste("NbinomWaldTest with:", ovr))
      ddsOVR <- DESeq2::nbinomWaldTest(ddsOVR)
      saveRDS(object = ddsOVR, file = ovr_dds_file)
      toc()
    }
    
  }
  
}
```

## Output log

>[1] "One vs rest comparisons:"
[1] "inv_vs_rest, prbc_vs_rest, lac_vs_rest, nonprbc_vs_rest"
[1] "One vs rest comparison with design:"
~ER + HER2 + inv_vs_rest
[1] "Results file at:"
[1] "/DATA/share/postpartumbc/data/Rds/17_dds_ovr_inv_vs_rest.Rds"
estimating size factors
using 'avgTxLength' from assays(dds), correcting for library size
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing
-- replacing outliers and refitting for 974 genes
-- DESeq argument 'minReplicatesForReplace' = 7 
-- original counts are preserved in counts(dds)
estimating dispersions
fitting model and testing
DESeq with: inv_vs_rest: 841.796 sec elapsed
[1] "One vs rest comparison with design:"
~ER + HER2 + prbc_vs_rest
[1] "Results file at:"
[1] "/DATA/share/postpartumbc/data/Rds/17_dds_ovr_prbc_vs_rest.Rds"
found results columns, replacing these
NbinomWaldTest with: prbc_vs_rest: 279.345 sec elapsed
[1] "One vs rest comparison with design:"
~ER + HER2 + lac_vs_rest
[1] "Results file at:"
[1] "/DATA/share/postpartumbc/data/Rds/17_dds_ovr_lac_vs_rest.Rds"
found results columns, replacing these
NbinomWaldTest with: lac_vs_rest: 266.679 sec elapsed
[1] "One vs rest comparison with design:"
~ER + HER2 + nonprbc_vs_rest
[1] "Results file at:"
[1] "/DATA/share/postpartumbc/data/Rds/17_dds_ovr_nonprbc_vs_rest.Rds"
found results columns, replacing these
NbinomWaldTest with: nonprbc_vs_rest: 260.084 sec elapsed

# Involuting vs rest {.tabset}

```{r load inv vs rest, collapse=T}
#list.files(here("data", "Rds"))[str_detect(list.files(here("data", "Rds")), "17_")]

dds = readRDS(here("data/Rds/17_dds_ovr_inv_vs_rest.Rds"))
ape = readRDS(here("data/Rds/17_ape_inv_vs_rest.Rds"))

design(dds)
dds$inv_vs_rest %>% levels()
ape@elementMetadata$description[4]

#Nicer column names
dds$inv_vs_rest <- dds$inv_vs_rest %>%
  dplyr::recode_factor(., "ppbc_inv" = "involuting")
vsd$inv_vs_rest <- vsd$inv_vs_rest %>%
  dplyr::recode_factor(., "ppbc_inv" = "involuting")

#Temporary fix until rerun makes this unnecessary
stopifnot(identical(colnames(dds), sample_data$sample_name))
dds$PPBC <- sample_data$PPBC
dds$survival <- sample_data$survival
```


## Summary

```{r inv vs rest report}
source(here("src", "deseq_report_functions.R"))

rep_inv_rest = deseq_report(ape, dds=dds, absl2fc = 0.5, title="Involuting vs rest",
                            intgroup="inv_vs_rest", groups = c("involuting", "rest"), 
                            colorby="survival",
                            top_vars = c("inv_vs_rest", "survival"),
                            top_colors = list(inv_vs_rest=ppbc_colors,
                                              survival = os_colors),
                            bottom_vars = c("ER", "HER2"),
                            bottom_colors = list(ER = er_colors,
                                                 HER2 = her2_colors),
                            legend_title = "PPBC"
)

```

## Top genes 

```{r inv vs rest top 30}
rep_inv_rest$significant_genes %>%
  select(gene_name, Type, padj, log2FoldChange, baseMean, description, everything()) %>%
  head(30)
```
Protein coding hits:

```{r inv vs rest protein coding hits}
rep_inv_rest$significant_genes %>%
  filter(Type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name)
  
```

## Heatmap

```{r inv vs rest heatmap, fig.height=6, fig.width=10}
rep_inv_rest$heatmap
```

## Volcano plot

```{r inv vs rest volcano}
rep_inv_rest$volcano_plot
```


## Box plots

Plot the top 5 most significant.

```{r inv vs rest beehives}
rep_inv_rest$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r n inv vs rest pathways}
print(paste("Number of pathways significant at fdr < 0.1:",
            (rep_inv_rest$pathways %>% bind_rows() %>%
               filter(fdr < 0.1) %>% nrow())))
```
Top 30 preview

```{r top 30 inv vs rest pathways}
rep_inv_rest$pathways %>% bind_rows() %>%
  filter(fdr < 0.1) %>% select(pathway,everything()) %>%
  arrange(fdr) %>% head(30)
```

Plot top pathways

```{r inv vs rest pathway plot, fig.height=6, fig.width=8}
#repLRT$pathway_plots

bind_rows(rep_inv_rest$pathways) %>%
  #filter(fdr < 0.1) %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Sig pathways fdr < 0.1: Involution vs rest"
      )
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r inv vs rest outlier volcano}
rep_inv_rest$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 6:

```{r inv vs rest outlier bees}
rep_inv_rest$outlier_bees[1:6]
```

# Pregnant vs rest {.tabset}

```{r load prbc vs rest, collapse=T}
#list.files(here("data", "Rds"))[str_detect(list.files(here("data", "Rds")), "17_")]

dds = readRDS(here("data/Rds/17_dds_ovr_prbc_vs_rest.Rds"))
ape = readRDS(here("data/Rds/17_ape_prbc_vs_rest.Rds"))

design(dds)
dds$prbc_vs_rest %>% levels()
ape@elementMetadata$description[4]

#Nicer column names
dds$prbc_vs_rest <- dds$prbc_vs_rest %>%
  dplyr::recode_factor(., "prbc" = "pregnant")
vsd$prbc_vs_rest <- vsd$prbc_vs_rest %>%
  dplyr::recode_factor(., "prbc" = "pregnant")

#Temporary fix until rerun makes this unnecessary
stopifnot(identical(colnames(dds), sample_data$sample_name))
dds$PPBC <- sample_data$PPBC
dds$survival <- sample_data$survival
```

## Summary

```{r prbc vs rest report}
source(here("src", "deseq_report_functions.R"))

rep_prbc_rest = deseq_report(ape, dds=dds, absl2fc = 0.5, title="Pregnant vs rest",
                            intgroup="prbc_vs_rest", groups = c("pregnant", "rest"), 
                            colorby="survival",
                            top_vars = c("prbc_vs_rest", "survival"),
                            top_colors = list(prbc_vs_rest=ppbc_colors,
                                              survival = os_colors),
                            bottom_vars = c("ER", "HER2"),
                            bottom_colors = list(ER = er_colors,
                                                 HER2 = her2_colors),
                            legend_title = "PPBC"
)

```

## Top genes 

```{r prbc vs rest top 30}
rep_prbc_rest$significant_genes %>%
  select(gene_name, Type, padj, log2FoldChange, baseMean, description, everything()) %>%
  head(30)
```
Protein coding hits:

```{r prbc vs rest protein coding hits}
rep_prbc_rest$significant_genes %>%
  filter(Type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name)
  
```

## Heatmap

```{r prbc vs rest heatmap, fig.height=6, fig.width=10}
rep_prbc_rest$heatmap
```

## Volcano plot

```{r prbc vs rest volcano}
rep_prbc_rest$volcano_plot
```


## Box plots

Plot the top 5 most significant.

```{r prbc vs rest beehives}
rep_prbc_rest$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r n prbc vs rest pathways}
print(paste("Number of pathways significant at fdr < 0.1:",
            (rep_prbc_rest$pathways %>% bind_rows() %>%
               filter(fdr < 0.1) %>% nrow())))
```
Top 30 preview

```{r top 30 prbc vs rest pathways}
rep_prbc_rest$pathways %>% bind_rows() %>%
  filter(fdr < 0.1) %>% select(pathway,everything()) %>%
  arrange(fdr) %>% head(30)
```

Plot top pathways

```{r prbc vs rest pathway plot, fig.height=6, fig.width=8}
#repLRT$pathway_plots

bind_rows(rep_prbc_rest$pathways) %>%
  #filter(fdr < 0.1) %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Sig pathways fdr < 0.1: Pregnant vs rest"
      )
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r prbc vs rest outlier volcano}
rep_prbc_rest$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 6:

```{r prbc vs rest outlier bees}
rep_prbc_rest$outlier_bees[1:6]
```

# Lactating vs rest {.tabset}

```{r load lac vs rest, collapse=T}
#list.files(here("data", "Rds"))[str_detect(list.files(here("data", "Rds")), "17_")]

dds = readRDS(here("data/Rds/17_dds_ovr_lac_vs_rest.Rds"))
ape = readRDS(here("data/Rds/17_ape_lac_vs_rest.Rds"))

design(dds)
dds$lac_vs_rest %>% levels()
ape@elementMetadata$description[4]

#Nicer column names
dds$lac_vs_rest <- dds$lac_vs_rest %>%
  dplyr::recode_factor(., "ppbc_lac" = "lactating")
vsd$lac_vs_rest <- vsd$lac_vs_rest %>%
  dplyr::recode_factor(., "ppbc_lac" = "lactating")

#Temporary fix until rerun makes this unnecessary
stopifnot(identical(colnames(dds), sample_data$sample_name))
dds$PPBC <- sample_data$PPBC
dds$survival <- sample_data$survival
```

## Summary

```{r lac vs rest report}
source(here("src", "deseq_report_functions.R"))

rep_lac_rest = deseq_report(ape, dds=dds, absl2fc = 0.5, title="Lactation vs rest",
                            intgroup="lac_vs_rest", groups = c("lactating", "rest"), 
                            colorby="survival",
                            top_vars = c("lac_vs_rest", "survival"),
                            top_colors = list(lac_vs_rest=ppbc_colors,
                                              survival = os_colors),
                            bottom_vars = c("ER", "HER2"),
                            bottom_colors = list(ER = er_colors,
                                                 HER2 = her2_colors),
                            legend_title = "PPBC"
)

```

## Top genes 

```{r lac vs rest top 30}
rep_lac_rest$significant_genes %>%
  select(gene_name, Type, padj, log2FoldChange, baseMean, description, everything()) %>%
  head(30)
```
Protein coding hits:

```{r lac vs rest protein coding hits}
rep_lac_rest$significant_genes %>%
  filter(Type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name)
  
```

## Heatmap

```{r lac vs rest heatmap, fig.height=6, fig.width=10}
rep_lac_rest$heatmap
```

## Volcano plot

```{r lac vs rest volcano}
rep_lac_rest$volcano_plot
```


## Box plots

Plot the top 5 most significant.

```{r lac vs rest beehives}
rep_lac_rest$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r n lac vs rest pathways}
print(paste("Number of pathways significant at fdr < 0.1:",
            (rep_lac_rest$pathways %>% bind_rows() %>%
               filter(fdr < 0.1) %>% nrow())))
```
Top 30 preview

```{r top 30 lac vs rest pathways}
rep_lac_rest$pathways %>% bind_rows() %>%
  filter(fdr < 0.1) %>% select(pathway,everything()) %>%
  arrange(fdr) %>% head(30)
```

Plot top pathways

```{r lac vs rest pathway plot, fig.height=6, fig.width=8}
#repLRT$pathway_plots

bind_rows(rep_lac_rest$pathways) %>%
  #filter(fdr < 0.1) %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Sig pathways fdr < 0.1: Lactating vs rest"
      )
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r lac vs rest outlier volcano}
rep_lac_rest$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 6:

```{r lac vs rest outlier bees}
rep_lac_rest$outlier_bees[1:6]
```

# Nulliparous vs rest {.tabset}

```{r load nonprbc vs rest, collapse=T}
#list.files(here("data", "Rds"))[str_detect(list.files(here("data", "Rds")), "17_")]

dds = readRDS(here("data/Rds/17_dds_ovr_nonprbc_vs_rest.Rds"))
ape = readRDS(here("data/Rds/17_ape_nonprbc_vs_rest.Rds"))

design(dds)
dds$nonprbc_vs_rest %>% levels()
ape@elementMetadata$description[4]

#Nicer column names
dds$nonprbc_vs_rest <- dds$nonprbc_vs_rest %>%
  dplyr::recode_factor(., "non_prbc" = "nulliparous")
vsd$nonprbc_vs_rest <- vsd$nonprbc_vs_rest %>%
  dplyr::recode_factor(., "non_prbc" = "nulliparous")

#Temporary fix until rerun makes this unnecessary
stopifnot(identical(colnames(dds), sample_data$sample_name))
dds$PPBC <- sample_data$PPBC
dds$survival <- sample_data$survival
```

## Summary

```{r nonprbc vs rest report}
source(here("src", "deseq_report_functions.R"))

rep_nonprbc_rest = deseq_report(ape, dds=dds, absl2fc = 0.5, title="Nulliparous vs rest",
                            intgroup="nonprbc_vs_rest", groups = c("nulliparous", "rest"), 
                            colorby="survival",
                            top_vars = c("nonprbc_vs_rest", "survival"),
                            top_colors = list(nonprbc_vs_rest=ppbc_colors,
                                              survival = os_colors),
                            bottom_vars = c("ER", "HER2"),
                            bottom_colors = list(ER = er_colors,
                                                 HER2 = her2_colors),
                            legend_title = "PPBC"
)
```

## Top genes 

```{r nonprbc vs rest top 30}
rep_nonprbc_rest$significant_genes %>%
  select(gene_name, Type, padj, log2FoldChange, baseMean, description, everything()) %>%
  head(30)
```
Protein coding hits:

```{r nonprbc vs rest protein coding hits}
rep_nonprbc_rest$significant_genes %>%
  filter(Type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name)
  
```

## Heatmap

```{r nonprbc vs rest heatmap, fig.height=6, fig.width=10}
rep_nonprbc_rest$heatmap
```

## Volcano plot

```{r nonprbc vs rest volcano}
rep_nonprbc_rest$volcano_plot
```


## Box plots

Plot the top 5 most significant.

```{r nonprbc vs rest beehives}
rep_nonprbc_rest$beehive_plots
```

## Preliminary pathway

A fast method for a general overview. See FlexGSEA for a properly rigorous analysis.

```{r n nonprbc vs rest pathways}
print(paste("Number of pathways significant at fdr < 0.1:",
            (rep_nonprbc_rest$pathways %>% bind_rows() %>%
               filter(fdr < 0.1) %>% nrow())))
```
Top 30 preview

```{r top 30 nonprbc vs rest pathways}
rep_nonprbc_rest$pathways %>% bind_rows() %>%
  filter(fdr < 0.1) %>% select(pathway,everything()) %>%
  arrange(fdr) %>% head(30)
```

Plot top pathways

```{r nonprbc vs rest pathway plot, fig.height=6, fig.width=8}
#repLRT$pathway_plots

bind_rows(rep_nonprbc_rest$pathways) %>%
  #filter(fdr < 0.1) %>%
  plot_enrichment(
        enrich_res = .,
        fdr = 0.1,
        max_nchar_path = 40,
        max_path = 20,
        title = "Sig pathways fdr < 0.1: Nulliparous vs rest"
      )
```

## Cook's outliers

Have a quick look at genes that had their padj set to NA because one or more samples exceeded the Cook's threshold

```{r nonprbc vs rest outlier volcano}
rep_nonprbc_rest$volc_outliers
```

Beehive plot of genes that fail Cooks filtering and are above the fold change threshold. Sorted by descending l2fc, top 6:

```{r nonprbc vs rest outlier bees}
rep_nonprbc_rest$outlier_bees[1:6]
```

# Save data

## LRT {.tabset}

Create a master results list.

```{r lrt res list}
#Get all report objects from environment
reslist = mget(ls(pattern="repLRT"))
names(reslist)
```

### Significance of genes

Save the individual results lists, without the figures.

```{r lrt sig genes}
#Save the results for all genes in an Excel file with multiple tabs
#The tab names will be the list names
resdf = lapply(reslist, function(x) x$annotated_results)
names(resdf) = paste(names(resdf),"all", sep="_")
openxlsx::write.xlsx(resdf, file = here("results", "diffex", "17_ERHER2_LRT_allgenes.xlsx"))

#Save the significant genes in a multi-tab excel file
resdf = lapply(reslist, function(x) x$significant_genes)
names(resdf) = paste("sig",names(resdf), sep="_")
#Also keep the thresholds used for significance
resdf[[length(resdf)+1]] = 
enframe(unlist(lapply(reslist, function(x) x$sig_threshold)), "name", "value") %>%
  mutate(groups = if_else(str_detect(name, "nolac"), "no lactation", "allgroups")) %>%
  mutate(name = str_remove(name, ".nolac")) %>%
  separate(name, into = c("comparison", "threshold"), sep="\\.")
names(resdf)[length(resdf)] = "thresholds"
openxlsx::write.xlsx(resdf, file = here("results", "diffex", "17_ERHER2_LRT_sig_genes.xlsx"))

rm(resdf)
```

### Pathway results

Fisher's exact pathway analyses

```{r lrt paths}
resdf = lapply(reslist, function(x) x$pathways)

#Returns a list of aggregated path results per comparison
pathlist = list()
pathlist[[names(resdf)[1]]] = resdf %>% map(1, .depth=1) %>% bind_rows( .id = "comp")
for (i in 2:length(resdf[[1]])){
  path = resdf %>% map(i, .depth=1) %>% bind_rows( .id = "comp")
  pathlist[[names(resdf)[i]]] = path
}

#Fix names
names(pathlist) = lapply(pathlist, function(x) unique(x$collection)) %>% unlist(use.names = F)
pathlist = lapply(pathlist, function(x) x[order(x$fdr),])

openxlsx::write.xlsx(pathlist, file = here("results", "diffex", "17_ERHER2_LRT_pathways.xlsx"))

rm(resdf)
```

### Heatmaps

```{r lrt heatmaps}
lrt_figdir <- here("results","diffex", "figs", "17_LRT_ERHER2")

dir.create(lrt_figdir, showWarnings = F)
dir.create(file.path(lrt_figdir, "heatmaps"), showWarnings = F)

hmlist = lapply(reslist, function(x) x$heatmap)
names(hmlist) = names(hmlist) %>% str_replace_all("rep", "hm")

for (i in 1:length(hmlist)){
  pdf(file.path(lrt_figdir, "heatmaps", paste0(names(hmlist)[i],".pdf")), width = 10, height = 8)
  print(hmlist[[i]])
  dev.off()
}

rm(hmlist)

```

### Volcano plot sig genes

Use caution when interpreting volcano plots for more than 2 groups.

```{r lrt vps}
volclist = lapply(reslist, function(x) x$volcano_plot)
names(volclist) = names(volclist) %>% str_replace("rep", "volc")

dir.create(file.path(lrt_figdir, "volcano_plots"), showWarnings = F)
lapply(names(volclist), 
       function(x) ggsave(filename=file.path(lrt_figdir,"volcano_plots",
                                             paste(x,".jpeg",sep="")), plot=volclist[[x]],
                          height = 6, width = 8))
rm(volclist)
```

### Volcano plot outliers

Volcano plots of significant genes are hard to interpret in an LRT with multiple levels. Use outliers only for diagnostics.

```{r lrt vp outliers}
outvplist = lapply(reslist, function(x) x$volc_outliers)
names(outvplist) = names(outvplist) %>% str_replace("rep", "volc")

dir.create(file.path(lrt_figdir,"volcano_plots"), showWarnings = F)
lapply(names(outvplist), 
       function(x) ggsave(filename=file.path(lrt_figdir, "volcano_plots",
                                             paste(x,".outliers.jpeg",sep="")), plot=outvplist[[x]],
                          height = 6, width = 8))
rm(outvplist)
```

### Pathway dotplots 

```{r lrt dp}
dir.create(file.path(lrt_figdir, "fisher_pathway"), showWarnings = F)

pathwaylist = lapply(reslist, function(x) x$pathway_plots)
for(i in 1:length(pathwaylist)){
  thiscomp = pathwaylist[[i]]
  compname = str_replace(str_remove(names(pathwaylist[i]), "rep_"),"_","_vs_")
  for (j in 1:length(thiscomp)){
    if (length(thiscomp) < 1){next}
    thissig = thiscomp[[j]]
    file=paste0(paste(compname, names(thiscomp[j]), sep="-"),".jpeg")
    #print(file)
    ggsave(filename=file.path(lrt_figdir, "fisher_pathway", file),
           plot=thissig, height = 6, width = 8)
  }
}


rm(pathwaylist)
```

## Pairwise comparisons {.tabset}

Retrieve all objects from environment starting with "rep_" that do NOT include "rest" in the variable name.

```{r pw reslist}
reslist <- mget(ls(pattern = "rep_")[str_detect(ls(pattern = "rep_"), "rest", negate = T)])
names(reslist)
stopifnot(length(reslist) == nrow(combos))
```


### Significance of genes

Save the individual results lists, without the figures.

```{r pw sig genews}
#Save the results for all genes in an Excel file with multiple tabs
#The tab names will be the list names

resdf = lapply(reslist, function(x) x$annotated_results)
names(resdf) = paste(names(resdf),"all", sep="_")
openxlsx::write.xlsx(resdf, file = here("results", "diffex", "17_ERHER2_pairwise_comparisons_allgenes.xlsx"))

#Save the significant genes in a multi-tab excel file
resdf = lapply(reslist, function(x) x$significant_genes)
names(resdf) = paste("sig",names(resdf), sep="_")
#Also keep the thresholds used for significance
resdf[[length(resdf)+1]] = enframe(unlist(lapply(reslist, function(x) x$sig_threshold)), "name", "value") %>%
                 separate(name, into = c("comparison", "threshold"), sep="\\.")
names(resdf)[length(resdf)] = "thresholds"

openxlsx::write.xlsx(resdf, file = here("results", "diffex", "17_ERHER2_pairwise_comparisons_sig_genes.xlsx"))
```

### Pathway results

Fisher's exact pathway analyses

```{r pw paths}
resdf = lapply(reslist, function(x) x$pathways)

#Returns a list of aggregated path results per comparison
pathlist = list()
pathlist[[names(resdf)[1]]] = resdf %>% map(1, .depth=1) %>% bind_rows( .id = "comp")
for (i in 2:length(resdf[[1]])){
  path = resdf %>% map(i, .depth=1) %>% bind_rows( .id = "comp")
  pathlist[[names(resdf)[i]]] = path
}

#Fix names
names(pathlist) = lapply(pathlist, function(x) unique(x$collection)) %>% unlist(use.names = F)
pathlist = lapply(pathlist, function(x) x[order(x$fdr),])

openxlsx::write.xlsx(pathlist, file = here("results", "diffex", "17_ERHER2_pairwise_comparisons_pathways.xlsx"))
rm(resdf)
```

### Heatmaps

`pairwise_figdir` defined in the Upset section.

```{r pw heatmaps}
dir.create(file.path(pairwise_figdir, "heatmaps"), showWarnings = F)

hmlist = lapply(reslist, function(x) x$heatmap)
names(hmlist) = names(hmlist) %>% str_replace_all("rep", "hm")

for (i in 1:length(hmlist)){
  pdf(file.path(pairwise_figdir, "heatmaps", paste0(names(hmlist)[i],".pdf")), width = 10, height = 8)
  print(hmlist[[i]])
  dev.off()
}

rm(hmlist)
```

### Volcano plots sig hits

```{r pw vps}
volclist = lapply(reslist, function(x) x$volcano_plot)
names(volclist) = names(volclist) %>% str_replace("rep", "volc")

dir.create(file.path(pairwise_figdir,"volcano_plots"), showWarnings = F)
lapply(names(volclist), 
       function(x) ggsave(filename=file.path(pairwise_figdir,"volcano_plots",
                                        paste(x,".jpeg",sep="")), plot=volclist[[x]],
                          height = 6, width = 8))
rm(volclist)
```

### Volcano plot outliers

```{r pw vp outliers}
outvplist = lapply(reslist, function(x) x$volc_outliers)
names(outvplist) = names(outvplist) %>% str_replace("rep", "volc")

dir.create(here(pairwise_figdir,"volcano_plots"), showWarnings = F)
lapply(names(outvplist), 
       function(x) ggsave(filename=file.path(pairwise_figdir,"volcano_plots",
                                        paste(x,".outliers.jpeg",sep="")), plot=outvplist[[x]],
                          height = 6, width = 8))
rm(outvplist)
```

### Pathway dotplots 

```{r pw dotplots}
dir.create(file.path(pairwise_figdir,"fisher_pathway"), showWarnings = F)

pathwaylist = lapply(reslist, function(x) x$pathway_plots)

for(i in 1:length(pathwaylist)){
  thiscomp = pathwaylist[[i]]
  compname = str_replace(str_remove(names(pathwaylist[i]), "rep_"),"_","_vs_")
  for (j in 1:length(thiscomp)){
    if (length(thiscomp) < 1){next}
    thissig = thiscomp[[j]]
    file=paste0(paste(compname, names(thiscomp[j]), sep="-"),".jpeg")
    #print(file)
    ggsave(filename=file.path(pairwise_figdir,"fisher_pathway",file),
           plot=thissig, height = 6, width = 8)
  }
}


rm(pathwaylist)
```

## One vs rest comparisons {.tabset}

```{r ovr reslist}
reslist = mget(ls(pattern="rep_")[str_detect(ls(pattern="rep_"), "rest")])
names(reslist)
```

### Significance of genes

Save the significant genes and results from each comparison.

```{r ovr sig genes}
#Save the results for all genes in an Excel file with multiple tabs

#The tab names will be the list names
resdf = lapply(reslist, function(x) x$annotated_results)
names(resdf) = paste(names(resdf),"all", sep="_")
openxlsx::write.xlsx(resdf, file = here("results", "diffex", "17_ERHER2_one_vs_rest_allgenes.xlsx"))

#Save the significant genes in a multi-tab excel file
resdf = lapply(reslist, function(x) x$significant_genes)
names(resdf) = paste("sig",names(resdf), sep="_")
#Also keep the thresholds used for significance
resdf[[length(resdf)+1]] = enframe(unlist(lapply(reslist, function(x) x$sig_threshold)), "name", "value") %>% #signficance and l2fc
                 separate(name, into = c("comparison", "threshold"), sep="\\.") %>%
  bind_rows(.,enframe(unlist(lapply(reslist, function(x) x$cooks_threshold)), #Cooks
                      "comparison", "value") %>% mutate(threshold = "cooks"))
names(resdf)[length(resdf)] = "thresholds"
write.xlsx(resdf, file = here("results", "diffex", "17_ERHER2_one_vs_rest_sig_genes.xlsx"))
rm(resdf)
```

### Pathway results

Fisher's exact pathway analyses

```{r ovr pathways}

resdf = lapply(reslist, function(x) x$pathways)

#Returns a list of aggregated path results per comparison
pathlist = list()
pathlist[[names(resdf)[1]]] = resdf %>% map(1, .depth=1) %>% bind_rows( .id = "comp")
for (i in 2:length(resdf[[1]])){
  path = resdf %>% map(i, .depth=1) %>% bind_rows( .id = "comp")
  pathlist[[names(resdf)[i]]] = path
}

#Fix names
names(pathlist) = lapply(pathlist, function(x) unique(x$collection)) %>% unlist(use.names = F)
pathlist = lapply(pathlist, function(x) x[order(x$fdr),])

openxlsx::write.xlsx(pathlist, file = here("results", "diffex", "17_ERHER2_one_vs_rest_pathways.xlsx"))

rm(resdf)
```

### Heatmaps

```{r ovr heatmap}
ovr_figdir <- here("results", "diffex", "figs", "17_one_vs_rest_ERHER2")

dir.create(file.path(ovr_figdir, "heatmaps"), showWarnings = F, recursive = T)

hmlist = lapply(reslist, function(x) x$heatmap)
names(hmlist) = names(hmlist) %>% str_replace_all("rep", "hm")

for (i in 1:length(hmlist)){
  pdf(file.path(ovr_figdir, "heatmaps", paste0(names(hmlist)[i],".pdf")), width = 10, height = 8)
  print(hmlist[[i]])
  dev.off()
}

rm(hm_list)
```

### Volcano plot sig hits

```{r ovr vp}
volclist = lapply(reslist, function(x) x$volcano_plot)
names(volclist) = names(volclist) %>% str_replace("rep", "volc")

dir.create(file.path(ovr_figdir,"volcano_plots"), showWarnings = F)
lapply(names(volclist), 
       function(x) ggsave(filename=file.path(ovr_figdir,"volcano_plots",
                                        paste(x,".jpeg",sep="")), plot=volclist[[x]],
                          height = 6, width = 8))
rm(volclist)
```

### Volcano plot outliers

```{r ovr vp outliers}
outvplist = lapply(reslist, function(x) x$volc_outliers)
names(outvplist) = names(outvplist) %>% str_replace("rep", "volc")

dir.create(file.path(ovr_figdir,"volcano_plots"), showWarnings = F)
lapply(names(outvplist), 
       function(x) ggsave(filename=file.path(ovr_figdir, "volcano_plots",
                                        paste(x,".outliers.jpeg",sep="")), plot=outvplist[[x]],
                          height = 6, width = 8))
rm(outvplist)
```

Remove unnecessary large objects

```{r remove extras}
rm(reslist)
rm(gene_sets)
```


```{r save workspace}
#Takes a loooong time
save.image(here("reports", "17_diffex_ER_HER2_status.RData"))
```































































































