---
title: "Comparison of differential expression formulas"
author: "Kat Moore"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float: yes
    toc_depth: 4
    highlight: tango
    df_print: paged
---

# Definition of formulas

Here we compare the results from the original differential expression formula (notebooks 6-8), hereafter referred to as the **bpam formula**, or as **formula x**:

`~ batch + PAM50 + study_group`

With the modified formula, hereafter referred to as the **eh formula**, or as **formula y**:

`~ ER + HER2 + study_group`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
rm(list = ls())
```

```{r packages, include=F}
# library(ComplexHeatmap)
# library(RColorBrewer)
library(here)
library(DESeq2)
library(limma)
library(readxl)
library(tidyverse)
```

```{r}
source(here::here("src", "general_R_tools.R"))
source(here::here("src", "deseq_report_functions.R"))
```

https://bioinformatics.stackexchange.com/questions/909/running-differential-expression-analyses-on-count-matrices-with-many-zeroes
https://support.bioconductor.org/p/117448/

# Load data

Gmt files for pathway analysis:

## Gene signatures

```{r gene signatures}
gene_sets = list(
  go_bp = flexgsea::read_gmt(here("data","external","gmt","c5.bp.v7.0.symbols.gmt")),
  hallmark = flexgsea::read_gmt(here("data","external","gmt","h.all.v7.0.symbols.gmt")),
  #c2_canon = flexgsea::read_gmt(here("data","external","gmt","c2.cp.v7.0.symbols.gmt")),
  c2_cgp = flexgsea::read_gmt(here("data","external","gmt","c2.cgp.v7.0.symbols.gmt"))
)
```

## Gene annotation

```{r}
gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
gx_annot = gx_annot %>% select(ensembl_gene_id = gene_id, gene_name, hgnc_symbol,
                               gene_type, description = gene_description) %>% distinct()
```



List all files in the excel results directory:

```{r}
xcel_dir <- here("results", "diffex")
list.files(xcel_dir, pattern = ".xlsx")
```

Colors for plotting

```{r load colors}
colors <- readRDS(here("data", "Rds", "06_heatmap_colors.Rds"))
```

## Batch and PAM formula {.tabset}

```{r}
bpam <- list()
bpam$formula <- "~ batch + PAM50 + study_group"
```

### LRT

```{r}
# readxl::excel_sheets(file.path(xcel_dir, "06_LRT_allgenes.xlsx"))

bpam$LRT <- readxl::read_excel(path = file.path(xcel_dir, "06_LRT_allgenes.xlsx"), sheet = "repLRT_all")

head(bpam$LRT)
```

### Pairwise

```{r}
# readxl::excel_sheets(file.path(xcel_dir, "07_pairwise_comparisons_allgenes.xlsx"))

bpam <- append(bpam, read_excel_tabs(file.path(xcel_dir, "07_pairwise_comparisons_allgenes.xlsx")))
names(bpam)
```

### One vs rest

```{r}
# readxl::excel_sheets(file.path(xcel_dir, "08_one_vs_rest_allgenes.xlsx"))

bpam <- append(bpam, read_excel_tabs(file.path(xcel_dir, "08_one_vs_rest_allgenes.xlsx")))
names(bpam)
```

## ER and HER2 formula {.tabset}

```{r}
eh <- list()
eh$formula <- "~ ER + HER2 + study_group"
```

### LRT

```{r}
# readxl::excel_sheets(file.path(xcel_dir, "06_LRT_allgenes.xlsx"))

eh$LRT <- readxl::read_excel(path = file.path(xcel_dir, "17_ERHER2_LRT_allgenes.xlsx"), sheet = "repLRT_all")

head(eh$LRT)
```

### Pairwise

```{r}
# readxl::excel_sheets(file.path(xcel_dir, "07_pairwise_comparisons_allgenes.xlsx"))

eh <- append(eh, read_excel_tabs(file.path(xcel_dir, "17_ERHER2_pairwise_comparisons_allgenes.xlsx")))
names(eh)
```

### One vs rest

```{r}
# readxl::excel_sheets(file.path(xcel_dir, "08_one_vs_rest_allgenes.xlsx"))

eh <- append(eh, read_excel_tabs(file.path(xcel_dir, "17_ERHER2_one_vs_rest_allgenes.xlsx")))
names(eh)
```

```{r, include = F}
stopifnot(all(names(bpam) == names(eh)))
```


## Note about duplicate IDs

Seems like the inclusion of hgnc symbol in the annotation later in the pipeline caused duplicate results of genes that have more than one hgnc symbol.

```{r}
eh_dups <- unique(eh$LRT[duplicated(eh$LRT$ensembl_gene_id),"ensembl_gene_id", drop=T])
eh$LRT %>% filter(ensembl_gene_id %in% eh_dups)

```

We can see based on the number of rows where it occurred in our pipeline.

```{r}
tibble(bpam = unlist(lapply(bpam, nrow)), eh = unlist(lapply(eh, nrow)))
```


This is going to cause problems for our Venn diagrams, so we'll need to fix it.

```{r}
fix_hgnc_dups <- function(df){
  require(dplyr)
  
  if("hgnc_symbol" %in% colnames(df)){
    df <- df %>% select(-hgnc_symbol) %>%
      distinct()
  }
  
  
  return(df)
}

names(eh)

#Ignore the first because it's the formula
eh[2:length(eh)] <- lapply(eh[2:length(eh)], function(x) fix_hgnc_dups(x))
bpam[2:length(bpam)] <- lapply(bpam[2:length(eh)], fix_hgnc_dups)

stopifnot(all(unlist(lapply(bpam, nrow)) == unlist(lapply(eh, nrow))))

```

# Comparing significance {.tabset}

Focus on:

-Number of sig genes in each formula
-N Genes gained and lost (names too?)
-Pathway analysis on setdiff genes?
-Common genes DOTHIS




# One vs rest comparisons

Keep to involution vs rest for now

```{r}
inv_vs_rest_diff = summarize_formula_differences(x = bpam$rep_inv_rest_all, y = eh$rep_inv_rest_all)
```

```{r}
limma::vennDiagram(inv_vs_rest_diff$vennCounts, names = c("bpam", "eh"),
                   main = "Involution vs rest")
```


## Gene lists {.tabset}

### Unique to bpam formula:

Protein coding genes:

```{r}
inv_vs_rest_diff$sig.x.not.y %>%
  filter(type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

Other genes:

IGHD and IGHA1 will be lost if we swith to the EH formula.

```{r}
inv_vs_rest_diff$sig.x.not.y %>%
  filter(!type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

```{r}
inv_vs_rest_diff$sig.x.not.y %>%
  filter(gene_name %in% c("IGHA1", "IGHD"))
```

### Unique to eh formula: 

Protein coding:

```{r}
inv_vs_rest_diff$sig.y.not.x %>%
  filter(type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

Other genes:

```{r}
inv_vs_rest_diff$sig.y.not.x %>%
  filter(!type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

### Common genes

Protein coding:

```{r}
inv_vs_rest_diff$sig.x.and.y %>%
  filter(type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```
Other genes:

```{r}
inv_vs_rest_diff$sig.x.and.y %>%
  filter(!type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

## Pathways {.tabset}

### Unique to bpam formula: {.tabset}

```{r}
inv_vs_rest_bpam_paths <- fisher_pathways(
  sig_genes = inv_vs_rest_diff$sig.x.not.y$gene_name,
  background_genes = inv_vs_rest_diff$gene.table$gene_name,
  list_signatures = gene_sets,
  fdr_cutoff = 0.05,
  verbose = T,
  collapse_rows=T)
```

#### GO-BP

```{r}
inv_vs_rest_bpam_paths %>% arrange(fdr, pval) %>%
  filter(collection == "go_bp") %>%
  head(20)
```

#### CGP

```{r}
inv_vs_rest_bpam_paths %>% arrange(fdr, pval) %>%
  filter(collection == "c2_cgp") %>%
  head(20)
```

### Unique to eh formula: {.tabset}

```{r}
inv_vs_rest_eh_paths <- fisher_pathways(
  sig_genes = inv_vs_rest_diff$sig.y.not.x$gene_name,
  background_genes = inv_vs_rest_diff$gene.table$gene_name,
  list_signatures = gene_sets,
  fdr_cutoff = 0.05,
  verbose = T,
  collapse_rows=T)

```

#### GO-BP

```{r}
inv_vs_rest_eh_paths %>% arrange(fdr, pval) %>%
  filter(collection == "go_bp") %>%
  head(20)
```

### Gains and losses

```{r, collapse=T}
pathway.x  = read_excel_tabs(file.path(xcel_dir, "08_one_vs_rest_pathways.xlsx")) %>%
  bind_rows()
#head(pathway.x)
pathway.x = pathway.x %>% filter(fdr < 0.05 & comp == "rep_inv_rest") %>% arrange(fdr)
print(paste("Significant pathways in x", nrow(pathway.x)))
print("")

pathway.y  = read_excel_tabs(file.path(xcel_dir, "17_ERHER2_one_vs_rest_pathways.xlsx")) %>%
  bind_rows()
#head(pathway.y)
pathway.y = pathway.y %>% filter(fdr < 0.05 & comp == "rep_inv_rest") %>% arrange(fdr)
print(paste("Significant pathways in y", nrow(pathway.y)))
print("")

print("Pathways lost if we switch from x to y:")
pathway.x$pathway[!pathway.x$pathway %in% pathway.y$pathway]
print("")
print("Pathways gained if we switch from x to y:")
pathway.y$pathway[!pathway.y$pathway %in% pathway.x$pathway]
```

## Difference plots {.tabset}

### Sig genes bpam only

```{r}
formula_difference_dotplots(inv_vs_rest_diff$sig.x.not.y,
                            stat = "padj",
                            padj.y.label.thresh = 0.25)
formula_difference_dotplots(inv_vs_rest_diff$sig.x.not.y,
                            stat = "l2fc", show.inverted.signs = T)

```

### Sig genes eh only

```{r}
formula_difference_dotplots(inv_vs_rest_diff$sig.y.not.x,
                            stat = "padj",
                            padj.x.label.thresh = 0.5)
formula_difference_dotplots(inv_vs_rest_diff$sig.y.not.x,
                            stat = "l2fc", show.inverted.signs = T)

```

### Significant in either formula

```{r}
plot_formula_differences(
  x = bpam$rep_inv_rest_all, y = eh$rep_inv_rest_all,
  plot.stat = "padj", sig.only = T,
  dp_title = "Involuting vs rest"
)
```

### All genes

```{r}
plot_formula_differences(
  x = bpam$rep_inv_rest_all, y = eh$rep_inv_rest_all,
  plot.stat = "padj", sig.only = F,
  dp_title = "Involuting vs rest"
)
```


# Pairwise comparisons

To limit the scope, look only at a few of the most interesting comparisons that include involution.

# Involuting vs nulliparous {.tabset}

```{r}
inv_vs_np_diff = summarize_formula_differences(x = bpam$rep_inv_nonprbc_all, y = eh$rep_inv_nonprbc_all)
```

```{r}
limma::vennDiagram(inv_vs_np_diff$vennCounts, names = c("bpam", "eh"),
                   main = "Involution vs nulliparous")
```

## Gene lists {.tabset}

### Unique to bpam formula:

Protein coding

```{r}
inv_vs_np_diff$sig.x.not.y %>%
  filter(type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```
Other genes: 

```{r}
inv_vs_np_diff$sig.x.not.y %>%
  filter(!type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```


### Unique to eh formula: 

Surprisingly, ESR1 is significant here.

Protein coding:

```{r}
inv_vs_np_diff$sig.y.not.x %>%
  filter(type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

Other genes:

```{r}
inv_vs_np_diff$sig.y.not.x %>%
  filter(!type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```
### Common genes

```{r}
inv_vs_np_diff$sig.x.and.y %>%
  filter(type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```
```{r}
inv_vs_np_diff$sig.x.and.y %>%
  filter(!type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

## Pathways {.tabset}

### Unique to bpam formula:

```{r}
fisher_pathways(
  sig_genes = inv_vs_np_diff$sig.x.not.y$gene_name,
  background_genes = inv_vs_np_diff$gene.table$gene_name,
  list_signatures = gene_sets,
  fdr_cutoff = 0.1,
  verbose = T,
  collapse_rows=T)
```

### Unique to eh formula: {.tabset}

```{r}
inv_vs_np_eh_paths <- fisher_pathways(
  sig_genes = inv_vs_np_diff$sig.y.not.x$gene_name,
  background_genes = inv_vs_np_diff$gene.table$gene_name,
  list_signatures = gene_sets,
  fdr_cutoff = 0.1,
  verbose = T,
  collapse_rows=T)

```
#### Hallmark

```{r}
inv_vs_np_eh_paths %>% arrange(fdr, pval) %>%
  filter(collection == "hallmark") %>%
  head(20)
```

#### GO-BP

```{r}
inv_vs_np_eh_paths %>% arrange(fdr, pval) %>%
  filter(collection == "go_bp") %>%
  head(20)
```

#### CGP

```{r}
inv_vs_np_eh_paths %>% arrange(fdr, pval) %>%
  filter(collection == "c2_cgp") %>%
  head(20)
```
### Gains and losses

```{r, collapse=T}
pathway.x  = read_excel_tabs(file.path(xcel_dir, "07_pairwise_comparisons_pathways.xlsx")) %>%
  bind_rows()
#head(pathway.x)
pathway.x = pathway.x %>% filter(fdr < 0.05 & comp == "rep_inv_nonprbc") %>% arrange(fdr)
print(paste("Significant pathways in x", nrow(pathway.x)))
print("")

pathway.y  = read_excel_tabs(file.path(xcel_dir, "17_ERHER2_pairwise_comparisons_pathways.xlsx")) %>%
  bind_rows()
#head(pathway.y)
pathway.y = pathway.y %>% filter(fdr < 0.05 & comp == "rep_inv_nonprbc") %>% arrange(fdr)
print(paste("Significant pathways in y", nrow(pathway.y)))
print("")

print("Pathways lost if we switch from x to y:")
pathway.x$pathway[!pathway.x$pathway %in% pathway.y$pathway]
print("")
print("Pathways gained if we switch from x to y:")
pathway.y$pathway[!pathway.y$pathway %in% pathway.x$pathway]
```
## Difference plots {.tabset}

### Sig genes bpam only

```{r}
formula_difference_dotplots(inv_vs_np_diff$sig.x.not.y,
                            stat = "padj",
                            padj.y.label.thresh = 0.25)
formula_difference_dotplots(inv_vs_np_diff$sig.x.not.y,
                            stat = "l2fc", show.inverted.signs = T)

```

### Sig genes eh only

```{r}
formula_difference_dotplots(inv_vs_np_diff$sig.y.not.x,
                            stat = "padj",
                            padj.x.label.thresh = 0.5)
formula_difference_dotplots(inv_vs_np_diff$sig.y.not.x,
                            stat = "l2fc", show.inverted.signs = T)

```

### Significant in either formula

```{r}
plot_formula_differences(
  x = bpam$rep_inv_nonprbc_all, y = eh$rep_inv_nonprbc_all,
  plot.stat = "padj", sig.only = T,
  dp_title = "Involuting vs nulliparous"
)
```

### All genes

```{r}
plot_formula_differences(
  x = bpam$rep_inv_nonprbc_all, y = eh$rep_inv_nonprbc_all,
  plot.stat = "padj", sig.only = F,
  dp_title = "Involuting vs nulliparous"
)
```

# Involuting vs pregnant {.tabset}

```{r}
inv_vs_pr_diff = summarize_formula_differences(x = bpam$rep_inv_prbc_all, y = eh$rep_inv_prbc_all)
```

```{r}
limma::vennDiagram(inv_vs_pr_diff$vennCounts, names = c("bpam", "eh"),
                   main = "Involution vs pregnant")
```

## Gene lists {.tabset}

### Unique to bpam formula:

Protein coding:

```{r}
inv_vs_pr_diff$sig.x.not.y %>%
  filter(type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

Other genes:

IGHD will be lost if we switch to the EH formula.

```{r}
inv_vs_pr_diff$sig.x.not.y %>%
  filter(!type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```


### Unique to eh formula: 

Protein coding genes:

```{r}
inv_vs_pr_diff$sig.y.not.x %>%
  filter(type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

Estrogen is not significant here.

```{r}
inv_vs_pr_diff$sig.y.not.x %>%
  filter(gene_name %in% c("ESR1"))
```

Other genes:

```{r}
inv_vs_pr_diff$sig.y.not.x %>%
  filter(!type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

## Pathways {.tabset}

### Unique to bpam formula: {.tabset}

```{r}
inv_vs_pr_bpam_paths <- fisher_pathways(
  sig_genes = inv_vs_pr_diff$sig.x.not.y$gene_name,
  background_genes = inv_vs_pr_diff$gene.table$gene_name,
  list_signatures = gene_sets,
  fdr_cutoff = 0.05,
  verbose = T,
  collapse_rows=T)
```

#### Hallmark

```{r}
inv_vs_pr_bpam_paths %>% arrange(fdr, pval) %>%
  filter(collection == "hallmark") %>%
  head(20)
```

#### CGP

```{r}
inv_vs_pr_bpam_paths %>% arrange(fdr, pval) %>%
  filter(collection == "c2_cgp") %>%
  head(20)
```

### Unique to eh formula: {.tabset}

```{r}
inv_vs_pr_eh_paths <- fisher_pathways(
  sig_genes = inv_vs_pr_diff$sig.y.not.x$gene_name,
  background_genes = inv_vs_pr_diff$gene.table$gene_name,
  list_signatures = gene_sets,
  fdr_cutoff = 0.05,
  verbose = T,
  collapse_rows=T)

```

#### Hallmark

```{r}
inv_vs_pr_eh_paths %>% arrange(fdr, pval) %>%
  filter(collection == "hallmark") %>%
  head(20)
```

#### GO-BP

```{r}
inv_vs_pr_eh_paths %>% arrange(fdr, pval) %>%
  filter(collection == "go_bp") %>%
  head(20)
```

#### CGP

```{r}
inv_vs_pr_eh_paths %>% arrange(fdr, pval) %>%
  filter(collection == "c2_cgp") %>%
  head(20)
```

## Difference plots {.tabset}

### Sig genes bpam only

```{r}
formula_difference_dotplots(inv_vs_pr_diff$sig.x.not.y,
                            stat = "padj",
                            padj.y.label.thresh = 0.25)
formula_difference_dotplots(inv_vs_pr_diff$sig.x.not.y,
                            stat = "l2fc", show.inverted.signs = T)

```

### Sig genes eh only

```{r}
formula_difference_dotplots(inv_vs_pr_diff$sig.y.not.x,
                            stat = "padj",
                            padj.x.label.thresh = 0.75)
formula_difference_dotplots(inv_vs_pr_diff$sig.y.not.x,
                            stat = "l2fc", show.inverted.signs = T)

```

### Significant in either formula

```{r}
plot_formula_differences(
  x = bpam$rep_inv_prbc_all, y = eh$rep_inv_prbc_all,
  plot.stat = "padj", sig.only = T,
  dp_title = "Involuting vs pregnant"
)
```

### All genes

```{r}
plot_formula_differences(
  x = bpam$rep_inv_prbc_all, y = eh$rep_inv_prbc_all,
  plot.stat = "padj", sig.only = F,
  dp_title = "Involuting vs pregnant"
)
```


# Likelihood ratio test {.tabset}

```{r}
lrt_diff = summarize_formula_differences(x = bpam$LRT, y = eh$LRT)
```

```{r}
limma::vennDiagram(lrt_diff$vennCounts)
```

## Gene lists {.tabset}

### Unique to bpam formula:

Protein coding

```{r}
lrt_diff$sig.x.not.y %>%
  filter(type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

Other genes:

```{r}
lrt_diff$sig.x.not.y %>%
  filter(!type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

### Unique to eh formula: 

Protein coding

```{r}
lrt_diff$sig.y.not.x %>%
  filter(type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

Other genes

```{r}
lrt_diff$sig.y.not.x %>%
  filter(!type %in% c("protein coding", "immune protein coding")) %>%
  pull(gene_name) %>% unique()
```

## Pathways {.tabset}

### Unique to bpam formula:

```{r}
fisher_pathways(
  sig_genes = lrt_diff$sig.x.not.y$gene_name,
  background_genes = eh$LRT$gene_name,
  list_signatures = gene_sets,
  fdr_cutoff = 0.05,
  verbose = T,
  collapse_rows=T)
```

### Unique to eh formula: {.tabset}

```{r}
lrt_eh_paths <- fisher_pathways(
  sig_genes = lrt_diff$sig.y.not.x$gene_name,
  background_genes = eh$LRT$gene_name,
  list_signatures = gene_sets,
  fdr_cutoff = 0.05,
  verbose = T,
  collapse_rows=T) %>%
  select(-signature, fdr)

```
#### Hallmark

```{r}
lrt_eh_paths %>% arrange(fdr, pval) %>%
  filter(collection == "hallmark") %>%
  head(20)
```

#### GO-BP

```{r}
lrt_eh_paths %>% arrange(fdr, pval) %>%
  filter(collection == "go_bp") %>%
  head(20)
```

#### CGP

```{r}
lrt_eh_paths %>% arrange(fdr, pval) %>%
  filter(collection == "c2_cgp") %>%
  head(20)
```

## Difference plots {.tabset}

### Sig genes bpam only

```{r}
source(here("src", "deseq_report_functions.R"))
formula_difference_dotplots(lrt_diff$sig.x.not.y,
                            stat = "padj",
                            padj.y.label.thresh = 0.25)
formula_difference_dotplots(lrt_diff$sig.x.not.y,
                            stat = "l2fc", show.inverted.signs = T)

```

### Sig genes eh only

```{r}
formula_difference_dotplots(lrt_diff$sig.y.not.x,
                            stat = "padj",
                            padj.x.label.thresh = Inf)
formula_difference_dotplots(lrt_diff$sig.y.not.x,
                            stat = "l2fc", show.inverted.signs = T)

```

### Significant in either formula

```{r}
plot_formula_differences(
  x = bpam$LRT, y = eh$LRT,
  plot.stat = "padj", sig.only = T,
  dp_title = "LRT"
)
```

### All genes

```{r}
plot_formula_differences(
  x = bpam$LRT, y = eh$LRT,
  plot.stat = "padj", sig.only = F,
  dp_title = "LRT"
)
```


# Goodness of fit

Focus on:
Deviance for each comparison
Dispersion plots
One comparison for test type (6 total)

DESeqDataSets for all results lists

```{r}
dds_df <- tibble(
  formula = c(rep("bpam", 3), rep("eh", 3)),
  test = rep(c("LRT", "Inv vs np", "Inv vs rest"), 2),
  filename = c(
    "06_ddsLRT_standard.Rds",
    "07_dds_apeglm_nonprbcref_standard.Rds",
    "08_dds_inv_vs_rest_standard.Rds",
    "17_dds_LRT.Rds",
    "17_dds_pairwise_ref_non_prbc.Rds",
    "17_dds_ovr_inv_vs_rest.Rds"),
  path = c(
    here("data", "Rds", "06_ddsLRT_standard.Rds"),
    here("data", "Rds", "07_dds_apeglm_nonprbcref_standard.Rds"),
    here("data", "Rds", "08_dds_inv_vs_rest_standard.Rds"),
    here("data", "Rds", "17_dds_LRT.Rds"),
    here("data", "Rds", "17_dds_pairwise_ref_non_prbc.Rds"),
    here("data", "Rds", "17_dds_ovr_inv_vs_rest.Rds")
  )
)

dds_df
```

```{r}
source(here("src", "deseq_report_functions.R"))

deseq_compare_gof <- function(dds.x, dds.y, name.x, name.y,
                              label.dev.thresh = Inf, label.basemean.thresh = Inf){
  res = list()
  
  stopifnot(all(rownames(dds.x) == rownames(dds.y)))
  
  dev.x = dds_dev(dds.x, print_median = T)
  dev.y = dds_dev(dds.y, print_median = T)
  
  res$devs.both  = tibble(ensembl_gene_id = rownames(dds.x),
                          dev.x = dev.x,
                          dev.y = dev.y) %>%
    left_join(.,select(gx_annot, ensembl_gene_id, gene_name),
              by = "ensembl_gene_id") %>%
    mutate(dev.diff = abs(dev.y - dev.x)) %>%
    arrange(desc(dev.diff))
  
  res$dev_hist <- compare_deviance_histogram(dds.x = dds.x, dds.y = dds.y, name.x = name.x, name.y = name.y)
  
  res$mean_dev <- compare_mean_deviance_trend(dds.x = dds.x, dds.y = dds.y, name.x = name.x, name.y = name.y,
                                              annot = gx_annot, label.dev.thresh = label.dev.thresh,
                                              label.basemean.thresh = label.basemean.thresh)
  res$mean_dev.x <- dds_mean_deviance_trend(dds.x, annot = gx_annot, title = name.x)
  res$mean_dev.y <- dds_mean_deviance_trend(dds.y, annot = gx_annot, title = name.x)
  
  return(res)
}
```


# Involution vs rest {.tabset}

```{r}
path.x = filter(dds_df, formula == "bpam" & test == "Inv vs rest")$path
path.y = filter(dds_df, formula == "eh" & test == "Inv vs rest")$path

print(path.x)
print(path.y)

dds.x = readRDS(path.x)
dds.y = readRDS(path.y)

gof_inv_vs_rest = deseq_compare_gof(dds.x, dds.y,
                            name.x = "Batch Pam50 Inv vs Rest", name.y = "ER HER2 Inv vs Rest")

```

## Deviance histogram

```{r}
gof_inv_vs_rest$dev_hist
```

## Mean-deviance trend

```{r}
gof_inv_vs_rest$mean_dev
```

## High deviance genes {.tabset}

Show some genes where the difference is high.


```{r}
for (i in 1:5){
  print(plot_high_deviance_gene(dds = dds.x, gene_name = gof_inv_vs_rest$devs.both$gene_name[i],
                                ensembl_gene_id = gof_inv_vs_rest$devs.both$ensembl_gene_id[i],
                                dev_df = gof_inv_vs_rest$devs.both))
}

```

# Involution vs nulliparous {.tabset}

```{r}
path.x = filter(dds_df, formula == "bpam" & test == "Inv vs np")$path
path.y = filter(dds_df, formula == "eh" & test == "Inv vs np")$path

print(path.x)
print(path.y)

dds.x = readRDS(path.x)
dds.y = readRDS(path.y)

gof_inv_vs_np = deseq_compare_gof(dds.x, dds.y,
                            name.x = "Batch Pam50 Inv vs Np", name.y = "ER HER2 Inv vs Np")

```

## Deviance histogram

```{r}
gof_inv_vs_np$dev_hist
```

## Mean-deviance trend

```{r}
gof_inv_vs_np$mean_dev
```

## High deviance genes {.tabset}

Show some genes where the difference is high.


```{r}
for (i in 1:5){
  print(plot_high_deviance_gene(dds = dds.x, gene_name = gof_inv_vs_np$devs.both$gene_name[i],
                                ensembl_gene_id = gof_inv_vs_np$devs.both$ensembl_gene_id[i],
                                dev_df = gof_inv_vs_rest$devs.bot))
}

```


# Likelihood ratio test {.tabset}

```{r}
path.x = filter(dds_df, formula == "bpam" & test == "LRT")$path
path.y = filter(dds_df, formula == "eh" & test == "LRT")$path

print(path.x)
print(path.y)

dds.x = readRDS(path.x)
dds.y = readRDS(path.y)

gof_lrt = deseq_compare_gof(dds.x, dds.y,
                            name.x = "Batch Pam50 LRT", name.y = "ER HER2 LRT")

```

## Deviance histogram

```{r}
gof_lrt$dev_hist
```

## Mean-deviance trend

```{r}
gof_lrt$mean_dev
```

## High deviance genes {.tabset}

Show some genes where the difference is high.


```{r}
for (i in 1:5){
  print(plot_high_deviance_gene(dds = dds.x, gene_name = gof_lrt$devs.both$gene_name[i],
                                ensembl_gene_id = gof_lrt$devs.both$ensembl_gene_id[i],
                                dev_df = gof_inv_vs_rest$devs.both))
}

```

# Problem genes {.tabset}

## BCMA

One of our "problem genes" from the B cell list. Sadly, is not significant even after changing formulas.

```{r}

source(here("src", "deseq_report_functions.R"))

plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000048462"),
                  title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                  color_fun = ggthemes::scale_color_colorblind(na.value = "grey50"))

plot_high_deviance_gene(dds = dds.x, gene_name = "TNFRSF17",
                        ensembl_gene_id = "ENSG00000048462",
                                dev_df = gof_inv_vs_rest$devs.both)
```

## TACI

Same story

```{r}
plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000240505"),
                  title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                  color_fun = ggthemes::scale_color_colorblind(na.value = "grey50"))
plot_high_deviance_gene(dds = dds.x, gene_name = "TNFRSF13B",
                        ensembl_gene_id = "ENSG00000240505",
                                dev_df = gof_inv_vs_rest$devs.both)
```

```{r}
save.image(here("reports", "17b_comparing_diffex_formulas.RData"))
```































































