---
title: "Comparing diffex formulas for PPBC - Genes for Lodewyk"
author: "Kat Moore"
date:  "`r Sys.Date()`"
output:   
  html_document:
    toc: yes
    toc_float: yes
    toc_depth: 4
    highlight: tango
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
rm(list = ls())
```

```{r packages, include=F}
# library(ComplexHeatmap)
# library(RColorBrewer)
library(here)
library(DESeq2)
library(limma)
library(readxl)
library(tidyverse)
```

```{r, include=F}
source(here::here("src", "general_R_tools.R"))
source(here::here("src", "deseq_report_functions.R"))
```


```{r}
load(here("reports", "17b_comparing_diffex_formulas.RData"))
```

# Introduction

The post-partum breast cancer project has 4 classes of tumor samples:

- Nulliparous (never pregnant)
- Pregnant
- Lactating
- Involution (breakdown of lactating tissue)

Although it is morphologically similar to nulliparous tissue, breast cancer diagnosed during lactation has a significantly poorer outcome.

In this notebook, we compare a series of genes significantly associated with involution using one of two formulas:

Formula x (also called **bPam**):
~ batch + PAM50 + PPBC

Formula y (also called **eh**):
~ ER + HER2 + PPBC

Up until now, formula x has been used for all analyses. 
However, we are considering formula y because of colinearity between covariates in formula x.
We exclude PR from formula y because its expression is highly associated with ER.

Below are some chi-square tests for association between potential covariates.

Formula x:

[1] "ChiSq pval between batch and PAM50 : 0.3263 , V: 0.32"
[1] "ChiSq pval between batch and study_group : 5e-04 , V: 0.58"
[1] "ChiSq pval between PAM50 and study_group : 0.058 , V: 0.31"

Formula y:

[1] "ChiSq pval between ER and HER2 : 0.4003 , V: 0.06"
[1] "ChiSq pval between ER and PR : 5e-04 , V: 0.76"
[1] "ChiSq pval between ER and study_group : 0.1429 , V: 0.16"
[1] "ChiSq pval between HER2 and PR : 0.5147 , V: 0.05"
[1] "ChiSq pval between HER2 and study_group : 0.96 , V: 0.04"
[1] "ChiSq pval between PR and study_group : 0.2734 , V: 0.14"

# Involution vs rest

"Rest" refers to pregnant, nulliparous and lactating samples pooled together.

- show 5 of each from unique and overlapping significance lists here

```{r}
limma::vennDiagram(inv_vs_rest_diff$vennCounts, names = c("bpam", "eh"),
                   main = "Involution vs rest")
```
```{r}
dds.x = readRDS("/DATA/share/postpartumbc/data/Rds/08_dds_inv_vs_rest_standard.Rds")
dds.y = readRDS("/DATA/share/postpartumbc/data/Rds/17_dds_ovr_inv_vs_rest.Rds")
```

## Formula x only {.tabset}

These genes are only significant when using:
~batch + PAM50 + PPBC(study group)

```{r, collapse = T}
#Unique to old formula
inv_vs_rest_diff$sig.x.not.y %>% filter(gene_name %in% c("IGHD", "IGHA1", "S100A7", "POTED", "TCN1")) %>%
  select(gene_name, ensembl_gene_id, baseMean, everything()) %>%
  arrange(padj.x)
```

### IGHA1

Immunoglobulin Heavy Constant Alpha 1

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000211895"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000211895"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000211895"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000211895"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### IGHD

Immunoglobulin Heavy Constant Delta

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000211898"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000211898"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000211898"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000211898"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### POTED

POTE Ankyrin Domain Family Member D

A cancer-testis antigen family member often observed upregulated in malignant breast and skin cancers.

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000166351"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000166351"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000166351"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000166351"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### TCN1

Transcobalamin 1 

This gene encodes a member of the vitamin B12-binding protein family. This family of proteins, alternatively referred to as R binders, is expressed in various tissues and secretions. This protein is a major constituent of secondary granules in neutrophils and facilitates the transport of cobalamin into cells.

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000134827"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000134827"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000134827"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000134827"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### S100A7

S100 Calcium Binding Protein A7 

The protein encoded by this gene is a member of the S100 family of proteins containing 2 EF-hand calcium-binding motifs. S100 proteins are localized in the cytoplasm and/or nucleus of a wide range of cells, and involved in the regulation of a number of cellular processes such as cell cycle progression and differentiation. S100 genes include at least 13 members which are located as a cluster on chromosome 1q21. This protein differs from the other S100 proteins of known structure in its lack of calcium binding ability in one EF-hand at the N-terminus. The protein is overexpressed in hyperproliferative skin diseases, exhibits antimicrobial activities against bacteria and induces immunomodulatory activities.


```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000143556"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000143556"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000143556"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000143556"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

## Formula y only {.tabset}

Genes with NA for padj were filtered out by DESeq2 because they exceed the default threshold for Cooks distance.

```{r, collapse = T}
#Unique to new formula
inv_vs_rest_diff$sig.y.not.x %>% filter(gene_name %in% c("IL17RB", "ICAM5", "IL23R", "IRS4", "TNFSF11")) %>%
  select(gene_name, ensembl_gene_id, baseMean, everything()) %>%
  arrange(padj.y)
```
### IRS

Insulin Receptor Substrate 4 "ENSG00000133124"

Acts as an interface between multiple growth factor receptors possessing tyrosine kinase activity, such as insulin receptor, IGF1R and FGFR1, and a complex network of intracellular signaling molecules containing SH2 domains. Involved in the IGF1R mitogenic signaling pathway. Promotes the AKT1 signaling pathway and BAD phosphorylation during insulin stimulation without activation of RPS6KB1 or the inhibition of apoptosis. Interaction with GRB2 enhances insulin-stimulated mitogen-activated protein kinase activity. May be involved in nonreceptor tyrosine kinase signaling in myoblasts. Plays a pivotal role in the proliferation/differentiation of hepatoblastoma cell through EPHB2 activation upon IGF1 stimulation. May play a role in the signal transduction in response to insulin and to a lesser extent in response to IL4 and GH on mitogenesis. Plays a role in growth, reproduction and glucose homeostasis. May act as negative regulators of the IGF1 signaling pathway by suppressing the function of IRS1 and IRS2.

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000133124"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000133124"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000133124"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000133124"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### IL17RB

Interleukin 17 Receptor B ENSG00000056736

Interleukin-17 receptor (IL-17R) is a cytokine receptor which belongs to new subfamily of receptors binding proinflammatory cytokine interleukin 17A, a member of IL-17 family ligands produced by T helper 17 cells (Th17).[2] IL-17R family consists of 5 members: IL-17RA, IL-17RB, IL-17RC, IL-17RD and IL-17RE. Functional IL-17R is a transmembrane receptor complex usually consisting of one IL-17RA, which is a founding member of the family, and second other family subunit, thus forming heteromeric receptor binding different ligands. IL-17A, a founding member of IL-17 ligand family binds to heteromeric IL-17RA/RC receptor complex.[3] 

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000056736"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000056736"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000056736"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000056736"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### ICAM5

Intercellular Adhesion Molecule 5 ENSG00000105376

ICAM5 (Intercellular Adhesion Molecule 5) is a Protein Coding gene. Diseases associated with ICAM5 include Holoprosencephaly and Breast Cancer. Among its related pathways are Class I MHC mediated antigen processing and presentation and Innate Immune System. Gene Ontology (GO) annotations related to this gene include integrin binding. An important paralog of this gene is ICAM3.

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000105376"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000105376"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000105376"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000105376"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### IL23R

Interleukin 23 Receptor ENSG00000162594

Associates with IL12RB1 to form the interleukin-23 receptor. Binds IL23 and mediates T-cells, NK cells and possibly certain macrophage/myeloid cells stimulation probably through activation of the Jak-Stat signaling cascade. IL23 functions in innate and adaptive immunity and may participate in acute response to infection in peripheral tissues. IL23 may be responsible for autoimmune inflammatory diseases and be important for tumorigenesis. 

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000133124"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000133124"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000133124"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000133124"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### TNFSF11

TNF Superfamily Member 11 ENSG00000120659

This gene encodes a member of the tumor necrosis factor (TNF) cytokine family which is a ligand for osteoprotegerin and functions as a key factor for osteoclast differentiation and activation. This protein was shown to be a dentritic cell survival factor and is involved in the regulation of T cell-dependent immune response. T cell activation was reported to induce expression of this gene and lead to an increase of osteoclastogenesis and bone loss. This protein was shown to activate antiapoptotic kinase AKT/PKB through a signaling complex involving SRC kinase and tumor necrosis factor receptor-associated factor (TRAF) 6, which indicated this protein may have a role in the regulation of cell apoptosis. Targeted disruption of the related gene in mice led to severe osteopetrosis and a lack of osteoclasts. The deficient mice exhibited defects in early differentiation of T and B lymphocytes, and failed to form lobulo-alveolar mammary structures during pregnancy.

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000133124"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000133124"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000133124"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000133124"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

## Common genes {.tabset}

```{r}
inv_vs_rest_diff$sig.x.and.y %>% filter(gene_name %in% c("CELSR2", "JCHAIN", "ESR2", "CXCL14", "MUC2")) %>%
  select(gene_name, ensembl_gene_id, baseMean, everything()) %>%
  arrange(padj.x, padj.y)
```

### MUC2

Mucin 2, Oligomeric Mucus/Gel-Forming

ENSG00000284971 (shown)
ENSG00000198788 (not shown, very similar)

Mucin protein family member highly upregulated in some mucinous tumors. Two isoforms, both highly expressed.

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000284971"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000284971"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000284971"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000284971"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### ESR2

Estrogen receptor 2 ENSG00000140009

Nuclear hormone receptor. Binds estrogens with an affinity similar to that of ESR1, and activates expression of reporter genes containing estrogen response elements (ERE) in an estrogen-dependent manner (PubMed:20074560).

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000140009"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000140009"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000140009"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000140009"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```


### CELSR2

Cadherin EGF LAG Seven-Pass G-Type Receptor 2 ENSG00000143126

The protein encoded by this gene is a member of the flamingo subfamily, part of the cadherin superfamily. The flamingo subfamily consists of nonclassic-type cadherins; a subpopulation that does not interact with catenins. The flamingo cadherins are located at the plasma membrane and have nine cadherin domains, seven epidermal growth factor-like repeats and two laminin A G-type repeats in their ectodomain. They also have seven transmembrane domains, a characteristic unique to this subfamily. It is postulated that these proteins are receptors involved in contact-mediated communication, with cadherin domains acting as homophilic binding regions and the EGF-like domains involved in cell adhesion and receptor-ligand interactions. The specific function of this particular member has not been determined.

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000143126"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000143126"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000143126"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000143126"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### JCHAIN

Joining Chain Of Multimeric IgA And IgM ENSG00000132465

Serves to link two monomer units of either IgM or IgA. In the case of IgM, the J chain-joined dimer is a nucleating unit for the IgM pentamer, and in the case of IgA it induces larger polymers. It also help to bind these immunoglobulins to secretory component.


```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000132465"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000132465"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000132465"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000132465"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### CXCL14

C-X-C Motif Chemokine Ligand 14	ENSG00000145824

This antimicrobial gene belongs to the cytokine gene family which encode secreted proteins involved in immunoregulatory and inflammatory processes. The protein encoded by this gene is structurally related to the CXC (Cys-X-Cys) subfamily of cytokines. Members of this subfamily are characterized by two cysteines separated by a single amino acid. This cytokine displays chemotactic activity for monocytes but not for lymphocytes, dendritic cells, neutrophils or macrophages. It has been implicated that this cytokine is involved in the homeostasis of monocyte-derived macrophages rather than in inflammation. Potent chemoattractant for neutrophils, and weaker for dendritic cells. Not chemotactic for T-cells, B-cells, monocytes, natural killer cells or granulocytes. Does not inhibit proliferation of myeloid progenitors in colony formation assays.

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000145824"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_rest_all, ensembl_gene_id == "ENSG00000145824"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000145824"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_rest_all, ensembl_gene_id == "ENSG00000145824"),
                                    title_string = "Inv vs rest", intgroup = "inv_vs_rest", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

# Involution vs nulliparous

Here we include a few more genes that were not already included in the inv vs rest section. 
The genes that would be gained if switching to the ER and HER2 formula are both highly expressed and particularly interesting.
Several of them are involved in wound healing and regeneration, which would be consistent with the original hypothesis...

```{r}
limma::vennDiagram(inv_vs_np_diff$vennCounts, names = c("bpam", "eh"),
                   main = "Involution vs nulliparous")
```
```{r}
dds.x = readRDS("/DATA/share/postpartumbc/data/Rds/07_dds_apeglm_nonprbcref_standard.Rds")
dds.y = readRDS("/DATA/share/postpartumbc/data/Rds/17_dds_pairwise_ref_non_prbc.Rds")
```


## Formula x only {.tabset}

```{r, collapse = T}
#Unique to old formula
inv_vs_np_diff$sig.x.not.y %>% filter(gene_name %in% c("PIGR", "AKAP6")) %>%
  select(gene_name, ensembl_gene_id, baseMean, everything()) %>%
  arrange(padj.x)
```

### PIGR

Polymeric Immunoglobulin Receptor ENSG00000162896

This gene is a member of the immunoglobulin superfamily. The encoded poly-Ig receptor binds polymeric immunoglobulin molecules at the basolateral surface of epithelial cells; the complex is then transported across the cell to be secreted at the apical surface.  The complex is then transported across the cell to be secreted at the apical surface. During this process a cleavage occurs that separates the extracellular (known as the secretory component) from the transmembrane segment.

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000162896"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000162896"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000162896"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000162896"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### AKAP6

ENSG00000151320

The A-kinase anchor proteins (AKAPs) are a group of structurally diverse proteins, which have the common function of binding to the regulatory subunit of protein kinase A (PKA) and confining the holoenzyme to discrete locations within the cell. 

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000151320"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000151320"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000151320"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000151320"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

## Formula y only {.tabset}

```{r, collapse = T}
#Unique to new formula
inv_vs_np_diff$sig.y.not.x %>% filter(gene_name %in% c("CLU", "IGF1R", "ESR1", "FLNB", "GREB1", "THSD4")) %>%
  select(gene_name, ensembl_gene_id, baseMean, everything()) %>%
  arrange(desc(baseMean))
```

### CLU

Clusterin ENSG00000120885

The protein encoded by this gene is a secreted chaperone that can under some stress conditions also be found in the cell cytosol. It has been suggested to be involved in several basic biological events such as cell death, tumor progression, and neurodegenerative disorders. Fits with the wound-healing hypothesis.

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000120885"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000120885"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000120885"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000120885"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### IGF1R

Insulin Like Growth Factor 1 Receptor  ENSG00000140443

This receptor binds insulin-like growth factor with a high affinity. It has tyrosine kinase activity. The insulin-like growth factor I receptor plays a critical role in transformation events. Cleavage of the precursor generates alpha and beta subunits. It is highly overexpressed in most malignant tissues where it functions as an anti-apoptotic agent by enhancing cell survival. Alternatively spliced transcript variants encoding distinct isoforms have been found for this gene. 

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000140443"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000140443"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000140443"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000140443"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### ESR1

ENSG00000091831

Somehow still significant for study group in spite of being part of the design formula???

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000091831"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000091831"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000091831"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000091831"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### FLNB

Filamin B ENSG00000136068

This gene encodes a member of the filamin family. The encoded protein interacts with glycoprotein Ib alpha as part of the process to repair vascular injuries. Fits with the wound-healing hypothesis.

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000136068"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000136068"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000136068"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000136068"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### GREB1

Growth Regulating Estrogen Receptor Binding 1 ENSG00000196208

This gene is an estrogen-responsive gene that is an early response gene in the estrogen receptor-regulated pathway. It is thought to play an important role in hormone-responsive tissues and cancer. Three alternatively spliced transcript variants encoding distinct isoforms have been found for this gene. [provided by RefSeq, Jul 2008]
GREB1 (Growth Regulating Estrogen Receptor Binding 1) is a Protein Coding gene. Diseases associated with GREB1 include Breast Cancer and Amelogenesis Imperfecta, Type Ig. Among its related pathways are Ectoderm Differentiation. An important paralog of this gene is GREB1L.
May play a role in estrogen-stimulated cell proliferation. Acts as a regulator of hormone-dependent cancer growth in breast and prostate cancers.

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000196208"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000196208"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000196208"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000196208"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

### THSD4

Thrombospondin Type 1 Domain Containing 4 ENSG00000187720

Gene Ontology (GO) annotations related to this gene include peptidase activity and **metalloendopeptidase activity**. Fits the wound-healing theory.
Promotes FBN1 matrix assembly. Attenuates TGFB signaling, possibly by accelerating the sequestration of large latent complexes of TGFB or active TGFB by FBN1 (maybe also FLNB??) microfibril assembly, thereby negatively regulating the expression of TGFB regulatory targets, such as POSTN (By similarity).

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000187720"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000187720"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000187720"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "ER",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.y, filter(eh$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000187720"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "HER2",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

## For Daniel
```{r}
lrt_diff$sig.x.not.y %>%
   filter(gene_name == "IGHV6-1")
```



```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000282745"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000282745"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```




IGLV2-5

```{r}
lrt_diff$sig.x.not.y %>%
   filter(gene_name == "IGLV2-5")
```

```{r, fig.width=10}
ggpubr::ggarrange(plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000253234"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "PAM50",
                                    color_fun = ggthemes::scale_color_colorblind(na.value = "grey50")),
                  
                  plot_gene_beehive(dds.x, filter(bpam$rep_inv_nonprbc_all, ensembl_gene_id == "ENSG00000253234"),
                                    groups_to_plot = c("ppbc_inv", "non_prbc"),
                                    title_string = "Inv vs np", intgroup = "study_group", colorby = "batch",
                                    color_fun = ggthemes::scale_color_hc(na.value = "grey50")),
                  ncol = 2)
```

```{r}
#test = prcomp(model.matrix(dds.y))

test = as.data.frame(colData(dds.y)) %>% select(PAM50)
model.matrix(test[,1, drop=T])
```


```{r}
colData(dds.y) %>%
  as.data.frame() %>%
  select(ER, HER2, PR) %>%
  cor()
```

```{r}
corvars_cat = t(combn(c("PAM50", "ER", "HER2", "PR"), 2))
corvars_cat
```

```{r}
sd = as.data.frame(colData(dds.y))

for (i in 1:nrow(corvars_cat)){
  thisx = corvars_cat[i,1]
  thisy = corvars_cat[i,2]
  tbl = table(sd[,thisx], sd[,thisy])
  chires =chisq.test(sd[,thisx], sd[,thisy], simulate.p.value = T)
  v = sqrt(chires$statistic / sum(tbl))
  print(paste("ChiSq pval between", thisx, "and", thisy, ":", round(chires$p.value, 4), ", V:", round(v,2)))
  
}
```


```{r}
sd$PAM50 %>% levels()
```
### Possible covariates

```{r}
test = colData(dds.y) %>%
  as.data.frame() %>%
  mutate(
    Basal = if_else(PAM50 == "Basal", 1, 0),
    HerPos = if_else(PAM50 == "Her2", 1, 0),
    LumB = if_else(PAM50 == "LumB", 1, 0),
    LumA = if_else(PAM50 == "LumA", 1, 0),
    Normal = if_else(PAM50 == "Normal", 1, 0)
    ) %>%
  select(Basal, HerPos, LumB, LumA, Normal, ER, HER2, PR)

res = prcomp(test)

res$x[,1:2] #Use these two PCs as covariates
res$rotation[,1:2]
```

```{r}
library(pheatmap)
colData(dds.y) %>%
  as.data.frame() %>%
  mutate(
    Basal = if_else(PAM50 == "Basal", 1, 0),
    HerPos = if_else(PAM50 == "Her2", 1, 0),
    LumB = if_else(PAM50 == "LumB", 1, 0),
    LumA = if_else(PAM50 == "LumA", 1, 0),
    Normal = if_else(PAM50 == "Normal", 1, 0)
    ) %>%
  select(Basal, HerPos, LumB, LumA, Normal, ER, HER2, PR) %>%
  cor() %>% round(2) %>%
  pheatmap(display_numbers = T)
```

```{r}
sd %>% select(PAM50, ER) %>% table()
sd %>% select(PAM50, HER2) %>% table()
sd %>% select(PAM50, PR) %>% table()
```


