---
title: "Clustering within involution samples"
Author: Kat Moore
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: yes
    toc_float: no
    df_print: paged
    highlight: kate
    canon: true
params:
  dds: "data/rnaseq/processed/08_dds_ovr_inv_vs_rest.Rds"
  vsd: "data/rnaseq/interim/08_vsd_ovr.Rds"
  gx_annot: "data/rnaseq/metadata/01_gene_annot.tsv"
  coxdata: "data/rnaseq/interim/04_survdata.Rds"
  tools: "src/rnaseq/deseq_report_functions.R"
  cp: "data/rnaseq/interim/color_palettes.Rds"
  sp: "data/rnaseq/interim/survival_colors.Rds"
  DEGhm: "figures/Fig1a_DEG_heatmap.pdf"
  IGbarplot: "figures/Fig1d_IG_cluster_barplot.pdf"
  IGkaplanOS: "figures/Fig1e_IG_cluster_KM.pdf"
  rtIGkaplanOS: "figures/Fig1e_risktable.csv"
  IGkaplanDRS: "figures/supfigs/Supfig9_IG_cluster_KM_DRS.pdf"
  rtIGkaplanDRS: "figures/supfigs/Fig1e_risktable.csv"
---

```{r, message=F}
library(DESeq2)
library(apeglm)
library(ggrepel)
library(ComplexHeatmap)
library(openxlsx)
library(scrime)
library(survival)
library(survminer)
library(ggrepel)
library(RColorBrewer)
library(here)
library(tidyverse)

theme_set(theme_bw())
```

A k-means clustering strategy on DEG genes within the PPBC project.

# Setup and pre-processing {.tabset}

This data has been filtered on the basis of having a gene-wise non-zero count in at least a third of the samples. 

```{r}
dds = readRDS(here(params$dds))
```

We will be changing study_group to match the groups plotted, so save the original in its own column.

```{r}
table(dds$study_group)
dds$study_group_orig <- dds$study_group
```

```{r}
source(here(params$tools))
```

Subset for just the involution samples.

```{r}
inv.dds = dds[,colnames(dds)[dds$study_group == "ppbcpw"]]
inv.dds$study_group = droplevels(inv.dds$study_group)
table(inv.dds$study_group)
```

## Gene metadata

```{r}
gx_annot <- read_tsv(here(params$gx_annot))
gx_annot = gx_annot %>%
  select(ensembl_gene_id = gene_id, gene_name,
         gene_type, description = gene_description) %>% distinct()
```

## Variance stabilizing transformation

Renders the data homoskedastic, useful for visualization purposes.

Ensure that we can set "study group" to be one vs rest columns when desired.

```{r}
vsd = readRDS(here(params$vsd))
```

## Sample annotation data

```{r}
sd = as.data.frame(colData(dds)) %>%
  select(sample_name, study_group, contains("rest"),
         PAM50, study_group_orig, everything())
head(sd)
```

## Symbol conversion and de-deuplication

```{r}
geneEx = rownames_to_column(as.data.frame(assay(vsd)), "ensembl_gene_id")
geneEx = right_join(select(gx_annot, gene_name, ensembl_gene_id),
                    geneEx, by = "ensembl_gene_id") %>%
    select(-ensembl_gene_id)
geneEx = summarize_expression_duplicate_ids(geneEx, id_column = "gene_name", verbose = T)

```

## Colors

```{r}
cp = readRDS(here(params$cp))
study_colors = cp$study_colors
pam_colors = cp$pam_colors
gene_colors = cp$gene_colors
ppbc_colors = cp$ppbc_colors
inv_colors = cp$study_colors[names(cp$study_colors) %in% c("rest", "ppbcpw")]

sp = readRDS(here(params$sp))
```

## Diffex results

Read in significant results from comparison involution vs non-prbc, and inv vs rest

```{r}
resDir <- file.path("results", "rnaseq", "diffex")
#readxl::excel_sheets(here(resDir, "07_pairwise_comparisons_allgenes.xlsx"))
res_invp = openxlsx::read.xlsx(here(resDir, "07_pairwise_comparisons_allgenes.xlsx"),
                               sheet="rep_inv_nonprbc_all")
#head(res_invp)

sig.invp = openxlsx::read.xlsx(here(resDir, "07_pairwise_comparisons_sig_genes.xlsx"),
                               sheet="sig_rep_inv_nonprbc")
#head(sig.invp)

res_invr = openxlsx::read.xlsx(here(resDir, "08_one_vs_rest_allgenes.xlsx"),
                               sheet="rep_inv_rest_all")
#head(res_invr)

#openxlsx::getSheetNames(here("results", "diffex", "08_one_vs_rest_sig_genes.xlsx"))
sig.invr = openxlsx::read.xlsx(here(resDir, "08_one_vs_rest_sig_genes.xlsx"),
                               sheet="sig_rep_inv_rest")
#head(sig.invp)
```

# Diffex genes

Thresholds applied:

```{r}
rbind(openxlsx::read.xlsx(here(resDir, "07_pairwise_comparisons_sig_genes.xlsx"),
                          sheet="thresholds") %>%
        filter(comparison == "rep_inv_nonprbc"),
      openxlsx::read.xlsx(here(resDir, "08_one_vs_rest_sig_genes.xlsx"),
                          sheet="thresholds") %>%
        filter(comparison == "rep_inv_rest"))
```

```{r, collapse=T}
print(paste("There are", nrow(sig.invp), "significant genes in the comparison involution vs non-prbc"))
print(paste("There are", nrow(sig.invr), "significant genes in the comparison involution vs rest"))
```

Top genes vs nulliparous

```{r}
sig.invr %>% head(30)
```

vs rest

```{r}
sig.invp %>% head(30)
```

# K-means Heatmaps

## PPBCpw vs NPBC

```{r, fig.height=8, fig.width=10}
set.seed(1)

deseq_heatmap(mat = geneEx, sampledata = mutate(sd, study_group = study_group_orig),
              sig_results = sig.invp, title = "PPBCpw vs NPBC, K-means",
              groups_to_plot =c("ppbcpw", "npbc"), maxn_rownames = Inf,
              column_km=3, row_km = 3)

```

## PPBCpw vs rest

```{r, fig.height=8, fig.width=10}
set.seed(1)

inv.hm = deseq_heatmap(mat = geneEx, sampledata = sd,
              sig_results = sig.invr,
              title = "Significant DEGs from PPBCpw vs rest, K-means clustering",
              maxn_rownames = Inf,
              #groups_to_plot =c("ppbcpw", "rest"),
              column_km=3, row_km = 3)

inv.hm
```

```{r}
clusterDir = file.path(here("results/rnaseq/clustering"))
dir.create(clusterDir, showWarnings = F)
pdf(file = file.path(clusterDir, "11_hm_clust_DEG_inv_vs_rest.pdf"))
inv.hm
dev.off()
```

## Figure 1A : Heatmap PPBCpw vs rest

Slight tweaks for manuscript figure. Show all non-ppbcpw as "rest", but `deseq_heatmap` does not set the colors properly with `inv_vs_rest`. This is because of some kind of very weird bug in ComplexHeamtap that prevents color hexes from being recognized...sometimes. It is not possible to perfectly match the colors, but it's close. This figure is otherwise identical to the one that shows all groups

```{r, fig.height=8, fig.width=10}
set.seed(1)

fig.hm = deseq_heatmap(
  mat = geneEx, sampledata = sd,
  top_vars = "inv_vs_rest",
  #top_colors = list(inv_vs_rest = inv_colors),
  #top_colors = list(inv_vs_rest = c("rest"="#F73859", "ppbcpw"="#46CDCF")), # does not work
  top_colors = list(inv_vs_rest = c("rest"="purple4", "ppbcpw"="darkturquoise")),
  sig_results = sig.invr,
  title = "Significant DEGs from PPBCpw vs rest, K-means clustering",
  maxn_rownames = Inf,
  intgroup = "inv_vs_rest",
  groups_to_plot =c("ppbcpw", "rest"),
  column_km=3, row_km = 3
  )

fig.hm
```

```{r}
figureDir = file.path(here("figures"))
dir.create(figureDir, showWarnings = F)
pdf(file = here(params$DEGhm), width = 10, height = 10)
fig.hm
dev.off()
```

```{r, fig.height=8, fig.width=10, eval=F,include=F}
# In which we try and fail to diagnose this mysterious problem.
create_hm_matrix <- function(mat, sampledata, sig_results,
                             maxn_genes = length(unique(sig_results$gene_name)),
                             row_scale = T){
  
  #Reduce genes to significant only
  sig_results = sig_results %>% arrange(padj)
  genestoplot = sig_results$gene_name[1:maxn_genes]
  mat = mat[rownames(mat) %in% genestoplot, ]
  
  stopifnot(identical(colnames(mat), sampledata$sample_name))
  
  #Row scale settings
  if (row_scale==T){
    mat = scrime::rowScales(mat)
  }
  
  #Ensure input is matrix
  mat = as.matrix(mat)
  
  mat
  
}

set.seed(1)

Heatmap(
  create_hm_matrix(mat = geneEx, sampledata = sd, sig_results = sig.invr),
  top_annotation = HeatmapAnnotation(
    df=as.data.frame(sd[,"inv_vs_rest", drop=F]),
    which="col",
    #col = list(inv_vs_rest = inv_colors), #does not work
    #col = list(inv_vs_rest = c("rest"="#F73859", "ppbcpw"="#46CDCF")) # does not work
    col = list(inv_vs_rest = c("rest"="purple4", "ppbcpw"="darkturquoise")) #does work for some reason
  ),
  # bottom_annotation = colBottom,
  # left_annotation = rowAnno,
  # heatmap_legend_param = hlp,
  show_row_names = T,
  show_column_names = F,
  cluster_rows = T,
  row_names_gp = gpar(fontsize = 8),
  column_names_gp = gpar(fontsize = 8),
  column_title = "Significant DEGs from PPBCpw vs rest, K-means clustering",
  column_km=3, row_km = 3
)
```

# Retrieve clustering

Extract the samples (columns) and genes (rows) found by k-means clustering.

## Summary column clusters

The column clusters can be split into a low (1), medium(2) and high(3) Ig-expressing groups

We can retrieve the gene/sample names either with row/column_order functions, which return indices:

```{r, fig.height = 7, fig.width= 10}
fig.hm = draw(fig.hm)
```

```{r, collapse=F}
#row_order(fig.hm)
#column_order(fig.hm) #Shows that `3` is high, `2` is medium and `1` is low
lapply(column_order(fig.hm), length)

col_clust = 
bind_rows(
  tibble(sample_name = colnames(geneEx)[column_order(fig.hm)$`1`],IG_col_cluster = "low", col_cluster=1),
  tibble(sample_name = colnames(geneEx)[column_order(fig.hm)$`2`],IG_col_cluster = "medium", col_cluster=2),
  tibble(sample_name = colnames(geneEx)[column_order(fig.hm)$`3`],IG_col_cluster = "high", col_cluster=3)
) %>% left_join(.,select(as.data.frame(colData(dds)), sample_name, study_group, PAM50), by="sample_name") %>%
  mutate(IG_col_cluster = factor(IG_col_cluster, levels = c("low", "medium", "high")))

addmargins(table(col_clust$study_group, col_clust$IG_col_cluster), 1)

nrow(col_clust)

#col_clust %>% arrange(sample_name)
```

Define colors for IG clusters

```{r}
IG_expression_colors = cp$tertile_colors
names(IG_expression_colors) = c("low", "medium", "high")
color_grid(IG_expression_colors)
```

## Figure 1D - Barplot IG clusters

```{r, fig.width=8}
study_ig = col_clust %>% ggplot(aes(x = IG_col_cluster, fill=study_group)) + geom_bar(position = "fill") +
  scale_fill_manual(values=study_colors) +
  ggtitle("Study group by IG expression cluster")


clust_ig = col_clust %>% ggplot(aes(fill = factor(IG_col_cluster, levels=c("high", "medium","low")), x=study_group)) +
  geom_bar(position = "fill") +
  #scale_fill_brewer(type="div", palette = 9, direction = 1) +
  scale_fill_manual(values = IG_expression_colors) +
  ggtitle("Division of IG expression clusters by study group") +
  labs(fill="IG_col_cluster")


study_ig
clust_ig

```

```{r}
pdf(file = here(params$IGbarplot),
    width = 10, height = 7)
study_ig
dev.off()
```

```{r}
pdf(file = file.path(clusterDir, "11_barplots_ig_clusters.pdf"),
    width = 8, height = 6)
study_ig
clust_ig
dev.off()
```

## Summary row clusters

Extract and store the row clusters:

```{r}
rowdend.hm = row_dend(fig.hm)

row_clust = 
bind_rows(
  tibble(gene_name = as.hclust(rowdend.hm$`1`)$labels,row_cluster = 1),
  tibble(gene_name = as.hclust(rowdend.hm$`2`)$labels,row_cluster = 2),
  tibble(gene_name = as.hclust(rowdend.hm$`3`)$labels,row_cluster = 3)
) %>%
  right_join(sig.invr,
             ., by = "gene_name") #%>% select(gene_name, immune_gene, row_cluster, description, everything())
```

```{r}
row_clust[row_clust$row_cluster==1,] %>% head()
row_clust[row_clust$row_cluster==2,] %>% head()
row_clust[row_clust$row_cluster==3,] %>% head()
```

```{r}
write.xlsx(list(row_clusters = row_clust, col_clusters = col_clust),
           file = file.path(clusterDir,"11_inv_clusters.xlsx"))
```

# Clusters as metagenes

This is a sanity check to make sure the clustering makes sense.

Remember that row cluster 3 is the IG genes.

```{r}
row_clust %>% filter(row_cluster == 3) %>%
  pull(gene_name)
```

Collapse clusters into single metagene by averaging experession, then test assocation between metagene and survival/subgroups

```{r}
metagene_mat = 
  rbind(colMeans(geneEx[rownames(geneEx) %in% row_clust[row_clust$row_cluster==1,"gene_name"], ]),
        colMeans(geneEx[rownames(geneEx) %in% row_clust[row_clust$row_cluster==2,"gene_name"], ]),
        colMeans(geneEx[rownames(geneEx) %in% row_clust[row_clust$row_cluster==3,"gene_name"], ])
  )

rownames(metagene_mat) = c("metagene1", "metagene2","metageneIG")
metagene_mat = scrime::rowScales(metagene_mat)

#pheatmap::pheatmap(metagene_mat, show_colnames = F, cluster_rows = F,
#                   annotation_col = data.frame(row.names = sd$sample_name,
#                                               study_group = sd$inv_vs_rest))

#Top column annotation
colTop <- HeatmapAnnotation(df=data.frame(row.names = sd$sample_name,
                                          study_group = sd$PPBC),
                            which="col",
                            col = list(study_group = ppbc_colors),
                            annotation_legend_param = list(list(title = "PPBC")))

#Bottom column annotation

colBottom <- HeatmapAnnotation(df=data.frame(row.names = sd$sample_name,
                                             PAM50 = sd$PAM50),
                               which="col",
                               col = list(PAM50 = pam_colors))

set.seed(1)
Heatmap(metagene_mat,column_km = 3,
        top_annotation = colTop,
        bottom_annotation = colBottom,
        #left_annotation = rowAnno,
        heatmap_legend_param = list(title="rowscaled vst"),
        show_row_names = T,
        show_column_names = F,
        cluster_rows = F,
        #row_names_gp = gpar(fontsize = row_size),
        #column_names_gp = gpar(fontsize = col_size),
        #column_title = title
)
```

Swap rows and columns.

```{r}

rowAnno <- HeatmapAnnotation(
  df=data.frame(row.names = sd$sample_name,
                study_group = sd$PPBC,
                PAM50 = sd$PAM50),
  which="row",
  col = list(study_group = ppbc_colors,
             PAM50 = pam_colors),
  annotation_legend_param = list(list(title = "PPBC"))
)


set.seed(1)
Heatmap(t(metagene_mat),row_km = 3,
        #top_annotation = colTop,
        #bottom_annotation = colBottom,
        left_annotation = rowAnno,
        heatmap_legend_param = list(title="colscaled vst"),
        show_row_names = F,
        show_column_names = T,
        cluster_rows = F,
        #row_names_gp = gpar(fontsize = row_size),
        #column_names_gp = gpar(fontsize = col_size),
        #column_title = title
)
```

# Kruskal wallis

Association between cluster status and groups

For inv vs rest

```{r}
stopifnot(all(colnames(metagene_mat) == sd$sample_name))

kruskal.test(x = metagene_mat[rownames(metagene_mat)== "metageneIG"], 
             g = sd$inv_vs_rest)
```

```{r}
kruskal.test(x = metagene_mat[rownames(metagene_mat)== "metageneIG"], 
             g = sd$study_group)
```


```{r}
kruskal.test(x = metagene_mat[rownames(metagene_mat)== "metageneIG"], 
             g = sd$PAM50)
```

```{r}
kruskal.test(x = sd$inv_vs_rest, 
             g = sd$PAM50)
```

# Chi-square tests

For study group

```{r}
chisq.test(as.matrix(table(col_clust$study_group, col_clust$IG_col_cluster)),
           simulate.p.value = T)
```

Recode as OVR

```{r}
col_clust %>%
  mutate(inv_vs_rest = ifelse(study_group=="ppbcpw", "inv", "rest")) %>%
  select(IG_col_cluster, inv_vs_rest) %>% table() %>%
  as.matrix() %>% chisq.test(., simulate.p.value = T)
```

For PAM50

```{r}
chisq.test(as.matrix(table(col_clust$PAM50, col_clust$IG_col_cluster)),
           simulate.p.value = T)
```

# Clustering involution samples

We will attempt clustering both before and after removal of molecular subtype as a confounding factor. See if the samples also cluster based on immune high-medium-low categories.

```{r}
inv.dds$IG_cluster = col_clust[match(colData(inv.dds)$sample_name,
                                     col_clust$sample_name), "IG_col_cluster"][[1]]

sd$IG_cluster = col_clust[match(sd$sample_name,col_clust$sample_name), "IG_col_cluster"][[1]]

table(inv.dds$IG_cluster, inv.dds$study_group)
```

## Inv heatmaps

Subset down to inv and define heat map annotations

```{r}
#Subst to inv
sd.inv = as.data.frame(colData(inv.dds))
geneEx.inv = geneEx[,colnames(geneEx) %in% sd.inv$sample_name]
#head(select(sd.inv, sample_name, IG_cluster))


#Top column annotation
ann_top = select(sd.inv, IG_cluster, death)
#Ensure that the order matches
ann_top = ann_top[match(colnames(geneEx.inv),rownames(ann_top)), ,drop=F]
colTop <- HeatmapAnnotation(df=ann_top, which="col",
                              col = list(IG_cluster=IG_expression_colors,
                                         death = sp$overall_survival))

#Bottom column annotation
ann_bottom = sd.inv[,"PAM50", drop=F]
colBottom <- HeatmapAnnotation(df=ann_bottom, which="col",
                               col = list(PAM50 = pam_colors))

# Row annotation
anno_rows = res_invr %>%
    select(gene_name, Type) %>%
    distinct() %>%
    filter(!duplicated(gene_name)) %>%
    column_to_rownames("gene_name")

anno_rows = anno_rows[match(rownames(geneEx.inv),rownames(anno_rows)), ,drop=F]
anno_rows = anno_rows[rownames(anno_rows) %in% sig.invr$gene_name,,drop=F]

#Essential that the order be the same!
rowAnno = HeatmapAnnotation(df=anno_rows, which="row", col=list(Type = gene_colors))

```

## Significant genes in involution vs rest

```{r, fig.height=7, fig.width=10}
set.seed(1)

invclust =  Heatmap(as.matrix(geneEx.inv[rownames(geneEx.inv) %in% sig.invr$gene_name,]),
          top_annotation = colTop,
          bottom_annotation = colBottom,
          left_annotation = rowAnno,
          heatmap_legend_param = list(title="vst counts"),
          show_row_names = T,
          show_column_names = F,
          cluster_rows = T,
          row_names_gp = gpar(fontsize = 8),
          column_names_gp = gpar(fontsize = 8),
          column_title = "Clustering involution samples, sig genes",
          column_km = 3, column_km_repeats=50, row_km = 3, row_km_repeats = 50)

invclust.row =  Heatmap(scrime::rowScales(geneEx.inv[rownames(geneEx.inv) %in% sig.invr$gene_name,]) %>% as.matrix,
          top_annotation = colTop,
          bottom_annotation = colBottom,
          left_annotation = rowAnno,
          heatmap_legend_param = list(title="rowscaled vst"),
          show_row_names = T,
          show_column_names = F,
          cluster_rows = T,
          row_names_gp = gpar(fontsize = 8),
          column_names_gp = gpar(fontsize = 8),
          column_title = "Clustering involution samples, sig genes",
          column_km = 3, column_km_repeats=50, row_km = 3, row_km_repeats = 50)

invclust
invclust.row

```

## Involution: Top variant genes

Look for patterns within the involution group taking the top variant genes.

```{r}
rv = rowVars(as.matrix(geneEx.inv))
select = order(rv, decreasing = TRUE)[seq_len(min(500, 
            length(rv)))]
varEx.vst = geneEx.inv[select,]
vargenes = rownames(varEx.vst)
dim(varEx.vst)
```

We just see PAM50.

```{r, fig.width=10}
set.seed(1)

# Row annotation
anno_rows = res_invr %>%
    select(gene_name, Type) %>%
    distinct() %>%
    filter(!duplicated(gene_name)) %>%
    column_to_rownames("gene_name")

anno_rows = anno_rows[match(rownames(varEx.vst),rownames(anno_rows)), ,drop=F]
#anno_rows = anno_rows[rownames(anno_rows) %in% sig.invr$gene_name,,drop=F]

#Essential that the order be the same!
rowAnno = HeatmapAnnotation(df=anno_rows, which="row", col=list(Type = gene_colors))
inv.var =  
Heatmap(as.matrix(varEx.vst),
          top_annotation = colTop,
          bottom_annotation = colBottom,
          left_annotation = rowAnno,
          heatmap_legend_param = list(title="vst counts"),
          show_row_names = F,
          show_column_names = F,
          cluster_rows = T,
          row_names_gp = gpar(fontsize = 8),
          column_names_gp = gpar(fontsize = 8),
          column_title = "Clustering involution samples, top 500 most variant genes vst",
          column_km = 3, column_km_repeats=50, row_km = 3, row_km_repeats = 50)

inv.var.row =  
Heatmap(as.matrix(scrime::rowScales(varEx.vst)),
          top_annotation = colTop,
          bottom_annotation = colBottom,
          left_annotation = rowAnno,
          heatmap_legend_param = list(title="vst counts"),
          show_row_names = F,
          show_column_names = F,
          cluster_rows = T,
          row_names_gp = gpar(fontsize = 8),
          column_names_gp = gpar(fontsize = 8),
          column_title = "Clustering involution samples, top 500 most variant genes vst rowscaled",
          column_km = 3, column_km_repeats=50, row_km = 3, row_km_repeats = 50)

inv.var
inv.var.row
```

# Survival analysis IG clusters

```{r}
surv <- readRDS(here(params$coxdata))
print(paste(nrow(surv), "samples have complete survival data"))
surv <- left_join(surv,
           select(col_clust, sample_name, IG = IG_col_cluster),
           by = "sample_name")

stopifnot(nrow(filter(surv, is.na(sample_name)))==0)
stopifnot(nrow(filter(surv, is.na(time_OS_months)))==0)

surv <- as.data.frame(surv)

saveRDS(surv, here("data/rnaseq/interim/11_ig_survdata.Rds"))
```

## Overall survival

Baseline: Study group

```{r}
survdiff(survival::Surv(time =time_OS_months, event = death) ~ study_group, data=surv)
```

### KM IG clusters

```{r}
ig_colors = cp$tertile_colors
names(ig_colors) = paste0("IG=", names(ig_colors))
ggsurvplot( 
    fit = survfit(Surv(time = time_OS_months, event=death) ~ IG, data = surv), 
    xlab = "Years", 
    ylab = "Overall survival probability",
    title = "Overall survival by IG cluster",
    palette = ig_colors,
    pval = T,
    ggtheme = theme_bw())
```

### KM IG clusters, facet by PAM50

Basal and Her2 are spread among them, but Lum A and B are almost universally low IG groups.

```{r}
table(sd$PAM50, sd$IG_cluster)
```

```{r}
ggsurvplot(
    fit = survfit(Surv(time=time_OS_months, event=death) ~ PAM50 + IG, data = surv), 
    xlab = "Months", 
    ylab = "Overall survival probability",
    title = "IG cluster by study group",
    pval = T,
    palette = cp$tertile_colors,
    ggtheme = theme_bw(),
    facet.by = "PAM50")


```

### KM IG clusters, facet by study group

```{r, fig.width=8}
sg_ig <-  ggsurvplot(
    fit = survfit(Surv(time=time_OS_months, event=death) ~ PPBC + IG, data = surv), 
    xlab = "Months", 
    ylab = "Overall survival probability",
    title = "IG cluster by study group",
    pval = T,
    palette = cp$tertile_colors,
    ggtheme = theme_bw(),
    facet.by = "PPBC")

sg_ig
```

### Figure 1E: KM study group, faceted by IG cluster

Remove lactation from survival curves for publication due to lack of power.

```{r, fig.width=8}
nolacsurv <- droplevels(as.data.frame(filter(surv, study_group != "ppbcdl")))

ig_kmOS <- ggsurvplot(
    fit = survfit(Surv(time=time_OS_months, event=death) ~ study_group + IG,
                  data = nolacsurv), 
    xlab = "Months", 
    ylab = "Overall survival probability",
    title = "Study group by IG cluster",
    pval = T, 
    facet.by = "IG",
    palette = study_colors[!names(study_colors) %in% c("ppbcdl", "rest")],
    ggtheme = theme_bw(),
    risk.table = T)

ig_kmOS
```

Risk table is hard to plot legibly.

### Risk table

```{r, fig.height=7, fig.width=12}
risktable <- function(ggkm, response, text.size = 8, ...){
  ggplot(ggkm$data, aes(time, study_group)) + 
    ggrepel::geom_text_repel(aes(label = n.risk)) +
    facet_wrap(as.formula(paste("~", response)), ...) +
    theme_bw() + 
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          strip.background = element_rect(fill="white", size = 1),
          axis.text.x = element_text(color="black", size = text.size+2),
          axis.text.y = element_text(color="black", size = text.size),
          axis.title = element_text(color="black", size = text.size+4),
          strip.text = element_text(color="black", size = text.size))+
    ggtitle("Risk Tables") +
    xlab("Time") + ylab("")
}

rt_ig_kmOS <- risktable(ig_kmOS, response = "IG", nrow=3)
rt_ig_kmOS
```

### Save output

```{r}
pdf(file = here(params$IGkaplanOS),
    width = 12, height = 7)
ig_kmOS
rt_ig_kmOS
dev.off()
```

Save risk table data separately.

```{r}
write_csv(ig_kmOS$data, here(params$rtIGkaplanOS))
```

### Multivariate Cox

```{r}
ig_cox = 
coxph(
  Surv(time=time_OS_months,
       event=death) ~ age_at_diagnosis + grade + stage + surgery + radiotherapy + strata(hormonetherapy) + chemotherapy + herceptin + strata(PAM50) + IG + study_group,
  data = surv)

summary(ig_cox)
```

## Distant recurrence free survival

### Univariate

```{r}
survdiff(survival::Surv(time = time_DRS_months, event = distant_recurrence) ~ IG, data=surv)
```

### KM DRS: IG clusters

```{r}
ig_colors = cp$tertile_colors
names(ig_colors) = paste0("IG=", names(ig_colors))

ggsurvplot(
    fit = survfit(Surv(time=time_DRS_months, event=distant_recurrence) ~ IG, data = surv), 
    xlab = "Months", 
    ylab = "DRS",
    title= "IG clusters: DRS (univariate)",
    pval = T,
    palette = ig_colors,
    ggtheme = theme_bw())
```

### KM DRS IG clusters, facet by study group

```{r, fig.width=10}
ggsurvplot(
    fit = survfit(Surv(time_DRS_months, event=distant_recurrence) ~ PPBC + IG, data = surv), 
    xlab = "Months", 
    ylab = "DRS probability",
    title = "DRS - IG clusters by PPBC",
    pval = T,
    facet.by = "PPBC",
    palette = cp$tertile_colors,
    ggtheme = theme_bw()
)
```

### Sup Figure S9: KM DRS study group, faceted by IG cluster

```{r, fig.width=10}
nolacsurv <- droplevels(as.data.frame(filter(surv, study_group != "ppbcdl")))

IGkaplanDRS <- ggsurvplot(
    fit = survfit(Surv(time=time_DRS_months, event=distant_recurrence) ~ study_group + IG,
                  data = nolacsurv), 
    xlab = "Months", 
    ylab = "DRS probability",
    title = "DRS - PPBC by IG cluster",
    pval = T, facet.by = "IG",
    palette = study_colors[!names(study_colors) %in% c("ppbcdl", "rest")],
    ggtheme = theme_bw())

IGkaplanDRS
```
### Risk table

```{r, fig.height=7, fig.width=12}
rtIGkaplanDRS <- risktable(IGkaplanDRS, response = "IG", nrow=3)
rtIGkaplanDRS
```

### Save output

```{r}
supFigDir <- file.path(here("figures/supfigs"))
dir.create(supFigDir, showWarnings = F)
pdf(file = here(params$IGkaplanDRS),width = 12, height = 7)
IGkaplanDRS
rtIGkaplanDRS
dev.off()
```

Save risk table data separately.

```{r}
write_csv(IGkaplanDRS$data, here(params$rtIGkaplanDRS))
```

### Multivariate Cox DRS

```{r}
ig_drs = coxph(
  Surv(time=time_DRS_months,
       event=distant_recurrence) ~ age_at_diagnosis + year_diagnosis + grade +
    stage + surgery + radiotherapy + strata(hormonetherapy) + chemotherapy +
    herceptin + strata(PAM50) + IG + PPBC,
  data = surv)

summary(ig_drs)
```

# Session info

```{r}
sessionInfo()
```
