---
title: "Differential expression for PAM50 subgroups"
author: "Kat Moore"
date: "`r Sys.Date()`"
output: 
   html_document:
    toc: yes
    toc_float: yes
    toc_depth: 4
    highlight: tango
    df_print: paged
params:
  comparison:  
    value: "ppbc_inv_vs_rest"
    input: select
    choices: [
      "ppbc_inv_vs_rest",
      "ppbc_inv_vs_non_prbc", 
      "ppbc_inv_vs_prbc"
      ]

---

This paramaterized Rmarkdown template is designed to show differential expression results for PPBC subsetted by PAM50 subgroup.
For each subgroup, the following comparisons were made:

* "ppbc_inv_vs_rest" : Involuting samples vs all others (lactating + nulliparous + pregnant)
* "ppbc_inv_vs_non_prbc" : Involuting samples vs nulliparous samples 
* "ppbc_inv_vs_prbc" : Involuting samples vs pregnant samples


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(DESeq2)
library(ggrepel)
library(survival)
library(survminer)
library(ComplexHeatmap)
library(RColorBrewer)
library(openxlsx)
library(here)
library(UpSetR)
library(ggpubr)
library(tictoc)
library(tidyverse)

theme_set(theme_bw())
```

This report will show results for the comparison *`r params$comparison`*.

Available subgroups:

* Basal
* Her2
* LumA
* LumB

The Normal subgroup was excluded due to insufficient samples.

# Load data {.tabset}

Read in relevant dds files.

```{r read dds}
psub <- c("Basal", "Her2", "LumA", "LumB")
resDir <- here("data", "Rds", "subgroup_diffex")
dds_files <- paste0("14_dds_", psub, "_", params$comparison, ".Rds")
ddsl = lapply(file.path(resDir, dds_files), readRDS)
names(ddsl) = psub
dds_files
```

Diffex formula of dds files:

```{r}
lapply(ddsl, design)
```

Auto-detect column of interest based on comparison name.
Print number of samples in each group.
Also ensure that there are no empty levels in study group.

```{r}
coi = if_else(params$comparison == "ppbc_inv_vs_rest", "inv_vs_rest", "study_group")
group = str_split(params$comparison, "_vs_")[[1]][1]
ref = str_split(params$comparison, "_vs_")[[1]][2]

fix_sg_levels <- function(x){
  x <- x[,colData(x)[,coi] %in% c(group, ref)]
  colData(x)[,coi] <- droplevels(colData(x)[,coi])
  return(x)
}
ddsl = lapply(ddsl, fix_sg_levels)
lapply(ddsl, function(x) table(colData(x)[,coi]))
```

Read in relevant apeglm shrunken results:

```{r}
ape_files <- paste0("14_ape_", psub, "_", params$comparison, ".Rds")
apel <- lapply(file.path(resDir, ape_files), readRDS)
names(apel) = psub
ape_files
```

Show the comparison column and levels from apeglm results.
This should be the same for every subgroup and match the file name.

```{r}
lapply(apel, function(x) x@elementMetadata$description[2])

stopifnot(length(unique(unlist(lapply(apel, function(x) x@elementMetadata$description[2])))) == 1)
```


## Gene sets

Gene annotation:

```{r}
gx_annot <- read_tsv(here("data/metadata/01_tx_annot.tsv"))
gx_annot = gx_annot %>% select(ensembl_gene_id = gene_id, gene_name,
                               gene_type, description = gene_description) %>% distinct()
```

For Fisher's exact tests

```{r}
gene_sets = list(
  go_bp = flexgsea::read_gmt(here("data","external","gmt","c5.bp.v7.0.symbols.gmt")),
  hallmark = flexgsea::read_gmt(here("data","external","gmt","h.all.v7.0.symbols.gmt")),
  c2_canon = flexgsea::read_gmt(here("data","external","gmt","c2.cp.v7.0.symbols.gmt")),
  c2_cgp = flexgsea::read_gmt(here("data","external","gmt","c2.cgp.v7.0.symbols.gmt"))
)
```


For downstream analyses, we'd like to know which of the differentially expressed genes are immunologically relevant.

An immune gene is defined as follows:

Either immune/immuno/interleukin is part of the gene name and description OR the gene is part of the the ImmPort database. See https://www.innatedb.com/redirect.do?go=resourcesGeneLists

The Immunology Database and Analysis Portal (ImmPort) system was developed under the Bioinformatics Integration Support Contract (BISC) Phase II by the Northrop Grumman Information Technology Health Solutions team for the NIH, NIAID, and DAIT. The principal investigator of the BISC project is Dr. Richard Scheuermann at University of Texas Southwestern Medical Center. The list of immunologically related genes in ImmPort is a collection of ~6,000 human genes, which was formed with the goal of retrieving all genes that have immune system-related functions. This list was generated using automatic searches of EntrezGene and Gene Ontology records using immunology-related keywords. The list was then manually curated by immunology experts examining various literature sources. 

```{r}
immune_gene_list <- read_csv(here("data", "external","gene-sets","InnateDB_genes.csv"))
head(immune_gene_list)
```


## Colors

```{r}
cp = readRDS(here("data", "Rds", "color_palettes.Rds"))
study_colors = cp$study_colors
pam_colors = cp$pam_colors
gene_colors = cp$gene_colors
ppbc_colors = cp$ppbc_colors

sp = readRDS(here("data/Rds/survival_colors.Rds"))

#Minor tweak to death colors to avoid white on white background
death_colors = sp$overall_survival
death_colors[names(death_colors) == 0] <- "lightgray"
#color_grid(death_colors)
```


## Vsts

Variance stabilizing transformations for visualizations.

```{r}
blind_vst = function(dds){
  design(dds) = formula(~ 1)
  vsd = vst(dds, blind=T)
  #mat = assay(vsd)
  return(vsd)
}

tic()
vsdl = lapply(ddsl, blind_vst)
toc()
```

Add a death column to all vsds and dds for plotting:

```{r}
add_death <- function(x){
  x$death <- factor(x$overall_survival, levels = c(0, 1))
  return(x)
}

ddsl = lapply(ddsl, add_death)
vsdl = lapply(vsdl, add_death)
```

Convert to gene symbols and deduplicate for heatmaps.

```{r gene symbols}
summarize_expression_duplicate_ids <- function(mat, id_column=NULL, f=colMeans, verbose=F){
  require(dplyr)
  
  #Easiest way to write functions with dplyr is to standardize the column name
  
  input = mat #Save input before doing anything to it in case there are no duplicates
  
  if(is.null(id_column)==T){
    mat = as.data.frame(mat)
    mat = rownames_to_column(mat, "symbol")
    id_column = "symbol"
  }
  
  if(id_column != "symbol"){
    colnames(mat)[colnames(mat)==id_column] <- "symbol"
  }
  
  if (sum(duplicated(mat$symbol)) == 0){
    print("No duplicate symbols")
    return(input)
  }
  
  if(verbose==T){
    print(paste("Starting with gene expression matrix containing", nrow(mat), "rows."))
  }
  
  #Make frequency table
  id_table <- as.data.frame(table(mat$symbol))
  
  #Identify duplicate genes
  dups <- id_table$Var1[id_table$Freq > 1]
  stopifnot(length(dups) == length(unique(dups)))
  if(verbose == T){
    print(paste("Number of genes with duplicate names:", length(dups)))
  }
  
  #Set aside rows with unique gene names
  nodup_df <- mat[!mat$symbol %in% dups,]
  
  #Set aside rows with duplicate ids
  dup_df <- mat[mat$symbol %in% dups,]
  stopifnot(nrow(nodup_df) + nrow(dup_df) == nrow(mat))
  
  #Sort by recurring id
  dup_df <- dup_df[order(dup_df$symbol),]
  
  if(verbose==T){
    print(paste("Number of rows with duplicate gene ids:", nrow(dup_df)))
  }
  
  #Mean expression fpkm of genes with the same symbol
  mean_exps <- matrix(ncol = ncol(dup_df)-1, nrow=0) #Empty matrix, -1 gene symbol column
  
  for (i in 1:length(unique(dup_df$symbol))){
    #Subset rows with same symbol, discard symbol column, then apply aggregate function
    exp <- f(as.matrix(dup_df[dup_df$symbol==unique(dup_df$symbol)[i], -1]))
    mean_exps <- rbind(mean_exps, exp)
  }
  stopifnot(nrow(mean_exps) == length(unique(dup_df$symbol)))
  
  rownames(mean_exps) <- unique(dup_df$symbol)
  mean_exps <- as.data.frame(mean_exps) %>% rownames_to_column("symbol")
  
  dedupped_df <- rbind(mean_exps, nodup_df)
  dedupped_df <- dedupped_df[order(dedupped_df$symbol),]
  
  stopifnot(length(unique(dedupped_df$symbol))==length(dedupped_df$symbol)) #All symbols should not be unique
  stopifnot(nrow(mat) - #starting number
              nrow(dup_df) + #rows with duplicate genes...
              length(dups) == #...which condense down into this many unique genes...
              nrow(dedupped_df)) #...should equal the number of rows in the final matrix
  
  if (verbose == T){
    print(paste("Number of genes after applying", substitute(f),  "to duplicate ids:", nrow(dedupped_df)))
  }
  
  rownames(dedupped_df) = NULL #Required for column_to_rownames
  dedupped_df = column_to_rownames(dedupped_df, "symbol")
  
  return(dedupped_df)
}

geneEx = lapply(vsdl, function(x) rownames_to_column(as.data.frame(assay(x)), "ensembl_gene_id") %>%
                  right_join(select(gx_annot, gene_name, ensembl_gene_id), ., by = "ensembl_gene_id") %>%
                  select(-ensembl_gene_id))
  
geneEx = lapply(geneEx, function(x) as.matrix(summarize_expression_duplicate_ids(x, id_column = "gene_name")))
```

# Calculate reports {.tabset}

In the following order, create a report for each of:

```{r}
psub
```

Thresholds:

```{r}
fdr = 0.05 #padj threshold
alfc = 0.5 #abs(l2fc) threshold
```

Title substring for plots:

```{r}
#Uses study group encoding
#ts = str_replace(params$comparison, "_vs_", " vs ") 

#First letter capital
CapStr <- function(y) {
  c <- strsplit(y, " ")[[1]]
  paste(toupper(substring(c, 1,1)), substring(c, 2),
      sep="", collapse=" ")
}

#Uses PPBC encoding
tdict = distinct(as.data.frame(colData(ddsl[[1]])[,c("PPBC", coi)]))
colnames(tdict)[2] <- "study_group"

ts = paste0(CapStr(as.character(tdict$PPBC[tdict$study_group == group])),
       " vs ",
       ifelse(ref == "rest", ref,
               as.character(tdict$PPBC[tdict$study_group == ref])))

ts #title substring

```

## Annotate results

Add gene name, consolidate biotypes, etc.

```{r annotate results}
annotate_results <- function(results_object, anno_df = gx_annot, immune_df = immune_gene_list,
                             mark_immune = T, mark_outliers = T, pthresh, lfcthresh){
  
  #Takes a res object from deseq, merges it with a gene annotation file, and an optional immune list to create a data frame sorted by adjusted p value
  #Type is an aggregate of immunoglobulin genes from gene type and genes with a known immune function from ImmPort
  require(tidyverse)
  
  anno_res <- results_object %>% as.data.frame() %>% rownames_to_column("ensembl_gene_id") %>%
    right_join(anno_df,., by = "ensembl_gene_id") %>% arrange(padj)
  
  if (mark_immune==T){ #Shows genes that defined as immune based on external list
    anno_res = anno_res %>% mutate(ImmPort_gene = if_else(
      ensembl_gene_id %in% immune_df$ensembl, T, F
    ))
    
    anno_res = anno_res %>%
      mutate(Type=case_when( #Order is hierarchical
        str_detect(string=gene_type,pattern="^IG") ~ "immunoglobulin",
        str_detect(string=gene_type,pattern="^TR") ~ "T cell receptor",
        ImmPort_gene == T ~ "immune protein coding",
        gene_type == "protein_coding" ~ "protein coding",
        TRUE ~ "other noncoding"))
    
    anno_res = anno_res %>% select(gene_name,gene_type,Type, everything())
  }
  
  #DESeq sets both the pval and padj to NA for genes above a Cooks distance threshold
  #Cook’s distance is a measure of how much a single sample is influencing the fitted coefficients for a gene,
  #and a large value of Cook’s distance is intended to indicate an outlier count.
  #If a row is filtered by automatic independent filtering, for having a low mean normalized count,
  # then only the adjusted p value will be set to NA.
  #Keep track of both possibilities
  
  if (mark_outliers == T){
    anno_res = anno_res %>%
      mutate(cooks_outlier = if_else(baseMean > 0 & is.na(pvalue), T, F)) %>%
      mutate(indepfilt = if_else(!is.na(pvalue) & is.na(padj), T, F))
  }
  
  #Significance column
  anno_res = anno_res %>%
     mutate(sig = padj <= !!pthresh & abs(log2FoldChange) >= !!lfcthresh)
  
  #Clean up text presentation, round numbers, reorder columns
  anno_res = anno_res %>% arrange(padj) %>%
    mutate(description = str_remove(description, "\\[.*\\]")) %>%
    mutate(padj = signif(padj, 2),
           log2FoldChange = round(log2FoldChange, 2),
           baseMean = round(baseMean, 1),
           lfcSE = signif(lfcSE, 2),
           pvalue = signif(pvalue, 2),
    ) %>%
    select(gene_name, gene_type, sig,
           padj,
           l2fc = log2FoldChange,
           lfcSE,
           baseMean,
           pvalue,
           description,
           everything())
  
  
  return(anno_res)
}

annotated_results <- lapply(apel, function(x) annotate_results(x, pthresh = fdr, lfcthresh = alfc))

#lapply(annotated_results, head)
```

## Significant results

Based on padj threshold `r fdr` and l2fc threshold `r alfc`.

```{r}
sig_results <- lapply(annotated_results, function(x) dplyr::filter(x, sig == T))
```

Significant genes by subgroup in `r params$comparison`:

```{r}
lapply(sig_results, nrow)
```

## Volcano plots

```{r volcano plots}
volcano_plot <- function(x, title="", path_save_fig = NULL, pthresh = 0.05, absl2fc=0.5,
                         shape_col=NULL, dedup_ids = T, colors = gene_colors){
  
  require(tidyverse)
  require(ggrepel)
  require(here)
  
  volcdf <- x %>%
    mutate(NegLog10FDR = -log10(padj),
           Significant = (padj < !!pthresh & abs(l2fc) > !!absl2fc),
           Color = if_else(
             sig == T, Type, "n.s."
           )) %>%
    filter(!is.na(Significant))
  
  if (dedup_ids == T){
    print("Summarizing -log10FDR and log2FoldChange by median for duplicate gene names in volcano plot")
    volcdf = volcdf %>% group_by(gene_name, Significant, Type, Color) %>%
      summarise(log2FoldChange = median(l2fc), NegLog10FDR = median(NegLog10FDR)) %>%
      arrange(desc(NegLog10FDR))
  }
  
  colors = c(colors,"n.s."="gray")
  
  volcplot <-ggplot(volcdf, aes(x=log2FoldChange, y=NegLog10FDR, color=Color)) +
    geom_point() +
    scale_color_manual(values=colors) +
    geom_hline(yintercept = -log10(pthresh), colour="darkred", linetype="dashed") +
    geom_vline(xintercept = absl2fc, color="darkred", linetype="dashed") +
    geom_vline(xintercept = -absl2fc, color="darkred", linetype="dashed") +
    theme_bw()
  
  volcplot <-
    volcplot +
    geom_text_repel(data=head(volcdf[volcdf$Significant==T,], 10),
                    aes(label=gene_name), show.legend = F) +
    ggtitle(title) +
    labs(y="-log10(FDR)")
  
  if (is.null(path_save_fig)==F){
    suppressMessages(ggsave(filename=path_save_fig, plot = volcplot))
  }
  
  if (is.null(shape_col)==F){
    volcplot = volcplot + aes(shape=get(shape_col)) + labs(shape=shape_col)
  }
  
  return(volcplot)
  
}

volcano_plots <- Map(volcano_plot, x = annotated_results, title = paste0(ts, ": ", names(annotated_results)))
#volcano_plots$Basal
```

## Pathways

```{r fisher pathways}
source(here("src", "enrichment-analysis-functions.R"))
pathway_cutoff = 0.1

tic("Fisher pathways")
pathways = lapply(psub, function(x){
  s = sig_results[[x]]
  d = annotated_results[[x]]
  fisher_pathways(
      sig_genes = s$gene_name,
      background_genes = d$gene_name,
      list_signatures = gene_sets,
      fdr_cutoff = pathway_cutoff,
      verbose = T,
      collapse_rows=T)
  })
names(pathways) = psub
toc()
```

## Heatmaps

Function for drawing heatmaps

```{r}
deseq_heatmap = function(mat = geneEx[[cur]],
                         sampledata = as.data.frame(colData(ddsl[[cur]])),
                         sig_results = sig_results[[cur]],
                         groups_to_plot=levels(as.data.frame(colData(ddsl[[cur]]))[,coi]),
                         title=ts,
                         intgroup = coi,
                         top_vars = "PPBC",
                         top_colors = list(PPBC=ppbc_colors),
                         bottom_vars = "death",
                         bottom_colors = list(death = death_colors),
                         row_colors = list(Type = gene_colors),
                         legend_title = NULL,
                         row_scale = T,
                         maxn_rownames = 50,
                         maxn_genes = 5000,
                         row_size = 8, col_size = 8,
                         show_col_names = F, ...){

  #Reduce genes to significant only
  sig_results = sig_results %>% arrange(padj)
  genestoplot = sig_results$gene_name[1:maxn_genes]
  mat = mat[rownames(mat) %in% genestoplot, ]
  
  #Reduce genes and sample data to compared groups only
  #sampledata = sampledata %>% dplyr::filter(study_group %in% groups_to_plot)
  sampledata = sampledata %>% dplyr::filter(!!as.symbol(intgroup) %in% groups_to_plot)
  mat = mat[,colnames(mat) %in% sampledata$sample_name]
  stopifnot(identical(colnames(mat), sampledata$sample_name))
  
  #Toggle row labels depending on size of heatmap
  if (nrow(mat) <= maxn_rownames){
    show_row_names = T
  } else {
    show_row_names = F
  }
  
  #Row scale settings
  if (row_scale==T){
    mat = scrime::rowScales(mat)
  }
  #Change legend according to whether input is scaled
  if (row_scale==T){
    hlp = list(title="rowscaled vst")
  } else {
    hlp = list(title="vst counts")
  }
  
  #Ensure input is matrix
  mat = as.matrix(mat)
  
  #Heatmap annotation
  ann_top = sampledata[,top_vars, drop=F]
  
  #Top column annotation
  colTop <- HeatmapAnnotation(df=ann_top, which="col",
                              col = top_colors
                              #annotation_legend_param = list(list(title = legend_title))
                              )
  #return(colTop)
  #Bottom column annotation
  ann_bottom = sampledata[,bottom_vars, drop=F]
  colBottom <- HeatmapAnnotation(df=ann_bottom, which="col", col = bottom_colors)
  
  # Row annotation
  anno_rows = sig_results %>%
    select(gene_name, Type) %>%
    distinct() %>%
    filter(!duplicated(gene_name)) %>%
    column_to_rownames("gene_name")
  
  #Essential that the order be the same!
  anno_rows = anno_rows[match(rownames(mat),rownames(anno_rows)), ,drop=F]
  
  rowAnno = HeatmapAnnotation(df=anno_rows, which="row",
                              col=row_colors
                              #col=list(Type = row_colors)
  )
  
  Heatmap(mat,
          top_annotation = colTop,
          bottom_annotation = colBottom,
          left_annotation = rowAnno,
          heatmap_legend_param = hlp,
          show_row_names = show_row_names,
          show_column_names = show_col_names,
          cluster_rows = T,
          row_names_gp = gpar(fontsize = row_size),
          column_names_gp = gpar(fontsize = col_size),
          column_title = title,
          ...)
}

heatmaps = list()
```


## Beehive plots

Function for plotting a given gene.

```{r}
plot_gene_simple = function(dds, ensembl_gene_id, gene_name, intgroup=coi, colorby="death"){
  require(DESeq2)
  require(tidyverse)
  
  #Normalizes by size factors and log transform the data with a pseudocount of 0.5
  d <- plotCounts(dds, gene=ensembl_gene_id, intgroup=c(intgroup,colorby), normalize=T, transform=T, returnData=TRUE)
  
  d[,intgroup] <- as.factor(d[,intgroup])
  d[,colorby] <- as.factor(d[,colorby])
  
  dplot = ggplot(d, aes(x=get(intgroup), y=count)) +
    geom_point(position=position_jitter(w=0.1,h=0),
               aes(color=get(colorby))) +
    geom_boxplot(alpha = 0) +
    scale_y_log10() + xlab(intgroup) + labs(color=colorby) +
    ggtitle(paste(gene_name,ensembl_gene_id, sep=":")) +
    scale_color_viridis_d() +
    theme_bw()
  
  return(dplot)
}

beehives = list()
```

# Overview

Number of differentially expressed gene for comparison `r params$comparison`

```{r}
lapply(sig_results, nrow)
```

Number of samples in each comparison:

```{r}
lapply(ddsl, function(x) table(colData(x)[,coi]))
```

Overlap between subgroup significant hits

```{r}
sig_genes = lapply(sig_results, function(x) x$ensembl_gene_id)
names(sig_genes) = names(sig_results)

upset(fromList(sig_genes), order.by = "freq", 
       #empty.intersections = "on",
       #nsets = length(sig_genes), 
       #set_size.show = TRUE,
       mainbar.y.label = ts,
       sets.bar.color = "blue", 
       main.bar.color = "black", 
       #set_size.scale_max = 1200,
       sets.x.label = "Sig genes")
```

# Basal {.tabset}

```{r}
cur <- "Basal"
```

Number of significant genes at padj threshold `r fdr` and absolute log2 fold change threshold `r alfc`

```{r}
nrow(sig_results[[cur]])
```

Significant hits:

```{r}
sig_results[[cur]]
```

Genes enriched in `r params$comparison` (names only, max 500):

```{r}
sig_results[[cur]] %>%
  filter(l2fc > 0) %>%
  pull(gene_name) %>%
  head(500)
```

Genes depleted  in `r params$comparison` (names only, max 500):

```{r}
sig_results[[cur]] %>%
  filter(l2fc < 0) %>%
  pull(gene_name) %>%
  head(500)
```

Number of significant genes that meet padj threshold `r fdr` but not abs(l2fc) threshold `r alfc`:

```{r}
annotated_results[[cur]] %>%
  arrange(padj) %>%
  mutate(fdr = padj <= !!fdr, l2fc = abs(l2fc) >= !!alfc) %>%
  select(fdr, l2fc)  %>% table()
```

Variance in expression for genes that hit the padj threshold but not the lfc threshold:

```{r}
annotated_results[[cur]] %>%
  arrange(padj) %>%
  filter(padj <= !!fdr, abs(l2fc) < !!alfc) %>%
  #select(gene_name, baseMean) %>%
  group_by(gene_type) %>%
  summarise(av_baseMean = mean(baseMean),
            sd_baseMean = sd(baseMean),
            count = n()) %>%
  arrange(desc(count))
```

## Volcano plot

```{r}
volcano_plots[[cur]]
```

## Heatmap

```{r, fig.height=6, fig.width=10}
heatmaps[[cur]] <- deseq_heatmap(mat = geneEx[[cur]],
                                 sig_results = sig_results[[cur]],
                                 title = paste(ts, cur, sep=": "))
heatmaps[[cur]]
```

## Beehive plots

Plot top 6 significant genes (or less if fewer are significant)

```{r}
sigcur = nrow(sig_results[[cur]])

if(sigcur > 0){
  bl <- list()

beemax <- ifelse(sigcur > 6,
                  6, sigcur)

for(i in 1:beemax){
  bl[[length(bl) + 1]] <- plot_gene_simple(
    dds=ddsl[[cur]],
    ensembl_gene_id=sig_results[[cur]]$ensembl_gene_id[i],
    gene_name=sig_results[[cur]]$gene_name[i],
  )
  }
} else {
  bl <- 0
}


beehives[[cur]] <- bl
beehives[[cur]]
```

## Pathway

A fast method for a general overview via Fisher's exact pathway analysis. 
See FlexGSEA for a properly rigorous analysis.

Pathways with Fisher's exact fdr < `r pathway_cutoff`

```{r}
pathways[[cur]]
```

For comparison `r params$comparison`

```{r, fig.height=6, fig.width=8}
if(nrow(pathways[[cur]])>0){
  pathways[[cur]] %>%
  plot_enrichment(
        enrich_res = .,
        fdr = pathway_cutoff,
        max_nchar_path = 40,
        max_path = 20,
        title = paste("Top 20 sig pathways fdr <", pathway_cutoff, ": ", cur)
      )
} else {
  print("No significant pathways")
}

```

# Her2 {.tabset}

```{r}
cur <- "Her2"
```

Number of significant genes at padj threshold `r fdr` and absolute log2 fold change threshold `r alfc`

```{r}
nrow(sig_results[[cur]])
```

Significant hits:

```{r}
sig_results[[cur]]
```

Genes enriched in `r params$comparison` (names only, max 500):

```{r}
sig_results[[cur]] %>%
  filter(l2fc > 0) %>%
  pull(gene_name) %>%
  head(500)
```

Genes depleted  in `r params$comparison` (names only, max 500):

```{r}
sig_results[[cur]] %>%
  filter(l2fc < 0) %>%
  pull(gene_name) %>%
  head(500)
```

Number of significant genes that meet padj threshold `r fdr` but not abs(l2fc) threshold `r alfc`:

```{r}
annotated_results[[cur]] %>%
  arrange(padj) %>%
  mutate(fdr = padj <= !!fdr, l2fc = abs(l2fc) >= !!alfc) %>%
  select(fdr, l2fc)  %>% table()
```

Variance in expression for genes that hit the padj threshold but not the lfc threshold:

```{r}
annotated_results[[cur]] %>%
  arrange(padj) %>%
  filter(padj <= !!fdr, abs(l2fc) < !!alfc) %>%
  #select(gene_name, baseMean) %>%
  group_by(gene_type) %>%
  summarise(av_baseMean = mean(baseMean),
            sd_baseMean = sd(baseMean),
            count = n()) %>%
  arrange(desc(count))
```

## Volcano plot

```{r}
volcano_plots[[cur]]
```

## Heatmap

```{r, fig.height=6, fig.width=10}
heatmaps[[cur]] <- deseq_heatmap(mat = geneEx[[cur]],
                                 sig_results = sig_results[[cur]],
                                 title = paste(ts, cur, sep=": "))
heatmaps[[cur]]
```

## Beehive plots

Plot top 6 significant genes (or less if fewer are significant)

```{r}
sigcur = nrow(sig_results[[cur]])

if(sigcur > 0){
  bl <- list()

beemax <- ifelse(sigcur > 6,
                  6, sigcur)

for(i in 1:beemax){
  bl[[length(bl) + 1]] <- plot_gene_simple(
    dds=ddsl[[cur]],
    ensembl_gene_id=sig_results[[cur]]$ensembl_gene_id[i],
    gene_name=sig_results[[cur]]$gene_name[i],
  )
  }
} else {
  bl <- 0
}


beehives[[cur]] <- bl
beehives[[cur]]
```

## Pathway

A fast method for a general overview via Fisher's exact pathway analysis. 
See FlexGSEA for a properly rigorous analysis.

Pathways with Fisher's exact fdr < `r pathway_cutoff`

```{r}
pathways[[cur]]
```

For comparison `r params$comparison`

```{r, fig.height=6, fig.width=8}
if(nrow(pathways[[cur]])>0){
  pathways[[cur]] %>%
  plot_enrichment(
        enrich_res = .,
        fdr = pathway_cutoff,
        max_nchar_path = 40,
        max_path = 20,
        title = paste("Top 20 sig pathways fdr <", pathway_cutoff, ": ", cur)
      )
} else {
  print("No significant pathways")
}

```

# LumA {.tabset}

```{r}
cur <- "LumA"
```

Number of significant genes at padj threshold `r fdr` and absolute log2 fold change threshold `r alfc`

```{r}
nrow(sig_results[[cur]])
```

Significant hits:

```{r}
sig_results[[cur]]
```

Genes enriched in `r params$comparison` (names only, max 500):

```{r}
sig_results[[cur]] %>%
  filter(l2fc > 0) %>%
  pull(gene_name) %>%
  head(500)
```

Genes depleted  in `r params$comparison` (names only, max 500):

```{r}
sig_results[[cur]] %>%
  filter(l2fc < 0) %>%
  pull(gene_name) %>%
  head(500)
```

Number of significant genes that meet padj threshold `r fdr` but not abs(l2fc) threshold `r alfc`:

```{r}
annotated_results[[cur]] %>%
  arrange(padj) %>%
  mutate(fdr = padj <= !!fdr, l2fc = abs(l2fc) >= !!alfc) %>%
  select(fdr, l2fc)  %>% table()
```

Variance in expression for genes that hit the padj threshold but not the lfc threshold:

```{r}
annotated_results[[cur]] %>%
  arrange(padj) %>%
  filter(padj <= !!fdr, abs(l2fc) < !!alfc) %>%
  #select(gene_name, baseMean) %>%
  group_by(gene_type) %>%
  summarise(av_baseMean = mean(baseMean),
            sd_baseMean = sd(baseMean),
            count = n()) %>%
  arrange(desc(count))
```

## Volcano plot

```{r}
volcano_plots[[cur]]
```

## Heatmap

```{r, fig.height=6, fig.width=10}
heatmaps[[cur]] <- deseq_heatmap(mat = geneEx[[cur]],
                                 sig_results = sig_results[[cur]],
                                 title = paste(ts, cur, sep=": "))
heatmaps[[cur]]
```

## Beehive plots

Plot top 6 significant genes (or less if fewer are significant)

```{r}
sigcur = nrow(sig_results[[cur]])

if(sigcur > 0){
  bl <- list()

beemax <- ifelse(sigcur > 6,
                  6, sigcur)

for(i in 1:beemax){
  bl[[length(bl) + 1]] <- plot_gene_simple(
    dds=ddsl[[cur]],
    ensembl_gene_id=sig_results[[cur]]$ensembl_gene_id[i],
    gene_name=sig_results[[cur]]$gene_name[i],
  )
  }
} else {
  bl <- 0
}


beehives[[cur]] <- bl
beehives[[cur]]
```

## Pathway

A fast method for a general overview via Fisher's exact pathway analysis. 
See FlexGSEA for a properly rigorous analysis.

Pathways with Fisher's exact fdr < `r pathway_cutoff`

```{r}
pathways[[cur]]
```

For comparison `r params$comparison`

```{r, fig.height=6, fig.width=8}
if(nrow(pathways[[cur]])>0){
  pathways[[cur]] %>%
  plot_enrichment(
        enrich_res = .,
        fdr = pathway_cutoff,
        max_nchar_path = 40,
        max_path = 20,
        title = paste("Top 20 sig pathways fdr <", pathway_cutoff, ": ", cur)
      )
} else {
  print("No significant pathways")
}

```


# LumB {.tabset}

```{r}
cur <- "LumB"
```

Number of significant genes at padj threshold `r fdr` and absolute log2 fold change threshold `r alfc`

```{r}
nrow(sig_results[[cur]])
```

Significant hits:

```{r}
sig_results[[cur]]
```

Genes enriched in `r params$comparison` (names only, max 500):

```{r}
sig_results[[cur]] %>%
  filter(l2fc > 0) %>%
  pull(gene_name) %>%
  head(500)
```

Genes depleted  in `r params$comparison` (names only, max 500):

```{r}
sig_results[[cur]] %>%
  filter(l2fc < 0) %>%
  pull(gene_name) %>%
  head(500)
```

Number of significant genes that meet padj threshold `r fdr` but not abs(l2fc) threshold `r alfc`:

```{r}
annotated_results[[cur]] %>%
  arrange(padj) %>%
  mutate(fdr = padj <= !!fdr, l2fc = abs(l2fc) >= !!alfc) %>%
  select(fdr, l2fc)  %>% table()
```

Variance in expression for genes that hit the padj threshold but not the lfc threshold:

```{r}
annotated_results[[cur]] %>%
  arrange(padj) %>%
  filter(padj <= !!fdr, abs(l2fc) < !!alfc) %>%
  #select(gene_name, baseMean) %>%
  group_by(gene_type) %>%
  summarise(av_baseMean = mean(baseMean),
            sd_baseMean = sd(baseMean),
            count = n()) %>%
  arrange(desc(count))
```

## Volcano plot

```{r}
volcano_plots[[cur]]
```

## Heatmap

```{r, fig.height=6, fig.width=10}
heatmaps[[cur]] <- deseq_heatmap(mat = geneEx[[cur]],
                                 sig_results = sig_results[[cur]],
                                 title = paste(ts, cur, sep=": "))
heatmaps[[cur]]
```

## Beehive plots

Plot top 6 significant genes (or less if fewer are significant)

```{r}
sigcur = nrow(sig_results[[cur]])

if(sigcur > 0){
  bl <- list()

beemax <- ifelse(sigcur > 6,
                  6, sigcur)

for(i in 1:beemax){
  bl[[length(bl) + 1]] <- plot_gene_simple(
    dds=ddsl[[cur]],
    ensembl_gene_id=sig_results[[cur]]$ensembl_gene_id[i],
    gene_name=sig_results[[cur]]$gene_name[i],
  )
  }
} else {
  bl <- 0
}


beehives[[cur]] <- bl
beehives[[cur]]
```

## Pathway

A fast method for a general overview via Fisher's exact pathway analysis. 
See FlexGSEA for a properly rigorous analysis.

Pathways with Fisher's exact fdr < `r pathway_cutoff`

```{r}
pathways[[cur]]
```

For comparison `r params$comparison`

```{r, fig.height=6, fig.width=8}
if(nrow(pathways[[cur]])>0){
  pathways[[cur]] %>%
  plot_enrichment(
        enrich_res = .,
        fdr = pathway_cutoff,
        max_nchar_path = 40,
        max_path = 20,
        title = paste("Top 20 sig pathways fdr <", pathway_cutoff, ": ", cur)
      )
} else {
  print("No significant pathways")
}

```

# Save data

Save the significant genes and results from each comparison.

```{r}
#Save the results for all genes in an Excel file with multiple tabs
f1 = paste0("14_subgroup_diffex_", params$comparison, "_allgenes.xlsx")
resdf = annotated_results
resdf$thresholds = tibble(metric = c("padj", "l2fc"),
                          cutoff = c(fdr, alfc))
#The tab names will be the list names
openxlsx::write.xlsx(resdf, file = here("results", "diffex", f1))

#Separate file for sig genes only
resdf = sig_results
resdf$thresholds = tibble(metric = c("padj", "l2fc"),
                          cutoff = c(fdr, alfc))
f2 = paste0("14_subgroup_diffex_", params$comparison, "_sig_genes.xlsx")
write.xlsx(resdf, file = here("results", "diffex", f2))
rm(resdf)
```

```{r}
sessionInfo()
```
