---
title: "Antibody isotypes in PPBC"
author: "Kat Moore"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 4
    df_print: paged
params:
  bx_annot: "data/rnaseq/processed/bx_annot.Rds"
  dds: "data/rnaseq/processed/08_dds_ovr_inv_vs_rest.Rds"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(here)
library(DESeq2)
library(edgeR)
library(survival)
library(survminer)
library(tidyverse)
library(ggpubr)

theme_set(theme_bw())
```

Supplementary plots regarding antibody isotype abundance, as indicated by RNA expression of IG heavy chains.

## Load data

Gene expression & annotation:

```{r load data}
dds <- readRDS(here(params$dds))
```

Gene annotation, including uniprot/entrez, where available.

```{r}
bx_annot <- readRDS(here(params$bx_annot))
head(bx_annot)
```


## Select IG heavy chains

Detect relevant genes. We should get:

α – IgA, further divided into subclasses IgA1 and IgA2
δ – IgD
γ – IgG, further divided into subclasses IgG1 to IgG4
ε – IgE
μ – IgM

```{r}
ag_annot <- bx_annot %>%
  filter(gene_type == "IG C gene") %>%
  filter(str_detect(tolower(description), "immuno.*heavy")) %>%
  arrange(gene_name)

unique(ag_annot$gene_name)
```

Filter the count matrix for these genes.

```{r}
igExpr <- assay(dds)[rownames(assay(dds)) %in% unique(ag_annot$ensembl_gene_id),]
```


## Pool isotype expression

Match gene to isotype

```{r}
ag_annot <- ag_annot %>%
  mutate(isotype = str_remove(str_remove(gene_name, "[[:digit:]]$"), "H"),
         .after = gene_name) %>%
  mutate(isotype = str_replace(isotype, "IG", "Ig"))

ag_annot
```

Combine annotation with expression.

```{r}
igDF <- igExpr %>%
  as.data.frame() %>%
  rownames_to_column("ensembl_gene_id") %>%
  right_join(ag_annot, ., by ="ensembl_gene_id") %>%
  #Not needed right now
  select(-gene_type, -description, -entrez_id, -uniprot_id) %>%
  pivot_longer(cols = -c(ensembl_gene_id, gene_name, isotype),
               names_to = "sample_name", values_to = "expr")

head(igDF)
```

Pool the isotype expression and reshape into matrix.

```{r}
igMat <- igDF %>%
  group_by(isotype, sample_name) %>%
  summarise(pooled_expr = sum(expr), .groups="drop") %>%
  pivot_wider(names_from = sample_name, values_from = pooled_expr) %>%
  column_to_rownames("isotype") %>%
  as.matrix()

igMat[,1:4]
```

## Absent isotypes

Not all antibody isotypes are present, the absent ones are below the minimum count threshold applied during differential expression and for Cox regressions.

These isotypes pass the minimum count threshold:

```{r}
ag_annot %>%
  filter(ensembl_gene_id %in% rownames(assay(dds))) %>%
  pull(isotype) %>% unique() %>% sort()
```

These isotypes are absent due to low abundance:

```{r}
ag_annot %>%
  filter(!isotype %in% filter(ag_annot, ensembl_gene_id %in% rownames(assay(dds)))$isotype) %>%
  pull(isotype) %>% unique()
```

## Library size normalization

Create a DGEList for easier normalization.

```{r}
dgeIg <- DGEList(counts = igMat,
                 #library size should be reflective of all genes, not just IG
                 lib.size = colSums(assay(dds)),
                 samples = as.data.frame(colData(dds)),
                 group = as.data.frame(colData(dds))$PPBC)

#dgeIg
```

We use TMM followed by log2 for normalization, just like we did for Cox regressions and box plots.

```{r}
normTMMlog2 <- function(object){
  object = calcNormFactors(object, method="TMM")
  object = cpm(object, log=T, normalized.lib.sizes=T)
  return(object)
}

dgeIg$samples$age_diagnosis <- dgeIg$samples$year_diagnosis-dgeIg$samples$year_birth

# Reshape into data frame for plotting
igNorm <- normTMMlog2(dgeIg) %>%
  as.data.frame() %>%
  rownames_to_column("isotype") %>%
  pivot_longer(cols = -c(isotype), names_to = "sample_name", values_to = "normExpr") %>%
  left_join(., select(dgeIg$samples, sample_name, study_group, PPBC,
                      #Include survival variables
                      death, time_OS_months,
                      time_DRS_months, distant_recurrence,
                      age_diagnosis, PAM50, 
                      grade, stage, surgery,
                      radiotherapy = RT, hormonetherapy = HT, chemotherapy = CT,
                      herceptin),
            by = "sample_name")

head(igNorm)
```

## Isotype box plot

```{r, fig.width=6, fig.height=5}
comp = list(c("involuting", "nulliparous"),
            c("involuting", "pregnant"))

igNorm %>%
  ggplot(aes(x=PPBC, y = normExpr)) +
  stat_compare_means(comparison = comp) +
  geom_boxplot(alpha = 0) +
  geom_jitter(height = 0, width = 0.2, aes(color = isotype)) +
  facet_wrap(~isotype, strip.position = "top", scales = "free_y") +
  #Prevents p values from being cut off
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.15))) +
  theme(legend.position = "none") +
  ggtitle("Normalized isotype expression in PPBC") +
  scale_color_viridis_d()
```

## Isotype bar plot

The totals also reflect the different number of sampels between groups.

```{r}
igNorm %>%
  ggplot(aes(x = PPBC, y = normExpr, fill = isotype)) +
  geom_bar(stat = "identity") +
  scale_fill_viridis_d() +
  ggtitle("Normalized isotype expression in PPBC")
```

## Isotype percentage bar plot

```{r}
igPerc <- igNorm %>%
  group_by(isotype, PPBC) %>%
  summarise(total_iso = sum(normExpr),
            .groups = "drop") %>%
  group_by(PPBC) %>%
  mutate(percent = total_iso/sum(total_iso))

igPerc %>%
  select(-total_iso) %>%
  mutate(percent = round(percent, 3) * 100) %>%
  pivot_wider(names_from = PPBC, values_from = percent) %>%
  column_to_rownames("isotype") %>%
  as.matrix()
```


```{r}
igPerc%>%
  ggplot(aes(x = PPBC, y = percent, fill = isotype)) +
  geom_bar(stat = "identity") +
  scale_fill_viridis_d() +
  ggtitle("Percentage isotype expression in PPBC")
```

## Reshape data for survival

```{r}
ig_ntile <- function(df, colname, n = 3){
  

  stopifnot(n %in% c(2,3,4))
  
  labels = switch(as.character(n),
                  "2" = c("low", "high"),
                  "3" = c("low", "medium", "high"),
                  "4" = c("Q1", "Q2", "Q3", "Q4")
  )
 
  igrank = paste0(colname, "_rank")
  
  #df[[igtile]] <- dplyr::ntile(df[[colname]], n)
  df[[igrank]] <- factor(dplyr::ntile(df[[colname]], n), labels = labels)
  
  #df <- df %>%
  #  relocate(contains("rank"), .after = all_of(colname))
  
  newdf <- as.data.frame(df[,igrank,drop=F])
  rownames(newdf) <- df$sample_name
  
  newdf
}

igSurv <- pivot_wider(igNorm, names_from = isotype, values_from = normExpr) %>%
  relocate(contains("Ig"), .after="sample_name")

igSurv <- lapply(unique(igNorm$isotype), function(x){ig_ntile(igSurv, x, 2)}) %>%
  bind_cols() %>%
  bind_cols(., igSurv) %>%
  relocate(sample_name, .before = everything())

head(igSurv)

stopifnot(all(rownames(igSurv) == igSurv$sample_name))
```


## Kaplan-meier

### IgA

```{r, fig.width=6, fig.height=4}
#OS
survfit(Surv(time=time_OS_months, event=death) ~ IgA_rank,
        data = as.data.frame(igSurv)) %>% 
  ggsurvplot(fit = ., facet.by = "PPBC",
             xlab = "Months", 
             ylab = "Overall survival probability",
             title = paste0("Normalized IgA expression and survival"),
             palette = c("gold1", 
                         #"darkorange", 
                         "red"),
             pval = T,
             ggtheme = theme_bw(),
             risk.table = F,
             cumevents = F, cumcensor = F,
             test.for.trend = F)

#DRS
survfit(Surv(time=time_DRS_months, event=distant_recurrence) ~ IgA_rank,
        data = as.data.frame(igSurv)) %>% 
  ggsurvplot(fit = ., facet.by = "PPBC",
             xlab = "Months", 
             ylab = "Distant recurrence probability",
             title = paste0("Normalized IgA expression and DRS"),
             palette = c("gold1", 
                         #"darkorange", 
                         "red"),
             pval = T,
             ggtheme = theme_bw(),
             risk.table = F,
             cumevents = F, cumcensor = F,
             test.for.trend = F)
```

### IgD

```{r, fig.width=6, fig.height=4}
survfit(Surv(time=time_OS_months, event=death) ~ IgD_rank,
        data = as.data.frame(igSurv)) %>% 
  ggsurvplot(fit = ., facet.by = "PPBC",
             xlab = "Months", 
             ylab = "Overall survival probability",
             title = paste0("Normalized IgD expression and survival"),
             palette = c("gold1", 
                         #"darkorange", 
                         "red"),
             pval = T,
             ggtheme = theme_bw(),
             risk.table = F,
             cumevents = F, cumcensor = F,
             test.for.trend = F)

#DRS
survfit(Surv(time=time_DRS_months, event=distant_recurrence) ~ IgD_rank,
        data = as.data.frame(igSurv)) %>% 
  ggsurvplot(fit = ., facet.by = "PPBC",
             xlab = "Months", 
             ylab = "Distant recurrence probability",
             title = paste0("Normalized IgD expression and DRS"),
             palette = c("gold1", 
                         #"darkorange", 
                         "red"),
             pval = T,
             ggtheme = theme_bw(),
             risk.table = F,
             cumevents = F, cumcensor = F,
             test.for.trend = F)
```

### IgG

```{r, fig.width=6, fig.height=4}
survfit(Surv(time=time_OS_months, event=death) ~ IgG_rank,
        data = as.data.frame(igSurv)) %>% 
  ggsurvplot(fit = ., facet.by = "PPBC",
             xlab = "Months", 
             ylab = "Overall survival probability",
             title = paste0("Normalized IgG expression and survival"),
             palette = c("gold1", 
                         #"darkorange", 
                         "red"),
             pval = T,
             ggtheme = theme_bw(),
             risk.table = F,
             cumevents = F, cumcensor = F,
             test.for.trend = F)

#DRS
survfit(Surv(time=time_DRS_months, event=distant_recurrence) ~ IgG_rank,
        data = as.data.frame(igSurv)) %>% 
  ggsurvplot(fit = ., facet.by = "PPBC",
             xlab = "Months", 
             ylab = "Distant recurrence probability",
             title = paste0("Normalized IgG expression and DRS"),
             palette = c("gold1", 
                         #"darkorange", 
                         "red"),
             pval = T,
             ggtheme = theme_bw(),
             risk.table = F,
             cumevents = F, cumcensor = F,
             test.for.trend = F)
```

### IgM

```{r, fig.width=6, fig.height=4}
survfit(Surv(time=time_OS_months, event=death) ~ IgM_rank,
        data = as.data.frame(igSurv)) %>% 
  ggsurvplot(fit = ., facet.by = "PPBC",
             xlab = "Months", 
             ylab = "Overall survival probability",
             title = paste0("Normalized IgM expression and survival"),
             palette = c("gold1", 
                         #"darkorange", 
                         "red"),
             pval = T,
             ggtheme = theme_bw(),
             risk.table = F,
             cumevents = F, cumcensor = F,
             test.for.trend = F)

#DRS
survfit(Surv(time=time_DRS_months, event=distant_recurrence) ~ IgM_rank,
        data = as.data.frame(igSurv)) %>% 
  ggsurvplot(fit = ., facet.by = "PPBC",
             xlab = "Months", 
             ylab = "Distant recurrence probability",
             title = paste0("Normalized IgM expression and DRS"),
             palette = c("gold1", 
                         #"darkorange", 
                         "red"),
             pval = T,
             ggtheme = theme_bw(),
             risk.table = F,
             cumevents = F, cumcensor = F,
             test.for.trend = F)
```

## CoxPH

Subset data to two major PPBC groups.

```{r}
igInv <- igSurv %>%
  filter(PPBC == "involuting")
nrow(igInv)

igNull <- igSurv %>%
  filter(PPBC == "nulliparous")

nrow(igNull)
```

### All PPBC groups

OS

```{r}
coxph(Surv(time=time_OS_months,
           event=death) ~ IgA + IgD + IgG + IgM,
      data = igSurv) %>%
  summary()
```

DRS

```{r}
coxph(
  Surv(time=time_DRS_months,
       event=distant_recurrence) ~ IgA + IgD + IgG + IgM,
  data = igSurv) %>%
  summary()
```

### Involuting

OS

```{r}
coxph(Surv(time=time_OS_months,
           event=death) ~ IgA + IgD + IgG + IgM,
      data = igInv) %>%
  summary()
```

DRS

```{r}
coxph(
  Surv(time=time_DRS_months,
       event=distant_recurrence) ~ IgA + IgD + IgG + IgM,
  data = igInv) %>%
  summary()
```

### Nulliparous

OS

```{r}
coxph(Surv(time=time_OS_months,
           event=death) ~ IgA + IgD + IgG + IgM,
      data = igNull) %>%
  summary()
```

DRS

```{r}
coxph(
  Surv(time=time_DRS_months,
       event=distant_recurrence) ~ IgA + IgD + IgG + IgM,
  data = igNull) %>%
  summary()
```

## Session info

```{r}
sessionInfo()
```
