---
title: "Gene unity setup"
author: "Kat Moore"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: yes
    toc_float: yes
    df_print: paged
    highlight: kate
    canon: true
params:
    coxres: "results/rnaseq/survival/12_cox_allgenes.xlsx"
    gx_annot: "data/rnaseq/processed/bx_annot.Rds"
    coxdata: "data/rnaseq/processed/12_coxdata.Rds"
    dds: "data/rnaseq/processed/08_dds_ovr_inv_vs_rest.Rds"
    diffex_lrt: "results/rnaseq/diffex/06_LRT_allgenes.xlsx"
    diffex_pw: "results/rnaseq/diffex/07_pairwise_comparisons_allgenes.xlsx"
    diffex_ovr: "results/rnaseq/diffex/08_one_vs_rest_allgenes.xlsx"
    inv_diffex: "results/rnaseq/diffex/09_diffex_involution_duration.xlsx"
    bf_diffex: "results/rnaseq/diffex/09b_diffex_breastfeeding_duration.xlsx"
---

This notebook will unify the various differential expression and survival analyses to identify a list of interesting genes for discussion in the manuscript.

```{r, message=F}
library(DESeq2)
library(survival)
library(ggthemes)
library(survminer)
library(pheatmap)
library(RColorBrewer)
library(edgeR)
library(openxlsx)
library(magrittr)
library(here)
library(ggpubr)
library(tidyverse)

theme_set(theme_bw())
```

We aggregate all of the previous results for summary and print the number of significant hits per analysis (where applicable).
The output can be used to produce individual gene reports.

# Count and metadata {.tabset}

In this tabset, raw results are loaded and processed for use elsewhere (shinyApp).
Click on the tabs if you want to inspect the data.

## Counts

Raw counts:

```{r}
dds = readRDS(here(params$dds))
```

## Annotation

A larger gene annotation data frame that includes uniprot and entrez ids, for webscraping:

```{r}
bx_annot <- readRDS(here(params$gx_annot))
head(bx_annot)
```

## Transposed matrix for Cox regressions

Also created in notebook 12. Gene expression columns are TMM-log2 normalized.

```{r}
coxdata = readRDS(here(params$coxdata))
colnames(coxdata)[1:30]
```

## Involution vs rest encoding

```{r}
coxdata$involution = as.numeric(coxdata$study_group == "ppbcpw")

coxdata = coxdata %>% select(sample_name:study_group, involution, everything())

table(coxdata$involution, coxdata$PPBC)
```

## Sample data

Separate data frame containing exclusively the sample data without gene expression.

```{r}
genecol = which(colnames(coxdata)=="ENSG00000000003")
sample_data = coxdata[,1:(genecol - 1)]

head(sample_data)
```

## Gene expression df

For plotting

```{r}
geneEx = rownames_to_column(as.data.frame(coxdata[,genecol:ncol(coxdata)]),"sample_name")

head(geneEx[,1:10])
```

In gene x sample format:

```{r}
ens_mat <- t(coxdata[,genecol:ncol(coxdata)])
#dim(ens_mat)
ens_mat[1:10,1:10]
```

# Diffex results

Divided into three sections:

* PPBC primary diffex (Likelihood ratio test, pairwise comparisons, and one vs rest comparisons)
* Subgroup (Involution vs rest, involution vs nulliparous and involution vs pregnant for Basal, Her2, and LumA/B samples)
* Involution duration (<6mo vs >6mo) and breastfeeding duration (<1 mo vs >1 mo)

## Main diffex:

```{r, collapse=T}
lrt <- list(readxl::read_excel(here(params$diffex_lrt),
                          sheet = "repLRT_all"))
names(lrt) <- "LRT"

pw <- here(params$diffex_pw) %>% 
  readxl::excel_sheets() %>% 
  purrr::set_names() %>% 
  map(readxl::read_excel, path = here(params$diffex_pw))
#names(pw)

ovr = here(params$diffex_ovr) %>% 
  readxl::excel_sheets() %>% 
  purrr::set_names() %>% 
  map(readxl::read_excel, path = here(params$diffex_ovr))
#names(ovr)

#Combine all results into single list.
res_list = append(lrt, pw)
res_list = append(res_list, ovr)
names(res_list) = str_remove_all(str_remove_all(names(res_list), "rep_"), "_all")
names(res_list) = str_replace(names(res_list), "_", "_vs_")
names(res_list)
```

Number of hits per group:

```{r}
sapply(res_list, function(x){
  x %>%
    filter(padj <= 0.05 & abs(log2FoldChange) > 0.5) %>%
    nrow()
}) %>% enframe("comparison", "significant genes")
```

Sanity check: Ensure that the number of genes matches the minimum count filter.
The "vs_rest" comparisons start with a few duplicates because we included an `hgnc_symbol` in those data frames.
Some genes have more than one hgnc symbol per ensembl gene id, which results in duplicates.

```{r}
min.counts = nrow(dds)
min.counts

#Compute the results of vs_rest separately because they have an extra column
stopifnot(unique(c(
  sapply(res_list[!str_detect(names(res_list), "vs_rest")], nrow),
  sapply(res_list[str_detect(names(res_list), "vs_rest")], function(x) {
    x %>% select(-hgnc_symbol) %>%
      distinct() %>% nrow()
  })
)) == min.counts)
```


### Important comparisons

Involution vs rest

```{r}
res_list$inv_vs_rest %>%
  filter(abs(log2FoldChange) > 0.5 & padj <= 0.05) %>%
  mutate(
    padj = signif(padj, 2),
    #pval = signif(pvalue, 2),
    log2FoldChange = round(log2FoldChange, 3),
    baseMean = round(baseMean, 1)
  ) %>% arrange(padj) %>%
  select(gene_name, padj, #pval, 
         log2FoldChange, baseMean,
         ensembl_gene_id, gene_type) %>%
  head(100)
```

Involution vs nulliparous

```{r}
res_list$inv_vs_nonprbc %>%
  filter(abs(log2FoldChange) > 0.5 & padj <= 0.05) %>%
  mutate(
    padj = signif(padj, 2),
    #pval = signif(pvalue, 2),
    log2FoldChange = round(log2FoldChange, 3),
    baseMean = round(baseMean, 1)
  ) %>% arrange(padj) %>%
  select(gene_name, padj, #pval, 
         log2FoldChange, baseMean,
         ensembl_gene_id, gene_type) %>%
  head(100)
```

## Subgroup analysis:

```{r}
read_excel_tabs = function(path){
  tab_names = readxl::excel_sheets(path)
  ldf = lapply(tab_names, function(x) readxl::read_excel(path, sheet = x))
  names(ldf) = tab_names
  ldf <- ldf[names(ldf)!="thresholds"]
  ldf <- bind_rows(ldf, .id = "subgroup")
  return(ldf)
}

sub_reslist <- list.files(here("results/rnaseq/diffex/"),
           pattern = "14_subgroup.*allgenes.xlsx") %>%
  set_names(str_remove(str_remove(., "14_"),
            "_allgenes.xlsx")) %>%
  map(~read_excel_tabs(here("results/rnaseq/diffex",.)))

names(sub_reslist)
#Manual fix
names(sub_reslist) <- str_replace(
  str_remove_all(names(sub_reslist), "diffex_|ppbc_"),
  "non_prbc", "nonprbc")
names(sub_reslist)
```

Rename to keep consistent (for functions).

```{r}
sub_reslist <- lapply(sub_reslist, function(x){
  rename(x, log2FoldChange = l2fc)
})
```

Hits per subgroup:

```{r}
sapply(sub_reslist, function(y){
  sapply(c("Basal", "Her2", "LumA", "LumB"),
         function(x) {
           sub = y[y$subgroup == x,]
           sub %>% dplyr::filter(padj <= 0.05 & abs(log2FoldChange) > 0.5) %>% nrow()
         })
})
```

## Involution and breastfeeding duration:

```{r}
read_excel_tabs = function(path){
  tab_names = readxl::excel_sheets(path)
  ldf = lapply(tab_names, function(x) readxl::read_excel(path, sheet = x))
  names(ldf) = tab_names
  ldf <- ldf[names(ldf)=="all"]
  ldf <- bind_rows(ldf)
  return(ldf)
}

#readxl::excel_sheets(here(params$bf_diffex))
bf <- readxl::read_excel(here(params$bf_diffex), sheet = "all") %>%
  mutate(diffex_duration = "breastfeeding", .before=everything()) %>%
  mutate(ref = "more than 1 month", group = "1 month or less")

inv <- readxl::read_excel(here(params$inv_diffex), sheet = "all") %>%
  mutate(diffex_duration = "involution", .before=everything()) %>%
  mutate(ref = "more than 6 months", group = "6 months or less")

invbf_reslist <- bind_rows(bf, inv) %>%
  rename(log2FoldChange = l2fc) %>%
  mutate(sig = ifelse(is.na(sig), F, sig))

invbf_reslist %>%
  select(diffex_duration, ref, group) %>% distinct()
```

Count threshold sanity check:

```{r}
stopifnot(
  unique(
    pull(
      count(group_by(
        distinct(select(invbf_reslist, -hgnc_symbol)),
        diffex_duration)),n)
    ) == min.counts
  ) 
```

Hits in each group (NAs are due to DESeq outlier handling:)

```{r}
invbf_reslist %>% group_by(diffex_duration, sig) %>%
  count()
```

Few results in breastfeeding only.
Involution results:

```{r}
invbf_reslist %>%
  filter(diffex_duration == "involution") %>%
  filter(sig == T) %>%
  mutate(
    padj = signif(padj, 2),
    #pval = signif(pvalue, 2),
    log2FoldChange = round(log2FoldChange, 3),
    baseMean = round(baseMean, 1)
  ) %>% arrange(padj) %>% select(diffex_duration, gene_name,
                                 padj, #pval, 
                                 log2FoldChange, baseMean,
                                 ensembl_gene_id,
                                 group, ref
  )  
```


# Genewise survival

```{r}
gs_sheets = readxl::excel_sheets(here(params$coxres))

#Excluding models run on involution samples only, as there were few hits.
#gs_sheets = gs_sheets[!str_detect(gs_sheets,"inv_only")]

gw_surv <- lapply(gs_sheets, function(x) readxl::read_excel(
  here(params$coxres), sheet = x))
names(gw_surv) = gs_sheets

gw_surv <- bind_rows(gw_surv, .id = "model")
gw_surv <- gw_surv %>%
  mutate(
    input = ifelse(str_detect(model, "inv_"), "ppbcpw samples", "all samples"),
    formula = if_else(
      str_detect(model, "uni"), "univariate", "multivariate"),
    outcome = if_else(
      str_detect(model, "os"), "overall survival", "distant recurrence"),
    family = ifelse(str_detect(model, "genewise"), "genewise", "interaction")
    ) %>%
  mutate(fdr = signif(fdr, 2),
         p.value = signif(p.value, 2),
         padj.inv_vs_rest = signif(padj.inv_vs_rest, 2),
         padj.inv_vs_nonprbc = signif(padj.inv_vs_nonprbc, 2),
         l2fc.inv_vs_rest = signif(l2fc.inv_vs_rest, 2),
         l2fc.inv_vs_nonprbc = signif(l2fc.inv_vs_nonprbc, 2))

gw_surv = gw_surv %>%
  select(model:fdr,p.value, everything())

# gw_surv = gw_surv %>%
#   mutate(model = recode(model,
#     "univ_cox_dr" = "univDRS",
#     "univ_cox_surv" = "univOS",
#     "multi_cox_dr" = "multiDRS",
#     "multi_cox_surv" = "multiOS"
#   ))

gw_surv %>%
  select(model, input, formula, outcome, family) %>% distinct() %>%
  as.matrix()

```

Count threshold sanity check:

```{r}
stopifnot(min.counts == length(unique(gw_surv$ensembl_gene_id)))
```

Top 50 hits on each model:

```{r}
for(u in unique(gw_surv$model)){
  gw_surv %>% filter(model == !!u) %>%
    arrange(fdr) %>% head(50) %>% print()
} 
```

Number of hits at various thresholds:

```{r}
thresh = seq(0.05, 0.3, by = 0.05)
model = unique(gw_surv$model)


sapply(thresh, function(y){
  sapply(model, function(x){
    z = gw_surv[gw_surv$model == x,]
    nrow(z[z$fdr <= y,])
  })
}) %>% magrittr::set_colnames(
  as.character(thresh)
)

rm(thresh)
rm(model)

```

# Save image

For loading in genewise reports.

```{r}
#First remove params, or they'll cause problems
rm(params)
save.image(here("data/rnaseq/processed/16_gene_report_environment.RData"))
```
